{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy.integrate\n",
    "import numpy as np\n",
    "from numba import njit, vectorize, float32, float64, boolean\n",
    "from numba.experimental import jitclass\n",
    "from matplotlib import pyplot as plt\n",
    "import multiprocessing\n",
    "import pandas as pd\n",
    "import holoviews as hv\n",
    "hv.extension('bokeh')\n",
    "from holoviews.operation.datashader import datashade\n",
    "\n",
    "# Ensure we don't hit floating point errors\n",
    "np.seterr('raise')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The problem is defined by the location of the currently active polymerases. Each polymerase is characterized by:\n",
    "1. It's location\n",
    "2. The length of the nascent mRNA tail ($x$)\n",
    "3. The DNA twist at this location ($\\phi$)\n",
    "4. The RNAC at this location ($\\theta$)\n",
    "\n",
    "We actually only need to track the first three; the fourth is completely determined by the first three.\n",
    "\n",
    "Given this state information, we can compute the relaxed supercoiling, which in turn gives us information on the "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "params = {\n",
    "    'mRNA_drag': 1/20, # pN nm^(alpha / 1)\n",
    "    'mRNA_exponent': 1, # the value of alpha\n",
    "    'DNA_twist_mobility': 10, # s pN nm\n",
    "    'RNAP_radius': 15, # nm\n",
    "    'RNAP_velocity': 20, # nm / s\n",
    "    'RNAP_torque_cutoff': 12, # pN nm\n",
    "    'RNAP_stall_torque_width': 3, #pN\n",
    "    'DNA_force': 1, # pN\n",
    "    'DNA_bend_plength': 50, # pN\n",
    "    'DNA_twist_plength': 95, # pN\n",
    "    'DNA_plectonome_twist_plength': 24, # pN\n",
    "    'temperature': 298, # K\n",
    "    'base_promoter_initiation_rate': 1 / 120, # 1 / sec\n",
    "    'topo_rate': 1 / 1200, # 1 / sec\n",
    "    'mRNA_deg_rate': 1 / 1200 # 1 / sec\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Most member variables are float64's except for left_bc_free/right_bc_free\n",
    "supercoiling_varspec = [(name,float64) for name in\n",
    "                       ['kb', 'w0', 'tau_s', 'tau_0', 'tau_p', 'sc_s', 'sc_p', 'rnac_r',\n",
    "                        'v_0', 't_c', 'stall_torque_width', 'chi', 'eta', 'alpha',\n",
    "                       'left_bc_position', 'right_bc_position', 'left_bc_val', 'right_bc_val']]\n",
    "supercoiling_varspec.append(('left_bc_free', boolean))\n",
    "supercoiling_varspec.append(('right_bc_free', boolean))\n",
    "\n",
    "@jitclass(supercoiling_varspec)\n",
    "class SupercoilingPolymeraseModel(object):\n",
    "    \"\"\"\n",
    "    Class that encapsulates all physical constants and\n",
    "    calculations needed to compute an ODE model for\n",
    "    supercoiling induced by transcription.\n",
    "    \n",
    "    It exposes the system_derivatives member function,\n",
    "    which can be used in an ODE solver to represent the motion\n",
    "    of each of the polymerases and the induced supercoiling.\n",
    "    \"\"\"\n",
    "    def __init__(self, A, C, P, T, f,\n",
    "                 v_0, critical_torque, stall_torque_width, rnac_r,\n",
    "                 chi, eta, alpha, left_bc_free, right_bc_free,\n",
    "                 left_bc_loc, right_bc_loc, left_bc_val, right_bc_val):\n",
    "        \"\"\"\n",
    "        Constructor that takes all relevant system constants and initalizes\n",
    "        internal parameters used by other functions.\n",
    "    \n",
    "        Args:\n",
    "        -----\n",
    "        A: DNA bend persistence length (nm)\n",
    "        C: DNA twist persistance length (nm)\n",
    "        P: DNA plectonome twist persistence length (nm)\n",
    "        T: Temperature of the system (K)\n",
    "        f: Constant force held on DNA (pN)\n",
    "        v_0: The maximum speed of the polymerase (nm / s)\n",
    "        critical_torque: Torque (in pN nm) at which the\n",
    "            polymerases begin to stall.\n",
    "        stall_torque_width: How wide the stall torque distribution is (pN)\n",
    "        rnac_r: The radius of a RNAP (used for hard-sphere replusion)\n",
    "        chi: The DNA twisting mobility (s pN nm)\n",
    "        eta: The The mRNA drag coefficent (pN (nm^(alpha - 1)))\n",
    "        alpha: The mRNA power-law scaling exponent.\n",
    "        left_bc_free: A boolean representing if the left BC is free.\n",
    "        right_bc_free: A boolean representing if the right BC is free.\n",
    "        left_bc_loc: If the left BC is not free, encodes the location of the BC.\n",
    "        right_bc_loc: If the right BC is not free, encodes the location of the BC.\n",
    "        left_bc_val: If the left BC is not free, encodes the value of phi at the BC.\n",
    "        right_bc_val: If the right BC is not free, encodes the value of phi at the BC.\n",
    "        \"\"\"\n",
    "        self.kb =  1380649 / 100000000 # pN nm / K\n",
    "        self.w0 =  1.85 # 1/nm\n",
    "\n",
    "        # Init torque model constants\n",
    "        c = self.kb * T * C * self.w0 ** 2 # pN\n",
    "        p = self.kb * T * P * self.w0 ** 2 # pN\n",
    "        g = f - np.sqrt(self.kb * T * f / A) # pN\n",
    "        cs = c * (1 - C / (4 * A) * np.sqrt(self.kb * T / (A * f))) # pN\n",
    "\n",
    "        # Compute critical torque model values\n",
    "        self.tau_s = cs / self.w0 # pN nm\n",
    "        self.tau_0 = np.sqrt(2 * p * g / (self.w0 ** 2 * (1 - p / cs))) # pN nm\n",
    "        self.tau_p = p / self.w0 # pN nm\n",
    "        self.sc_s = 1 / cs * np.sqrt(2 * p * g  / (1 - p / cs)) # dimless\n",
    "        self.sc_p = 1 / p * np.sqrt(2 * p * g / (1 - p / cs)) # dimless\n",
    "        \n",
    "        self.rnac_r = rnac_r\n",
    "        self.v_0 = v_0\n",
    "        self.t_c = critical_torque\n",
    "        self.stall_torque_width = stall_torque_width\n",
    "        self.chi = chi\n",
    "        self.eta = eta\n",
    "        self.alpha = alpha\n",
    "        self.left_bc_free = left_bc_free\n",
    "        self.right_bc_free = right_bc_free\n",
    "        self.left_bc_position = left_bc_loc\n",
    "        self.right_bc_position = right_bc_loc\n",
    "        self.left_bc_val = left_bc_val\n",
    "        self.right_bc_val = right_bc_val\n",
    "    \n",
    "    def evaluate_twist(self, x, states):\n",
    "        \"\"\"\n",
    "        Given the current polymerase states,\n",
    "        computes the excess DNA twist phi at each\n",
    "        of the points in x.\n",
    "        \n",
    "        Args:\n",
    "        -----\n",
    "        x: An ndarray representing the points to evaluate phi at.\n",
    "        states: A (3N,) shape ndarray encoding (location, mRNA, phi) for each polymerase.\n",
    "            This vector is assumed to be sorted in increasing order of location.\n",
    "        \n",
    "        Returns:\n",
    "        --------\n",
    "        An ndarray matching the shape of x, where twist has been evaluated.\n",
    "        \"\"\"\n",
    "        augmented_loc = np.concatenate((np.array([self.left_bc_position]),\n",
    "                                        states[::3],\n",
    "                                        np.array([self.right_bc_position])))\n",
    "        augmented_phi = np.concatenate((np.array([self.left_bc_val]),\n",
    "                                        states[2::3],\n",
    "                                        np.array([self.right_bc_val])))\n",
    "        # Replace these with free BCs if needed\n",
    "        if self.left_bc_free:\n",
    "            # Step w0 away from BC, to avoid divide by zero errors.\n",
    "            augmented_loc[0] = augmented_loc[1] - 1\n",
    "            augmented_phi[0] = augmented_phi[1]\n",
    "        if self.right_bc_free:\n",
    "            augmented_loc[-1] = augmented_loc[-2] + 1\n",
    "            augmented_phi[-1] = augmented_phi[-2]\n",
    "        \n",
    "        return np.interp(x, augmented_loc, augmented_phi)\n",
    "    \n",
    "    def torque_response(self, sc):\n",
    "        \"\"\"\n",
    "        Given a ndarray of supercoiling values, calculates\n",
    "        the torque response on each using vectorized Numpy functions.\n",
    "        \n",
    "        The torque response is a function of the provided\n",
    "        DNA persistence lengths, the temperature and force,\n",
    "    \n",
    "        Args:\n",
    "        -----\n",
    "        sc: A ndarray containing supercoiling density values.\n",
    "        \n",
    "        Returns:\n",
    "        --------\n",
    "        The calculated torque, using a two-phase constant-force\n",
    "        torque equation.\n",
    "        \"\"\"\n",
    "        abs_sc = np.abs(sc)\n",
    "        result = self.tau_0 * np.sign(sc)\n",
    "        result[abs_sc < self.sc_s] = self.tau_s * sc[abs_sc < self.sc_s]\n",
    "        result[abs_sc >= self.sc_p] = self.tau_p * sc[abs_sc >= self.sc_p]\n",
    "        return result\n",
    "    \n",
    "    def polymerase_velocity(self, sc_behind, sc_ahead):\n",
    "        \"\"\"\n",
    "        Given supercoiling densities behind and ahead of\n",
    "        each polymerase, returns the velocity of each\n",
    "        polymerase.\n",
    "        \n",
    "        Args:\n",
    "        -----\n",
    "        sc_behind: An ndarray containing the supercoiling\n",
    "            density behind each polymerase.\n",
    "        sc_ahead: An ndarray containing the supercoiling\n",
    "            density ahead of each polymerase.\n",
    "        \n",
    "        Returns:\n",
    "        --------\n",
    "        The velocity of each polymerase (in nm/s).\n",
    "        \"\"\"\n",
    "        # Restrict exponential argument to 10, to ensure we don't overflow\n",
    "        return self.v_0 / (\n",
    "            (1 + np.exp(np.minimum(10.0,\n",
    "                           (np.abs(self.torque_response(sc_behind)) - self.t_c) / self.stall_torque_width))) *\n",
    "            (1 + np.exp(np.minimum(10.0,\n",
    "                           (np.abs(self.torque_response(sc_ahead)) - self.t_c) / self.stall_torque_width))))\n",
    "    \n",
    "    def system_derivatives(self, states, polymerase_directions):\n",
    "        \"\"\"\n",
    "        Given the state of the system, in terms of a location, nascant mRNA length,\n",
    "        and local excess DNA rotation, computes the derivatives of all of these states.\n",
    "    \n",
    "        Args:\n",
    "        -----\n",
    "        states: A (3N,) shape ndarray encoding (location, mRNA, phi) for each polymerase.\n",
    "            This vector is assumed to be sorted in increasing order of location\n",
    "        polymerase_directions: A (N,) shape ndarray encoding +-1, depending on the\n",
    "            direction that the polymerase is moving.\n",
    "\n",
    "        Returns:\n",
    "        --------\n",
    "        A 3N-length vector encoding the time derivatives of each of the states.\n",
    "        \"\"\"\n",
    "        augmented_loc = np.concatenate((np.array([self.left_bc_position]),\n",
    "                                        states[::3],\n",
    "                                        np.array([self.right_bc_position])))\n",
    "        augmented_phi = np.concatenate((np.array([self.left_bc_val]),\n",
    "                                        states[2::3],\n",
    "                                        np.array([self.right_bc_val])))\n",
    "        # Replace these with free BCs if needed\n",
    "        if self.left_bc_free:\n",
    "            # Step w0 away from BC, to avoid divide by zero errors.\n",
    "            augmented_loc[0] = augmented_loc[1] - 1\n",
    "            augmented_phi[0] = augmented_phi[1]\n",
    "        if self.right_bc_free:\n",
    "            augmented_loc[-1] = augmented_loc[-2] + 1\n",
    "            augmented_phi[-1] = augmented_phi[-2]\n",
    "        # Compute supercoiling in each N + 1 region\n",
    "        supercoiling = np.diff(augmented_phi) / (np.diff(augmented_loc) * -self.w0)\n",
    "        rnac_torque = self.torque_response(supercoiling[1:]) - self.torque_response(supercoiling[:-1])\n",
    "        velocities = polymerase_directions * self.polymerase_velocity(supercoiling[1:], supercoiling[:-1])\n",
    "        # Calculate inter-polymerase distances (augmented with large distances)\n",
    "        inter_distances = np.concatenate((np.array([np.inf]),\n",
    "                                          np.diff(augmented_loc[1:-1]),\n",
    "                                          np.array([np.inf])))\n",
    "        # Halt if the polymerase is within the 2 RNAC radii in the direction it is going\n",
    "        velocities[(polymerase_directions > 0) & (inter_distances[1:] < 2 * self.rnac_r)] = 0\n",
    "        velocities[(polymerase_directions < 0) & (inter_distances[:-1] < 2 * self.rnac_r)] = 0\n",
    "        \n",
    "        drag = self.eta * (states[1::3] ** self.alpha)\n",
    "        angular_changes =  drag * velocities * self.w0 / (self.chi + drag) - rnac_torque / (self.chi + drag)\n",
    "\n",
    "        derivatives = np.zeros(states.shape)\n",
    "        derivatives[::3] = velocities\n",
    "        derivatives[1::3] = polymerase_directions * velocities # Ensure that transcript size is strictly positive\n",
    "        derivatives[2::3] = angular_changes\n",
    "        return derivatives\n",
    "\n",
    "def bind_supercoiling_model(params, bcs):\n",
    "    \"\"\"\n",
    "    Given the system parameters and boundary conditions,\n",
    "    returns a PolymeraseSupercoilingModel representing the system.\n",
    "    \n",
    "    Args:\n",
    "    -----\n",
    "    params: A dictionary containing the following keys:\n",
    "        mRNA_drag:            pN nm^(alpha / 1)\n",
    "        mRNA_exponent:        the value of alpha\n",
    "        DNA_twist_mobility:   s pN nm\n",
    "        RNAP_radius:           nm\n",
    "        RNAP_velocity:        nm / s\n",
    "        RNAP_torque_cutoff:   pN nm\n",
    "        RNAP_stall_torque_width: pN\n",
    "        DNA_force:            pN\n",
    "        DNA_bend_plength:     pN\n",
    "        DNA_twist_plength:    pN\n",
    "        DNA_plectonome_twist_plength: pN\n",
    "        temperature:          K\n",
    "    bcs: A 2-tuple encoding (left_bc, right_bc). A BC is either\n",
    "        the string \"free\", or a tuple of (location, value).\n",
    "    \n",
    "    Returns:\n",
    "    --------\n",
    "    A SupercoilingPolymeraseModel with all relevant values set.\n",
    "    \"\"\"\n",
    "    # Calculate BCs\n",
    "    left_bc_free = bcs[0] == 'free'\n",
    "    right_bc_free = bcs[1] == 'free'\n",
    "    left_bc = (0, 0)\n",
    "    right_bc = (0, 0)\n",
    "    if not left_bc_free:\n",
    "        left_bc = bcs[0]\n",
    "    if not right_bc_free:\n",
    "        right_bc = bcs[1]\n",
    "    \n",
    "    return SupercoilingPolymeraseModel(\n",
    "            params['DNA_bend_plength'], params['DNA_twist_plength'],\n",
    "            params['DNA_plectonome_twist_plength'], params['temperature'],\n",
    "            params['DNA_force'], params['RNAP_velocity'],\n",
    "            params['RNAP_torque_cutoff'], params['RNAP_stall_torque_width'],\n",
    "            params['RNAP_radius'],\n",
    "            params['DNA_twist_mobility'], params['mRNA_drag'],\n",
    "            params['mRNA_exponent'], left_bc_free, right_bc_free,\n",
    "            left_bc[0], right_bc[0], left_bc[1], right_bc[1])\n",
    "\n",
    "@njit\n",
    "def sample_discrete(probs):\n",
    "    \"\"\"Randomly sample an index with probability given by probs.\"\"\"\n",
    "    # Generate random number\n",
    "    q = np.random.rand()\n",
    "    \n",
    "    # Find index\n",
    "    i = 0\n",
    "    p_sum = 0.0\n",
    "    while p_sum < q:\n",
    "        p_sum += probs[i]\n",
    "        i += 1\n",
    "    return i - 1\n",
    "\n",
    "def attach_on_call(func):\n",
    "    \"\"\"Decorator for adding a stochastic event.\"\"\"\n",
    "    def wrap(self):\n",
    "        self.stochastic_events.append(func(self))\n",
    "    return wrap\n",
    "\n",
    "class SupercoilingSimulation(object):\n",
    "    \"\"\"\n",
    "    A base class that encodes a Gillepse-algorithm method\n",
    "    for stochastically simulating transcript initiation, alongside\n",
    "    a supercoiling model.\n",
    "    \n",
    "    This code tracks the location of polymerases as they move over\n",
    "    the genome. Events such as polymerase addition and removal are\n",
    "    natively tracked using built-in functions. Other genomic activity\n",
    "    such as topoisomerase activity can be extended with this model.\n",
    "    \n",
    "    A decorator @attach_on_call is provided that allows arbitrary hooking\n",
    "    into the event system. Events can be predicated on an ODE\n",
    "    integrated state or as a Gillepsie-style stochastic event.\n",
    "    \"\"\"\n",
    "    \n",
    "    def __init__(self, params, bcs, genes):\n",
    "        \"\"\"\n",
    "        Takes the problem formulation and initalizes the\n",
    "        simulation internals.\n",
    "        \n",
    "        Additionally initalizes an stochastic_events list, that contains\n",
    "        events in the form of tuples\n",
    "        (base_rate, state_multiplier_func,state_mutate_function)\n",
    "        \n",
    "        Args:\n",
    "        -----\n",
    "        params: A dictionary containing all relevant physical paramteres. See\n",
    "            the docstring of `bind_supercoiling_model` for details.\n",
    "            The extra parameters expected by this method are:\n",
    "                base_promoter_initiation_rate: Rate that promoters add polymerases (1/sec)\n",
    "        bcs: A 2-tuple encoding (left_bc, right_bc). A BC is either\n",
    "            the string \"free\", or a tuple of (location, value).\n",
    "        genes: An optional list of 2-tuples encoding (gene_start, gene_end, promoter_strength).\n",
    "            The gene direction is inferred from these values.\n",
    "        \"\"\"\n",
    "        self.model = bind_supercoiling_model(params, bcs)\n",
    "        self.genes = [(g[0], g[1]) for g in genes]\n",
    "        self.base_promoter_rate = params['base_promoter_initiation_rate']\n",
    "        self.topo_rate = params['topo_rate']\n",
    "        self.mrna_deg_rate = params['mRNA_deg_rate']\n",
    "        \n",
    "        self.stochastic_events = []\n",
    "        \n",
    "        for g in genes:\n",
    "            self.add_stochastic_gene_init_event(*g)\n",
    "    \n",
    "    @attach_on_call\n",
    "    def enable_topo_relaxation(self):\n",
    "        \"\"\"\n",
    "        Returns the stochastic event information\n",
    "        required to relax DNA.\n",
    "        \n",
    "        Returns:\n",
    "        --------\n",
    "        A tuple (base_rate, state_dependent_rate, mutate_func).\n",
    "        Here, the base rate is set to the relaxation rate constant\n",
    "        and the state_dependent_rate is always set to 1. Mutate_func is\n",
    "        the key function that removes supercoiling from a gene pair.\n",
    "        \"\"\"\n",
    "        state_dependent_func = lambda x: 1.0\n",
    "        def remove_supercoiling(self, t, state):\n",
    "            if len(state) == 0:\n",
    "                return\n",
    "            # Sort genes by physical location\n",
    "            sorted_genes = sorted(self.genes, key=lambda x: min(x))\n",
    "            # We now rewrite all polymerases that are sitting on the range\n",
    "            # (idx).start to (idx + 1).end\n",
    "            # We do this by removing all relevant polymerases from state, then\n",
    "            # using the model.compute_twist function to recalculate, at each\n",
    "            # of the polymerases to rewrite\n",
    "            if len(sorted_genes) == 1:\n",
    "                lower_bound = -np.inf\n",
    "                upper_bound = np.inf\n",
    "            else:\n",
    "                relax_after_idx = np.random.choice(range(len(sorted_genes) - 1))\n",
    "                lower_bound = min(sorted_genes[relax_after_idx])\n",
    "                upper_bound = max(sorted_genes[relax_after_idx + 1])\n",
    "            polymerase_rewrite_idx = np.where(\n",
    "                (state[::3] > lower_bound) & (state[::3] < upper_bound))\n",
    "            polymerase_positions = state[::3][polymerase_rewrite_idx]\n",
    "            if len(polymerase_positions) == 0:\n",
    "                # No polymerases in this range: just return\n",
    "                return\n",
    "            # Exploit polymerases always being sorted\n",
    "            new_state = np.delete(state, range(np.min(polymerase_rewrite_idx) * 3,\n",
    "                                               (np.max(polymerase_rewrite_idx) + 1) * 3))\n",
    "            self.simstate[2::3][polymerase_rewrite_idx] = \\\n",
    "                self.model.evaluate_twist(polymerase_positions, new_state)\n",
    "            \n",
    "            if 'topo' not in self.event_times:\n",
    "                self.event_times['topo'] = []\n",
    "            self.event_times['topo'].append(t)\n",
    "        \n",
    "        return (self.topo_rate, state_dependent_func, remove_supercoiling)\n",
    "    \n",
    "    def add_polymerase(self, location, direction, termination_func, event_functions=[]):\n",
    "        \"\"\"\n",
    "        Adds a polymerase to the current state. A termination function is passed\n",
    "        which specifies when the polymerase should be removed from the simulation.\n",
    "        \n",
    "        Additional event functions can be passed which encode when extra ODE events\n",
    "        should be triggered, in addition to a function\n",
    "        specifying what happens when that event occurs. This can be used to simulate\n",
    "        things such as transcript cleavage, stalling, etc.\n",
    "        \n",
    "        Args:\n",
    "        -----\n",
    "        location: A float encoding the start location\n",
    "        direction: An integer, either +1 or -1, to indicate polymerase motion direction\n",
    "        termination_func: A function that takes two arguments, returns 0 when termination should occur\n",
    "            X: the 3N number of states encoding (position, transcript_length, excess_twist)\n",
    "            i: The index of the polymerase you belong to\n",
    "        event_functions: A list of tuples (event_func, mutate_func)\n",
    "            event_func: A function with the same signature as above.\n",
    "            mutate_func: A function that takes the time, state, and \n",
    "               polymerase index and mutates them somehow, returning nothing.\n",
    "        \"\"\"\n",
    "        new_twist = self.model.evaluate_twist(location, self.simstate)\n",
    "        # Find the correct insertion index\n",
    "        possible_inserts = np.append(self.simstate[::3], np.inf)\n",
    "        insert_idx = np.where(possible_inserts > location)[0][0]\n",
    "        # Insert new polymerase state into the correct location\n",
    "        self.simstate = np.insert(self.simstate, insert_idx * 3, [location, 0, new_twist])\n",
    "        self.polymerase_directions = np.insert(self.polymerase_directions, insert_idx, direction)\n",
    "        \n",
    "        # Add ODE functions\n",
    "        def terminate_transcription(self, t, state, i):\n",
    "            self.simstate = np.delete(self.simstate, np.s_[(i * 3):((i + 1) * 3)])\n",
    "            self.polymerase_directions = np.delete(self.polymerase_directions, i)\n",
    "            del self.simstate_ode_funcs[i]\n",
    "            # Add event time\n",
    "            if 'genes' not in self.event_times:\n",
    "                self.event_times['genes'] = {}\n",
    "            if location not in self.event_times['genes']:\n",
    "                self.event_times['genes'][location] = []\n",
    "            self.event_times['genes'][location].append(t)\n",
    "                \n",
    "        self.simstate_ode_funcs.insert(insert_idx,\n",
    "                                       [(termination_func, terminate_transcription)] + event_functions)\n",
    "\n",
    "    def add_stochastic_gene_init_event(self, gene_start, gene_end, strength, state_func=lambda x: 1.0):\n",
    "        \"\"\"\n",
    "        Initalizes a simple gene which has a state-independent initiation\n",
    "        rate, and whose spawned polymerases debind exactly at the gene end.\n",
    "        \n",
    "        Polymerases are not allowed to add if there is a polymerase already\n",
    "        occupying the starting site.\n",
    "        \n",
    "        Args:\n",
    "        -----\n",
    "        gene_start: The location of the gene starting location\n",
    "        gene_end: The location of the gene ending location\n",
    "        strength: The state-independent multiplier on the base polymerase addition rate\n",
    "        state_func: An optional function that takes the state of the system (3N vector)\n",
    "            and returns an additional state-dependent multiplier\n",
    "        \n",
    "        Side effects:\n",
    "        -------------\n",
    "        Adds this gene to the stochastic_events struct.\n",
    "        \"\"\"\n",
    "        direction = np.sign(gene_end - gene_start)\n",
    "        def mutate_func(self, t, state):\n",
    "            # Our mutate function simply calls add_polymerase, with a lambda that\n",
    "            # triggers an event when the polymerase reaches the gene end\n",
    "            self.add_polymerase(gene_start, direction, lambda x, i: x[3*i] - gene_end)\n",
    "        # Define a function that stops polymerase addition if there is already a polymerase\n",
    "        # occupying the site\n",
    "        reject_occupied_site = lambda x: np.min(\n",
    "            np.abs(np.concatenate((np.array([np.inf]), x[::3])) - gene_start)\n",
    "        ) > self.model.rnac_r\n",
    "        final_state_func = lambda x: reject_occupied_site(x) * state_func(x)\n",
    "        self.stochastic_events.append((self.base_promoter_rate * strength,\n",
    "                                       final_state_func,\n",
    "                                       mutate_func))\n",
    "        \n",
    "    def simulate(self, tspan):\n",
    "        \"\"\"\n",
    "        Given a timespan, integrates using a Gillepse-inspired ODE\n",
    "        solution.\n",
    "        \n",
    "        Args:\n",
    "        -----\n",
    "        tspan: A tuple containing (t_start, t_end) [s]\n",
    "        \n",
    "        Returns:\n",
    "        --------\n",
    "        A list of tuples, containing (times, states) for each integration interval.\n",
    "        \"\"\"\n",
    "        # Reset sim state\n",
    "        self.simstate = np.zeros(0)\n",
    "        self.simstate_ode_funcs = []\n",
    "        self.polymerase_directions = np.zeros(0)\n",
    "        self.event_times = {}\n",
    "        result = []\n",
    "        \n",
    "        last_t = tspan[0]\n",
    "        \n",
    "        while last_t < tspan[1]:\n",
    "            # Make a Gillepse draw to find the next event to occur\n",
    "            gillepse_mean_time = 1 / np.sum([x[0] for x in self.stochastic_events])\n",
    "            next_attempt_time = last_t + np.random.exponential(gillepse_mean_time)\n",
    "            \n",
    "            while last_t < next_attempt_time:\n",
    "                if len(self.simstate) > 0:\n",
    "                    # Do an ODE simulation\n",
    "                    stop_events = []\n",
    "                    mutate_funcs = []\n",
    "                    for polymerase_idx, events in enumerate(self.simstate_ode_funcs):\n",
    "                        # Weird lambda currying. MWE that explains the problem:\n",
    "                        # [(lambda curried_i:(lambda x: x * curried_i))(i) for i in range(10)]\n",
    "                        # vs \n",
    "                        # [lambda x: x * i for i in range(10)]\n",
    "                        # https://stackoverflow.com/a/34021333\n",
    "                        stop_events += [(lambda curried:\n",
    "                                        lambda t, x: event[0](x,curried))(\n",
    "                            polymerase_idx) for event in events]\n",
    "                        mutate_funcs += [(lambda curried:\n",
    "                                        lambda s, t, x: event[1](s, t, x,curried))(\n",
    "                            polymerase_idx) for event in events]\n",
    "                    for i in range(len(stop_events)):\n",
    "                        stop_events[i].terminal = True\n",
    "                    ode_result = scipy.integrate.solve_ivp(\n",
    "                        lambda t,y:self.model.system_derivatives(y,\n",
    "                                                self.polymerase_directions),\n",
    "                        (last_t, next_attempt_time),\n",
    "                        self.simstate,\n",
    "                        events=stop_events,\n",
    "                        method='RK45')\n",
    "                    \n",
    "                    if ode_result.status == -1:\n",
    "                        print(ode_result.message)\n",
    "                        result.append((ode_result.y, ode_result.y))\n",
    "                        return result\n",
    "                    result.append((ode_result.t, ode_result.y))\n",
    "                    last_t = ode_result.t[-1]\n",
    "                    self.simstate = ode_result.y[:,-1]\n",
    "\n",
    "                    if ode_result.status == 1:\n",
    "                        # A polymerase hit a stop event site. See which one it is:\n",
    "                        event_idx = np.where(np.array(\n",
    "                            [len(x) for x in ode_result.t_events]) > 0)[0][0]\n",
    "                        # Mutate our state\n",
    "                        mutate_funcs[event_idx](self, last_t, self.simstate)\n",
    "                else:\n",
    "                    # Just advance the time directly; no polymerases right now\n",
    "                    last_t = next_attempt_time\n",
    "\n",
    "            # Perform a Gillepsie draw\n",
    "            event_probs = gillepse_mean_time * np.array([x[0] for x in self.stochastic_events])\n",
    "            \n",
    "            #self.init_mean_time = 1 / np.sum(self.promoter_strength * self.base_promoter_rate)\n",
    "            #self.init_probability = self.promoter_strength * self.base_promoter_rate * self.init_mean_time\n",
    "            next_stochastic_idx = sample_discrete(event_probs)\n",
    "            # Apply the state-specific state value to see if we perform this event\n",
    "            if np.random.random() < self.stochastic_events[next_stochastic_idx][1](self.simstate):\n",
    "                # Perform the stochastic state mutation\n",
    "                self.stochastic_events[next_stochastic_idx][2](self, last_t, self.simstate)\n",
    "        return (result, self.event_times)\n",
    "\n",
    "    def postprocess_run(self, sim_run, domain_points=1000, domain_endpoints=None):\n",
    "        \"\"\"\n",
    "        Given the raw run data, calculates parameters\n",
    "        of interest. Namely, this is the supercoiling density,\n",
    "        the number of mRNAs over time (calculated using the degradation rate)\n",
    "        \n",
    "        Args:\n",
    "        -----\n",
    "        sim_run: A tuple of the form (raw, event_times):\n",
    "            raw: A list of tuples, of the form (time_ndarray, 3N-state_ndarray)\n",
    "            event_times: A dictionary containing details on when events occured\n",
    "        domain_ponts: The number of points used to discritize space\n",
    "        domain_endpoints: (optional) A tuple containing (domain_start, domain_end)\n",
    "            to explicitly discritize over.\n",
    "        \n",
    "        Returns:\n",
    "        --------\n",
    "        A dictionary containing the following key-value pairs:\n",
    "            raw: The input sim_run\n",
    "            time: A single ndarray containing all timepoints\n",
    "            x_domain: A vector of the locations that were sampled\n",
    "            excess_twist: A sample of the excess twist, phi, across the domain for each timepoint\n",
    "            sc_density: The supercoiling density across the domain for each timepoint\n",
    "            gene_expression: The expression of each gene over time (calculated with a secondary\n",
    "                Gillepsie run)\n",
    "        \"\"\"\n",
    "        raw, events = sim_run\n",
    "        result = {'raw': raw}\n",
    "        \n",
    "        result['time'] = np.concatenate(list(x[0] for x in raw))\n",
    "        min_x = min([np.min(x[1][::3]) for x in raw])\n",
    "        max_x = max([np.max(x[1][::3]) for x in raw])\n",
    "        if domain_endpoints is None:\n",
    "            domain_endpoints = (min_x, max_x)\n",
    "        result['x_domain'] = np.linspace(*domain_endpoints, domain_points)\n",
    "        result['excess_twist'] = np.zeros((len(result['time']), domain_points))\n",
    "        compute_idx = 0\n",
    "        for run in raw:\n",
    "            for i in range(len(run[0])):\n",
    "                result['excess_twist'][compute_idx, :] = self.model.evaluate_twist(\n",
    "                    result['x_domain'], run[1][:,i])\n",
    "                compute_idx += 1\n",
    "        dx = (max_x - min_x) / domain_points\n",
    "        result['sc_density'] = np.diff(result['excess_twist']) / (dx * -self.model.w0)\n",
    "        \n",
    "        # Now compute mRNA at each timepoint\n",
    "        # The propensity of mRNA degrading is mrna_deg_rate * num_of_mrna\n",
    "        gene_expression = np.zeros((len(result['time']),len(events['genes'])))\n",
    "        gene_idx = 0\n",
    "        for gene_start, creation_events in events['genes'].items():\n",
    "            time_accum = np.array([creation_events[0]])\n",
    "            mRNA_accum = np.array([1])\n",
    "            creation_idx = 1\n",
    "            while time_accum[-1] < result['time'][-1]:\n",
    "                # Draw the time to the next degradation\n",
    "                if mRNA_accum[-1] > 0:\n",
    "                    next_deg = time_accum[-1] + \\\n",
    "                        np.random.exponential(1.0 / (mRNA_accum[-1] * self.mrna_deg_rate))\n",
    "                else:\n",
    "                    next_deg = result['time'][-1]\n",
    "                # Check if a creation event occurs\n",
    "                if creation_idx < len(creation_events):\n",
    "                    if next_deg < creation_events[creation_idx]:\n",
    "                        # Degradation occured!\n",
    "                        time_accum = np.append(time_accum, np.array([next_deg]))\n",
    "                        mRNA_accum = np.append(mRNA_accum, np.array([mRNA_accum[-1] - 1]))\n",
    "                        continue\n",
    "                    # Otherwise, a creation event occurs\n",
    "                    time_accum = np.append(time_accum, np.array([creation_events[creation_idx]]))\n",
    "                    mRNA_accum = np.append(mRNA_accum, np.array(mRNA_accum[-1] + 1))\n",
    "                    creation_idx += 1\n",
    "                else:\n",
    "                    time_accum = np.append(time_accum, np.array([next_deg]))\n",
    "                    mRNA_accum = np.append(mRNA_accum, np.array([mRNA_accum[-1] - 1]))\n",
    "\n",
    "            gene_expression[:,gene_idx] = np.interp(result['time'], time_accum, mRNA_accum)\n",
    "            gene_idx += 1\n",
    "        result['gene_expression'] = gene_expression\n",
    "        return result\n",
    "\n",
    "def expression_single_run(params, bcs, genes, gene_names, sim_time, i=0):\n",
    "    sim = SupercoilingSimulation(params, bcs, genes)\n",
    "    sim.enable_topo_relaxation()\n",
    "    result = sim.postprocess_run(sim.simulate((sim_time[0], sim_time[1])))\n",
    "    times = np.linspace(*sim_time)\n",
    "    id_val = np.ones(times.shape, dtype=int) * i\n",
    "    df_result = pd.DataFrame(data={'id': id_val, 'time': times})\n",
    "    for i, name in enumerate(gene_names):\n",
    "        df_result[name] = np.interp(times, result['time'], result['gene_expression'][:,i])\n",
    "    return df_result\n",
    "    \n",
    "def bulk_simulation(params, bcs, genes, gene_names, sim_time, n_runs):\n",
    "    \"\"\"\n",
    "    Given the relevant parameters to a supercoiling simulation,\n",
    "    runs several rounds of simulations in order to reach aggregate averages.\n",
    "    \n",
    "    Args:\n",
    "    -----\n",
    "    params, bcs, genes: Parameters required by the SupercoilingSimulation constructor\n",
    "    gene_names: Names attached to each gene, used when generating the aggregate dataframe.\n",
    "    sim_time: A tuple containing (start_time, end_time, n_points) over which the simulation should be run.\n",
    "    n_runs: The number of runs to aggregate\n",
    "    \n",
    "    Returns:\n",
    "    --------\n",
    "    A single dataframe containing the columns:\n",
    "        run_id: Integer containing which run it came from\n",
    "        time: The time value of the datapoint\n",
    "        gene_names: A series of columns with name = each of the names in gene_names\n",
    "    \"\"\"\n",
    "    with multiprocessing.Pool() as p:\n",
    "        runs = [p.apply_async(expression_single_run, args=(params, bcs, genes, gene_names, sim_time, i))\n",
    "                for i in range(n_runs)]\n",
    "        p.close()\n",
    "        p.join()\n",
    "        return pd.concat([r.get() for r in runs])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "result = bulk_simulation(params, ('free', (10000, 0)), [(1000, 700, 1), (1700, 1300, 1)],\n",
    "                ['upstream', 'downstream'], (0, 10000, 1000), 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "combo_plots = hv.Scatter(result, 'time', 'upstream') * hv.Scatter(result, 'time', 'downstream')\n",
    "datashade(combo_plots)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sim = SupercoilingSimulation(params, ('free', (10000, 0)), [(1000, 700, 1),  (1700, 1300, 1)])\n",
    "sim.enable_topo_relaxation()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "postprocess['gene_expression'][:,1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.random.seed(1)\n",
    "example_run = sim.simulate((0,10000))\n",
    "for segment in example_run[0]:\n",
    "    plt.plot(segment[0], segment[1][::3,:].T, 'c')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "postprocess = sim.postprocess_run(example_run)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(postprocess['time'].shape)\n",
    "print(postprocess['x_domain'].shape)\n",
    "print(postprocess['excess_twist'].shape)\n",
    "print(postprocess['sc_density'].shape)\n",
    "postprocess['gene_expression'].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.plot(postprocess['time'], postprocess['gene_expression'])\n",
    "for topo_event in example_run[1]['topo']:\n",
    "    plt.axvline(topo_event)\n",
    "plt.show()\n",
    "plt.imshow(postprocess['excess_twist'].T)\n",
    "plt.colorbar()\n",
    "plt.show()\n",
    "plt.imshow(postprocess['sc_density'].T)\n",
    "plt.colorbar()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.plot(np.linspace(0,.05,100),sim.model.polymerase_velocity(np.zeros((100,)), np.linspace(0,.05,100)))\n",
    "plt.show()\n",
    "plt.plot(np.linspace(0,20,100), 1 / (1 + np.exp((np.linspace(0,20,100) - 12)/.1)))\n",
    "plt.show()\n",
    "plt.plot(np.linspace(0,.1,100), sim.model.torque_response(np.linspace(0,.1,100)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
