{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib\n",
    "import matplotlib.cm as cm\n",
    "import matplotlib.gridspec as gridspec\n",
    "import h5py\n",
    "import pandas as pd\n",
    "from pandas.api.types import CategoricalDtype\n",
    "import numpy as np\n",
    "import svgutils.transform as sg\n",
    "import scipy.signal\n",
    "import scipy.stats\n",
    "from typing import Union\n",
    "\n",
    "from pathlib import Path\n",
    "import re\n",
    "\n",
    "import itertools\n",
    "\n",
    "import rushd as rd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Helper functions, color palettes, etc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_external_svgs(final_filename, base_svg_filename, subfigures_dict, external_svg_mapping):\n",
    "    # Inspired by https://jetholt.com/micro/programmatically-merging-svg-files/\n",
    "    # Units from https://www.w3.org/TR/css-values/#absolute-lengths\n",
    "    def unit_to_pixels(unit_str):\n",
    "        value = float(re.search(r'[0-9\\.]+', unit_str).group())\n",
    "        unit_mapping = {\n",
    "            'cm': 96.0 / 2.54,\n",
    "            'mm': 96.0 / 2.54 / 10.0,\n",
    "            'Q':  96.0 / 2.54 / 40.0,\n",
    "            'in': 96.0,\n",
    "            'pc': 96.0 / 6.0,\n",
    "            'pt': 96.0 / 72.0,\n",
    "            'px': 1.0\n",
    "        }\n",
    "        for unit, conversion in unit_mapping.items():\n",
    "            if unit_str.endswith(unit):\n",
    "                return value * conversion\n",
    "        # Default to pixels\n",
    "        return value\n",
    "    def pixels_to_base_units(svg, xy):\n",
    "        # Converts (xy) units to whatever base units.\n",
    "        raw_value = float(re.search(r'[0-9\\.]+', svg.get_size()[0]).group())\n",
    "        pixel_value = unit_to_pixels(svg.get_size()[0])\n",
    "\n",
    "        return (xy[0] * raw_value / pixel_value, xy[1] * raw_value / pixel_value)\n",
    "\n",
    "    def svg_size(svg):\n",
    "        size_tup = svg.get_size()\n",
    "        return (unit_to_pixels(size_tup[0]), unit_to_pixels(size_tup[1]))\n",
    "    base_svg = sg.fromfile(base_svg_filename)\n",
    "    # Get the size of the base SVG\n",
    "    base_size = svg_size(base_svg)\n",
    "    for subfig, mapping in external_svg_mapping:\n",
    "        external_svg = sg.fromfile(mapping['filename'])\n",
    "        external_size = svg_size(external_svg)\n",
    "        \n",
    "        if 'extents' in mapping:\n",
    "            extents = mapping['extents']\n",
    "        else:\n",
    "            extents = (0, 0, 1, 1) # left, bottom, width, height\n",
    "\n",
    "        # Calculate subfigure extents in pixel-space\n",
    "        subfig_bbox = subfigures_dict[subfig].bbox.get_points()\n",
    "        fig_bbox = subfigures_dict[subfig].figure.bbox.get_points()\n",
    "\n",
    "        fig_space_extents = (\n",
    "            subfig_bbox[0,0] + extents[0] * (subfig_bbox[1,0] - subfig_bbox[0,0]), # left\n",
    "            subfig_bbox[0,1] + extents[1] * (subfig_bbox[1,1] - subfig_bbox[0,1]), # bottom\n",
    "            (subfig_bbox[1,0] - subfig_bbox[0,0]) * (extents[2]), # width\n",
    "            (subfig_bbox[1,1] - subfig_bbox[0,1]) * (extents[3]) # height\n",
    "        )\n",
    "        pixel_extents = (\n",
    "            (fig_space_extents[0] - fig_bbox[0,0]) / (fig_bbox[1,0] - fig_bbox[0,0]) * base_size[0],\n",
    "            (fig_space_extents[1] - fig_bbox[0,1]) / (fig_bbox[1,1]  - fig_bbox[0,1]) * base_size[1],\n",
    "            (fig_space_extents[2]) / (fig_bbox[1,0] - fig_bbox[0,0]) * base_size[0],\n",
    "            (fig_space_extents[3]) / (fig_bbox[1,1] - fig_bbox[0,1]) * base_size[1]\n",
    "        )\n",
    "        scale_factor = min(pixel_extents[2] / external_size[0], pixel_extents[3] / external_size[1])\n",
    "        print(f'Base pixel size:{base_size}\\nsubfig {subfig}:\\n\\tsize:{external_size}\\n\\tfig_space_extents:{fig_space_extents}\\n\\tpixel_extents:{pixel_extents}\\n\\tscale factor:{scale_factor}')\n",
    "        # First, scale the  external SVG then move to the correct location.\n",
    "        external_elem = external_svg.getroot()\n",
    "        # Figure out the base units, and convert to that.\n",
    "        offset = pixels_to_base_units(base_svg, (pixel_extents[0], base_size[1] - pixel_extents[1] - pixel_extents[3]))\n",
    "        external_elem.moveto(*offset, scale_factor)\n",
    "        base_svg.append(external_elem)\n",
    "    base_svg.save(final_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def draw_arrows(subfig, ax, x1, x2, y1, y2, dy, dir1, dir2, c1, c2):\n",
    "    subfig.add_artist(matplotlib.lines.Line2D([x1, x1], [y1+0.02, y2],\n",
    "        linewidth=2.5, color=c1, solid_capstyle='projecting', transform=ax.transAxes))\n",
    "    subfig.add_artist(matplotlib.lines.Line2D([x2, x2], [y1+0.02, y2],\n",
    "        linewidth=2.5, color=c2, solid_capstyle='projecting', transform=ax.transAxes))\n",
    "    subfig.add_artist(matplotlib.patches.FancyArrow(x1, y2, dy * dir1, 0,\n",
    "        width=0.025, color=c1, transform=ax.transAxes))\n",
    "    subfig.add_artist(matplotlib.patches.FancyArrow(x2, y2, dy * dir2, 0,\n",
    "        width=0.025, color=c2, transform=ax.transAxes))\n",
    "    subfig.add_artist(matplotlib.lines.Line2D([x1 - 0.1 if dir1 > 0 else x1 - 0.1 - dy, x2 + 0.1 if dir2 < 0 else x2 + 0.1 + dy], [y1, y1],\n",
    "        linewidth=2, color='#262626', transform=ax.transAxes))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_genes_realistic(gene_tuples, gene_colors, gene_dirs, length, ax):\n",
    "    xax_transform = ax.get_xaxis_transform()\n",
    "    ax.plot([0, length * 0.34], [1.1, 1.1], color=main_palette['axis_gray'], transform=xax_transform, clip_on=False)\n",
    "    for gene, gene_dir, color in zip(gene_tuples, gene_dirs, gene_colors):\n",
    "        base = gene[0] - (length * 0.34 * 0.01) if gene_dir > 0 else gene[1] + (length * 0.34 * 0.01)\n",
    "        ax.plot(gene, [1.1, 1.1], linewidth=3.5, solid_capstyle='butt', color=color, transform=xax_transform, clip_on=False)\n",
    "        ax.plot([base, base], [1.1, 1.33], linewidth=1, solid_capstyle='projecting', color=color, transform=xax_transform, clip_on=False)\n",
    "        ax.plot([base, base + gene_dir * (length * 0.34 * 0.06)], [1.33, 1.33], linewidth=1, solid_capstyle='projecting', color=color, transform=xax_transform, clip_on=False)\n",
    "        ax.add_artist(matplotlib.patches.FancyArrow(base + gene_dir * (length * 0.34 * 0.06), 1.33, gene_dir, 0,\n",
    "            width=0.1, head_width=0.1, head_length=length * 0.34 * 0.05, color=color, transform=xax_transform, clip_on=False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def transform_xcoords(transform, xcoords):\n",
    "    return transform.transform([(x,0) for x in xcoords])[:,0]\n",
    "def transform_ycoords(transform, ycoords):\n",
    "    return transform.transform([(0,y) for y in ycoords])[:,1]\n",
    "def plot_genes_unified(ax: matplotlib.axes.Axes, gene_locations, gene_colors, plot_gene_body=True, units='axes', base_line_extents=(0,1), base_offset_inches=0.1, plot_dx=None, plot_inducible=None):\n",
    "    arrow_height_inches = 0.1\n",
    "    arrow_width_inches = 0.06\n",
    "    arrow_genebody_offset_inches = 0.02\n",
    "    inducible_bar_width_inches = 0.131\n",
    "    inducible_bar_height_inches = 0.09\n",
    "    inducible_bar_woffset = 0.015\n",
    "    inducible_bar_hoffset = 0.02\n",
    "    arrow_head_size_inches = 0.05\n",
    "    fig = ax.get_figure()\n",
    "    figinches_transform = fig.dpi_scale_trans\n",
    "    axes_to_figinches = ax.transAxes + figinches_transform.inverted()\n",
    "    xaxis_blended_to_axes = ax.get_xaxis_transform() + ax.transAxes.inverted()\n",
    "    # Calculate useful numbers in figinches space\n",
    "    base_y = transform_ycoords(axes_to_figinches, [1.0])[0] + base_offset_inches\n",
    "    # Plot the gene line\n",
    "    ax.plot(\n",
    "        transform_xcoords(axes_to_figinches,base_line_extents),\n",
    "        [base_y, base_y],\n",
    "        color=main_palette['axis_gray'], transform=figinches_transform, linewidth=1.0, clip_on=False\n",
    "    )\n",
    "    # Plot each gene\n",
    "    for gene_idx, (gene, gene_color) in enumerate(zip(gene_locations, gene_colors)):\n",
    "        dir = 1 if gene[1] > gene[0] else -1\n",
    "        if units == 'axes':\n",
    "            gene_loc = transform_xcoords(axes_to_figinches, gene)\n",
    "        elif units == 'xaxis':\n",
    "            gene_loc = transform_xcoords(xaxis_blended_to_axes + axes_to_figinches, gene)\n",
    "        if plot_gene_body:\n",
    "            ax.plot(gene_loc, [base_y, base_y],\n",
    "                linewidth=3, solid_capstyle='butt', color=gene_color, transform=figinches_transform, clip_on=False)\n",
    "        arrow_xinches = gene_loc[0] - (dir * arrow_genebody_offset_inches)\n",
    "        # Plot gene arrow\n",
    "        ax.plot(\n",
    "            [arrow_xinches, arrow_xinches, arrow_xinches + (dir * arrow_width_inches)],\n",
    "            [base_y, base_y + arrow_height_inches, base_y + arrow_height_inches],\n",
    "            linewidth=1, solid_capstyle='projecting', color=gene_color, transform=figinches_transform, clip_on=False)\n",
    "        ax.add_artist(matplotlib.patches.FancyArrow(arrow_xinches + (dir * arrow_width_inches), base_y + arrow_height_inches,\n",
    "            dir * 0.0001, 0, width=0, head_width=arrow_head_size_inches, head_length=arrow_head_size_inches, color=gene_color, transform=figinches_transform, clip_on=False))\n",
    "        if plot_inducible == gene_idx:\n",
    "            tri_ax = ax.inset_axes([\n",
    "                arrow_xinches + (-(inducible_bar_width_inches + inducible_bar_woffset) if dir == 1 else inducible_bar_woffset),\n",
    "                base_y + inducible_bar_hoffset,\n",
    "                inducible_bar_width_inches,\n",
    "                inducible_bar_height_inches], transform=figinches_transform)\n",
    "            n_grad = 200\n",
    "            base_im = np.ones((n_grad,n_grad)) - np.tril(np.tile(\n",
    "                        np.linspace(0.6, 0.0, n_grad),\n",
    "                        (n_grad,1)))\n",
    "            final_im = base_im if dir == -1 else np.flip(base_im, axis=1)\n",
    "            tri_ax.imshow(\n",
    "                scipy.ndimage.gaussian_filter(final_im, sigma=2),\n",
    "                interpolation='none', cmap='gray', vmin=0, vmax=1, aspect='auto')\n",
    "            tri_ax.set_axis_off()\n",
    "            tri_ax.set_xlim((-40,240))\n",
    "            tri_ax.set_ylim((240, -40))\n",
    "\n",
    "    if plot_dx is not None:\n",
    "        first_loc = transform_xcoords(axes_to_figinches if units == 'axes' else xaxis_blended_to_axes + axes_to_figinches, gene_locations[plot_dx[0]])\n",
    "        sec_loc = transform_xcoords(axes_to_figinches if units == 'axes' else xaxis_blended_to_axes + axes_to_figinches, gene_locations[plot_dx[1]])\n",
    "        ax.plot([max(first_loc), min(sec_loc)], [base_y - 2.5 * arrow_genebody_offset_inches] * 2,\n",
    "            linewidth=1, solid_capstyle='butt', color='k', transform=figinches_transform, clip_on=False)\n",
    "        ax.add_artist(matplotlib.patches.FancyArrow(max(first_loc) + arrow_head_size_inches, base_y - 2.5 * arrow_genebody_offset_inches,\n",
    "            -1 * 0.0001, 0, width=0, head_width=arrow_head_size_inches, head_length=arrow_head_size_inches, overhang=0.2, color='k', transform=figinches_transform, clip_on=False))\n",
    "        ax.add_artist(matplotlib.patches.FancyArrow(min(sec_loc) - arrow_head_size_inches, base_y - 2.5 * arrow_genebody_offset_inches,\n",
    "            1 * 0.0001, 0, width=0, head_width=arrow_head_size_inches, head_length=arrow_head_size_inches, overhang=0.2, color='k', transform=figinches_transform, clip_on=False))\n",
    "        ax.text((max(first_loc) + min(sec_loc)) / 2.0, base_y - 3.7 * arrow_genebody_offset_inches, 'Δx', va='top', ha='center', fontsize=6, transform=figinches_transform, clip_on=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def summarize_escape_times(df):\n",
    "    sorted_escapes = sorted(df.escape_time)\n",
    "    counts = np.arange(1, len(sorted_escapes) + 1)\n",
    "    interp_time = np.linspace(0, 25000, 500)\n",
    "    escape_fraction =  np.interp(interp_time, sorted_escapes, counts) / len(sorted_escapes)\n",
    "    return pd.DataFrame({\n",
    "        'time': interp_time,\n",
    "        'escape_fraction': escape_fraction,\n",
    "        'stable_fraction': 1.0 - escape_fraction\n",
    "    })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_burst_dynamics(inter_rnap_times, burst_separation_time):\n",
    "    burst_mask = inter_rnap_times > burst_separation_time\n",
    "    inter_burst_times = np.concatenate(([burst_separation_time],inter_rnap_times[burst_mask]))\n",
    "    inter_burst_idx = np.concatenate(([0],np.where(burst_mask)[0], [len(inter_rnap_times)]))\n",
    "    burst_size = np.diff(inter_burst_idx)\n",
    "    raw_times = np.cumsum(inter_rnap_times)\n",
    "    burst_duration = raw_times[inter_burst_idx[1:] - 1] - raw_times[inter_burst_idx[:-1]]\n",
    "    return {\n",
    "        'inter_burst_time': inter_burst_times,\n",
    "        'burst_size': burst_size,\n",
    "        'burst_duration': burst_duration\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Edited from https://github.com/mwaskom/seaborn/blob/473ebefc797143371c2c096e7c4a70e8469bd5be/seaborn/distributions.py#L207=\n",
    "def custom_cmap_from_base(color, sat_multipliers=(0,1.5), lightness_multipliers=(0,1.5)):\n",
    "    import matplotlib.colors\n",
    "    import colorsys\n",
    "    sat_lower, sat_upper = sat_multipliers\n",
    "    l_lower, l_upper = lightness_multipliers\n",
    "    h, l, s = colorsys.rgb_to_hls(*matplotlib.colors.to_rgb(color))\n",
    "    xx = np.linspace(-1, 1, int(1.15 * 256))[:256]\n",
    "    ramp = np.zeros((256, 3))\n",
    "    ramp[:,0] = h\n",
    "    ramp[:,1] = (l_lower * l) + (l_upper * l - l_lower * l) * np.linspace(0,1,256)\n",
    "    ramp[:,2] = (sat_lower * s) + (sat_upper * s - sat_lower * s) * (1.0 + np.cos(xx)) / 2.0\n",
    "    colors = np.clip([colorsys.hls_to_rgb(*hls) for hls in ramp], 0, 1)\n",
    "    return matplotlib.colors.ListedColormap(colors[::-1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "main_palette = {\n",
    "    'tandem': '#225A9B',\n",
    "    'tandem_alt': '#19D2BF',\n",
    "    'convergent': '#FFB133',\n",
    "    'divergent': '#FE484E',\n",
    "    'gray': '#666666',\n",
    "    'reporter_only': '#666666',\n",
    "    'light_gray': '#888888',\n",
    "    'axis_gray': '#262626',\n",
    "    'black': 'k',\n",
    "    'polymerase_color': '#7FCDC4'\n",
    "}\n",
    "# OG Beach towel\n",
    "#    'tandem': '#227C9D',\n",
    "#    'tandem_alt': '#17C3B2',\n",
    "#    'convergent': '#FFB133',\n",
    "#    'divergent': '#FE6D73',\n",
    "font_sizes = {\n",
    "    'subpanel_label': 10,\n",
    "    'base_size': 8,\n",
    "    'colorbar_title': 7,\n",
    "    'subaxis_title': 7,\n",
    "    'line_annotation': 7,\n",
    "    'data_annotation': 6,\n",
    "}\n",
    "sns.set_style('ticks')\n",
    "sns.set_context('paper', font_scale=1.0, rc={'font.size': font_sizes['base_size'], 'font.family': 'sans-serif', 'font.sans-serif':['Helvetica Neue']})\n",
    "plt.rcParams.update({'axes.labelsize': font_sizes['base_size'], 'xtick.labelsize': font_sizes['base_size'], 'ytick.labelsize': font_sizes['base_size']})\n",
    "light_palette = {k:v + '28' for k, v in main_palette.items()}\n",
    "no_yellow_viridis = matplotlib.colors.ListedColormap(cm.get_cmap('viridis', 256)(np.linspace(0,0.8,256)))\n",
    "no_white_greens = matplotlib.colors.ListedColormap(cm.get_cmap('Greens', 256)(np.linspace(0.2,1.0,256)))\n",
    "k_formatter = matplotlib.ticker.FuncFormatter(lambda x, _: f'{x:.0f}' if abs(x) < 1000 else f'{x/1000:.0f}k')\n",
    "fold_formatter = matplotlib.ticker.FuncFormatter(lambda x, _: f'{x}x')\n",
    "hours_formatter = matplotlib.ticker.FuncFormatter(lambda x,_: f'{x/3600:.0f}')\n",
    "hours_1f_formatter = matplotlib.ticker.FuncFormatter(lambda x,_: f'{x/3600:.1f}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if (rd.datadir/'projects'/'tangles'/'simulations'/'fig_base_model_summary.gzip.0').exists():\n",
    "    base_model_df = pd.read_parquet(\n",
    "        rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/'fig_base_model_summary.gzip.0', 'fig_base_model'), memory_map=True\n",
    "    )\n",
    "    for i in range(1,8):\n",
    "        base_model_df = base_model_df.append(\n",
    "            pd.read_parquet(rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/f'fig_base_model_summary.gzip.{i}', 'fig_base_model'), memory_map=True),\n",
    "            ignore_index=True\n",
    "        )\n",
    "    print(f'Loaded {len(base_model_df)} total simulation runs from Parquet cached file')\n",
    "else:\n",
    "    base_model_df = []\n",
    "    with h5py.File(\n",
    "        rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/'modeling_model_bc_topo_selection_summaries.h5', 'fig_base_model_cache'), 'r') as summaries:\n",
    "        for group in summaries.keys():\n",
    "            group_df = []\n",
    "            for run in summaries[group].values():\n",
    "                if not run.attrs['comment'].decode('utf-8').startswith('fig.bm'):\n",
    "                    continue\n",
    "                group_df.append(pd.DataFrame(data={\n",
    "                    'inducible_mRNA': run['final_mRNA'][0,:],\n",
    "                    'constant_mRNA': run['final_mRNA'][1,:]}))\n",
    "                group_df[-1]['induction'] = run.attrs['gene.base_rate'][0]\n",
    "                group_df[-1]['condition'] = run.attrs['comment'].decode('utf-8')\n",
    "                group_df[-1]['is_plasmid'] = run.attrs['bcs.is_circular']\n",
    "                for term in [\n",
    "                    'rates.sc_dependent', 'coeff.sigma_squared',\n",
    "                    'coeff.mRNA_drag_exponent', 'coeff.mRNA_drag_coeff',\n",
    "                    'rnap.max_velocity', 'rnap.stall_torque', 'rnap.stall_width']:\n",
    "                    group_df[-1][term] = run.attrs[term]\n",
    "                gene_endpoints = sorted(np.concatenate((\n",
    "                    run.attrs['gene.start'],\n",
    "                    run.attrs['gene.end'])))\n",
    "                group_df[-1]['spacing'] = int((gene_endpoints[2] - gene_endpoints[1]) / 0.34)\n",
    "            base_model_df.append(pd.concat(group_df, ignore_index=True))\n",
    "            print(f'Loaded {len(base_model_df[-1])} simulations...')\n",
    "    base_model_df = pd.concat(base_model_df, ignore_index=True)\n",
    "    print(f'Loaded {len(base_model_df)} total simulation runs!')\n",
    "    base_model_df['condition'] = pd.Categorical(base_model_df['condition'])\n",
    "    for i, split in enumerate(np.array_split(base_model_df, 8)):\n",
    "        split.to_parquet(rd.outfile(rd.datadir/'projects'/'tangles'/'simulations'/f'fig_base_model_summary.gzip.{i}', 'fig_base_model_cache'), compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if (rd.datadir/'projects'/'tangles'/'simulations'/'fig_spacing_sweep.gzip').exists():\n",
    "    spacing_sweep_df = pd.read_parquet(\n",
    "        rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/'fig_spacing_sweep.gzip', 'fig_base_model')\n",
    "    )\n",
    "    print(f'Loaded {len(spacing_sweep_df)} total simulation runs from Parquet cached file')\n",
    "else:\n",
    "    spacing_sweep_df = []\n",
    "    with h5py.File(\n",
    "        rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/'modeling_spacing_sweep_summaries.h5', 'fig_base_model_cache'), 'r') as summaries:\n",
    "        for group in summaries.keys():\n",
    "            group_df = []\n",
    "            for run in summaries[group].values():\n",
    "                if not run.attrs['comment'].decode('utf-8').startswith('fig.bm'):\n",
    "                    continue\n",
    "                group_df.append(pd.DataFrame(data={\n",
    "                    'inducible_mRNA': run['final_mRNA'][0,:],\n",
    "                    'constant_mRNA': run['final_mRNA'][1,:]}))\n",
    "                group_df[-1]['induction'] = run.attrs['gene.base_rate'][0]\n",
    "                group_df[-1]['condition'] = run.attrs['comment'].decode('utf-8')\n",
    "                group_df[-1]['is_plasmid'] = run.attrs['bcs.is_circular']\n",
    "                for term in [\n",
    "                    'rates.sc_dependent', 'coeff.sigma_squared',\n",
    "                    'coeff.mRNA_drag_exponent', 'coeff.mRNA_drag_coeff',\n",
    "                    'rnap.max_velocity', 'rnap.stall_torque', 'rnap.stall_width']:\n",
    "                    group_df[-1][term] = run.attrs[term]\n",
    "                gene_endpoints = sorted(np.concatenate((\n",
    "                    run.attrs['gene.start'],\n",
    "                    run.attrs['gene.end'])))\n",
    "                group_df[-1]['spacing'] = int((gene_endpoints[2] - gene_endpoints[1]) / 0.34)\n",
    "            spacing_sweep_df.append(pd.concat(group_df, ignore_index=True))\n",
    "            print(f'Loaded {len(spacing_sweep_df[-1])} simulations...')\n",
    "    spacing_sweep_df = pd.concat(spacing_sweep_df, ignore_index=True)\n",
    "    print(f'Loaded {len(spacing_sweep_df)} total simulation runs!')\n",
    "    spacing_sweep_df['condition'] = pd.Categorical(spacing_sweep_df['condition'])\n",
    "    spacing_sweep_df.to_parquet(rd.outfile(rd.datadir/'projects'/'tangles'/'simulations'/'fig_spacing_sweep.gzip', 'fig_base_model_cache'), compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if (rd.datadir/'projects'/'tangles'/'simulations'/'fig_sc_density.gzip').exists() and (rd.datadir/'projects'/'tangles'/'simulations'/'fig_bursting.gzip').exists():\n",
    "    sc_density_df = pd.read_parquet(\n",
    "        rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/'fig_sc_density.gzip', 'fig_base_model'), memory_map=True\n",
    "    )\n",
    "    print(f'Loaded {len(sc_density_df)} total simulation runs from Parquet cached file')\n",
    "    bursting_df = pd.read_parquet(\n",
    "        rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/'fig_bursting.gzip', 'fig_bursting'), memory_map=True\n",
    "    )\n",
    "    print(f'Loaded {len(bursting_df)} total simulation runs from Parquet cached file')\n",
    "else:\n",
    "    sc_density_df = []\n",
    "    bursting_df = []\n",
    "    sc_examples = {}\n",
    "    sc_burst_condition_dtype = CategoricalDtype([\n",
    "    \"fig.bm.sc_bursts.tandem_reporter_upstream\",\n",
    "    \"fig.bm.sc_bursts.tandem_reporter_downstream\",\n",
    "    \"fig.bm.sc_bursts.convergent\",\n",
    "    \"fig.bm.sc_bursts.divergent\",\n",
    "    \"fig.bm_examples.sc_bursts.tandem_reporter_upstream\",\n",
    "    \"fig.bm_examples.sc_bursts.tandem_reporter_downstream\",\n",
    "    \"fig.bm_examples.sc_bursts.convergent\",\n",
    "    \"fig.bm_examples.sc_bursts.divergent\",\n",
    "    \"fig.toggles.sc_bursts.tandem_reporter_upstream\",\n",
    "    \"fig.toggles.sc_bursts.tandem_reporter_downstream\",\n",
    "    \"fig.toggles.sc_bursts.convergent\",\n",
    "    \"fig.toggles.sc_bursts.divergent\",\n",
    "    ])\n",
    "    with h5py.File(\n",
    "        rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/'modeling_sc_bursting_combined.h5', 'fig_base_model'), 'r') as summaries:\n",
    "        for group in summaries.keys():\n",
    "            group_df = []\n",
    "            for run in summaries[group].values():\n",
    "                condition = run.attrs['comment'].decode('utf-8')\n",
    "                bcs_length = run.attrs['bcs.length']\n",
    "                is_plasmid = run.attrs['bcs.is_circular']\n",
    "                alpha = run.attrs['coeff.sigma_squared']\n",
    "                sc_dependent = run.attrs['rates.sc_dependent']\n",
    "                induction = run.attrs['gene.base_rate'][0]\n",
    "\n",
    "                data_dict = {\n",
    "                    'condition': condition,\n",
    "                    'bcs_length': bcs_length,\n",
    "                    'is_plasmid': is_plasmid,\n",
    "                    'alpha': alpha,\n",
    "                    'sc_dependent': sc_dependent,\n",
    "                    'induction': induction\n",
    "                }\n",
    "                if condition.startswith('fig.bm_examples'):\n",
    "                    data_dict['step_time'] = run.attrs['step_time']\n",
    "                    data_dict['induction'] = run.attrs['step_induction']\n",
    "                if condition.startswith('fig.toggles'):\n",
    "                    data_dict['hill_n'] = run.attrs['hill_coeff']\n",
    "                    data_dict['k_factor']   = run.attrs['K_factor']\n",
    "                    data_dict['mRNA_deg_factor'] = run.attrs['mRNA_deg_factor']\n",
    "\n",
    "                # Save the average supercoiling density per condition\n",
    "                frozen_meta = tuple(sorted(data_dict.items()))\n",
    "                if frozen_meta not in sc_examples:\n",
    "                    sc_examples[frozen_meta] = {\n",
    "                        'sc_density': np.array(run['sc_density']),\n",
    "                        'sc_time': np.array(run['sc_times']),\n",
    "                        'count': 1\n",
    "                    }\n",
    "                else:\n",
    "                    sc_examples[frozen_meta]['sc_density'] += np.array(run['sc_density'])\n",
    "                    sc_examples[frozen_meta]['count'] += 1\n",
    "                \n",
    "                if sc_examples[frozen_meta]['count'] > 20:\n",
    "                    continue\n",
    "                \n",
    "                raw_events = pd.DataFrame({\n",
    "                    'event_type': run['event_type'],\n",
    "                    'event_time': run['event_times'],\n",
    "                    'event_gene': run['event_genes']\n",
    "                })\n",
    "                intergene_1_time = np.diff(raw_events[(raw_events.event_gene == 1) & (raw_events.event_type == 1)].event_time)\n",
    "                intergene_2_time = np.diff(raw_events[(raw_events.event_gene == 2) & (raw_events.event_type == 1)].event_time)\n",
    "                for burst_threshold in [10, 20, 30, 60]:\n",
    "                    if len(intergene_1_time) > 0:\n",
    "                        group_df.append(pd.DataFrame({\n",
    "                            **data_dict,\n",
    "                            'inter_rnap_gene': 1,\n",
    "                            'burst_threshold': burst_threshold,\n",
    "                            **calculate_burst_dynamics(intergene_1_time, burst_threshold),\n",
    "                        }))\n",
    "                        group_df[-1]['condition'] = pd.Categorical(group_df[-1]['condition'], dtype=sc_burst_condition_dtype)\n",
    "                    if len(intergene_2_time) > 0:\n",
    "                        group_df.append(pd.DataFrame({\n",
    "                            **data_dict,\n",
    "                            'inter_rnap_gene': 2,\n",
    "                            'burst_threshold': burst_threshold,\n",
    "                            **calculate_burst_dynamics(intergene_2_time, burst_threshold),\n",
    "                        }))\n",
    "                        group_df[-1]['condition'] = pd.Categorical(group_df[-1]['condition'], dtype=sc_burst_condition_dtype)\n",
    "            if len(group_df) > 0:\n",
    "                bursting_df.append(pd.concat(group_df, ignore_index=True))\n",
    "            print(f'Loaded {len(bursting_df[-1])} simulations...')\n",
    "    bursting_df = pd.concat(bursting_df, ignore_index=True)\n",
    "    print(f'Loaded {len(bursting_df)} total simulation runs!')\n",
    "    sc_density_df = pd.concat([\n",
    "        pd.DataFrame({\n",
    "            **{k:v for k,v in meta},\n",
    "            'sc_density': sc['sc_density'][time_idx,:] / sc['count'],\n",
    "            'distance': np.linspace(0, dict(meta)['bcs_length'], sc['sc_density'].shape[1]),\n",
    "            'time': time\n",
    "        })\n",
    "    for meta, sc in sc_examples.items() for time_idx, time in enumerate(sc['sc_time'])], ignore_index=True)\n",
    "    sc_density_df.condition = pd.Categorical(sc_density_df.condition, dtype=sc_burst_condition_dtype)\n",
    "    print(f'Loaded {len(sc_density_df)} total simulation runs!')\n",
    "    sc_density_df.to_parquet(rd.outfile(rd.datadir/'projects'/'tangles'/'simulations'/'fig_sc_density.gzip', 'fig_sc_bursting_cache'), compression='gzip')\n",
    "    bursting_df.to_parquet(rd.outfile(rd.datadir/'projects'/'tangles'/'simulations'/'fig_bursting.gzip', 'fig_sc_bursting_cache'), compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc_density_df['norm_induction'] = sc_density_df.induction * 120.0\n",
    "available_induction = sc_density_df[sc_density_df.condition == 'fig.bm.sc_bursts.convergent'].induction\n",
    "selected_induction = [np.min(available_induction), 1/120.0, np.max(available_induction)]\n",
    "bursting_df_selected = bursting_df[\n",
    "    ((bursting_df.induction) == selected_induction[0]) |\n",
    "    ((bursting_df.induction) == selected_induction[1]) |\n",
    "    ((bursting_df.induction) == selected_induction[2])\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with h5py.File(rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/'sigma_sweep.h5', 'fig_supp_alpha'), 'r') as sigma_h5:\n",
    "    alpha_sweep_df = pd.concat([\n",
    "        pd.DataFrame({\n",
    "            'sigma': sigma_h5['sweep']['sigma'],\n",
    "            'alpha': sigma_h5['sweep']['s2_coeff'][a_idx],\n",
    "            'energy': sigma_h5['sweep']['values'][a_idx,:]\n",
    "        }) for a_idx in range(len(sigma_h5['sweep']['s2_coeff']))\n",
    "    ])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Randomly downsample (replace later)\n",
    "#base_model_df = base_model_df.sample(frac=0.3, random_state=2022)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Model comparison and explanation figure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model_palette = {\n",
    "    'gray': main_palette['gray'],\n",
    "    'light_gray': main_palette['light_gray'],\n",
    "    'axis_gray': main_palette['axis_gray'],\n",
    "    '0th_order': main_palette['light_gray'],\n",
    "    '1st_order': main_palette['light_gray'],\n",
    "    '2nd_order': main_palette['polymerase_color'],\n",
    "    'black': main_palette['black']\n",
    "    }\n",
    "# Plot three desired energy curves\n",
    "fig_energy_curves = plt.figure(figsize=(2.5,1.5))\n",
    "fig_energy_ax = fig_energy_curves.gca()\n",
    "\n",
    "fig_energy_ax.axvline(0, color=main_palette['axis_gray'])\n",
    "\n",
    "for alpha, color in zip([0.0, 0.025], [model_palette['1st_order'], model_palette['2nd_order']]):\n",
    "    closest_alpha_df = alpha_sweep_df[\n",
    "        np.abs(alpha_sweep_df.alpha - alpha) == min(np.abs(alpha_sweep_df.alpha - alpha))\n",
    "    ].copy()\n",
    "    closest_alpha_df['energy'] += 10\n",
    "    fig_energy_ax.plot(closest_alpha_df['sigma'], closest_alpha_df['energy'], color=color)\n",
    "fig_energy_ax.set_xlim([-0.2,0.2])\n",
    "fig_energy_ax.set_ylim([-250,250])\n",
    "sns.despine(ax=fig_energy_ax, left=False, bottom=True, top=False, right=True)\n",
    "fig_energy_ax.set_xticks([-0.2,0.2])\n",
    "fig_energy_ax.set_yticks([-275,275])\n",
    "fig_energy_ax.tick_params(direction='inout', length=4, pad=1, labelsize=8)\n",
    "fig_energy_ax.yaxis.tick_left()\n",
    "fig_energy_ax.xaxis.tick_top()\n",
    "fig_energy_ax.spines['top'].set_position('zero')\n",
    "fig_energy_ax.spines['left'].set_position('zero')\n",
    "fig_energy_ax.text(0.22, 0, 'σ', fontsize=8, verticalalignment='center')\n",
    "fig_energy_ax.text(0, 320, 'E (pN nm)', fontsize=8, horizontalalignment='center')\n",
    "fig_energy_ax.text(0.025, -60, '1st order', fontsize=8, color=model_palette['1st_order'])\n",
    "fig_energy_ax.text(0.025, -120, '2nd order', fontsize=8, color=model_palette['2nd_order'])\n",
    "fig_energy_curves.savefig('test.svg')\n",
    "fig_energy_curves.savefig(rd.outfile(rd.rootdir/'writeups'/'figures'/'modeling_paper'/'energy_curves.svg'), bbox_inches='tight')\n",
    "fig_energy_curves.savefig(rd.outfile(rd.rootdir/'writeups'/'figures'/'modeling_paper'/'energy_curves.pdf'), bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Base model behavior figure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create the overall figure, gridspec, and add subfigure labels.\n",
    "fig_bm = plt.figure(figsize=(6.30,8.2))\n",
    "fig_bm_gridspec = gridspec.GridSpec(5, 2, figure=fig_bm,\n",
    "    wspace=0.1, hspace=0.15, height_ratios=[0.5, 0.7, 1.4, 0.8, 0.8])\n",
    "fig_bm_subfigures = {\n",
    "    'A': fig_bm.add_subfigure(fig_bm_gridspec[0,:]),\n",
    "    'B': fig_bm.add_subfigure(fig_bm_gridspec[1,:]),\n",
    "    'C': fig_bm.add_subfigure(fig_bm_gridspec[2,0]),\n",
    "    'D': fig_bm.add_subfigure(fig_bm_gridspec[2,1]),\n",
    "    'E': fig_bm.add_subfigure(fig_bm_gridspec[3,:]),\n",
    "    'F': fig_bm.add_subfigure(fig_bm_gridspec[4,:]),\n",
    "}\n",
    "for label, subfig in fig_bm_subfigures.items():\n",
    "    subfig.add_artist(matplotlib.text.Text(x=0, y=1, text=f'{label}.', fontsize=font_sizes['subpanel_label'], fontweight='bold', verticalalignment='top',transform=subfig.transSubfigure))\n",
    "fig_bm_palette = {\n",
    "    'fig.bm.tandem_reporter_upstream': main_palette['tandem'],\n",
    "    'fig.bm.tandem_reporter_downstream': main_palette['tandem_alt'],\n",
    "    'fig.bm.convergent': main_palette['convergent'],\n",
    "    'fig.bm.divergent': main_palette['divergent'],\n",
    "    'fig.bm.sc_bursts.tandem_reporter_upstream': main_palette['tandem'],\n",
    "    'fig.bm.sc_bursts.tandem_reporter_downstream': main_palette['tandem_alt'],\n",
    "    'fig.bm.sc_bursts.convergent': main_palette['convergent'],\n",
    "    'fig.bm.sc_bursts.divergent': main_palette['divergent'],\n",
    "    'independent': main_palette['light_gray'],\n",
    "    'dependent': main_palette['gray'],\n",
    "    'gray': main_palette['gray'],\n",
    "    'light_gray': main_palette['light_gray'],\n",
    "    'axis_gray': main_palette['axis_gray'],\n",
    "    '0th_order': main_palette['light_gray'],\n",
    "    '1st_order': main_palette['light_gray'],\n",
    "    '2nd_order': 'k',\n",
    "    'reporter_only': main_palette['reporter_only'],\n",
    "    'black': main_palette['black']\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot BCs\n",
    "df_bm_bcs = base_model_df[\n",
    "    (\n",
    "        (np.abs(base_model_df['coeff.sigma_squared'] - 0.025) < 0.001) &\n",
    "        (base_model_df['rates.sc_dependent'] == 1.0)\n",
    "    ) & (\n",
    "        (base_model_df['condition'] == 'fig.bm.tandem_reporter_upstream') |\n",
    "        (base_model_df['condition'] == 'fig.bm.tandem_reporter_downstream') |\n",
    "        (base_model_df['condition'] == 'fig.bm.convergent') |\n",
    "        (base_model_df['condition'] == 'fig.bm.divergent')\n",
    "    )\n",
    "].copy()\n",
    "df_bm_bcs['condition'] = df_bm_bcs['condition'].cat.remove_unused_categories()\n",
    "df_bm_bcs['norm_induction'] = df_bm_bcs['induction'] / (1.0 / 120.0)\n",
    "df_bm_bcs['constant_mRNA'] = df_bm_bcs['constant_mRNA'] / 250.0\n",
    "fig_bm_subfigures['B'].subplots_adjust(left=0.35, bottom=0.36, right=0.9, top=0.8)\n",
    "fig_bm_bcs_axes = fig_bm_subfigures['B'].subplots(ncols=2, gridspec_kw={'wspace': 0.5})\n",
    "for ax in fig_bm_bcs_axes:\n",
    "    ax.axvline(1.0, ymax=0.9, ls=':', color=main_palette['gray'])\n",
    "sns.lineplot(data=df_bm_bcs[df_bm_bcs['is_plasmid']==0.0],\n",
    "             x='norm_induction', y='constant_mRNA', hue='condition',\n",
    "             palette=fig_bm_palette, legend=False, linewidth=1,\n",
    "             ax=fig_bm_bcs_axes[1])\n",
    "sns.despine(ax=fig_bm_bcs_axes[1])\n",
    "sns.lineplot(data=df_bm_bcs[df_bm_bcs['is_plasmid']==1.0],\n",
    "             x='norm_induction', y='constant_mRNA', hue='condition',\n",
    "             palette=fig_bm_palette, legend=False, linewidth=1,\n",
    "             ax=fig_bm_bcs_axes[0])\n",
    "sns.despine(ax=fig_bm_bcs_axes[0])\n",
    "for ax in fig_bm_bcs_axes:\n",
    "    ax.set_xlabel('Fold induction')\n",
    "    ax.set_ylabel('Reporter output')\n",
    "    ax.set_xlim([0.0, 2.0])\n",
    "    ax.set_xticks([0.0, 1.0, 2.0])\n",
    "    ax.set_xticklabels(['0x', '1x', '2x'])\n",
    "    ax.set_ylim([0.0, 1.4])\n",
    "\n",
    "fig_bm_subfigures['B'].text(0.79, 0.85, 'Linear', ha='center', fontsize=font_sizes['base_size'])\n",
    "fig_bm_subfigures['B'].text(0.46, 0.85, 'Circular', ha='center', fontsize=font_sizes['base_size'])\n",
    "fig_bm.savefig('test.svg')\n",
    "del df_bm_bcs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot distribution data\n",
    "fig_bm_noise_axes = []\n",
    "for is_plasmid, subfig, noise_offset in zip([0,1], [fig_bm_subfigures[x] for x in 'DC'], [[-0.06,0], [0.3,0]]):\n",
    "    df_bm_dist = base_model_df[\n",
    "        (\n",
    "            base_model_df['is_plasmid'] == is_plasmid\n",
    "        ) & (\n",
    "            (base_model_df['condition'] == 'fig.bm.tandem_reporter_upstream') |\n",
    "            (base_model_df['condition'] == 'fig.bm.tandem_reporter_downstream') |\n",
    "            (base_model_df['condition'] == 'fig.bm.convergent') |\n",
    "            (base_model_df['condition'] == 'fig.bm.divergent')\n",
    "        ) & (\n",
    "            np.abs(base_model_df['induction'] - 8.0e-3) < 1e-4\n",
    "        ) & (\n",
    "            (base_model_df['rates.sc_dependent'] == 1.0) &\n",
    "            (np.abs(base_model_df['coeff.sigma_squared'] - 0.025) < 0.001)\n",
    "        )\n",
    "    ].copy()\n",
    "    df_bm_dist['condition'] = df_bm_dist['condition'].cat.remove_unused_categories()\n",
    "    df_bm_dist['norm_reporter']  = df_bm_dist['constant_mRNA'] / 340.0\n",
    "    df_bm_dist['norm_inducible'] = df_bm_dist['inducible_mRNA'] / 340.0\n",
    "    subfig.subplots_adjust(left=0.2, bottom=0.25, right=0.9, top=0.8)\n",
    "    fig_bm_dist_axes = subfig.subplots(ncols=2, sharey=False, gridspec_kw={'wspace': 0.5, 'width_ratios':[5, 1]})\n",
    "    for condition, noise_yloc in zip(['fig.bm.convergent', 'fig.bm.divergent', 'fig.bm.tandem_reporter_downstream', 'fig.bm.tandem_reporter_upstream'], [1,1,0.9,1.1]):\n",
    "        df_plot_dist = df_bm_dist[df_bm_dist.condition == condition]\n",
    "        # Compute distribution noise\n",
    "        ext_noise = np.var((df_plot_dist.norm_inducible + df_plot_dist.norm_reporter) / 2.0)\n",
    "        int_noise = np.var((df_plot_dist.norm_inducible - df_plot_dist.norm_reporter) / 2.0)\n",
    "        fig_bm_dist_axes[1].plot(noise_yloc, int_noise / ext_noise, 'o', color=fig_bm_palette[condition])\n",
    "        sns.kdeplot(data=df_plot_dist,\n",
    "                    x='norm_inducible', y='norm_reporter', color=fig_bm_palette[condition], alpha=0.9, linewidths=0.8,\n",
    "                    ax=fig_bm_dist_axes[0], levels=5 if is_plasmid else 8)\n",
    "    sns.despine(ax=fig_bm_dist_axes[0])\n",
    "    fig_bm_dist_axes[0].set_aspect('equal')\n",
    "    sns.despine(ax=fig_bm_dist_axes[1], bottom=True)\n",
    "    # Configure the noise ratio plot\n",
    "    fig_bm_dist_axes[1].xaxis.set_visible(False)\n",
    "    fig_bm_dist_axes[1].text(0.5, 0.95, 'Noise\\nratio', ha='center', ma='center', fontsize=font_sizes['data_annotation'], transform=fig_bm_dist_axes[1].transAxes)\n",
    "    fig_bm_dist_axes[1].set_xlim([0.8,1.2])\n",
    "    fig_bm_dist_axes[1].set_yscale('log')\n",
    "    fig_bm_dist_axes[0].set_xlim([0.0, 1.0])\n",
    "    fig_bm_dist_axes[0].set_ylim([0.0, 1.0])\n",
    "    fig_bm_dist_axes[0].set_xticks(np.linspace(0.0,1.0,5))\n",
    "    fig_bm_dist_axes[0].set_yticks(np.linspace(0.0,1.0,5))\n",
    "\n",
    "    fig_bm_dist_axes[0].set_xlabel('Inducible gene output')\n",
    "    fig_bm_dist_axes[0].set_ylabel('Reporter output')\n",
    "\n",
    "    o_x = noise_offset[0]\n",
    "    o_y = noise_offset[1]\n",
    "    fig_bm_dist_axes[0].annotate(text='', xytext=(0.25 + o_x, 0.05 + o_y), xy=(0.1 + o_x, 0.2 + o_y),\n",
    "        arrowprops={'facecolor': 'black', 'edgecolor': '#000000ff', 'width': 0.8, 'headwidth': 5.0, 'headlength': 5.0, 'shrink': 0.0, 'capstyle': 'round'})\n",
    "    fig_bm_dist_axes[0].annotate(text='', xytext=(0.25 + o_x, 0.05 + o_y), xy=(0.4 + o_x, 0.2 + o_y),\n",
    "        arrowprops={'facecolor': 'black', 'edgecolor': '#000000ff', 'width': 0.8, 'headwidth': 5.0, 'headlength': 5.0, 'shrink': 0.0, 'capstyle': 'round'})\n",
    "    fig_bm_dist_axes[0].text(0.18 + o_x, 0.05 + o_y, 'int.', horizontalalignment='right', transform=fig_bm_dist_axes[0].transData, fontsize=font_sizes['data_annotation'], fontstyle='italic')\n",
    "    fig_bm_dist_axes[0].text(0.32 + o_x, 0.05 + o_y, 'ext.', horizontalalignment='left', transform=fig_bm_dist_axes[0].transData, fontsize=font_sizes['data_annotation'], fontstyle='italic')\n",
    "    subfig.text(0.25 + o_x, 0.20 + o_y, 'noise', horizontalalignment='center', transform=fig_bm_dist_axes[0].transData, fontsize=font_sizes['data_annotation'], fontstyle='italic')\n",
    "    fig_bm_noise_axes.append(fig_bm_dist_axes[1])\n",
    "fig_bm_subfigures['D'].text(0.5, 0.85, 'Linear boundary conditions', fontsize=font_sizes['base_size'], ha='center')\n",
    "fig_bm_subfigures['C'].text(0.5, 0.85, 'Circular boundary conditions', fontsize=font_sizes['base_size'], ha='center')\n",
    "fig_bm_noise_axes[0].set_ylim((0.1, 50))\n",
    "fig_bm_noise_axes[0].set_yticks([0.1,1.0,10.0])\n",
    "fig_bm_noise_axes[0].set_yticklabels(['0.1','1.0', '10.0'])\n",
    "fig_bm_noise_axes[1].set_ylim((0.01, 0.5))\n",
    "fig_bm_noise_axes[1].set_yticks([0.01,0.1])\n",
    "fig_bm_noise_axes[1].set_yticklabels(['0.01','0.1'])\n",
    "fig_bm.savefig('test.svg')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot overall sc density as a function of induction\n",
    "fig_bm_subfigures['E'].subplots_adjust(left=0.1, bottom=0.35, right=0.85, top=0.75)\n",
    "fig_sc_density_axes = fig_bm_subfigures['E'].subplots(nrows=1, ncols=4, sharey=True, sharex=True, gridspec_kw={'hspace': 0.5})\n",
    "conditions = [f'fig.bm.sc_bursts.{x}' for x in ['convergent', 'divergent', 'tandem_reporter_upstream', 'tandem_reporter_downstream']]\n",
    "induction_levels = [0.0, 1.0]\n",
    "for cond_idx, induction_idx in itertools.product(range(len(conditions)), range(len(induction_levels))):\n",
    "    ax = fig_sc_density_axes[cond_idx]\n",
    "    cond = conditions[cond_idx]\n",
    "    induction = induction_levels[induction_idx]\n",
    "    ax.axhline(y=0, color=main_palette['axis_gray'])\n",
    "    if induction_idx == 1:\n",
    "        sns.lineplot(data=sc_density_df[\n",
    "            (sc_density_df.condition == cond) &\n",
    "            (sc_density_df.norm_induction == induction) &\n",
    "            (sc_density_df.alpha == 0.025) &\n",
    "            (np.abs(sc_density_df.time - 15000) < 1)\n",
    "        ], x='distance', y='sc_density',\n",
    "        color=fig_bm_palette[cond],\n",
    "        ls='-',\n",
    "        ax=ax, legend=None)\n",
    "    else:\n",
    "        #sns.lineplot(data=sc_density_df[\n",
    "        #    (sc_density_df.condition == cond) &\n",
    "        #    (sc_density_df.norm_induction == induction) &\n",
    "        #    (sc_density_df.alpha == 0.025) &\n",
    "        #    (np.abs(sc_density_df.time - 15000) < 1)\n",
    "        #], x='distance', y='sc_density',\n",
    "        #color=fig_bm_palette[cond],\n",
    "        #ls='-', alpha=0.5,\n",
    "        #ax=ax, legend=None)\n",
    "        sns.lineplot(data=sc_density_df[\n",
    "            (sc_density_df.condition == cond) &\n",
    "            (sc_density_df.norm_induction == induction) &\n",
    "            (sc_density_df.alpha == 0.025) &\n",
    "            (np.abs(sc_density_df.time - 15000) < 1)\n",
    "        ], x='distance', y='sc_density',\n",
    "        color=fig_bm_palette['reporter_only'],\n",
    "        ls=':',\n",
    "        ax=ax, legend=None)\n",
    "    ax.set_xlim([0,3400])\n",
    "    ax.set_yticks([-0.3,0.0, 0.3])\n",
    "    ax.set_xticks([x * 0.34 for x in [0,5000,10000]])\n",
    "    ax.set_xticklabels(['0', '5', '10'])\n",
    "    ax.set_xlabel('')\n",
    "    ax.text(100, 0.39, 'Uninduced', color=fig_bm_palette['reporter_only'], fontsize=font_sizes['data_annotation'])\n",
    "    ax.text(100, 0.27, 'Induced', color=fig_bm_palette[cond], fontsize=font_sizes['data_annotation'])\n",
    "    xax_transform = ax.get_xaxis_transform()\n",
    "    if induction_idx == 0:\n",
    "        if cond.endswith('tandem_reporter_upstream'):\n",
    "            plot_genes_unified(ax, [(3000 * 0.34, 4000 * 0.34), (6000 * 0.34, 7000 * 0.34)], [fig_bm_palette[cond], fig_bm_palette['gray']], plot_gene_body=True, units='xaxis', plot_inducible=1)\n",
    "        if cond.endswith('tandem_reporter_downstream'):\n",
    "            plot_genes_unified(ax, [(3000 * 0.34, 4000 * 0.34), (6000 * 0.34, 7000 * 0.34)], [fig_bm_palette['gray'], fig_bm_palette[cond]], plot_gene_body=True, units='xaxis', plot_inducible=0)\n",
    "        if cond.endswith('convergent'):\n",
    "            plot_genes_unified(ax, [(3000 * 0.34, 4000 * 0.34), (7000 * 0.34, 6000 * 0.34)], [fig_bm_palette['gray'], fig_bm_palette[cond]], plot_gene_body=True, units='xaxis', plot_inducible=0)\n",
    "        if cond.endswith('divergent'):\n",
    "            plot_genes_unified(ax, [(4000 * 0.34, 3000 * 0.34), (6000 * 0.34, 7000 * 0.34)], [fig_bm_palette['gray'], fig_bm_palette[cond]], plot_gene_body=True, units='xaxis', plot_inducible=0)\n",
    "    sns.despine(ax=ax)\n",
    "\n",
    "fig_sc_density_axes[0].set_ylabel(\"Supercoiling density\")\n",
    "fig_bm_subfigures['E'].text(0.5, 0.13, 'Genomic distance (kb)', ha='center')\n",
    "fig_bm.savefig('test.svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot delta spacing subfigure\n",
    "df_bm_spacing = spacing_sweep_df[\n",
    "    (\n",
    "        spacing_sweep_df['is_plasmid'] == 0\n",
    "    ) & (\n",
    "        (spacing_sweep_df['condition'] == 'fig.bm.spacing.tandem_reporter_upstream') |\n",
    "        (spacing_sweep_df['condition'] == 'fig.bm.spacing.tandem_reporter_downstream') |\n",
    "        (spacing_sweep_df['condition'] == 'fig.bm.spacing.convergent') |\n",
    "        (spacing_sweep_df['condition'] == 'fig.bm.spacing.divergent')\n",
    "    ) & (\n",
    "        (spacing_sweep_df['rates.sc_dependent'] == 1.0) &\n",
    "        (spacing_sweep_df['coeff.sigma_squared'] == 0.025)\n",
    "    ) & (\n",
    "        (np.abs(spacing_sweep_df['induction'] - 0.0221 / 120.0) < 0.001 / 120.0)|\n",
    "        (np.abs(spacing_sweep_df['induction'] - 0.3562 / 120.0) < 0.001 / 120.0)|\n",
    "        (np.abs(spacing_sweep_df['induction'] - 1.172 / 120.0) < 0.001 / 120.0) |\n",
    "        (np.abs(spacing_sweep_df['induction'] - 1.429 / 120.0) < 0.001 / 120.0) |\n",
    "        (np.abs(spacing_sweep_df['induction'] - 2.592 / 120.0) < 0.001 / 120.0)\n",
    "    )\n",
    "].copy()\n",
    "df_bm_spacing['kb_spacing'] = df_bm_spacing['spacing'] / 1000\n",
    "df_bm_spacing['norm_induction'] = df_bm_spacing['induction'] / (1.0 / 120.0)\n",
    "df_bm_spacing['norm_reporter']  = df_bm_spacing['constant_mRNA'] / 300.0\n",
    "df_bm_spacing['condition'] = df_bm_spacing['condition'].cat.remove_unused_categories()\n",
    "max_spacing_median = df_bm_spacing.groupby(['condition', 'norm_induction', 'spacing']).median()['norm_reporter'].xs(10000, level='spacing')\n",
    "df_bm_spacing['delta_reporter'] = df_bm_spacing.groupby(['condition', 'norm_induction', 'spacing'])['norm_reporter'].transform(lambda col: col - max_spacing_median.xs(col.name[0], level='condition').xs(col.name[1]))\n",
    "\n",
    "fig_bm_subfigures['F'].subplots_adjust(left=0.1, bottom=0.35, right=0.85, top=0.75)\n",
    "fig_bm_spacing_axes = fig_bm_subfigures['F'].subplots(ncols=4, sharey=True)\n",
    "for condition, ax in zip(['fig.bm.spacing.convergent', 'fig.bm.spacing.divergent', 'fig.bm.spacing.tandem_reporter_upstream', 'fig.bm.spacing.tandem_reporter_downstream'], fig_bm_spacing_axes):\n",
    "    ax.axhline(0, color=fig_bm_palette['gray'], linestyle=':')\n",
    "    sns.lineplot(data=df_bm_spacing[df_bm_spacing['condition']==condition], x='kb_spacing', y='delta_reporter', hue='norm_induction',\n",
    "        palette=no_white_greens, hue_norm=matplotlib.colors.Normalize(vmin=0, vmax=2),\n",
    "        legend=False, ax=ax, ci=None)\n",
    "    #ax.set_facecolor(light_palette[condition.replace('_','.').split('.')[2]])\n",
    "    sns.despine(ax=ax)\n",
    "    ax.set_xscale('log')\n",
    "    ax.set_xlabel('')\n",
    "    ax.set_xlim([0.5,10])\n",
    "    ax.set_yticks([-0.2,0,0.1])\n",
    "    ax.set_ylim([-0.25,0.1])\n",
    "\n",
    "plot_genes_unified(fig_bm_spacing_axes[0], [(0.3, 0.4), (0.7, 0.6)], [fig_bm_palette[c] for c in ['gray', 'fig.bm.convergent']], base_line_extents=(0.1,0.9), plot_dx=(0,1), plot_inducible=0)\n",
    "plot_genes_unified(fig_bm_spacing_axes[1], [(0.4, 0.3), (0.6, 0.7)], [fig_bm_palette[c] for c in ['gray', 'fig.bm.divergent']], base_line_extents=(0.1,0.9), plot_dx=(0,1), plot_inducible=0)\n",
    "plot_genes_unified(fig_bm_spacing_axes[2], [(0.3, 0.4), (0.6, 0.7)], [fig_bm_palette[c] for c in ['fig.bm.tandem_reporter_upstream', 'gray']], base_line_extents=(0.1,0.9), plot_dx=(0,1), plot_inducible=1)\n",
    "plot_genes_unified(fig_bm_spacing_axes[3], [(0.3, 0.4), (0.6, 0.7)], [fig_bm_palette[c] for c in ['gray', 'fig.bm.tandem_reporter_downstream']], base_line_extents=(0.1,0.9), plot_dx=(0,1), plot_inducible=0)\n",
    "    \n",
    "# Add colorbar\n",
    "fig_bm_spacing_cbar = fig_bm_subfigures['F'].add_axes([0.87, 0.35, 0.02, 0.4])\n",
    "fig_bm_subfigures['F'].colorbar(cm.ScalarMappable(norm=matplotlib.colors.Normalize(vmin=0, vmax=2), cmap=no_white_greens), cax=fig_bm_spacing_cbar)\n",
    "fig_bm_spacing_cbar.set_yticks([0,1,2])\n",
    "fig_bm_spacing_cbar.yaxis.set_major_formatter(fold_formatter)\n",
    "fig_bm_spacing_cbar.set_title('Fold\\ninduction', fontsize=font_sizes['colorbar_title'], loc='left')\n",
    "fig_bm_spacing_axes[0].set_ylabel('ΔReporter output')\n",
    "fig_bm_subfigures['F'].text(0.5, 0.12, 'Inter-gene spacing (kb)', ha='center')\n",
    "fig_bm.savefig('test.svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add external images\n",
    "fig_bm.savefig(rd.outfile(rd.rootdir/'writeups'/'figures'/'modeling_paper'/'fig_base_model_mpl.svg','fig_base_model'))\n",
    "fig_bm.savefig(rd.outfile(rd.rootdir/'writeups'/'figures'/'modeling_paper'/'fig_base_model_mpl.pdf','fig_base_model'))\n",
    "#add_external_svgs(\n",
    "#    rd.outfile(rd.rootdir/'writeups'/'figures'/'modeling_paper'/'fig_base_model.svg', 'fig_base_model_assemble'),\n",
    "#    rd.infile(rd.rootdir/'writeups'/'figures'/'modeling_paper'/'fig_base_model_mpl.svg', 'fig_base_model_assemble'), fig_bm_subfigures, [\n",
    "#    ('A', {\n",
    "#        'filename': rd.infile(rd.rootdir/'writeups'/'figures'/'modeling_paper'/'fig_1a.svg', 'fig_base_model_assemble'),\n",
    "#        'extents': (0.05, 0.05, 0.5, 0.9)\n",
    "#    }),\n",
    "#])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Supplemental figure - torque and energy response"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig_supp_torque_energy = plt.figure(figsize=(7,2))\n",
    "fig_supp_torque_energy_gridspec = gridspec.GridSpec(1, 2, figure=fig_supp_torque_energy,\n",
    "    wspace=0.1, hspace=0.15)\n",
    "fig_supp_torque_energy_subfigures = {\n",
    "    'A': fig_supp_torque_energy.add_subfigure(fig_supp_torque_energy_gridspec[0,0]),\n",
    "    'B': fig_supp_torque_energy.add_subfigure(fig_supp_torque_energy_gridspec[0,1]),\n",
    "}\n",
    "for label, subfig in fig_supp_torque_energy_subfigures.items():\n",
    "    subfig.add_artist(matplotlib.text.Text(x=0, y=1, text=f'{label}.', fontsize=font_sizes['subpanel_label'], fontweight='bold', verticalalignment='top',transform=subfig.transSubfigure))\n",
    "    subfig.subplots_adjust(left=0.2, bottom=0.28, right=0.95, top=0.95)\n",
    "energy_df = pd.read_csv(rd.infile(rd.rootdir/'data'/'rnap_delta_bubble_energy.csv', 'fig_supp_torque_energy'), names=['sigma','energy_bubble', 'energy_nobubble'])\n",
    "energy_df['raw_torque'] = energy_df.energy_nobubble / (1.2 * 2 * np.pi)\n",
    "torque_ax = fig_supp_torque_energy_subfigures['A'].subplots(ncols=1)\n",
    "energy_ax = fig_supp_torque_energy_subfigures['B'].subplots(ncols=1)\n",
    "\n",
    "energy_ax.axvline(x=0, color=main_palette['gray'], label='_')\n",
    "energy_ax.axhline(y=0, color=main_palette['gray'], label='_')\n",
    "torque_ax.axvline(x=0, color=main_palette['gray'], label='_')\n",
    "torque_ax.axhline(y=0, color=main_palette['gray'], label='_')\n",
    "sns.lineplot(data=energy_df, x='sigma', y='raw_torque', ax=torque_ax, color='k')\n",
    "sns.lineplot(data=energy_df, x='sigma', y='energy_nobubble', ax=energy_ax, color=main_palette['gray'])\n",
    "sns.lineplot(data=energy_df, x='sigma', y='energy_bubble', ax=energy_ax, color=main_palette['polymerase_color'])\n",
    "#energy_ax.legend(['Unwinding only', 'Unwinding\\n+ local melting'])\n",
    "energy_ax.text(-0.15, 220, 'Unwinding only', color=main_palette['gray'])\n",
    "energy_ax.text(-0.15, 110, 'Unwinding and\\nlocal melting', color=main_palette['polymerase_color'])\n",
    "energy_ax.set_xlabel('Supercoilng density, σ')\n",
    "energy_ax.set_ylabel('Energy (pN nm)')\n",
    "energy_ax.set_xlim([-0.2,0.2])\n",
    "torque_ax.set_xlabel('Supercoiling density, σ')\n",
    "torque_ax.set_ylabel('Torque (pN nm)')\n",
    "torque_ax.set_xlim([-0.2,0.2])\n",
    "sns.despine(ax=energy_ax)\n",
    "sns.despine(ax=torque_ax)\n",
    "fig_supp_torque_energy.savefig('test.svg')\n",
    "fig_supp_torque_energy.savefig(rd.outfile(rd.rootdir/'writeups'/'figures'/'modeling_paper'/'fig_supp_torque_energy.svg', 'fig_supp_torque_energy'))\n",
    "fig_supp_torque_energy.savefig(rd.outfile(rd.rootdir/'writeups'/'figures'/'modeling_paper'/'fig_supp_torque_energy.pdf', 'fig_supp_torque_energy'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Supplemental Figure - Alpha sweep\n",
    "\n",
    "Shows how the behavior of the base model changes as we change the value of $\\alpha$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot sweeping alpha\n",
    "df_alpha_sweep = base_model_df[\n",
    "    (\n",
    "        base_model_df.is_plasmid == 0.0\n",
    "    ) & (\n",
    "        (base_model_df['condition'] == 'fig.bm.tandem_reporter_upstream') |\n",
    "        (base_model_df['condition'] == 'fig.bm.tandem_reporter_downstream') |\n",
    "        (base_model_df['condition'] == 'fig.bm.convergent') |\n",
    "        (base_model_df['condition'] == 'fig.bm.divergent')\n",
    "    ) & (\n",
    "        base_model_df['rates.sc_dependent'] == 1.0\n",
    "    ) & (\n",
    "        base_model_df['coeff.sigma_squared'] > 0\n",
    "    )\n",
    "].copy()\n",
    "df_alpha_sweep['condition'] = df_alpha_sweep['condition'].cat.remove_unused_categories()\n",
    "df_alpha_sweep['norm_induction'] = df_alpha_sweep['induction'] / (1.0 / 120.0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "alpha_values = [0.0, 0.01, 0.02, 0.025, 0.03, 0.0316, 0.05]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig_supp_alpha = plt.figure(figsize=(6,7.5))\n",
    "fig_supp_alpha_gridspec = gridspec.GridSpec(5, 1, figure=fig_supp_alpha,\n",
    "    wspace=0.1, hspace=0.15)\n",
    "fig_supp_alpha_subfigures = {\n",
    "    'A': fig_supp_alpha.add_subfigure(fig_supp_alpha_gridspec[0,:]),\n",
    "    'B': fig_supp_alpha.add_subfigure(fig_supp_alpha_gridspec[1,:]),\n",
    "    'C': fig_supp_alpha.add_subfigure(fig_supp_alpha_gridspec[2,:]),\n",
    "    'D': fig_supp_alpha.add_subfigure(fig_supp_alpha_gridspec[3,:]),\n",
    "    'E': fig_supp_alpha.add_subfigure(fig_supp_alpha_gridspec[4,:]),\n",
    "}\n",
    "for label, subfig in fig_supp_alpha_subfigures.items():\n",
    "    subfig.add_artist(matplotlib.text.Text(x=0, y=1, text=f'{label}.', fontsize=font_sizes['subpanel_label'], fontweight='bold', verticalalignment='top',transform=subfig.transSubfigure))\n",
    "    subfig.subplots_adjust(left=0.1, bottom=0.3, right=0.95, top=0.8)\n",
    "unique_alphas = sorted(base_model_df['coeff.sigma_squared'].unique())\n",
    "fig_supp_alpha_induction = fig_supp_alpha_subfigures['A'].subplots(ncols=3, sharey=True)\n",
    "for ax, induction in zip(fig_supp_alpha_induction,[0.1, 1.0, 2.0]):\n",
    "    df_at_induction = df_alpha_sweep[np.abs(df_alpha_sweep.norm_induction - induction) == min(np.abs(df_alpha_sweep.norm_induction - induction))]\n",
    "    sns.lineplot(data=df_at_induction, x='coeff.sigma_squared', y='constant_mRNA', hue='condition',\n",
    "                palette=fig_bm_palette, ci=None, ax=ax, legend=None)\n",
    "    sns.despine(ax=ax)\n",
    "    ax.set_xlim([0.01, 0.1])\n",
    "    ax.set_xscale('log')\n",
    "    ax.set_xticks([0.01,0.1])\n",
    "    ax.xaxis.set_minor_formatter('')\n",
    "    ax.set_xlabel('')\n",
    "    for tick, align in zip(ax.xaxis.get_major_ticks(), ['left', 'right']):\n",
    "        tick.label1.set_horizontalalignment(align)\n",
    "    \n",
    "    ax.axvline(0.025, color=fig_bm_palette['2nd_order'])\n",
    "    ax.text(0.5, 1.05, f'{induction:0.0f}x induction', transform=ax.transAxes, fontsize=font_sizes['subaxis_title'], ha='center')\n",
    "fig_supp_alpha_induction[1].set_xlabel('α')\n",
    "fig_supp_alpha_induction[0].set_ylabel('Reporter counts')\n",
    "fig_supp_energy_axes = fig_supp_alpha_subfigures['B'].subplots(ncols=len(alpha_values), sharex=True, sharey=True)\n",
    "fig_supp_linear_axes = fig_supp_alpha_subfigures['C'].subplots(ncols=len(alpha_values), sharex=True, sharey=True)\n",
    "fig_supp_circular_axes = fig_supp_alpha_subfigures['D'].subplots(ncols=len(alpha_values), sharex=True, sharey=True)\n",
    "fig_supp_dist_axes = fig_supp_alpha_subfigures['E'].subplots(ncols=len(alpha_values), sharex=True, sharey=True)\n",
    "# Plot Fig1b\n",
    "for alpha, e_ax, linear_ax, circular_ax, dist_ax in zip(\n",
    "    alpha_values,\n",
    "    fig_supp_energy_axes,\n",
    "    fig_supp_linear_axes,\n",
    "    fig_supp_circular_axes,\n",
    "    fig_supp_dist_axes\n",
    "):\n",
    "    df_supp_alpha = base_model_df[\n",
    "        (\n",
    "            base_model_df['rates.sc_dependent'] == 1.0\n",
    "        ) & (\n",
    "            (base_model_df['condition'] == 'fig.bm.tandem_reporter_downstream') |\n",
    "            (base_model_df['condition'] == 'fig.bm.tandem_reporter_upstream') |\n",
    "            (base_model_df['condition'] == 'fig.bm.convergent') |\n",
    "            (base_model_df['condition'] == 'fig.bm.divergent')\n",
    "        ) & (\n",
    "            (np.abs(base_model_df['coeff.sigma_squared'] - alpha) < 0.0001)\n",
    "        )\n",
    "    ].copy()\n",
    "    e_ax.axhline(0, color=main_palette['gray'])\n",
    "    e_ax.axvline(0, color=main_palette['gray'])\n",
    "\n",
    "    closest_alpha_df = alpha_sweep_df[\n",
    "        np.abs(alpha_sweep_df.alpha - alpha) == min(np.abs(alpha_sweep_df.alpha - alpha))\n",
    "    ]\n",
    "    \n",
    "    e_ax.plot(closest_alpha_df['sigma'], closest_alpha_df['energy'], color='k')\n",
    "    e_ax.text(x=0.5, y=1.1, s=f'α={alpha:0.1e}', horizontalalignment='center', transform=e_ax.transAxes)\n",
    "\n",
    "    df_supp_alpha['condition'] = df_supp_alpha['condition'].cat.remove_unused_categories()\n",
    "    df_supp_alpha['norm_induction'] = df_supp_alpha['induction'] / (1.0 / 120.0)\n",
    "    sns.lineplot(data=df_supp_alpha[df_supp_alpha.is_plasmid == 0.0],\n",
    "                x='norm_induction', y='constant_mRNA', hue='condition', palette=fig_bm_palette, linewidth=1.5,\n",
    "                legend=False, ax=linear_ax, ci=None)\n",
    "    sns.lineplot(data=df_supp_alpha[df_supp_alpha.is_plasmid == 1.0],\n",
    "                x='norm_induction', y='constant_mRNA', hue='condition', palette=fig_bm_palette, linewidth=1.5,\n",
    "                legend=False, ax=circular_ax, ci=None)\n",
    "    sns.kdeplot(data=df_supp_alpha[(df_supp_alpha.is_plasmid == 0.0) & (np.abs(df_supp_alpha.induction - 8.0e-3) == np.min(np.abs(df_supp_alpha.induction - 8.0e-3)))],\n",
    "                x='inducible_mRNA', y='constant_mRNA', hue='condition', palette=fig_bm_palette, ax=dist_ax, levels=4, legend=None)\n",
    "    linear_ax.set_xlabel('')\n",
    "    circular_ax.set_xlabel('')\n",
    "    dist_ax.set_xlabel('')\n",
    "    sns.despine(ax=e_ax)\n",
    "    sns.despine(ax=linear_ax)\n",
    "    sns.despine(ax=circular_ax)\n",
    "    sns.despine(ax=dist_ax)\n",
    "fig_supp_energy_axes[3].set_xlabel('σ')\n",
    "fig_supp_linear_axes[3].set_xlabel('Inducible output')\n",
    "fig_supp_circular_axes[3].set_xlabel('Inducible output')\n",
    "fig_supp_dist_axes[3].set_xlabel('Inducible output')\n",
    "fig_supp_energy_axes[0].set_ylabel('Energy (pN nm)')\n",
    "fig_supp_linear_axes[0].set_ylabel('Reporter output')\n",
    "fig_supp_circular_axes[0].set_ylabel('Reporter output')\n",
    "fig_supp_dist_axes[0].set_ylabel('Reporter output')\n",
    "fig_supp_linear_axes[3].text(x=0.5, y=1.1, s='Linear BCs', horizontalalignment='center', transform=fig_supp_linear_axes[3].transAxes)\n",
    "fig_supp_circular_axes[3].text(x=0.5, y=1.1, s='Circular BCs', horizontalalignment='center', transform=fig_supp_circular_axes[3].transAxes)\n",
    "fig_supp_alpha.savefig('test.svg')\n",
    "fig_supp_alpha.savefig(rd.outfile(rd.rootdir/'writeups'/'figures'/'modeling_paper'/'fig_supp_alpha_sweep.svg', 'fig_supp_alpha'))\n",
    "fig_supp_alpha.savefig(rd.outfile(rd.rootdir/'writeups'/'figures'/'modeling_paper'/'fig_supp_alpha_sweep.pdf', 'fig_supp_alpha'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Supplemental figure - Zero, first, second order model comparison with alpha sweep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create the overall figure, gridspec, and add subfigure labels.\n",
    "fig_order_model_comparison = plt.figure(figsize=(6.30,2))\n",
    "fig_order_model_comparison_gridspec = gridspec.GridSpec(1, 2, figure=fig_order_model_comparison,\n",
    "    wspace=0.1, hspace=0.15)\n",
    "fig_order_model_comparison_subfigures = {\n",
    "    'A': fig_order_model_comparison.add_subfigure(fig_order_model_comparison_gridspec[0,0]),\n",
    "    'B': fig_order_model_comparison.add_subfigure(fig_order_model_comparison_gridspec[0,1]),\n",
    "}\n",
    "for label, subfig in fig_order_model_comparison_subfigures.items():\n",
    "    subfig.add_artist(matplotlib.text.Text(x=0, y=1, text=f'{label}.', fontsize=font_sizes['subpanel_label'],\n",
    "        fontweight='bold', verticalalignment='top',transform=subfig.transSubfigure))\n",
    "fig_order_model_comparison_palette = {\n",
    "    'fig.bm.tandem_reporter_upstream': main_palette['tandem'],\n",
    "    'fig.bm.tandem_reporter_downstream': main_palette['tandem_alt'],\n",
    "    'fig.bm.convergent': main_palette['convergent'],\n",
    "    'fig.bm.divergent': main_palette['divergent'],\n",
    "    'independent': main_palette['light_gray'],\n",
    "    'dependent': main_palette['gray'],\n",
    "    'gray': main_palette['gray'],\n",
    "    'light_gray': main_palette['light_gray'],\n",
    "    'axis_gray': main_palette['axis_gray'],\n",
    "    '0th_order': main_palette['light_gray'],\n",
    "    '1st_order': main_palette['light_gray'],\n",
    "    '2nd_order': 'k'\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig_order_model_comparison_subfigures['B'].subplots_adjust(left=0.2, bottom=0.3, right=0.95, top=0.90)\n",
    "fig_supp_bm_alpha_ax = fig_order_model_comparison_subfigures['B'].subplots(ncols=1, sharey=True)\n",
    "df_at_induction = df_alpha_sweep[np.abs(df_alpha_sweep.norm_induction - 1.0) == min(np.abs(df_alpha_sweep.norm_induction - 1.0))]\n",
    "sns.lineplot(data=df_at_induction, x='coeff.sigma_squared', y='constant_mRNA', hue='condition',\n",
    "            palette=fig_bm_palette, ci=None, ax=fig_supp_bm_alpha_ax, legend=None)\n",
    "sns.despine(ax=fig_supp_bm_alpha_ax)\n",
    "fig_supp_bm_alpha_ax.set_xlim([0.01, 0.1])\n",
    "fig_supp_bm_alpha_ax.set_xscale('log')\n",
    "fig_supp_bm_alpha_ax.set_xticks([0.01,0.1])\n",
    "fig_supp_bm_alpha_ax.xaxis.set_minor_formatter('')\n",
    "fig_supp_bm_alpha_ax.set_xlabel('')\n",
    "for tick, align in zip(fig_supp_bm_alpha_ax.xaxis.get_major_ticks(), ['left', 'right']):\n",
    "    tick.label1.set_horizontalalignment(align)\n",
    "\n",
    "fig_supp_bm_alpha_ax.axvline(0.025, color=main_palette['light_gray'], linestyle=':')\n",
    "fig_supp_bm_alpha_ax.set_xlabel('α')\n",
    "fig_supp_bm_alpha_ax.set_ylabel('Reporter counts')\n",
    "fig_order_model_comparison.savefig('test.svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot output by alpha value\n",
    "df_supp_bm_order = base_model_df[\n",
    "    (\n",
    "        base_model_df['is_plasmid'] == 0\n",
    "    ) & (\n",
    "        (base_model_df['condition'] == 'fig.bm.tandem_reporter_upstream') |\n",
    "        (base_model_df['condition'] == 'fig.bm.tandem_reporter_downstream') |\n",
    "        (base_model_df['condition'] == 'fig.bm.convergent') |\n",
    "        (base_model_df['condition'] == 'fig.bm.divergent')\n",
    "    ) & (\n",
    "        (base_model_df['rates.sc_dependent'] == 0.0) |\n",
    "        (\n",
    "            (base_model_df['rates.sc_dependent'] == 1.0) &\n",
    "            (\n",
    "                (base_model_df['coeff.sigma_squared'] == 0.0) |\n",
    "                (np.abs(base_model_df['coeff.sigma_squared'] - 0.025) < 0.001)\n",
    "            )\n",
    "        )\n",
    "    )\n",
    "].copy()\n",
    "df_supp_bm_order['sc_condition'] = 'independent'\n",
    "df_supp_bm_order.loc[\n",
    "    (df_supp_bm_order['rates.sc_dependent'] == 1.0) &\n",
    "    (df_supp_bm_order['coeff.sigma_squared'] == 0.0), 'sc_condition'] = 'dependent'\n",
    "df_supp_bm_order.loc[\n",
    "    (df_supp_bm_order['rates.sc_dependent'] == 1.0) &\n",
    "    (np.abs(df_supp_bm_order['coeff.sigma_squared'] - 0.025) < 0.001), 'sc_condition'] = 'dependent_squared'\n",
    "df_supp_bm_order['condition'] = df_supp_bm_order['condition'].cat.remove_unused_categories()\n",
    "df_supp_bm_order_normed = (\n",
    "    df_supp_bm_order.set_index(\n",
    "        [\n",
    "            'condition',\n",
    "            'sc_condition',\n",
    "            'induction'\n",
    "        ])['constant_mRNA'] # Select the constant_mRNA column\n",
    "    / df_supp_bm_order.groupby([\n",
    "        'condition',\n",
    "        'sc_condition',\n",
    "        'induction']).mean().xs( # Divide by the zero-induction, tandem case\n",
    "            sorted(df_supp_bm_order['induction'].unique())[1], level='induction')['constant_mRNA']).reset_index()\n",
    "df_supp_bm_order_normed['norm_induction'] = df_supp_bm_order_normed['induction'] / (1.0 / 120.0)\n",
    "# Tweak subplot layout\n",
    "fig_order_model_comparison_subfigures['A'].subplots_adjust(left=0.2, bottom=0.25, right=0.95, top=0.8)\n",
    "fig_order_compare_axes = fig_order_model_comparison_subfigures['A'].subplots(ncols=4, sharey=True)\n",
    "dashes = {'independent': (2.8,1.2), 'dependent': '', 'dependent_squared': ''}\n",
    "for ax, condition in zip(fig_order_compare_axes, ['fig.bm.tandem_reporter_upstream', 'fig.bm.tandem_reporter_downstream', 'fig.bm.convergent', 'fig.bm.divergent']):\n",
    "    fig_order_model_comparison_palette['dependent_squared'] = fig_order_model_comparison_palette[condition]\n",
    "    sns.lineplot(data=df_supp_bm_order_normed[df_supp_bm_order_normed.condition == condition],\n",
    "                x='norm_induction', y='constant_mRNA', style='sc_condition', hue='sc_condition', dashes=dashes, palette=fig_order_model_comparison_palette, linewidth=1.5,\n",
    "                legend=False, ax=ax, ci=None)\n",
    "    sns.despine(ax=ax)\n",
    "    ax.set_xlim([0.0, 2.5])\n",
    "    ax.set_xticks([0.0, 1.0, 2.0])\n",
    "    ax.set_xticklabels(['0', '1', '2'])\n",
    "    #ax.set_ylim([0.0, 1.4])\n",
    "# Set labels\n",
    "fig_order_compare_axes[0].set_ylabel('Relative\\nreporter output')\n",
    "fig_order_compare_axes[0].set_xlabel('')\n",
    "fig_order_compare_axes[1].set_xlabel('Fold induction')\n",
    "fig_order_compare_axes[2].set_xlabel('')\n",
    "fig_order_compare_axes[3].set_xlabel('')\n",
    "\n",
    "for ax, cond, locs in zip(fig_order_compare_axes,\n",
    "        ['fig.bm.tandem_reporter_upstream', 'fig.bm.tandem_reporter_downstream','fig.bm.convergent', 'fig.bm.divergent'],\n",
    "        [(1.8,0.85,1.8, 0.7,0.2,1.15),\n",
    "         (1.8,0.85,0.55,0.8,0.2, 0.25),\n",
    "         (1.8,0.85,0.8,0.8,0.4, 0.25),\n",
    "         (1.8,0.85,1.8,0.7,0.15,0.65)]):\n",
    "    ax.text(locs[0], locs[1], '0°', color=fig_order_model_comparison_palette['independent'], fontsize=font_sizes['line_annotation'])\n",
    "    ax.text(locs[2], locs[3], '1°', color=fig_order_model_comparison_palette['dependent'], fontsize=font_sizes['line_annotation'])\n",
    "    ax.text(locs[4], locs[5], '2°', color=fig_order_model_comparison_palette[cond], fontsize=font_sizes['line_annotation'])\n",
    "# Add annotations\n",
    "fig_order_model_comparison_subfigures['A'].text(0.5, 1, 'Upstream\\nTandem', horizontalalignment='center', transform=fig_order_compare_axes[0].transAxes, fontsize=font_sizes['subaxis_title'])\n",
    "fig_order_model_comparison_subfigures['A'].text(0.5, 1, 'Downstream\\nTandem', horizontalalignment='center', transform=fig_order_compare_axes[1].transAxes, fontsize=font_sizes['subaxis_title'])\n",
    "fig_order_model_comparison_subfigures['A'].text(0.5, 1, 'Convergent', horizontalalignment='center', transform=fig_order_compare_axes[2].transAxes, fontsize=font_sizes['subaxis_title'])\n",
    "fig_order_model_comparison_subfigures['A'].text(0.5, 1, 'Divergent', horizontalalignment='center', transform=fig_order_compare_axes[3].transAxes, fontsize=font_sizes['subaxis_title'])\n",
    "plot_genes_unified(fig_order_compare_axes[0], [(0.3, 0.4), (0.6, 0.7)], [fig_order_model_comparison_palette[c] for c in ['fig.bm.tandem_reporter_upstream', 'gray']], base_line_extents=(0.1,0.9), base_offset_inches=0.2)\n",
    "plot_genes_unified(fig_order_compare_axes[1], [(0.3, 0.4), (0.6, 0.7)], [fig_order_model_comparison_palette[c] for c in ['gray', 'fig.bm.tandem_reporter_downstream']], base_line_extents=(0.1,0.9), base_offset_inches=0.2)\n",
    "plot_genes_unified(fig_order_compare_axes[2], [(0.3, 0.4), (0.7, 0.6)], [fig_order_model_comparison_palette[c] for c in ['gray', 'fig.bm.convergent']], base_line_extents=(0.1,0.9), base_offset_inches=0.2)\n",
    "plot_genes_unified(fig_order_compare_axes[3], [(0.4, 0.3), (0.6, 0.7)], [fig_order_model_comparison_palette[c] for c in ['gray', 'fig.bm.divergent']], base_line_extents=(0.1,0.9), base_offset_inches=0.2)\n",
    "\n",
    "fig_order_model_comparison.savefig('test.svg')\n",
    "#del df_supp_bm_order"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig_order_model_comparison.savefig(rd.outfile(rd.rootdir/'writeups'/'figures'/'modeling_paper'/'fig_supp_order_model_comparison.svg', 'fig_base_model'))\n",
    "fig_order_model_comparison.savefig(rd.outfile(rd.rootdir/'writeups'/'figures'/'modeling_paper'/'fig_supp_order_model_comparison.pdf', 'fig_base_model'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Supplemental figure - energy and torque for modified scenarios"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with h5py.File(rd.infile(rd.rootdir/'output'/'energy_perturbation_plots.h5'), 'r') as perturb_h5:\n",
    "    perturbation_df = []\n",
    "    for group in perturb_h5.keys():\n",
    "        perturbation_df.append(pd.concat([\n",
    "            pd.DataFrame({\n",
    "                'sigma': perturb_h5[group]['sigma'],\n",
    "                'alpha': perturb_h5[group]['s2_coeff'][a_idx],\n",
    "                'energy': perturb_h5[group]['initiation_energy'][a_idx,:],\n",
    "                'torque': perturb_h5[group]['torque'][a_idx,:],\n",
    "                'torque_perturb_type': perturb_h5[group].attrs['torque_perturb'].decode(),\n",
    "                'rnap_perturb': perturb_h5[group].attrs['rnap_perturb'].decode()\n",
    "            }) for a_idx in range(len(perturb_h5[group]['s2_coeff']))\n",
    "        ]))\n",
    "    perturbation_df = pd.concat(perturbation_df, ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "perturbation_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "supp_fig_perturbations = plt.figure(figsize=(5.50,5))\n",
    "supp_fig_perturbations_gridspec = gridspec.GridSpec(3, 1, figure=supp_fig_perturbations,\n",
    "    wspace=0.1, hspace=0.15)\n",
    "supp_fig_perturbations_subfigures = {\n",
    "    'A': supp_fig_perturbations.add_subfigure(supp_fig_perturbations_gridspec[0,0]),\n",
    "    'B': supp_fig_perturbations.add_subfigure(supp_fig_perturbations_gridspec[1,0]),\n",
    "}\n",
    "for label, subfig in supp_fig_perturbations_subfigures.items():\n",
    "    subfig.add_artist(matplotlib.text.Text(x=0, y=1, text=f'{label}.', fontsize=font_sizes['subpanel_label'],\n",
    "        fontweight='bold', verticalalignment='top',transform=subfig.transSubfigure))\n",
    "\n",
    "supp_fig_perturbations_subfigures['A'].subplots_adjust(left=0.1, bottom=0.3, right=0.95, top=0.90)\n",
    "torque_plots = supp_fig_perturbations_subfigures['A'].subplots(nrows=1, ncols=2, sharex=True, sharey=True, gridspec_kw={'hspace':0.4, 'wspace':0.4})\n",
    "supp_fig_perturbations_subfigures['B'].subplots_adjust(left=0.1, bottom=0.3, right=0.95, top=0.90)\n",
    "energy_plots = supp_fig_perturbations_subfigures['B'].subplots(nrows=3, ncols=len(alpha_values), sharex=True, sharey=True, gridspec_kw={'hspace':0.4, 'wspace':0.4})\n",
    "\n",
    "# Draw axes lines\n",
    "for ax in itertools.chain(torque_plots.flatten(), energy_plots.flatten()):\n",
    "    ax.axhline(0, color=main_palette['gray'])\n",
    "    ax.axvline(0, color=main_palette['gray'])\n",
    "    sns.despine(ax=ax)\n",
    "\n",
    "torque_df = perturbation_df[\n",
    "    (perturbation_df.alpha == 0.0) & (perturbation_df.torque_perturb_type=='original') & (perturbation_df.rnap_perturb=='original')\n",
    "]\n",
    "torque_plots[0].plot(torque_df.sigma, torque_df.torque, color='k')\n",
    "torque_df = perturbation_df[\n",
    "    (perturbation_df.alpha == 0.0) & (perturbation_df.torque_perturb_type=='buffering') & (perturbation_df.rnap_perturb=='original')\n",
    "]\n",
    "torque_plots[1].axvspan(0,0.031, color=main_palette['light_gray'])\n",
    "torque_plots[1].plot(torque_df.sigma, torque_df.torque, color='k')\n",
    "\n",
    "for idx, alpha in enumerate(alpha_values):\n",
    "    alpha_df = perturbation_df[\n",
    "        np.abs(perturbation_df.alpha - alpha) == min(np.abs(perturbation_df.alpha - alpha))\n",
    "    ]\n",
    "    original_df = alpha_df[(alpha_df.torque_perturb_type=='original')&(alpha_df.rnap_perturb=='original')]\n",
    "    buffer_df = alpha_df[(alpha_df.torque_perturb_type=='buffering')&(alpha_df.rnap_perturb=='original')]\n",
    "    well_df = alpha_df[(alpha_df.torque_perturb_type=='original')&(alpha_df.rnap_perturb=='energy_well')]\n",
    "\n",
    "    \n",
    "    energy_plots[0,idx].plot(original_df.sigma, original_df.energy, color='k')\n",
    "    energy_plots[1,idx].plot(buffer_df.sigma, buffer_df.energy, color='k')\n",
    "    energy_plots[2,idx].plot(well_df.sigma, well_df.energy, color='k')\n",
    "\n",
    "for t_ax in torque_plots:\n",
    "    t_ax.set_ylim(-30,30)\n",
    "for e_ax in energy_plots.flatten():\n",
    "    e_ax.set_ylim(-275,275)\n",
    "#for t_ax in itertools.chain(original_plots[0,1:], buffering_plots[0,1:], energy_well_plots[0,1:]):\n",
    "#    t_ax.set_yticklabels([])\n",
    "#for e_ax in itertools.chain(original_plots[1,1:], buffering_plots[1,1:], energy_well_plots[1,1:]):\n",
    "#    e_ax.set_yticklabels([])\n",
    "supp_fig_perturbations.savefig('test.svg', bbox_inches='tight')\n",
    "plt.show(supp_fig_perturbations)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Supplemental figure - hyperparameter sweep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if (rd.datadir/'projects'/'tangles'/'simulations'/'fig_supp_hyperparams.gzip').exists():\n",
    "    hyperparameters_df = pd.read_parquet(\n",
    "        rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/'fig_supp_hyperparams.gzip', 'fig_supp_hyperparams')\n",
    "    )\n",
    "    print(f'Loaded {len(hyperparameters_df)} total simulation runs from Parquet cached file')\n",
    "else:\n",
    "    hyperparameters_df = []\n",
    "    with h5py.File(\n",
    "        rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/'modeling_hyperparameters_summaries.h5', 'fig_supp_hyperparams_cache'), 'r') as summaries:\n",
    "        for group in summaries.keys():\n",
    "            group_df = []\n",
    "            for run in summaries[group].values():\n",
    "                group_df.append(pd.DataFrame(data={\n",
    "                    'inducible_mRNA': run['final_mRNA'][0,:],\n",
    "                    'constant_mRNA': run['final_mRNA'][1,:]}))\n",
    "                group_df[-1]['induction'] = run.attrs['gene.base_rate'][0]\n",
    "                group_df[-1]['condition'] = run.attrs['comment'].decode('utf-8')\n",
    "                group_df[-1]['is_plasmid'] = run.attrs['bcs.is_circular']\n",
    "                for term in [\n",
    "                    'rates.sc_dependent', 'coeff.sigma_squared',\n",
    "                    'coeff.mRNA_drag_exponent', 'coeff.mRNA_drag_coeff',\n",
    "                    'rnap.max_velocity', 'rnap.stall_torque', 'rnap.stall_width']:\n",
    "                    group_df[-1][term] = run.attrs[term]\n",
    "                gene_endpoints = sorted(np.concatenate((\n",
    "                    run.attrs['gene.start'],\n",
    "                    run.attrs['gene.end'])))\n",
    "                group_df[-1]['spacing'] = int((gene_endpoints[2] - gene_endpoints[1]) / 0.34)\n",
    "            hyperparameters_df.append(pd.concat(group_df, ignore_index=True))\n",
    "            print(f'Loaded {len(hyperparameters_df[-1])} simulations...')\n",
    "    hyperparameters_df = pd.concat(hyperparameters_df, ignore_index=True)\n",
    "    print(f'Loaded {len(hyperparameters_df)} total simulation runs!')\n",
    "    hyperparameters_df['condition'] = pd.Categorical(hyperparameters_df['condition'])\n",
    "    hyperparameters_df.to_parquet(rd.outfile(rd.datadir/'projects'/'tangles'/'simulations'/'fig_supp_hyperparams.gzip', 'fig_supp_hyperparams_cache'), compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig_supp_hyperparam = plt.figure(figsize=(4.50,5))\n",
    "fig_supp_hyperparam_gridspec = gridspec.GridSpec(4, 1, figure=fig_supp_hyperparam,\n",
    "    wspace=0.1, hspace=0.15)\n",
    "fig_supp_hyperparam_subfigures = {\n",
    "    'A': fig_supp_hyperparam.add_subfigure(fig_supp_hyperparam_gridspec[0,:]),\n",
    "    'B': fig_supp_hyperparam.add_subfigure(fig_supp_hyperparam_gridspec[1,:]),\n",
    "    'C': fig_supp_hyperparam.add_subfigure(fig_supp_hyperparam_gridspec[2,:]),\n",
    "    'D': fig_supp_hyperparam.add_subfigure(fig_supp_hyperparam_gridspec[3,:]),\n",
    "}\n",
    "for label, subfig in fig_supp_hyperparam_subfigures.items():\n",
    "    subfig.add_artist(matplotlib.text.Text(x=0, y=1, text=f'{label}.', fontsize=font_sizes['subpanel_label'], fontweight='bold', verticalalignment='top',transform=subfig.transSubfigure))\n",
    "    subfig.subplots_adjust(left=0.1, bottom=0.3, right=0.95, top=0.7)\n",
    "fig_supp_hyperparam_palette = {\n",
    "    'hyperparams.tandem_reporter_upstream': main_palette['tandem'],\n",
    "    'hyperparams.tandem_reporter_downstream': main_palette['tandem_alt'],\n",
    "    'hyperparams.convergent': main_palette['convergent'],\n",
    "    'hyperparams.divergent': main_palette['divergent'],\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def hyper_extract_df(base_df, drag_coeff=1/20, drag_exp=1, stall_torque=12, stall_width=3):\n",
    "    result = base_df[\n",
    "        (base_df['coeff.mRNA_drag_coeff'] == drag_coeff) &\n",
    "        (base_df['coeff.mRNA_drag_exponent'] == drag_exp) &\n",
    "        (base_df['rnap.stall_torque'] == stall_torque) &\n",
    "        (base_df['rnap.stall_width'] == stall_width)\n",
    "    ].copy()\n",
    "    result['norm_induction'] = result['induction'] / (1.0 / 120.0)\n",
    "    result['constant_mRNA'] = result['constant_mRNA'] / 250.0\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "drag_coeffs = [1/50, 1/20, 1/10, 1/5]\n",
    "drag_exp = [1.0, 2.0, 3.0]\n",
    "stall_torque = [8, 10, 12, 14, 16]\n",
    "stall_width = [1,3,5]\n",
    "fig_hyper_dcoeff_axes = fig_supp_hyperparam_subfigures['A'].subplots(ncols=len(drag_coeffs), sharex=True, sharey=True)\n",
    "fig_hyper_dexp_axes = fig_supp_hyperparam_subfigures['B'].subplots(ncols=len(drag_exp), sharex=True, sharey=True)\n",
    "fig_hyper_storque_axes = fig_supp_hyperparam_subfigures['C'].subplots(ncols=len(stall_torque), sharex=True, sharey=True)\n",
    "fig_hyper_swidth_axes = fig_supp_hyperparam_subfigures['D'].subplots(ncols=len(stall_width), sharex=True, sharey=True)\n",
    "for ax, dcoeff in zip(fig_hyper_dcoeff_axes,drag_coeffs):\n",
    "    sns.lineplot(data=hyper_extract_df(hyperparameters_df, drag_coeff=dcoeff),\n",
    "            x='norm_induction', y='constant_mRNA', hue='condition', palette=fig_supp_hyperparam_palette,\n",
    "            ax=ax, legend=None)\n",
    "    sns.despine(ax=ax)\n",
    "    ax.set_xlabel('')\n",
    "    ax.text(0.5, 1.0, f'η={dcoeff} (pN nm^(β - 1))', ha='center', fontsize=font_sizes['data_annotation'],\n",
    "            transform=ax.transAxes)\n",
    "fig_supp_hyperparam_subfigures['A'].text(0.53, 0.85, \"mRNA drag coefficent\", ha='center', fontsize=font_sizes['subaxis_title'])\n",
    "for ax, dexp in zip(fig_hyper_dexp_axes,drag_exp):\n",
    "    sns.lineplot(data=hyper_extract_df(hyperparameters_df, drag_exp=dexp),\n",
    "            x='norm_induction', y='constant_mRNA', hue='condition', palette=fig_supp_hyperparam_palette,\n",
    "            ax=ax, legend=None)\n",
    "    sns.despine(ax=ax)\n",
    "    ax.set_xlabel('')\n",
    "    ax.text(0.5, 1.0, f'β={dexp}', ha='center', transform=ax.transAxes)\n",
    "fig_supp_hyperparam_subfigures['B'].text(0.53, 0.85, \"mRNA drag exponent\", ha='center', fontsize=font_sizes['subaxis_title'])\n",
    "for ax, storque in zip(fig_hyper_storque_axes,stall_torque):\n",
    "    sns.lineplot(data=hyper_extract_df(hyperparameters_df, stall_torque=storque),\n",
    "            x='norm_induction', y='constant_mRNA', hue='condition', palette=fig_supp_hyperparam_palette,\n",
    "            ax=ax, legend=None)\n",
    "    sns.despine(ax=ax)\n",
    "    ax.set_xlabel('')\n",
    "    ax.text(0.5, 1.0, f'τ_s={storque} (pN nm)', ha='center',\n",
    "            fontsize=font_sizes['data_annotation'], transform=ax.transAxes)\n",
    "fig_supp_hyperparam_subfigures['C'].text(0.53, 0.85, \"Stall torque\", ha='center', fontsize=font_sizes['subaxis_title'])\n",
    "for ax, swidth in zip(fig_hyper_swidth_axes,stall_width):\n",
    "    sns.lineplot(data=hyper_extract_df(hyperparameters_df, stall_width=swidth),\n",
    "            x='norm_induction', y='constant_mRNA', hue='condition', palette=fig_supp_hyperparam_palette,\n",
    "            ax=ax, legend=None)\n",
    "    sns.despine(ax=ax)\n",
    "    ax.set_xlabel('')\n",
    "    ax.text(0.5, 1.0, f'τ_w={swidth} (pN nm)', ha='center',\n",
    "            fontsize=font_sizes['data_annotation'], transform=ax.transAxes)\n",
    "fig_supp_hyperparam_subfigures['D'].text(0.53, 0.85, \"Stall width\", ha='center', fontsize=font_sizes['subaxis_title'])\n",
    "fig_hyper_dcoeff_axes[0].set_ylabel('Reporter output')\n",
    "fig_hyper_dexp_axes[0].set_ylabel('Reporter output')\n",
    "fig_hyper_storque_axes[0].set_ylabel('Reporter output')\n",
    "fig_hyper_swidth_axes[0].set_ylabel('Reporter output')\n",
    "fig_supp_hyperparam_subfigures['A'].text(0.5, 0.06, 'Fold induction', ha='center')\n",
    "fig_supp_hyperparam_subfigures['B'].text(0.5, 0.06, 'Fold induction', ha='center')\n",
    "fig_supp_hyperparam_subfigures['C'].text(0.5, 0.06, 'Fold induction', ha='center')\n",
    "fig_supp_hyperparam_subfigures['D'].text(0.5, 0.06, 'Fold induction', ha='center')\n",
    "fig_supp_hyperparam.savefig('test.svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig_supp_hyperparam.savefig(rd.outfile(rd.rootdir/'writeups'/'figures'/'modeling_paper'/'fig_supp_hyperparam.svg', 'fig_supp_hyperparams'))\n",
    "fig_supp_hyperparam.savefig(rd.outfile(rd.rootdir/'writeups'/'figures'/'modeling_paper'/'fig_supp_hyperparam.pdf', 'fig_supp_hyperparams'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Supplemental figure - modifying burst threshold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create the overall figure, gridspec, and add subfigure labels.\n",
    "fig_supp_burst_threshold = plt.figure(figsize=(7.5,3.5))\n",
    "fig_supp_burst_threshold_gridspec = gridspec.GridSpec(2, 1, figure=fig_supp_burst_threshold,\n",
    "    wspace=0.1, hspace=0.1, height_ratios=[1.0, 1.0])\n",
    "fig_supp_burst_threshold_subfigures = {\n",
    "    'A': fig_supp_burst_threshold.add_subfigure(fig_supp_burst_threshold_gridspec[0,:]),\n",
    "    'B': fig_supp_burst_threshold.add_subfigure(fig_supp_burst_threshold_gridspec[1,:]),\n",
    "}\n",
    "for label, subfig in fig_supp_burst_threshold_subfigures.items():\n",
    "    subfig.add_artist(matplotlib.text.Text(x=0, y=1, text=f'{label}.', fontsize=font_sizes['subpanel_label'], fontweight='bold', verticalalignment='top',transform=subfig.transSubfigure))\n",
    "fig_supp_burst_threshold_palette = {\n",
    "    'fig.bm.sc_bursts.tandem_reporter_upstream': main_palette['tandem'],\n",
    "    'fig.bm.sc_bursts.tandem_reporter_downstream': main_palette['tandem_alt'],\n",
    "    'fig.bm.sc_bursts.convergent': main_palette['convergent'],\n",
    "    'fig.bm.sc_bursts.divergent': main_palette['divergent'],\n",
    "    'gray': main_palette['gray'],\n",
    "    'background_gray': '#DDDDDD'\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot overall burst duration as a function of induction and condition\n",
    "fig_supp_burst_threshold_subfigures['A'].subplots_adjust(left=0.08, bottom=0.29, right=0.95, top=0.9)\n",
    "fig_burst_threshold_size_axes = fig_supp_burst_threshold_subfigures['A'].subplots(ncols=4, sharey=True, sharex=True)\n",
    "bm_supp_bursting_df = bursting_df_selected[\n",
    "    (bursting_df_selected.inter_rnap_gene == 2) &\n",
    "    (bursting_df_selected.burst_size > 0) &\n",
    "    (bursting_df_selected.burst_duration > 0) &\n",
    "    (bursting_df_selected.inter_burst_time > 0) &\n",
    "    (\n",
    "        ((bursting_df_selected.condition) == conditions[0]) |\n",
    "        ((bursting_df_selected.condition) == conditions[1]) |\n",
    "        ((bursting_df_selected.condition) == conditions[2]) |\n",
    "        ((bursting_df_selected.condition) == conditions[3]) \n",
    "    ) & (\n",
    "        bursting_df_selected.induction == selected_induction[1]\n",
    "    )\n",
    "].copy()\n",
    "bm_supp_bursting_df.condition = bm_supp_bursting_df.condition.cat.remove_unused_categories()\n",
    "for ax, threshold in zip(fig_burst_threshold_size_axes, [10,20,30,60]):\n",
    "    sns.kdeplot(data=bm_supp_bursting_df[bm_supp_bursting_df.burst_threshold == threshold], x='burst_size', hue='condition',\n",
    "        common_norm=False,\n",
    "        palette=fig_supp_burst_threshold_palette, ax=ax, legend=None, log_scale=True)\n",
    "    ax.set_xlim([1, 3e3])\n",
    "    ax.set_xlabel('')\n",
    "    ax.text(0.5, 0.98, f'{threshold}s threshold', ha='center', fontsize=font_sizes['subaxis_title'], transform=ax.transAxes)\n",
    "    sns.despine(ax=ax)\n",
    "fig_supp_burst_threshold_subfigures['A'].text(0.5, 0.06, 'Burst size', ha='center')\n",
    "fig_supp_burst_threshold.savefig('test.svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot inter-burst time as a function of induction\n",
    "fig_supp_burst_threshold_subfigures['B'].subplots_adjust(left=0.08, bottom=0.29, right=0.95, top=0.9)\n",
    "fig_burst_threshold_interburst_axes = fig_supp_burst_threshold_subfigures['B'].subplots(ncols=4, sharey=True, sharex=True)\n",
    "for ax, threshold in zip(fig_burst_threshold_interburst_axes, [10,20,30,60]):\n",
    "    sns.kdeplot(data=bm_supp_bursting_df[bm_supp_bursting_df.burst_threshold == threshold], x='inter_burst_time', hue='condition',\n",
    "    common_norm=False,\n",
    "    palette=fig_supp_burst_threshold_palette, ax=ax, legend=None, log_scale=True)\n",
    "    #ax.set_xticklabels(['0x', '1x', '2x'])\n",
    "    ax.set_xlim([1,5e3])\n",
    "    ax.set_xlabel('')\n",
    "    ax.text(0.5, 0.98, f'{threshold}s threshold', ha='center', fontsize=font_sizes['subaxis_title'], transform=ax.transAxes)\n",
    "    sns.despine(ax=ax)\n",
    "fig_supp_burst_threshold_subfigures['B'].text(0.5, 0.06, 'Inter-burst time (s)', ha='center')\n",
    "fig_supp_burst_threshold.savefig('test.svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig_supp_burst_threshold.savefig(rd.outfile(rd.rootdir/'writeups'/'figures'/'modeling_paper'/'fig_supp_burst_threshold.svg', 'fig_sc_bursting'))\n",
    "fig_supp_burst_threshold.savefig(rd.outfile(rd.rootdir/'writeups'/'figures'/'modeling_paper'/'fig_supp_burst_threshold.pdf', 'fig_sc_bursting'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Supplemental figure - sc density as a function of induction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot overall sc density as a function of induction\n",
    "fig_supp_sc_density = plt.figure(figsize=(6.5,4.5))\n",
    "fig_supp_sc_density.subplots_adjust(left=0.1, bottom=0.12, right=0.85, top=0.92)\n",
    "induction_levels = [0.0, 0.1, 0.5, 1.0, 1.5, 2.0]\n",
    "fig_supp_sc_density_axes = fig_supp_sc_density.subplots(nrows=len(induction_levels), ncols=4, sharey=True, sharex=True, gridspec_kw={'hspace': 0.3})\n",
    "conditions = [f'fig.bm.sc_bursts.{x}' for x in ['tandem_reporter_upstream', 'tandem_reporter_downstream', 'convergent', 'divergent']]\n",
    "for cond_idx, induction_idx in itertools.product(range(len(conditions)), range(len(induction_levels))):\n",
    "    ax = fig_supp_sc_density_axes[induction_idx, cond_idx]\n",
    "    cond = conditions[cond_idx]\n",
    "    induction = induction_levels[induction_idx]\n",
    "    sns.lineplot(data=sc_density_df[\n",
    "        (sc_density_df.condition == cond) &\n",
    "        (sc_density_df.norm_induction == induction) &\n",
    "        (sc_density_df.alpha == 0.025) &\n",
    "        (np.abs(sc_density_df.time - 15000) < 1)\n",
    "    ], x='distance', y='sc_density',\n",
    "       color=fig_bm_palette[cond],\n",
    "       ax=ax, legend=None)\n",
    "    ax.axhline(y=0, color=main_palette['axis_gray'])\n",
    "    ax.set_xlim([0,3400])\n",
    "    ax.set_yticks([-0.2,0.2])\n",
    "    ax.set_xticks([x * 0.34 for x in [0,5000,10000]])\n",
    "    ax.set_xticklabels(['0', '5', '10'])\n",
    "    ax.set_xlabel('')\n",
    "    sns.despine(ax=ax)\n",
    "    if induction_idx == 0:\n",
    "        if cond.endswith('tandem_reporter_upstream'):\n",
    "            plot_genes_unified(ax, [(3000 * 0.34, 4000 * 0.34), (6000 * 0.34, 7000 * 0.34)], [fig_bm_palette[cond], fig_bm_palette['gray']], plot_gene_body=True, units='xaxis', plot_inducible=1)\n",
    "        if cond.endswith('tandem_reporter_downstream'):\n",
    "            plot_genes_unified(ax, [(3000 * 0.34, 4000 * 0.34), (6000 * 0.34, 7000 * 0.34)], [fig_bm_palette['gray'], fig_bm_palette[cond]], plot_gene_body=True, units='xaxis', plot_inducible=0)\n",
    "        if cond.endswith('convergent'):\n",
    "            plot_genes_unified(ax, [(3000 * 0.34, 4000 * 0.34), (7000 * 0.34, 6000 * 0.34)], [fig_bm_palette['gray'], fig_bm_palette[cond]], plot_gene_body=True, units='xaxis', plot_inducible=0)\n",
    "        if cond.endswith('divergent'):\n",
    "            plot_genes_unified(ax, [(4000 * 0.34, 3000 * 0.34), (6000 * 0.34, 7000 * 0.34)], [fig_bm_palette['gray'], fig_bm_palette[cond]], plot_gene_body=True, units='xaxis', plot_inducible=0)\n",
    "for ax, induction in zip(fig_supp_sc_density_axes[:,-1], induction_levels):\n",
    "    ax.text(1.05, 0.5, f'{induction}x induction', va='center', transform=ax.transAxes, fontsize=font_sizes['data_annotation'])\n",
    "for ax in fig_supp_sc_density_axes[:,0]:\n",
    "    ax.set_ylabel(\"σ\")\n",
    "    ax.set_ylabel(\"σ\")\n",
    "fig_supp_sc_density.text(0.5, 0.03, 'Genomic distance (kb)', ha='center')\n",
    "fig_supp_sc_density.savefig('test.svg')\n",
    "fig_supp_sc_density.savefig(rd.outfile(rd.rootdir/'writeups'/'figures'/'modeling_paper'/'fig_supp_sc_density_induction.svg', 'fig_sc_bursting'))\n",
    "fig_supp_sc_density.savefig(rd.outfile(rd.rootdir/'writeups'/'figures'/'modeling_paper'/'fig_supp_sc_density_induction.pdf', 'fig_sc_bursting'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Single-cell behavior figure "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if (rd.datadir/'projects'/'tangles'/'simulations'/'fig_single_cell_behavior.gzip').exists():\n",
    "    fig_sc_behavior_df = pd.read_parquet(rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/'fig_single_cell_behavior.gzip', 'fig_single_cell_behavior'), memory_map=True)\n",
    "    print(f'Loaded {len(fig_sc_behavior_df)} total datapoints from Parquet cached file')\n",
    "else:\n",
    "    with h5py.File(\n",
    "        rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/'modeling_bm_examples_combined.h5', 'fig_single_cell_behavior_cache'), 'r') as fig2_raw:\n",
    "        fig_sc_behavior_df = []\n",
    "        i = 0\n",
    "        for group in fig2_raw.keys():\n",
    "            group_df = []\n",
    "            for run in fig2_raw[group].values():\n",
    "                run_shape = run['mRNA'].shape\n",
    "                for subrun in range(run_shape[2]):\n",
    "                    # downsample\n",
    "                    if subrun % 3 != 0:\n",
    "                        continue\n",
    "                    inducible = run['mRNA'][:,0,:]\n",
    "                    constant  = run['mRNA'][:,1,:]\n",
    "                    group_df.append(pd.DataFrame(data={\n",
    "                        'inducible_mRNA': inducible[:,subrun],\n",
    "                        'constant_mRNA':  constant[:,subrun],\n",
    "                        'time': run['time']}))\n",
    "                    group_df[-1]['induction'] = run.attrs['step_induction']\n",
    "                    group_df[-1]['condition'] = run.attrs['comment'].decode('utf-8')\n",
    "                    group_df[-1]['idx'] = i\n",
    "                    i += 1\n",
    "                    gene_endpoints = sorted(np.concatenate((\n",
    "                        run.attrs['gene.start'],\n",
    "                        run.attrs['gene.end'])))\n",
    "                    group_df[-1]['spacing'] = int((gene_endpoints[2] - gene_endpoints[1]) / 0.34)\n",
    "                    for term in [\n",
    "                        'step_time', 'coeff.sigma_squared',]:\n",
    "                        group_df[-1][term] = run.attrs[term]\n",
    "            fig_sc_behavior_df.append(pd.concat(group_df, ignore_index=True))\n",
    "            if len(fig_sc_behavior_df) > 10:\n",
    "                fig_sc_behavior_df = [pd.concat(fig_sc_behavior_df, ignore_index=True)]\n",
    "            print(f'Loaded {len(fig_sc_behavior_df[-1])} datapoints...')\n",
    "        fig_sc_behavior_df = pd.concat(fig_sc_behavior_df, ignore_index=True)\n",
    "        print(f'Loaded {len(fig_sc_behavior_df)} total datapoints!')\n",
    "        fig_sc_behavior_df.to_parquet(rd.outfile(rd.datadir/'projects'/'tangles'/'simulations'/'fig_single_cell_behavior.gzip', 'fig_single_cell_behavior_cache'), compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Prep fig_sc_behavior\n",
    "# For fig_sc_behavior, find three single runs (tandem, convergent, divergent) at maximum induction\n",
    "run_ids = {condition: fig_sc_behavior_df[\n",
    "    (fig_sc_behavior_df['condition']==condition)\n",
    "    & (np.abs(fig_sc_behavior_df['induction'] - 1.0) < 0.01)\n",
    "    & (fig_sc_behavior_df['coeff.sigma_squared'] == 0.025)]['idx'].iloc[0] for condition in fig_sc_behavior_df['condition'].unique()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def df_crosscorr(df):\n",
    "    constant_normed  = (df['constant_mRNA'] - df['constant_mRNA'].mean())\n",
    "    inducible_normed = (df['inducible_mRNA'] - df['inducible_mRNA'].mean())\n",
    "    geom_norm = np.sqrt(\n",
    "        scipy.signal.correlate(constant_normed, constant_normed, mode='valid') *\n",
    "        scipy.signal.correlate(inducible_normed, inducible_normed, mode='valid'))\n",
    "    geom_norm = geom_norm if geom_norm > 1e-10 else 1.0\n",
    "    return pd.DataFrame({\n",
    "        'lags':scipy.signal.correlation_lags(len(constant_normed), len(inducible_normed)),\n",
    "        'cross_corr':scipy.signal.correlate(inducible_normed, constant_normed) / geom_norm\n",
    "    })\n",
    "fig_sc_behavior_convolutions = fig_sc_behavior_df[\n",
    "    (fig_sc_behavior_df.time > 15000)\n",
    "    & (\n",
    "        (np.abs(fig_sc_behavior_df.induction - 0.464) < 0.01) |\n",
    "        (np.abs(fig_sc_behavior_df.induction - 1.0) < 0.01) |\n",
    "        (np.abs(fig_sc_behavior_df.induction - 2.61) < 0.01)\n",
    "    )\n",
    "    & (fig_sc_behavior_df['coeff.sigma_squared'] == 0.025)].groupby([\n",
    "        'condition', 'coeff.sigma_squared', 'induction', 'idx'\n",
    "    ]).apply(df_crosscorr).reset_index().groupby([\n",
    "        'condition', 'induction', 'lags']).mean().reset_index()\n",
    "time_delta = np.diff(fig_sc_behavior_df['time'].unique())[0]\n",
    "fig_sc_behavior_convolutions['time'] = fig_sc_behavior_convolutions['lags'] * time_delta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create the overall figure, gridspec, and add subfigure labels.\n",
    "fig_sc_behavior = plt.figure(figsize=(6.0,8.0))\n",
    "fig_sc_behavior_gridspec = gridspec.GridSpec(4, 6, figure=fig_sc_behavior,\n",
    "    wspace=0.1, hspace=0.1, height_ratios=[1.1, 1.0, 1.0, 1.0])\n",
    "fig_sc_behavior_subfigures = {\n",
    "    'A': fig_sc_behavior.add_subfigure(fig_sc_behavior_gridspec[0,:4]),\n",
    "    'B': fig_sc_behavior.add_subfigure(fig_sc_behavior_gridspec[1,:4]),\n",
    "    'C': fig_sc_behavior.add_subfigure(fig_sc_behavior_gridspec[0,4:]),\n",
    "    'D': fig_sc_behavior.add_subfigure(fig_sc_behavior_gridspec[1,4:]),\n",
    "    'E': fig_sc_behavior.add_subfigure(fig_sc_behavior_gridspec[2,:4]),\n",
    "    'F': fig_sc_behavior.add_subfigure(fig_sc_behavior_gridspec[3,:3]),\n",
    "    'G': fig_sc_behavior.add_subfigure(fig_sc_behavior_gridspec[2,3:]),\n",
    "    'H': fig_sc_behavior.add_subfigure(fig_sc_behavior_gridspec[3,3:])\n",
    "}\n",
    "for label, subfig in fig_sc_behavior_subfigures.items():\n",
    "    subfig.add_artist(matplotlib.text.Text(x=0, y=1, text=f'{label}.', fontsize=font_sizes['subpanel_label'], fontweight='bold', verticalalignment='top',transform=subfig.transSubfigure))\n",
    "fig_sc_behavior_palette = {\n",
    "    'fig.bm_examples.sc_bursts.tandem_reporter_upstream': main_palette['tandem'],\n",
    "    'fig.bm_examples.sc_bursts.tandem_reporter_downstream': main_palette['tandem_alt'],\n",
    "    'fig.bm_examples.sc_bursts.convergent': main_palette['convergent'],\n",
    "    'fig.bm_examples.sc_bursts.divergent': main_palette['divergent'],\n",
    "    'fig.bm_examples.tandem_reporter_upstream': main_palette['tandem'],\n",
    "    'fig.bm_examples.tandem_reporter_downstream': main_palette['tandem_alt'],\n",
    "    'fig.bm_examples.convergent': main_palette['convergent'],\n",
    "    'fig.bm_examples.divergent': main_palette['divergent'],\n",
    "    'fig.bm.sc_bursts.tandem_reporter_upstream': main_palette['tandem'],\n",
    "    'fig.bm.sc_bursts.tandem_reporter_downstream': main_palette['tandem_alt'],\n",
    "    'fig.bm.sc_bursts.convergent': main_palette['convergent'],\n",
    "    'fig.bm.sc_bursts.divergent': main_palette['divergent'],\n",
    "    'gray': main_palette['gray'],\n",
    "    'background_gray': '#DDDDDD',\n",
    "    'black': main_palette['black'],\n",
    "    'reporter_only': main_palette['reporter_only']\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot single-cell example behavior and SC density\n",
    "fig_sc_behavior_subfigures['A'].subplots_adjust(left=0.15, bottom=0.17, right=0.9, top=0.85)\n",
    "fig_scb_example_axes = fig_sc_behavior_subfigures['A'].subplots(ncols=2, nrows=2, sharey=True, sharex=True, gridspec_kw={'hspace':0.97}).flatten()\n",
    "for ax, (cond, id) in zip(np.concatenate((fig_scb_example_axes[2:], fig_scb_example_axes[:2])), run_ids.items()):\n",
    "    run = fig_sc_behavior_df[fig_sc_behavior_df.idx == id].copy()\n",
    "    run['normed_constant'] = run['constant_mRNA'] / 300\n",
    "    run['normed_inducible']= run['inducible_mRNA'] / 300\n",
    "    ax.axvspan(10000, max(run['time']), color=fig_sc_behavior_palette['background_gray'])\n",
    "    sns.lineplot(data=run, x='time', y='normed_constant', color=fig_sc_behavior_palette[cond], linewidth=1, legend=False, ax=ax)\n",
    "    sns.lineplot(data=run, x='time', y='normed_inducible', color=fig_sc_behavior_palette['gray'], linewidth=1, legend=False, ax=ax)\n",
    "    sns.despine(ax=ax)\n",
    "    ax.set_xlabel('')\n",
    "    ax.set_ylabel('')\n",
    "    ax.set_xlim([0,40000])\n",
    "    ax.set_xticks([0, 10000, 3600*5, 3600*10])\n",
    "    ax.xaxis.set_major_formatter(hours_1f_formatter)\n",
    "    ax.text(10000, 1.2, 'Induced', fontsize=font_sizes['data_annotation'])\n",
    "    ax.set_ylim([-0.02,1.4])\n",
    "    if cond.endswith('tandem_reporter_upstream'):\n",
    "        plot_genes_unified(ax, [(0.3, 0.4), (0.6, 0.7)], [fig_sc_behavior_palette[cond], fig_sc_behavior_palette['gray']], plot_inducible=1)\n",
    "    if cond.endswith('tandem_reporter_downstream'):\n",
    "        plot_genes_unified(ax, [(0.3, 0.4), (0.6, 0.7)], [fig_sc_behavior_palette['gray'], fig_sc_behavior_palette[cond]], plot_inducible=0)\n",
    "    if cond.endswith('convergent'):\n",
    "        plot_genes_unified(ax, [(0.3, 0.4), (0.7, 0.6)], [fig_sc_behavior_palette['gray'], fig_sc_behavior_palette[cond]], plot_inducible=0)\n",
    "    if cond.endswith('divergent'):\n",
    "        plot_genes_unified(ax, [(0.4, 0.3), (0.6, 0.7)], [fig_sc_behavior_palette['gray'], fig_sc_behavior_palette[cond]], plot_inducible=0)\n",
    "    ax.yaxis.labelpad=12\n",
    "fig_scb_example_axes[0].set_ylabel('Gene output')\n",
    "fig_scb_example_axes[2].set_ylabel('Gene output')\n",
    "fig_sc_behavior_subfigures['A'].text(0.5, 0.03, \"Time (hr)\", horizontalalignment='center')\n",
    "\n",
    "fig_sc_behavior.savefig('test.svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot sc-density behavior-cell example behavior\n",
    "fig_sc_behavior_subfigures['B'].subplots_adjust(left=0.15, bottom=0.19, right=0.9, top=0.8)\n",
    "fig_sc_density_axes = fig_sc_behavior_subfigures['B'].subplots(nrows=2, ncols=2, sharey=True, gridspec_kw={'hspace':0.97}).flatten()\n",
    "conditions = [f'fig.bm_examples.sc_bursts.{x}' for x in ['convergent', 'divergent', 'tandem_reporter_upstream', 'tandem_reporter_downstream']]\n",
    "times = [8421, 40000.0]\n",
    "for t_idx, cond_idx in itertools.product(range(len(times)), range(len(conditions))):\n",
    "    ax = fig_sc_density_axes[cond_idx]\n",
    "    cond = conditions[cond_idx]\n",
    "    ax.axhline(y=0, color=main_palette['axis_gray'])\n",
    "    sns.lineplot(data=sc_density_df[\n",
    "        (sc_density_df.condition == cond) &\n",
    "        (sc_density_df.alpha == 0.025) &\n",
    "        (sc_density_df.induction == 1.0) &\n",
    "        (np.abs(sc_density_df.time - times[t_idx]) < 1)\n",
    "    ], x='distance', y='sc_density', ax=ax,\n",
    "    color=fig_sc_behavior_palette[cond if t_idx == 1 else 'reporter_only'],\n",
    "    linestyle='-' if t_idx == 1 else ':')\n",
    "    ax.set_xlim([0,3400])\n",
    "    ax.set_xticks([x * 0.34 for x in [0,5000,10000]])\n",
    "    ax.set_yticks([-0.3, 0, 0.3])\n",
    "    ax.text(100, 0.35, 'Uninduced', color=fig_sc_behavior_palette['reporter_only'], fontsize=font_sizes['data_annotation'])\n",
    "    ax.text(1200, 0.35, 'Induced', color=fig_sc_behavior_palette[cond], fontsize=font_sizes['data_annotation'])\n",
    "    ax.set_xticklabels(['0', '5', '10'])\n",
    "    xax_transform = ax.get_xaxis_transform()\n",
    "    if t_idx == 0:\n",
    "        if cond.endswith('tandem_reporter_upstream'):\n",
    "            plot_genes_unified(ax, [(3000 * 0.34, 4000 * 0.34), (6000 * 0.34, 7000 * 0.34)], [fig_sc_behavior_palette[cond], fig_sc_behavior_palette['gray']], plot_gene_body=True, units='xaxis', plot_inducible=1)\n",
    "        if cond.endswith('tandem_reporter_downstream'):\n",
    "            plot_genes_unified(ax, [(3000 * 0.34, 4000 * 0.34), (6000 * 0.34, 7000 * 0.34)], [fig_sc_behavior_palette['gray'], fig_sc_behavior_palette[cond]], plot_gene_body=True, units='xaxis', plot_inducible=0)\n",
    "        if cond.endswith('convergent'):\n",
    "            plot_genes_unified(ax, [(3000 * 0.34, 4000 * 0.34), (7000 * 0.34, 6000 * 0.34)], [fig_sc_behavior_palette['gray'], fig_sc_behavior_palette[cond]], plot_gene_body=True, units='xaxis', plot_inducible=0)\n",
    "        if cond.endswith('divergent'):\n",
    "            plot_genes_unified(ax, [(4000 * 0.34, 3000 * 0.34), (6000 * 0.34, 7000 * 0.34)], [fig_sc_behavior_palette['gray'], fig_sc_behavior_palette[cond]], plot_gene_body=True, units='xaxis', plot_inducible=0)\n",
    "    sns.despine(ax=ax, bottom=True if t_idx == 0 else False)\n",
    "    ax.set_xlabel('')\n",
    "    ax.set_ylabel('')\n",
    "for ax in fig_sc_density_axes[:2]:\n",
    "    ax.set_xticklabels([])\n",
    "fig_sc_behavior_subfigures['B'].text(0.05, 0.5, \"Supercoiling density\", rotation=90, va='center', horizontalalignment='center')\n",
    "fig_sc_behavior_subfigures['B'].text(0.5, 0.05, \"Genomic distance (kb)\", va='center', horizontalalignment='center')\n",
    "fig_sc_behavior.savefig('test.svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig_sc_behavior_subfigures['C'].subplots_adjust(left=0.25, bottom=0.17, right=0.9, top=0.85)\n",
    "fig_sc_behavior_c_axes = fig_sc_behavior_subfigures['C'].subplots(nrows=2, gridspec_kw={'hspace':0.97})\n",
    "fig_sc_behavior_subfigures['D'].subplots_adjust(left=0.25, bottom=0.19, right=0.9, top=0.8)\n",
    "fig_sc_behavior_d_axes = fig_sc_behavior_subfigures['D'].subplots(nrows=2, gridspec_kw={'hspace':0.97})\n",
    "fig_sc_behavior_cd_axes = np.concatenate((fig_sc_behavior_d_axes, fig_sc_behavior_c_axes))\n",
    "fig_burst_size_axes = fig_sc_behavior_cd_axes[2:]\n",
    "fig_interburst_time_axes = fig_sc_behavior_cd_axes[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot overall burst duration as a function of induction and condition\n",
    "conditions = [f'fig.bm.sc_bursts.{x}' for x in ['tandem_reporter_upstream', 'tandem_reporter_downstream', 'convergent', 'divergent']]\n",
    "bm_bursting_df = bursting_df_selected[\n",
    "    (bursting_df_selected.burst_threshold == 30) &\n",
    "    (bursting_df_selected.inter_rnap_gene == 2) &\n",
    "    (bursting_df_selected.burst_size > 0) &\n",
    "    (bursting_df_selected.burst_duration > 0) &\n",
    "    (bursting_df_selected.inter_burst_time > 0) &\n",
    "    (\n",
    "        ((bursting_df_selected.condition) == conditions[0]) |\n",
    "        ((bursting_df_selected.condition) == conditions[1]) |\n",
    "        ((bursting_df_selected.condition) == conditions[2]) |\n",
    "        ((bursting_df_selected.condition) == conditions[3]) \n",
    "    )\n",
    "].copy()\n",
    "bm_bursting_df.condition = bm_bursting_df.condition.cat.remove_unused_categories()\n",
    "sns.kdeplot(data=bm_bursting_df[(bm_bursting_df.condition == 'fig.bm.sc_bursts.tandem_reporter_upstream') & (bm_bursting_df.induction == 0.0)], x='burst_size',\n",
    "    common_norm=False, color=fig_sc_behavior_palette['reporter_only'], linestyle=':',\n",
    "    ax=fig_burst_size_axes[1], legend=None, log_scale=True)\n",
    "sns.kdeplot(data=bm_bursting_df[(bm_bursting_df.condition == 'fig.bm.sc_bursts.tandem_reporter_upstream') & (bm_bursting_df.induction == 2.0 / 120.0)], x='burst_size',\n",
    "    common_norm=False, color=fig_sc_behavior_palette['fig.bm.sc_bursts.tandem_reporter_upstream'],\n",
    "    ax=fig_burst_size_axes[1], legend=None, log_scale=True)\n",
    "sns.kdeplot(data=bm_bursting_df[(bm_bursting_df.condition == 'fig.bm.sc_bursts.tandem_reporter_downstream') & (bm_bursting_df.induction == 2.0 / 120.0)], x='burst_size',\n",
    "    common_norm=False, color=fig_sc_behavior_palette['fig.bm.sc_bursts.tandem_reporter_downstream'],\n",
    "    ax=fig_burst_size_axes[1], legend=None, log_scale=True)\n",
    "sns.kdeplot(data=bm_bursting_df[(bm_bursting_df.condition == 'fig.bm.sc_bursts.convergent') & (bm_bursting_df.induction == 0.0)], x='burst_size',\n",
    "    common_norm=False, color=fig_sc_behavior_palette['reporter_only'], linestyle=':',\n",
    "    ax=fig_burst_size_axes[0], legend=None, log_scale=True)\n",
    "sns.kdeplot(data=bm_bursting_df[(bm_bursting_df.condition == 'fig.bm.sc_bursts.convergent') & (bm_bursting_df.induction == 2.0 / 120.0)], x='burst_size',\n",
    "    common_norm=False, color=fig_sc_behavior_palette['fig.bm.sc_bursts.convergent'],\n",
    "    ax=fig_burst_size_axes[0], legend=None, log_scale=True)\n",
    "sns.kdeplot(data=bm_bursting_df[(bm_bursting_df.condition == 'fig.bm.sc_bursts.divergent') & (bm_bursting_df.induction == 2.0 / 120.0)], x='burst_size',\n",
    "    common_norm=False, color=fig_sc_behavior_palette['fig.bm.sc_bursts.divergent'],\n",
    "    ax=fig_burst_size_axes[0], legend=None, log_scale=True)\n",
    "for ax in fig_burst_size_axes:\n",
    "    ax.set_xlim([1, 3e3])\n",
    "    ax.set_xticks([1,10,100,1000])\n",
    "    ax.set_xlabel('mRNA counts')\n",
    "    ax.xaxis.labelpad = 1\n",
    "    sns.despine(ax=ax)\n",
    "fig_burst_size_axes[0].text(0.5,1.2,'Reporter burst size', ha='center', fontsize=font_sizes['base_size'], transform=fig_burst_size_axes[0].transAxes)\n",
    "fig_burst_size_axes[1].text(8,2,'Reporter only', color=main_palette['reporter_only'], fontsize=font_sizes['line_annotation'])\n",
    "fig_burst_size_axes[1].text(8,1.6,'Upstream Tandem', color=main_palette['tandem'], fontsize=font_sizes['line_annotation'])\n",
    "fig_burst_size_axes[1].text(8,1.2,'Downstream Tandem', color=main_palette['tandem_alt'], fontsize=font_sizes['line_annotation'])\n",
    "fig_burst_size_axes[0].text(100,1.5,'Reporter only', color=main_palette['reporter_only'], fontsize=font_sizes['line_annotation'])\n",
    "fig_burst_size_axes[0].text(100,1.2,'Convergent', color=main_palette['convergent'], fontsize=font_sizes['line_annotation'])\n",
    "fig_burst_size_axes[0].text(100,0.9,'Divergent', color=main_palette['divergent'], fontsize=font_sizes['line_annotation'])\n",
    "fig_sc_behavior.savefig('test.svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot inter-burst time as a function of induction\n",
    "bm_bursting_df = bursting_df_selected[\n",
    "    (bursting_df_selected.burst_threshold == 30) &\n",
    "    (bursting_df_selected.inter_rnap_gene == 2) &\n",
    "    (bursting_df_selected.burst_size > 0) &\n",
    "    (bursting_df_selected.burst_duration > 0) &\n",
    "    (bursting_df_selected.inter_burst_time > 0) &\n",
    "    (\n",
    "        ((bursting_df_selected.condition) == conditions[0]) |\n",
    "        ((bursting_df_selected.condition) == conditions[1]) |\n",
    "        ((bursting_df_selected.condition) == conditions[2]) |\n",
    "        ((bursting_df_selected.condition) == conditions[3]) \n",
    "    )\n",
    "].copy()\n",
    "bm_bursting_df.condition = bm_bursting_df.condition.cat.remove_unused_categories()\n",
    "sns.kdeplot(data=bm_bursting_df[(bm_bursting_df.condition == 'fig.bm.sc_bursts.tandem_reporter_upstream') & (bm_bursting_df.induction == 0.0)], x='inter_burst_time',\n",
    "    common_norm=False, color=fig_sc_behavior_palette['reporter_only'], linestyle=':',\n",
    "    ax=fig_interburst_time_axes[1], legend=None, log_scale=True)\n",
    "sns.kdeplot(data=bm_bursting_df[(bm_bursting_df.condition == 'fig.bm.sc_bursts.tandem_reporter_upstream') & (bm_bursting_df.induction == 2.0 / 120.0)], x='inter_burst_time',\n",
    "    common_norm=False, color=fig_sc_behavior_palette['fig.bm.sc_bursts.tandem_reporter_upstream'],\n",
    "    ax=fig_interburst_time_axes[1], legend=None, log_scale=True)\n",
    "sns.kdeplot(data=bm_bursting_df[(bm_bursting_df.condition == 'fig.bm.sc_bursts.tandem_reporter_downstream') & (bm_bursting_df.induction == 2.0 / 120.0)], x='inter_burst_time',\n",
    "    common_norm=False, color=fig_sc_behavior_palette['fig.bm.sc_bursts.tandem_reporter_downstream'],\n",
    "    ax=fig_interburst_time_axes[1], legend=None, log_scale=True)\n",
    "sns.kdeplot(data=bm_bursting_df[(bm_bursting_df.condition == 'fig.bm.sc_bursts.convergent') & (bm_bursting_df.induction == 0.0)], x='inter_burst_time',\n",
    "    common_norm=False, color=fig_sc_behavior_palette['reporter_only'], linestyle=':',\n",
    "    ax=fig_interburst_time_axes[0], legend=None, log_scale=True)\n",
    "sns.kdeplot(data=bm_bursting_df[(bm_bursting_df.condition == 'fig.bm.sc_bursts.convergent') & (bm_bursting_df.induction == 2.0 / 120.0)], x='inter_burst_time',\n",
    "    common_norm=False, color=fig_sc_behavior_palette['fig.bm.sc_bursts.convergent'],\n",
    "    ax=fig_interburst_time_axes[0], legend=None, log_scale=True)\n",
    "sns.kdeplot(data=bm_bursting_df[(bm_bursting_df.condition == 'fig.bm.sc_bursts.divergent') & (bm_bursting_df.induction == 2.0 / 120.0)], x='inter_burst_time',\n",
    "    common_norm=False, color=fig_sc_behavior_palette['fig.bm.sc_bursts.divergent'],\n",
    "    ax=fig_interburst_time_axes[0], legend=None, log_scale=True)\n",
    "for ax in fig_interburst_time_axes:\n",
    "    ax.set_xlim([10, 1e3])\n",
    "    ax.set_xlabel('Time (s)')\n",
    "    ax.xaxis.labelpad = 0\n",
    "    sns.despine(ax=ax)\n",
    "fig_interburst_time_axes[0].text(0.5,1.2,'Reporter inter-burst time', ha='center', fontsize=font_sizes['base_size'], transform=fig_interburst_time_axes[0].transAxes)\n",
    "fig_sc_behavior.savefig('test.svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot single-cell cross correlation\n",
    "# Calculate average convolution\n",
    "fig_sc_behavior_subfigures['F'].subplots_adjust(left=0.2, bottom=0.3, right=0.9, top=0.85)\n",
    "fig_sc_crosscorr_ax = fig_sc_behavior_subfigures['F'].subplots(ncols=1)\n",
    "fig_sc_crosscorr_ax.axhline(y=0, color='k')\n",
    "fig_sc_crosscorr_ax.axvline(x=0, color='#888888', linestyle=':')\n",
    "sns.lineplot(data=fig_sc_behavior_convolutions[fig_sc_behavior_convolutions.induction == 1.0],\n",
    "    x='time', y='cross_corr', linewidth=1.5,\n",
    "    hue='condition', palette=fig_sc_behavior_palette,\n",
    "    legend=None, ax=fig_sc_crosscorr_ax)\n",
    "sns.despine(ax=fig_sc_crosscorr_ax)\n",
    "fig_sc_crosscorr_ax.set_xlabel('Time offset (hr)')\n",
    "fig_sc_crosscorr_ax.set_ylabel('Cross-correlation')\n",
    "fig_sc_crosscorr_ax.set_xlim([-3600*4,3600*4])\n",
    "fig_sc_crosscorr_ax.set_xticks([-3600*4,0,3600*4])\n",
    "fig_sc_crosscorr_ax.xaxis.set_major_formatter(hours_formatter)\n",
    "fig_sc_crosscorr_ax.set_ylim([-1.0,0.6])\n",
    "fig_sc_behavior.savefig('test.svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot reporter distribution plots\n",
    "fig_sc_noise_df = fig_sc_behavior_df[\n",
    "    (fig_sc_behavior_df['coeff.sigma_squared'] == 0.025) &\n",
    "    (np.abs(fig_sc_behavior_df['induction'] - 1.0) < 0.01)\n",
    "].copy()\n",
    "fig_sc_noise_df['normed_constant'] = fig_sc_noise_df['constant_mRNA'] / 300\n",
    "fig_sc_noise_std = fig_sc_noise_df.groupby(['condition', 'time']).std().reset_index()\n",
    "fig_sc_noise_mean = fig_sc_noise_df.groupby(['condition', 'time']).mean().reset_index()\n",
    "pre_induction_noise = fig_sc_noise_std[\n",
    "    (fig_sc_noise_mean.time < 10000) &\n",
    "    (fig_sc_noise_mean.time > 1000)]['normed_constant'].mean()\n",
    "fig_sc_behavior_subfigures['G'].subplots_adjust(left=0.25, bottom=0.2, right=0.93, top=0.7)\n",
    "fig_sc_noise_dist_axes = fig_sc_behavior_subfigures['G'].subplots(ncols=2, nrows=2, sharex=False, sharey=True, gridspec_kw={'hspace': 0.97}).flatten()\n",
    "for condition, ax in zip(\n",
    "        ['fig.bm_examples.convergent', 'fig.bm_examples.divergent', 'fig.bm_examples.tandem_reporter_upstream', 'fig.bm_examples.tandem_reporter_downstream'],\n",
    "        fig_sc_noise_dist_axes):\n",
    "    dist_df = fig_sc_noise_df[(fig_sc_noise_df.condition == condition)]\n",
    "    for t, ls, color in zip([5000, 35000], [':', '-'], ['reporter_only', condition]):\n",
    "        min_sep = min(np.abs(dist_df.time - t))\n",
    "        sns.kdeplot(data=dist_df[np.abs(dist_df.time - t) == min_sep], x='normed_constant', ax=ax,\n",
    "            color=fig_sc_behavior_palette[color], legend=None, linestyle=ls)\n",
    "    sns.despine(ax=ax)\n",
    "    ax.set_xlabel('')\n",
    "    ax.set_xlim([0,1.2])\n",
    "    if condition.endswith('tandem_reporter_upstream'):\n",
    "        plot_genes_unified(ax, [(0.35, 0.45), (0.65, 0.75)], [fig_sc_behavior_palette[c] for c in [condition, 'gray']], base_line_extents=(0.1,0.9), base_offset_inches=0.1, plot_inducible=1)\n",
    "    if condition.endswith('tandem_reporter_downstream'):\n",
    "        plot_genes_unified(ax, [(0.35, 0.45), (0.65, 0.75)], [fig_sc_behavior_palette[c] for c in ['gray', condition]], base_line_extents=(0.1,0.9), base_offset_inches=0.1, plot_inducible=0)\n",
    "    if condition.endswith('convergent'):\n",
    "        plot_genes_unified(ax, [(0.25, 0.38), (0.75, 0.62)], [fig_sc_behavior_palette[c] for c in ['gray', condition]], base_line_extents=(0.1,0.9), base_offset_inches=0.1, plot_inducible=0)\n",
    "    if condition.endswith('divergent'):\n",
    "        plot_genes_unified(ax, [(0.4, 0.3), (0.6, 0.7)], [fig_sc_behavior_palette[c] for c in ['gray', condition]], base_line_extents=(0.1,0.9), base_offset_inches=0.1, plot_inducible=0)\n",
    "for ax in fig_sc_noise_dist_axes[:2]:\n",
    "    ax.set_xticklabels([])\n",
    "fig_sc_behavior_subfigures['G'].text(0.55, 0.05, \"Reporter output\", horizontalalignment='center')\n",
    "fig_sc_noise_dist_axes[0].plot([0.32, 0.36], [0.88, 0.88], ls=':', color=main_palette['reporter_only'], transform=fig_sc_behavior_subfigures['G'].transSubfigure, clip_on=False)\n",
    "for i, c in enumerate([f'fig.bm_examples.{x}' for x in ['convergent', 'divergent', 'tandem_reporter_upstream', 'tandem_reporter_downstream']]):\n",
    "    fig_sc_noise_dist_axes[0].plot([0.6, 0.64], [0.91 - (0.02 * i)] * 2, ls='-', color=fig_sc_behavior_palette[c], transform=fig_sc_behavior_subfigures['G'].transSubfigure, clip_on=False)\n",
    "fig_sc_behavior_subfigures['G'].text(0.37, 0.88, 'Uninduced', ha='left', va='center')\n",
    "fig_sc_behavior_subfigures['G'].text(0.65, 0.88,  'Induced', ha='left', va='center')\n",
    "fig_sc_behavior.savefig('test.svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot overall noise behavior\n",
    "fig_sc_behavior_subfigures['H'].subplots_adjust(left=0.25, bottom=0.3, right=0.93, top=0.8)\n",
    "fig_sc_noise_ax = fig_sc_behavior_subfigures['H'].subplots(ncols=1)\n",
    "fig_sc_noise_ax.axvspan(10000, max(fig_sc_noise_df['time']), color=fig_sc_behavior_palette['background_gray'])\n",
    "fig_sc_noise_ax.axhline(y=pre_induction_noise, color=fig_sc_behavior_palette['gray'], linestyle=':')\n",
    "sns.lineplot(data=fig_sc_noise_std, x='time', y='normed_constant', hue='condition',\n",
    "    palette=fig_sc_behavior_palette, linewidth=1, estimator=None, ax=fig_sc_noise_ax, legend=None)\n",
    "sns.despine(ax=fig_sc_noise_ax)\n",
    "fig_sc_noise_ax.text(10000, 0.25, 'Induced', fontsize=font_sizes['data_annotation'])\n",
    "fig_sc_noise_ax.set_ylabel('Standard deviation')\n",
    "fig_sc_noise_ax.set_xlabel('Time (hr)')\n",
    "fig_sc_noise_ax.set_xlim([0, 40000])\n",
    "fig_sc_noise_ax.set_xticks([0,10000,3600*5, 3600*10])\n",
    "fig_sc_noise_ax.xaxis.set_major_formatter(hours_1f_formatter)\n",
    "fig_sc_noise_ax.set_ylim([0, 0.28])\n",
    "fig_sc_behavior.savefig('test.svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig_sc_behavior.savefig(rd.outfile(rd.rootdir/'writeups'/'figures'/'modeling_paper'/'fig_sc_behavior_mpl.svg', 'fig_single_cell_behavior'))\n",
    "fig_sc_behavior.savefig(rd.outfile(rd.rootdir/'writeups'/'figures'/'modeling_paper'/'fig_sc_behavior_mpl.pdf', 'fig_single_cell_behavior'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Supplemental figure - Distribution of final states with single-cell examples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "example_dist_supp_fig = plt.figure(figsize=(4,4))\n",
    "example_dist_supp_ax = example_dist_supp_fig.add_subplot()\n",
    "g = sns.kdeplot(data=fig_sc_behavior_df[\n",
    "    (fig_sc_behavior_df['coeff.sigma_squared'] == 0.025) &\n",
    "    (fig_sc_behavior_df['time'] == 40000) &\n",
    "    (fig_sc_behavior_df['induction'] == 1.0)\n",
    "], x='inducible_mRNA', y='constant_mRNA', hue='condition', palette=fig_sc_behavior_palette, ax=example_dist_supp_ax)\n",
    "for ltext, label in zip(g.get_legend().texts, ['Upstream Tandem', 'Downstream Tandem', 'Convergent', 'Divergent']):\n",
    "    ltext.set_text(label)\n",
    "g.get_legend()._set_loc(3)\n",
    "sns.despine(ax=example_dist_supp_ax)\n",
    "example_dist_supp_ax.set_xlim([0,350])\n",
    "example_dist_supp_ax.set_ylim([0,350])\n",
    "example_dist_supp_ax.set_xlabel('Inducible gene counts')\n",
    "example_dist_supp_ax.set_ylabel('Reporter gene counts')\n",
    "example_dist_supp_fig.savefig('test.svg')\n",
    "example_dist_supp_fig.savefig(rd.outfile(rd.rootdir/'writeups'/'figures'/'modeling_paper'/'fig_supp_example_distributions.svg', 'fig_single_cell_behavior'))\n",
    "example_dist_supp_fig.savefig(rd.outfile(rd.rootdir/'writeups'/'figures'/'modeling_paper'/'fig_supp_example_distributions.pdf', 'fig_single_cell_behavior'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Supplemental figure - Reporter distributions for different alpha, induction value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot reporter distribution plots\n",
    "reporter_high_alpha_supp_fig = plt.figure(figsize=(3, 2))\n",
    "fig_sc_noise_supp_alpha_df = fig_sc_behavior_df[\n",
    "    (fig_sc_behavior_df['coeff.sigma_squared'] == 0.03) &\n",
    "    (np.abs(fig_sc_behavior_df['induction'] - 1.77) < 0.01)\n",
    "].copy()\n",
    "fig_sc_noise_supp_alpha_df['normed_constant'] = fig_sc_noise_supp_alpha_df['constant_mRNA'] / 300\n",
    "pre_induction_noise = fig_sc_noise_std[\n",
    "    (fig_sc_noise_mean.time < 10000) &\n",
    "    (fig_sc_noise_mean.time > 1000)]['normed_constant'].mean()\n",
    "reporter_high_alpha_supp_fig.subplots_adjust(left=0.2, bottom=0.2, right=0.93, top=0.65)\n",
    "fig_sc_noise_dist_axes = reporter_high_alpha_supp_fig.subplots(ncols=2, nrows=2, sharex=False, sharey=True, gridspec_kw={'hspace': 0.97}).flatten()\n",
    "for condition, ax in zip(\n",
    "        ['fig.bm_examples.convergent', 'fig.bm_examples.divergent', 'fig.bm_examples.tandem_reporter_upstream', 'fig.bm_examples.tandem_reporter_downstream'],\n",
    "        fig_sc_noise_dist_axes):\n",
    "    dist_df = fig_sc_noise_supp_alpha_df[(fig_sc_noise_supp_alpha_df.condition == condition)]\n",
    "    for t, ls, color in zip([5000, 35000], [':', '-'], ['reporter_only', condition]):\n",
    "        min_sep = min(np.abs(dist_df.time - t))\n",
    "        sns.kdeplot(data=dist_df[np.abs(dist_df.time - t) == min_sep], x='normed_constant', ax=ax,\n",
    "            color=fig_sc_behavior_palette[color], legend=None, linestyle=ls)\n",
    "    sns.despine(ax=ax)\n",
    "    ax.set_xlabel('')\n",
    "    ax.set_xlim([0,1.2])\n",
    "    if condition.endswith('tandem_reporter_upstream'):\n",
    "        plot_genes_unified(ax, [(0.35, 0.45), (0.65, 0.75)], [fig_sc_behavior_palette[c] for c in [condition, 'gray']], base_line_extents=(0.1,0.9), base_offset_inches=0.1, plot_inducible=1)\n",
    "    if condition.endswith('tandem_reporter_downstream'):\n",
    "        plot_genes_unified(ax, [(0.35, 0.45), (0.65, 0.75)], [fig_sc_behavior_palette[c] for c in ['gray', condition]], base_line_extents=(0.1,0.9), base_offset_inches=0.1, plot_inducible=0)\n",
    "    if condition.endswith('convergent'):\n",
    "        plot_genes_unified(ax, [(0.25, 0.38), (0.75, 0.62)], [fig_sc_behavior_palette[c] for c in ['gray', condition]], base_line_extents=(0.1,0.9), base_offset_inches=0.1, plot_inducible=0)\n",
    "    if condition.endswith('divergent'):\n",
    "        plot_genes_unified(ax, [(0.4, 0.3), (0.6, 0.7)], [fig_sc_behavior_palette[c] for c in ['gray', condition]], base_line_extents=(0.1,0.9), base_offset_inches=0.1, plot_inducible=0)\n",
    "for ax in fig_sc_noise_dist_axes[:2]:\n",
    "    ax.set_xticklabels([])\n",
    "reporter_high_alpha_supp_fig.text(0.55, 0.05, \"Reporter output\", horizontalalignment='center')\n",
    "fig_sc_noise_dist_axes[0].plot([0.32, 0.36], [0.84, 0.84], ls=':', color=main_palette['reporter_only'], transform=reporter_high_alpha_supp_fig.transSubfigure, clip_on=False)\n",
    "for i, c in enumerate([f'fig.bm_examples.{x}' for x in ['convergent', 'divergent', 'tandem_reporter_upstream', 'tandem_reporter_downstream']]):\n",
    "    fig_sc_noise_dist_axes[0].plot([0.6, 0.64], [0.87 - (0.02 * i)] * 2, ls='-', color=fig_sc_behavior_palette[c], transform=reporter_high_alpha_supp_fig.transSubfigure, clip_on=False)\n",
    "reporter_high_alpha_supp_fig.text(0.37, 0.84, 'Uninduced', ha='left', va='center')\n",
    "reporter_high_alpha_supp_fig.text(0.65, 0.84,  'Induced', ha='left', va='center')\n",
    "reporter_high_alpha_supp_fig.text(0.55, 0.95,  'α=0.03, 1.8x induction', ha='center', va='center')\n",
    "reporter_high_alpha_supp_fig.savefig('test.svg')\n",
    "reporter_high_alpha_supp_fig.savefig(rd.outfile(rd.rootdir/'writeups'/'figures'/'modeling_paper'/'fig_supp_sc_reporter_output_high_alpha_induction.svg', 'fig_single_cell_behavior'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Toggle switch behavior figure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load toggle examples dataset\n",
    "if (rd.datadir/'projects'/'tangles'/'simulations'/'fig_toggles_examples.gzip').exists():\n",
    "    fig_toggles_examples = pd.read_parquet(rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/'fig_toggles_examples.gzip', 'fig3'))\n",
    "    print(f'Loaded {len(fig_toggles_examples)} total example datapoints from Parquet cached file')\n",
    "else:\n",
    "    with h5py.File(\n",
    "        rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/'modeling_toggles_combined.h5', 'fig_toggles_examples_cache'), 'r') as fig3_raw:\n",
    "        n_examples = 5\n",
    "        example_counts = {}\n",
    "        fig_toggles_examples = []\n",
    "        i = 1\n",
    "        for group in fig3_raw.keys():\n",
    "            for run in fig3_raw[group].values():\n",
    "                run_shape = run['mRNA'].shape\n",
    "                condition = run.attrs['comment'].decode('utf-8')\n",
    "                if not condition.startswith('fig.toggles'):\n",
    "                    continue\n",
    "                sc_dependent = run.attrs['rates.sc_dependent']\n",
    "                is_plasmid = run.attrs['bcs.is_circular']\n",
    "                hill_n     = run.attrs['hill_coeff']\n",
    "                mRNA_deg_factor = run.attrs['mRNA_deg_factor']\n",
    "                alpha = run.attrs['coeff.sigma_squared']\n",
    "                k_val      = run.attrs['K_val']\n",
    "                k_factor   = run.attrs['K_factor']\n",
    "                mRNA_deg_rate = run.attrs['rates.mRNA_degradation']\n",
    "                if 'time' in run.keys():\n",
    "                    time = run['time']\n",
    "                else:\n",
    "                    time = np.linspace(0.0, 50000.0, 500)\n",
    "                c_tup = (condition, alpha, hill_n, k_val, k_factor, sc_dependent, is_plasmid, mRNA_deg_rate)\n",
    "                if c_tup not in example_counts or example_counts[c_tup] < n_examples:\n",
    "                    fig_toggles_examples.append(pd.DataFrame(data={\n",
    "                        'condition': condition,\n",
    "                        'sc_dependent': sc_dependent,\n",
    "                        'is_plasmid': is_plasmid,\n",
    "                        'alpha': alpha,\n",
    "                        'mRNA_deg_rate': mRNA_deg_rate,\n",
    "                        'mRNA_deg_factor': mRNA_deg_factor,\n",
    "                        'hill_n': hill_n,\n",
    "                        'k_val': k_val,\n",
    "                        'k_factor': k_factor,\n",
    "                        'time': time,\n",
    "                        'mRNA_1': run['mRNA'][:,0,0],\n",
    "                        'mRNA_2': run['mRNA'][:,1,0],\n",
    "                        'idx': i\n",
    "                    }))\n",
    "                    i += 1\n",
    "                    if c_tup not in example_counts:\n",
    "                        example_counts[c_tup] = 1\n",
    "                    else:\n",
    "                        example_counts[c_tup] += 1\n",
    "        fig_toggles_examples = pd.concat(fig_toggles_examples)\n",
    "        print(f'Loaded {len(fig_toggles_examples)} example datapoints!')\n",
    "        fig_toggles_examples.to_parquet(rd.outfile(rd.datadir/'projects'/'tangles'/'simulations'/'fig_toggles_examples.gzip', 'fig_toggle_examples_cache'), compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if (rd.datadir/'projects'/'tangles'/'simulations'/'fig_toggles_basin_examples.gzip').exists():\n",
    "    fig_toggles_basin_examples = pd.read_parquet(rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/'fig_toggles_basin_examples.gzip', 'fig_toggles'))\n",
    "    print(f'Loaded {len(fig_toggles_basin_examples)} total example datapoints from Parquet cached file')\n",
    "else:\n",
    "    with h5py.File(\n",
    "        rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/'modeling_toggles_combined.h5', 'fig_toggles_basin_examples_cache'), 'r') as fig_toggles_raw:\n",
    "        n_examples = 1000\n",
    "        example_counts = {}\n",
    "        fig_toggles_basin_examples = []\n",
    "        i = 1\n",
    "        for group in fig_toggles_raw.keys():\n",
    "            group_dfs = []\n",
    "            for run in fig_toggles_raw[group].values():\n",
    "                run_shape = run['mRNA'].shape\n",
    "                condition = run.attrs['comment'].decode('utf-8')\n",
    "                if not condition.startswith('fig.toggles'):\n",
    "                    continue\n",
    "                sc_dependent = run.attrs['rates.sc_dependent']\n",
    "                is_plasmid = run.attrs['bcs.is_circular']\n",
    "                alpha = run.attrs['coeff.sigma_squared']\n",
    "                hill_n     = run.attrs['hill_coeff']\n",
    "                mRNA_deg_factor = run.attrs['mRNA_deg_factor']\n",
    "                k_val      = run.attrs['K_val']\n",
    "                k_factor   = run.attrs['K_factor']\n",
    "                mRNA_deg_rate = run.attrs['rates.mRNA_degradation']\n",
    "                if 'time' in run.keys():\n",
    "                    time = run['time']\n",
    "                else:\n",
    "                    time = np.linspace(0.0, 50000.0, 500)\n",
    "                c_tup = (condition, hill_n, alpha, k_val, k_factor, sc_dependent, is_plasmid, mRNA_deg_rate)\n",
    "                mRNA = run['mRNA'][:,:,:]\n",
    "                if c_tup not in example_counts or example_counts[c_tup] < n_examples:\n",
    "                    mRNA_1 = scipy.interpolate.interp1d(time, mRNA[:,0,:], axis=0)\n",
    "                    mRNA_2 = scipy.interpolate.interp1d(time, mRNA[:,1,:], axis=0)\n",
    "                    mRNA_1_points = mRNA_1([10200, 12500, 17500, 25000])\n",
    "                    mRNA_2_points = mRNA_2([10200, 12500, 17500, 25000])\n",
    "                    group_dfs.append(pd.DataFrame(data={\n",
    "                        'condition': condition,\n",
    "                        'sc_dependent': sc_dependent,\n",
    "                        'is_plasmid': is_plasmid,\n",
    "                        'alpha': alpha,\n",
    "                        'mRNA_deg_rate': mRNA_deg_rate,\n",
    "                        'mRNA_deg_factor': mRNA_deg_factor,\n",
    "                        'hill_n': hill_n,\n",
    "                        'k_val': k_val,\n",
    "                        'k_factor': k_factor,\n",
    "                        'mRNA_1_at_10200': mRNA_1_points[0,:],\n",
    "                        'mRNA_2_at_10200': mRNA_2_points[0,:],\n",
    "                        'mRNA_1_at_12500': mRNA_1_points[1,:],\n",
    "                        'mRNA_2_at_12500': mRNA_2_points[1,:],\n",
    "                        'mRNA_1_at_17500': mRNA_1_points[2,:],\n",
    "                        'mRNA_2_at_17500': mRNA_2_points[2,:],\n",
    "                        'mRNA_1_at_25000': mRNA_1_points[3,:],\n",
    "                        'mRNA_2_at_25000': mRNA_2_points[3,:],\n",
    "                    }))\n",
    "                    i += 1\n",
    "                    if c_tup not in example_counts:\n",
    "                        example_counts[c_tup] = run_shape[2]\n",
    "                    else:\n",
    "                        example_counts[c_tup] += run_shape[2]\n",
    "            fig_toggles_basin_examples.append(pd.concat(group_dfs, ignore_index=True))\n",
    "            print(f'Loaded {len(fig_toggles_basin_examples[-1])} examples from group')\n",
    "        fig_toggles_basin_examples = pd.concat(fig_toggles_basin_examples, ignore_index=True)\n",
    "        print(f'Loaded {len(fig_toggles_basin_examples)} example datapoints!')\n",
    "        fig_toggles_basin_examples.to_parquet(rd.outfile(rd.datadir/'projects'/'tangles'/'simulations'/'fig_toggles_basin_examples.gzip', 'fig_toggles_basin_examples_cache'), compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if (rd.datadir/'projects'/'tangles'/'simulations'/'fig_toggles_summaries.gzip').exists():\n",
    "    fig_toggles_df = pd.read_parquet(rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/'fig_toggles_summaries.gzip', 'fig_toggles'))\n",
    "    print(f'Loaded {len(fig_toggles_df)} simulation runs from Parquet cached file')\n",
    "else:\n",
    "    pass\n",
    "    with h5py.File(\n",
    "        rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/'modeling_toggles_combined.h5', 'fig_toggles_cache'), 'r') as fig_toggles_raw:\n",
    "        fig_toggles_df = []\n",
    "        i = 0\n",
    "        # Extract the Spearman correlation between the two mRNA values\n",
    "        # alongside storing the mean-centered correlation and auto-correlation\n",
    "        # of the two.\n",
    "        for group in fig_toggles_raw.keys():\n",
    "            group_df = []\n",
    "            for run in fig_toggles_raw[group].values():\n",
    "                run_shape = run['mRNA'].shape\n",
    "                condition = run.attrs['comment'].decode('utf-8')\n",
    "                if not condition.startswith('fig.toggles'):\n",
    "                    continue\n",
    "                hill_n    = run.attrs['hill_coeff']\n",
    "                alpha     = run.attrs['coeff.sigma_squared']\n",
    "                k_val     = run.attrs['K_val']\n",
    "                k_factor  = run.attrs['K_factor']\n",
    "                mRNA_deg_factor = run.attrs['mRNA_deg_factor']\n",
    "                sc_dependent = run.attrs['rates.sc_dependent']\n",
    "                is_plasmid = run.attrs['bcs.is_circular']\n",
    "                mRNA_deg_rate = run.attrs['rates.mRNA_degradation']\n",
    "                if 'time' in run.keys():\n",
    "                    time = run['time']\n",
    "                else:\n",
    "                    time = np.linspace(0.0, 50000.0, 500)\n",
    "                escape_times = np.zeros(run_shape[2])\n",
    "                mRNA = run['mRNA'][:,:,:]\n",
    "                for subrun in range(run_shape[2]):\n",
    "                    mRNA_1 = mRNA[:,0,subrun]\n",
    "                    mRNA_2 = mRNA[:,1,subrun]\n",
    "                    basin_escape = np.nonzero(mRNA_2 > mRNA_1)[0]\n",
    "                    if len(basin_escape) > 0:\n",
    "                        escape_times[subrun] = time[basin_escape[0]]\n",
    "                    else:\n",
    "                        escape_times[subrun] = np.inf\n",
    "                group_df.append(pd.DataFrame(data={\n",
    "                    'condition': condition,\n",
    "                    'sc_dependent': sc_dependent,\n",
    "                    'is_plasmid': is_plasmid,\n",
    "                    'alpha': alpha,\n",
    "                    'mRNA_deg_rate': mRNA_deg_rate,\n",
    "                    'hill_n': hill_n,\n",
    "                    'k_val': k_val,\n",
    "                    'k_factor': k_factor,\n",
    "                    'mRNA_deg_factor': mRNA_deg_factor,\n",
    "                    'escape_time': escape_times}))\n",
    "            fig_toggles_df.append(pd.concat(group_df, ignore_index=True))\n",
    "            print(f'Loaded {len(fig_toggles_df[-1])} simulations...')\n",
    "        fig_toggles_df = pd.concat(fig_toggles_df, ignore_index=True)\n",
    "        print(f'Loaded {len(fig_toggles_df)} total simulations!')\n",
    "        fig_toggles_df.to_parquet(rd.outfile(rd.datadir/'projects'/'tangles'/'simulations'/'fig_toggles_summaries.gzip', 'fig_toggles_cache'), compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load topo dataset\n",
    "if (rd.datadir/'projects'/'tangles'/'simulations'/'fig_toggles_topo_examples.gzip').exists():\n",
    "    topo_examples = pd.read_parquet(rd.datadir/'projects'/'tangles'/'simulations'/'fig_toggles_topo_examples.gzip')\n",
    "    print(f'Loaded {len(topo_examples)} total example datapoints from Parquet cached file')\n",
    "else:\n",
    "    with h5py.File(\n",
    "        rd.datadir/'projects'/'tangles'/'simulations'/'modeling_toggles_topo_combined.h5', 'r') as fig_toggles_topo_raw:\n",
    "        n_examples = 5\n",
    "        example_counts = {}\n",
    "        topo_examples = []\n",
    "        i = 1\n",
    "        for group in fig_toggles_topo_raw.keys():\n",
    "            for run in fig_toggles_topo_raw[group].values():\n",
    "                run_shape = run['mRNA'].shape\n",
    "                condition = run.attrs['comment'].decode('utf-8')\n",
    "                sc_dependent = run.attrs['rates.sc_dependent']\n",
    "                is_plasmid = run.attrs['bcs.is_circular']\n",
    "                alpha      = run.attrs['coeff.sigma_squared']\n",
    "                hill_n     = run.attrs['hill_coeff']\n",
    "                mRNA_deg_factor = run.attrs['mRNA_deg_factor']\n",
    "                topo_factor = run.attrs['topo_factor']\n",
    "                k_val      = run.attrs['K_val']\n",
    "                k_factor   = run.attrs['K_factor']\n",
    "                mRNA_deg_rate = run.attrs['rates.mRNA_degradation']\n",
    "                if 'time' in run.keys():\n",
    "                    time = run['time']\n",
    "                else:\n",
    "                    time = np.linspace(0.0, 50000.0, 500)\n",
    "                c_tup = (condition, hill_n, alpha, k_val, k_factor, topo_factor, mRNA_deg_factor)\n",
    "                if c_tup not in example_counts or example_counts[c_tup] < n_examples:\n",
    "                    topo_examples.append(pd.DataFrame(data={\n",
    "                        'condition': condition,\n",
    "                        'sc_dependent': sc_dependent,\n",
    "                        'is_plasmid': is_plasmid,\n",
    "                        'alpha': alpha,\n",
    "                        'mRNA_deg_rate': mRNA_deg_rate,\n",
    "                        'mRNA_deg_factor': mRNA_deg_factor,\n",
    "                        'topo_factor': topo_factor,\n",
    "                        'hill_n': hill_n,\n",
    "                        'k_val': k_val,\n",
    "                        'k_factor': k_factor,\n",
    "                        'time': time,\n",
    "                        'mRNA_1': run['mRNA'][:,0,0],\n",
    "                        'mRNA_2': run['mRNA'][:,1,0],\n",
    "                        'idx': i\n",
    "                    }))\n",
    "                    i += 1\n",
    "                    if c_tup not in example_counts:\n",
    "                        example_counts[c_tup] = 1\n",
    "                    else:\n",
    "                        example_counts[c_tup] += 1\n",
    "        topo_examples = pd.concat(topo_examples)\n",
    "        print(f'Loaded {len(topo_examples)} example datapoints!')\n",
    "        topo_examples.to_parquet(rd.datadir/'projects'/'tangles'/'simulations'/'fig_toggles_topo_examples.gzip', compression='gzip')\n",
    "if (rd.datadir/'projects'/'tangles'/'simulations'/'fig_toggles_topo_summaries.gzip').exists():\n",
    "    topo_df = pd.read_parquet(rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/'fig_toggles_topo_summaries.gzip', 'fig_toggles'))\n",
    "    print(f'Loaded {len(fig_toggles_df)} simulation runs from Parquet cached file')\n",
    "else:\n",
    "    pass\n",
    "    with h5py.File(\n",
    "        rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/'modeling_toggles_topo_combined.h5', 'fig_toggles_topo_cache'), 'r') as fig_toggles_topo_raw:\n",
    "        topo_df = []\n",
    "        i = 0\n",
    "        # Extract the Spearman correlation between the two mRNA values\n",
    "        # alongside storing the mean-centered correlation and auto-correlation\n",
    "        # of the two.\n",
    "        for group in fig_toggles_topo_raw.keys():\n",
    "            group_df = []\n",
    "            for run in fig_toggles_topo_raw[group].values():\n",
    "                run_shape = run['mRNA'].shape\n",
    "                condition = run.attrs['comment'].decode('utf-8')\n",
    "                alpha     = run.attrs['coeff.sigma_squared']\n",
    "                hill_n    = run.attrs['hill_coeff']\n",
    "                k_val     = run.attrs['K_val']\n",
    "                k_factor  = run.attrs['K_factor']\n",
    "                mRNA_deg_factor = run.attrs['mRNA_deg_factor']\n",
    "                topo_factor = run.attrs['topo_factor']\n",
    "                sc_dependent = run.attrs['rates.sc_dependent']\n",
    "                is_plasmid = run.attrs['bcs.is_circular']\n",
    "                mRNA_deg_rate = run.attrs['rates.mRNA_degradation']\n",
    "                if 'time' in run.keys():\n",
    "                    time = run['time']\n",
    "                else:\n",
    "                    time = np.linspace(0.0, 50000.0, 500)\n",
    "                escape_times = np.zeros(run_shape[2])\n",
    "                mRNA = run['mRNA'][:,:,:]\n",
    "                for subrun in range(run_shape[2]):\n",
    "                    mRNA_1 = mRNA[:,0,subrun]\n",
    "                    mRNA_2 = mRNA[:,1,subrun]\n",
    "                    basin_escape = np.nonzero(mRNA_2 > mRNA_1)[0]\n",
    "                    if len(basin_escape) > 0:\n",
    "                        escape_times[subrun] = time[basin_escape[0]]\n",
    "                    else:\n",
    "                        escape_times[subrun] = np.inf\n",
    "                group_df.append(pd.DataFrame(data={\n",
    "                    'condition': condition,\n",
    "                    'sc_dependent': sc_dependent,\n",
    "                    'is_plasmid': is_plasmid,\n",
    "                    'alpha': alpha,\n",
    "                    'mRNA_deg_rate': mRNA_deg_rate,\n",
    "                    'topo_factor': topo_factor,\n",
    "                    'hill_n': hill_n,\n",
    "                    'k_val': k_val,\n",
    "                    'k_factor': k_factor,\n",
    "                    'mRNA_deg_factor': mRNA_deg_factor,\n",
    "                    'escape_time': escape_times}))\n",
    "            topo_df.append(pd.concat(group_df, ignore_index=True))\n",
    "            print(f'Loaded {len(topo_df[-1])} simulations...')\n",
    "        topo_df = pd.concat(topo_df, ignore_index=True)\n",
    "        print(f'Loaded {len(topo_df)} total simulations!')\n",
    "        topo_df.to_parquet(rd.outfile(rd.datadir/'projects'/'tangles'/'simulations'/'fig_toggles_topo_summaries.gzip', 'fig_toggles_topo_cache'), compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Summarize escapes\n",
    "fig_toggles_escapes = fig_toggles_df.groupby([\n",
    "    'condition', 'alpha', 'mRNA_deg_factor', 'k_factor', 'hill_n']).apply(\n",
    "        summarize_escape_times).reset_index()\n",
    "topo_escapes = topo_df.groupby([\n",
    "    'condition', 'alpha', 'mRNA_deg_factor', 'k_factor', 'hill_n', 'topo_factor']).apply(\n",
    "        summarize_escape_times).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Summarize half-lives\n",
    "def summarize_half_lives(df, offset=0.0):\n",
    "    half_life_times = sorted(df[df.escape_fraction > 0.5]['time'])\n",
    "    return half_life_times[0]  - offset if len(half_life_times) != 0 else max(df['time']) - offset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig_toggles_halflives = fig_toggles_escapes.groupby(['condition', 'alpha', 'mRNA_deg_factor', 'k_factor', 'hill_n']).apply(lambda df: summarize_half_lives(df, 10000.0)).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Identify examples\n",
    "toggle_examples_available = fig_toggles_examples[\n",
    "    (fig_toggles_examples.time == 25000.0) &\n",
    "    (fig_toggles_examples.alpha == 0.025) &\n",
    "    (np.abs(fig_toggles_examples.mRNA_deg_factor - 10) < 0.001) &\n",
    "    (np.abs(fig_toggles_examples.k_factor - 1.0) < 0.01) &\n",
    "    (fig_toggles_examples.hill_n==2.0)\n",
    "]\n",
    "print(toggle_examples_available)\n",
    "toggle_example_mapping = {k: toggle_examples_available[toggle_examples_available.condition == k]['idx'].iloc[0]\n",
    "    for k in toggle_examples_available.condition.unique()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Print output for claims made in main text\n",
    "escape_percent_df = fig_toggles_escapes[\n",
    "    (fig_toggles_escapes.alpha == 0.025) &\n",
    "    (np.abs(fig_toggles_escapes.mRNA_deg_factor - 10) < 0.001) &\n",
    "    (np.abs(fig_toggles_escapes.k_factor - 1.0) < 0.01) &\n",
    "    (fig_toggles_escapes.hill_n == 1.0) &\n",
    "    (np.abs(fig_toggles_escapes.time - 11800) == np.min(np.abs(fig_toggles_escapes.time - 11800)))\n",
    "]\n",
    "print(escape_percent_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create the overall figure, gridspec, and add subfigure labels.\n",
    "fig_toggles = plt.figure(figsize=(6.0,9.0))\n",
    "fig_toggles_gridspec = gridspec.GridSpec(4, 3, figure=fig_toggles,\n",
    "    wspace=0.1, hspace=0.1, height_ratios=[1,4.0,1.0, 1.0])\n",
    "fig_toggles_subfigures = {\n",
    "    'A': fig_toggles.add_subfigure(fig_toggles_gridspec[0,:]),\n",
    "    'B': fig_toggles.add_subfigure(fig_toggles_gridspec[1,:]),\n",
    "    'C': fig_toggles.add_subfigure(fig_toggles_gridspec[2,0]),\n",
    "    'D': fig_toggles.add_subfigure(fig_toggles_gridspec[2,1]),\n",
    "    'E': fig_toggles.add_subfigure(fig_toggles_gridspec[2,2]),\n",
    "    'F': fig_toggles.add_subfigure(fig_toggles_gridspec[3,:]),\n",
    "}\n",
    "for label, subfig in fig_toggles_subfigures.items():\n",
    "    subfig.add_artist(matplotlib.text.Text(x=0, y=1,\n",
    "        text=f'{label}.', fontsize=font_sizes['subpanel_label'], fontweight='bold', verticalalignment='top',transform=subfig.transSubfigure))\n",
    "fig_toggles_palette = {\n",
    "    'fig.toggles.tandem_reporter_upstream': main_palette['tandem'],\n",
    "    'fig.toggles.tandem_reporter_downstream': main_palette['tandem_alt'],\n",
    "    'fig.toggles.convergent': main_palette['convergent'],\n",
    "    'fig.toggles.divergent': main_palette['divergent'],\n",
    "    'fig.toggles.topo.tandem_reporter_upstream': main_palette['tandem'],\n",
    "    'fig.toggles.topo.tandem_reporter_downstream': main_palette['tandem_alt'],\n",
    "    'fig.toggles.topo.convergent': main_palette['convergent'],\n",
    "    'fig.toggles.topo.divergent': main_palette['divergent'],\n",
    "    'fig.toggles.sc_bursts.tandem_reporter_upstream': main_palette['tandem'],\n",
    "    'fig.toggles.sc_bursts.tandem_reporter_downstream': main_palette['tandem_alt'],\n",
    "    'fig.toggles.sc_bursts.convergent': main_palette['convergent'],\n",
    "    'fig.toggles.sc_bursts.divergent': main_palette['divergent'],\n",
    "    'gray': main_palette['gray'],\n",
    "    'background_gray': '#DDDDDD'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot basin examples\n",
    "fig_toggle_basins_df = fig_toggles_basin_examples[\n",
    "    (fig_toggles_basin_examples.alpha == 0.025) &\n",
    "    (np.abs(fig_toggles_basin_examples.mRNA_deg_factor - 10) < 0.001) &\n",
    "    (np.abs(fig_toggles_basin_examples.k_factor - 1.0) < 0.01) &\n",
    "    (fig_toggles_basin_examples.hill_n==2.0)\n",
    "].copy()\n",
    "fig_toggles_subfigures['B'].subplots_adjust(left=0.1, bottom=0.1, right=0.85, top=0.9)\n",
    "fig_toggles_basins_axes = fig_toggles_subfigures['B'].subplots(nrows=4, ncols=4)\n",
    "basin_times = [10200,12500,17500,25000]\n",
    "conditions = ['fig.toggles.convergent', 'fig.toggles.divergent', 'fig.toggles.tandem_reporter_upstream', 'fig.toggles.tandem_reporter_downstream']\n",
    "for t_idx, cond_idx in itertools.product(range(len(basin_times)), range(len(conditions))):\n",
    "    t = basin_times[t_idx]\n",
    "    cond = conditions[cond_idx]\n",
    "    ax = fig_toggles_basins_axes[cond_idx, t_idx]\n",
    "    sub_df = fig_toggle_basins_df[fig_toggle_basins_df.condition == cond]\n",
    "    # Count the amount of mRNA in each state\n",
    "    mean_state = np.mean(sub_df[f'mRNA_1_at_{t}'] > sub_df[f'mRNA_2_at_{t}'])\n",
    "    sns.kdeplot(data=sub_df,x=f'mRNA_1_at_{t}', y=f'mRNA_2_at_{t}', cmap=custom_cmap_from_base(fig_toggles_palette[cond], (1.1, 0.8), (0.9, 1.4)),\n",
    "        cut=0, clip=[0,40],fill=True, levels=30, ax=ax)\n",
    "    ax.axline((0,0), slope=1, linestyle=':', color=main_palette['gray'])\n",
    "    sns.despine(ax=ax)\n",
    "    ax.set_aspect('equal')\n",
    "    ax.set_xlim([-2,40])\n",
    "    ax.set_ylim([-2,40])\n",
    "    ax.set_xticks([0,25])\n",
    "    ax.set_yticks([0,25])\n",
    "    ax.set_xlabel('')\n",
    "    ax.set_ylabel('')\n",
    "    if t_idx != 0:\n",
    "        ax.set_yticklabels([])\n",
    "    if cond_idx != 3:\n",
    "        ax.set_xticklabels([])\n",
    "    ax.text(0.25, 0.65, f'{(1.0 - mean_state) * 100:.1f}%', transform=ax.transAxes, fontsize=font_sizes['data_annotation'])\n",
    "    ax.text(0.65, 0.25, f'{mean_state * 100:.1f}%', transform=ax.transAxes, fontsize=font_sizes['data_annotation'])\n",
    "\n",
    "    if t_idx == 3:\n",
    "        if cond.endswith('tandem_reporter_upstream'):\n",
    "            plot_genes_unified(ax, [(1.2, 1.3), (1.5, 1.6)], [fig_toggles_palette[cond], fig_toggles_palette['gray']], base_line_extents=(1.1, 1.7), base_offset_inches=-0.4)\n",
    "            ax.text(1.2, 0.47, 'Upstream\\nTandem', va='top', fontsize=font_sizes['subaxis_title'], transform=ax.transAxes)\n",
    "            ax.text(1.15, 0.68, 'A', va='top', color=fig_toggles_palette[cond], fontsize=font_sizes['subaxis_title'], transform=ax.transAxes)\n",
    "            ax.text(1.72, 0.68, 'B', va='top', color=fig_toggles_palette['gray'], fontsize=font_sizes['subaxis_title'], transform=ax.transAxes)\n",
    "        if cond.endswith('tandem_reporter_downstream'):\n",
    "            plot_genes_unified(ax, [(1.2, 1.3), (1.5, 1.6)], [fig_toggles_palette['gray'], fig_toggles_palette[cond]], base_line_extents=(1.1, 1.7), base_offset_inches=-0.4)\n",
    "            ax.text(1.2, 0.47, 'Downstream\\nTandem', va='top', fontsize=font_sizes['subaxis_title'], transform=ax.transAxes)\n",
    "            ax.text(1.72, 0.68, 'A', va='top', color=fig_toggles_palette[cond], fontsize=font_sizes['subaxis_title'], transform=ax.transAxes)\n",
    "            ax.text(1.15, 0.68, 'B', va='top', color=fig_toggles_palette['gray'], fontsize=font_sizes['subaxis_title'], transform=ax.transAxes)\n",
    "        if cond.endswith('convergent'):\n",
    "            plot_genes_unified(ax, [(1.2, 1.3), (1.6, 1.5)], [fig_toggles_palette['gray'], fig_toggles_palette[cond]], base_line_extents=(1.1, 1.7), base_offset_inches=-0.4)\n",
    "            ax.text(1.2, 0.47, 'Convergent', va='top', fontsize=font_sizes['subaxis_title'], transform=ax.transAxes)\n",
    "            ax.text(1.75, 0.68, 'A', va='top', color=fig_toggles_palette[cond], fontsize=font_sizes['subaxis_title'], transform=ax.transAxes)\n",
    "            ax.text(1.15, 0.68, 'B', va='top', color=fig_toggles_palette['gray'], fontsize=font_sizes['subaxis_title'], transform=ax.transAxes)\n",
    "        if cond.endswith('divergent'):\n",
    "            plot_genes_unified(ax, [(1.3, 1.2), (1.5, 1.6)], [fig_toggles_palette['gray'], fig_toggles_palette[cond]], base_line_extents=(1.1, 1.7), base_offset_inches=-0.4)\n",
    "            ax.text(1.2, 0.47, 'Divergent', va='top', fontsize=font_sizes['subaxis_title'], transform=ax.transAxes)\n",
    "            ax.text(1.72, 0.68, 'A', va='top', color=fig_toggles_palette[cond], fontsize=font_sizes['subaxis_title'], transform=ax.transAxes)\n",
    "            ax.text(1.18, 0.68, 'B', va='top', color=fig_toggles_palette['gray'], fontsize=font_sizes['subaxis_title'], transform=ax.transAxes)\n",
    "\n",
    "fig_toggles_subfigures['B'].text(0.5, 0.04, \"Gene A mRNA count\", horizontalalignment='center')\n",
    "fig_toggles_subfigures['B'].text(0.05, 0.5, \"Gene B mRNA count\", horizontalalignment='center', verticalalignment='center', rotation=90)\n",
    "\n",
    "fig_toggles_basin_time_ax = fig_toggles_subfigures['B'].add_axes((0.1, 0.93, 0.75, 0.1))\n",
    "fig_toggles_basin_time_ax.yaxis.set_visible(False)\n",
    "fig_toggles_basin_time_ax.set_xlim([0,8])\n",
    "fig_toggles_basin_time_ax.set_xticks([0.8, 1, 3, 5, 7])\n",
    "fig_toggles_basin_time_ax.set_xticklabels(['Induction\\n2.8 hr', '2.83 hr', '3.5 hr', '5.0 hr', '7.0 hr'])\n",
    "fig_toggles_basin_time_ax.xaxis.get_major_ticks()[0]._apply_params(tickdir='in')\n",
    "basin_time_old_tick_loc = fig_toggles_basin_time_ax.xaxis.get_majorticklabels()[0].get_position()\n",
    "fig_toggles_basin_time_ax.xaxis.get_majorticklabels()[0].set_position((basin_time_old_tick_loc[0], basin_time_old_tick_loc[1] + 0.75))\n",
    "fig_toggles_subfigures['B'].text(0.5, 0.96, \"Time\", horizontalalignment='center')\n",
    "fig_toggles_basin_time_ax.plot(1, 0, \">k\", transform=fig_toggles_basin_time_ax.transAxes, clip_on=False)\n",
    "sns.despine(ax=fig_toggles_basin_time_ax, left=True)\n",
    "\n",
    "fig_toggles.savefig('test.svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot average escape percentage\n",
    "fig_toggle_escapes_df = fig_toggles_escapes[\n",
    "    (fig_toggles_escapes.alpha == 0.025) &\n",
    "    (np.abs(fig_toggles_escapes.mRNA_deg_factor - 10) < 0.001) &\n",
    "    (np.abs(fig_toggles_escapes.k_factor - 1.0) < 0.01) &\n",
    "    (fig_toggles_escapes.hill_n == 2.0)\n",
    "]\n",
    "fig_toggles_subfigures['C'].subplots_adjust(left=0.35, bottom=0.35, right=0.95, top=0.9)\n",
    "fig_toggle_escapes_ax = fig_toggles_subfigures['C'].subplots(ncols=1)\n",
    "sns.lineplot(data=fig_toggle_escapes_df, x='time', y='stable_fraction', hue='condition', palette=fig_toggles_palette, linewidth=1.5, ax=fig_toggle_escapes_ax, legend=None)\n",
    "sns.despine(ax=fig_toggle_escapes_ax)\n",
    "fig_toggle_escapes_ax.set_xlabel('Time (hr)')\n",
    "fig_toggle_escapes_ax.set_ylabel('Stable percent')\n",
    "fig_toggle_escapes_ax.set_xlim([9000,25000])\n",
    "fig_toggle_escapes_ax.set_xticks([10000,20000])\n",
    "fig_toggle_escapes_ax.set_xticklabels(['3','6'])\n",
    "fig_toggle_escapes_ax.set_yticks([0,0.5,1.0])\n",
    "fig_toggle_escapes_ax.set_yticklabels(['0%', '50%', '100%'])\n",
    "fig_toggle_escapes_ax.xaxis.labelpad = 2\n",
    "fig_toggles.savefig('test.svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot burst dynamics\n",
    "fig_toggles_subfigures['D'].subplots_adjust(left=0.35, bottom=0.35, right=0.95, top=0.9)\n",
    "fig_toggles_burst_size_ax = fig_toggles_subfigures['D'].subplots(ncols=1)\n",
    "burst_df = bursting_df[\n",
    "    (bursting_df.alpha == 0.025) &\n",
    "    (bursting_df.burst_threshold == 30) &\n",
    "    (bursting_df.inter_rnap_gene == 1) &\n",
    "    (np.abs(bursting_df.mRNA_deg_factor - 10) < 0.001) &\n",
    "    (np.abs(bursting_df.k_factor - 1.0) < 0.01) &\n",
    "    (bursting_df.hill_n == 2.0) &\n",
    "    (bursting_df.burst_size > 0) &\n",
    "    (bursting_df.inter_burst_time > 0) &\n",
    "    (\n",
    "        (bursting_df.condition == 'fig.toggles.sc_bursts.tandem_reporter_upstream') |\n",
    "        (bursting_df.condition == 'fig.toggles.sc_bursts.tandem_reporter_downstream') |\n",
    "        (bursting_df.condition == 'fig.toggles.sc_bursts.convergent') |\n",
    "        (bursting_df.condition == 'fig.toggles.sc_bursts.divergent')\n",
    "    )\n",
    "].copy()\n",
    "burst_df.condition = burst_df.condition.cat.remove_unused_categories()\n",
    "\n",
    "sns.kdeplot(data=burst_df, x='burst_size', hue='condition', palette=fig_toggles_palette, common_norm=False, log_scale=True,\n",
    "        legend=None, ax=fig_toggles_burst_size_ax)\n",
    "sns.despine(ax=fig_toggles_burst_size_ax)\n",
    "fig_toggles_burst_size_ax.set_xlim([0.5,1500])\n",
    "fig_toggles_burst_size_ax.set_xticks([1, 10, 100, 1000])\n",
    "fig_toggles_burst_size_ax.xaxis.labelpad = 2\n",
    "#fig_toggles_burst_size_ax.set_xticks([1,100])\n",
    "fig_toggles_burst_size_ax.set_xlabel('Gene A burst size')\n",
    "fig_toggles.savefig('test.svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot half lives as a function of mRNA degradation rate\n",
    "fig_toggles_deg_rate_df = fig_toggles_halflives[\n",
    "    (fig_toggles_halflives.alpha == 0.025) &\n",
    "    (np.abs(fig_toggles_halflives.k_factor - 1.0) < 0.01) &\n",
    "    (fig_toggles_halflives.hill_n == 2.0)\n",
    "].copy()\n",
    "fig_toggles_subfigures['E'].subplots_adjust(left=0.35, bottom=0.35, right=0.95, top=0.9)\n",
    "fig_toggles_deg_rate_ax = fig_toggles_subfigures['E'].subplots(ncols=1)\n",
    "fig_toggles_deg_rate_df['rel_mRNA_deg_factor'] = fig_toggles_deg_rate_df.mRNA_deg_factor / 10.0\n",
    "lp_3e = sns.lineplot(data=fig_toggles_deg_rate_df, x='rel_mRNA_deg_factor', y=0, hue='condition', palette=fig_toggles_palette, legend=None, ax=fig_toggles_deg_rate_ax)\n",
    "lp_3e.set(xscale='log')\n",
    "sns.despine(ax=fig_toggles_deg_rate_ax)\n",
    "fig_toggles_deg_rate_ax.xaxis.labelpad = 2\n",
    "fig_toggles_deg_rate_ax.set_xlim([.1,10])\n",
    "fig_toggles_deg_rate_ax.set_ylim([0, 15500])\n",
    "fig_toggles_deg_rate_ax.set_xticks([0.1,1,10])\n",
    "fig_toggles_deg_rate_ax.yaxis.set_major_formatter(k_formatter)\n",
    "fig_toggles_deg_rate_ax.set_yticks(list(range(0,20000,5000)))\n",
    "fig_toggles_deg_rate_ax.set_xticklabels(['0.1x', '1x', '10x'])\n",
    "fig_toggles_deg_rate_ax.set_xlabel('mRNA degradation rate')\n",
    "fig_toggles_deg_rate_ax.set_ylabel('Half life (s)')\n",
    "fig_toggles.savefig('test.svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot stability as a function of topoisomerase rate\n",
    "fig_toggles_topo_df = topo_escapes[\n",
    "    (topo_escapes.alpha == 0.025) &\n",
    "    (np.abs(topo_escapes.mRNA_deg_factor - 10) < 0.001) &\n",
    "    (np.abs(topo_escapes.k_factor - 1.0) < 0.01) &\n",
    "    (topo_escapes.hill_n == 2.0)\n",
    "].copy()\n",
    "fig_toggles_subfigures['F'].subplots_adjust(left=0.3, bottom=0.25, right=0.86, top=0.6)\n",
    "fig_toggles_topo_axes = fig_toggles_subfigures['F'].subplots(ncols=4, sharey=True)\n",
    "for c, ax in zip(['fig.toggles.topo.convergent', 'fig.toggles.topo.divergent', 'fig.toggles.topo.tandem_reporter_upstream', 'fig.toggles.topo.tandem_reporter_downstream'], fig_toggles_topo_axes):\n",
    "    if c.endswith('convergent') or c.endswith('divergent'):\n",
    "        lookup_c = c.replace('fig.toggles.topo', 'fig5')\n",
    "    else:\n",
    "        lookup_c = c\n",
    "    sns.lineplot(data=fig_toggles_topo_df[fig_toggles_topo_df.condition == lookup_c], x='time', y='stable_fraction', hue='topo_factor',\n",
    "        palette=no_white_greens, estimator=None, hue_norm=matplotlib.colors.LogNorm(vmin=1.0, vmax=300), legend=None, ax=ax)\n",
    "    sns.despine(ax=ax)\n",
    "    ax.set_xlabel('')\n",
    "    ax.set_xlim([10000,20000])\n",
    "    ax.set_xticks([10000,20000])\n",
    "    ax.xaxis.set_major_formatter(hours_formatter)\n",
    "    if c.endswith('tandem_reporter_upstream'):\n",
    "        plot_genes_unified(ax, [(0.35, 0.45), (0.65, 0.75)], [fig_toggles_palette[c] for c in [c, 'gray']], base_line_extents=(0.1,0.9))\n",
    "    if c.endswith('tandem_reporter_downstream'):\n",
    "        plot_genes_unified(ax, [(0.35, 0.45), (0.65, 0.75)], [fig_toggles_palette[c] for c in ['gray', c]], base_line_extents=(0.1,0.9))\n",
    "    if c.endswith('convergent'):\n",
    "        plot_genes_unified(ax, [(0.25, 0.38), (0.75, 0.62)], [fig_toggles_palette[c] for c in ['gray', c]], base_line_extents=(0.1,0.9))\n",
    "    if c.endswith('divergent'):\n",
    "        plot_genes_unified(ax, [(0.4, 0.3), (0.6, 0.7)], [fig_toggles_palette[c] for c in ['gray', c]], base_line_extents=(0.1,0.9))\n",
    "fig_toggles_topo_axes[0].set_ylabel('Stable percent')\n",
    "fig_toggles_topo_axes[0].yaxis.labelpad = 0\n",
    "fig_toggles_topo_axes[0].set_yticks([0,1])\n",
    "fig_toggles_topo_axes[0].set_yticklabels(['0%', '100%'])\n",
    "fig_toggles_subfigures['F'].text(0.6, 0.84, \"Circuit architectures\", fontsize=font_sizes['subaxis_title'], horizontalalignment='center')\n",
    "fig_toggles_subfigures['F'].text(0.6, 0.02, \"Time (hr)\", horizontalalignment='center')\n",
    "fig_toggles_topo_cbar_ax = fig_toggles_subfigures['F'].add_axes([0.89, 0.25, 0.02, 0.35])\n",
    "fig_toggles_subfigures['F'].colorbar(cm.ScalarMappable(norm=matplotlib.colors.LogNorm(vmin=1.0, vmax=300), cmap=no_white_greens), cax=fig_toggles_topo_cbar_ax)\n",
    "fig_toggles_topo_cbar_ax.set_yticks([1,10,100])\n",
    "fig_toggles_topo_cbar_ax.set_yticklabels([f'{x}x' for x in [1,10,100]])\n",
    "fig_toggles_topo_cbar_ax.set_title('REPLACEME\\nINPOST', color='r', fontsize=8)\n",
    "fig_toggles.savefig('test.svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig_toggles.savefig(rd.outfile(rd.rootdir / 'writeups' / 'figures' / 'modeling_paper' / 'fig_toggles_mpl.svg', 'fig_toggles'))\n",
    "fig_toggles.savefig(rd.outfile(rd.rootdir / 'writeups' / 'figures' / 'modeling_paper' / 'fig_toggles_mpl.pdf', 'fig_toggles'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Supplemental figure - Toggle switch behavior as a function of Hill coefficient"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot toggles over time, as a function of Hill coefficient\n",
    "fig_toggle_hill = plt.figure(figsize=(5.5, 2))\n",
    "fig_toggles_n_df = fig_toggles_escapes[\n",
    "    (fig_toggles_escapes.alpha == 0.025) &\n",
    "    (np.abs(fig_toggles_escapes.mRNA_deg_factor - 10) < 0.001) &\n",
    "    (np.abs(fig_toggles_escapes.k_factor - 1.0) < 0.01) &\n",
    "    ((fig_toggles_escapes.hill_n)  - np.trunc(fig_toggles_escapes.hill_n) == 0.0)\n",
    "].copy()\n",
    "\n",
    "fig_toggle_hill.subplots_adjust(left=0.1, bottom=0.35, right=0.9, top=0.8)\n",
    "fig_toggles_n_axes = fig_toggle_hill.subplots(ncols=3, sharey=True)\n",
    "for hill_n, ax in zip([1, 2, 4], fig_toggles_n_axes):\n",
    "    sns.lineplot(data=fig_toggles_n_df[fig_toggles_n_df.hill_n == hill_n], x='time', y='stable_fraction', hue='condition',\n",
    "        palette=fig_toggles_palette, estimator=None, ax=ax, legend=None)\n",
    "    ax.plot([10000, 10000], [0, 1], color=main_palette['light_gray'], linestyle=':')\n",
    "    ax.set_xlim([9000,25000])\n",
    "    ax.set_ylim([0,1])\n",
    "    ax.set_xticks([10000,20000])\n",
    "    ax.xaxis.set_major_formatter(hours_formatter)\n",
    "    ax.set_xlabel('')\n",
    "    sns.despine(ax=ax)\n",
    "    ax.text(0.5, 0.9, f'n={hill_n:0.1f}', transform=ax.transAxes, fontsize=font_sizes['subaxis_title'], ha='center')\n",
    "fig_toggles_n_axes[0].set_ylabel('Stable fraction')\n",
    "fig_toggle_hill.text(0.55, 0.15, \"Time (hr)\", horizontalalignment='center')\n",
    "fig_toggle_hill.savefig('test.svg')\n",
    "fig_toggle_hill.savefig(rd.outfile(rd.rootdir / 'writeups' / 'figures' / 'modeling_paper' / 'fig_supp_toggles_hill_n.svg', 'fig_toggles'))\n",
    "fig_toggle_hill.savefig(rd.outfile(rd.rootdir / 'writeups' / 'figures' / 'modeling_paper' / 'fig_supp_toggles_hill_n.pdf', 'fig_toggles'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Supplemental figure - Extra toggle switch behavior diagrams"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig_supp_toggle_sc_palette = {\n",
    "    'fig.toggles.sc_bursts.tandem_reporter_upstream': main_palette['tandem'],\n",
    "    'fig.toggles.sc_bursts.tandem_reporter_downstream': main_palette['tandem_alt'],\n",
    "    'fig.toggles.sc_bursts.convergent': main_palette['convergent'],\n",
    "    'fig.toggles.sc_bursts.divergent': main_palette['divergent'],\n",
    "    'gray': main_palette['gray']\n",
    "}\n",
    "fig_supp_extra_toggles = plt.figure(figsize=(6.0,3))\n",
    "fig_supp_extra_toggles_gridspec = gridspec.GridSpec(2, 2, figure=fig_toggles,\n",
    "    wspace=0.1, hspace=0.1, width_ratios=[3,1])\n",
    "fig_supp_extra_toggles_subfigures = {\n",
    "    'A': fig_supp_extra_toggles.add_subfigure(fig_supp_extra_toggles_gridspec[0,0]),\n",
    "    'B': fig_supp_extra_toggles.add_subfigure(fig_supp_extra_toggles_gridspec[0,1]),\n",
    "    'C': fig_supp_extra_toggles.add_subfigure(fig_supp_extra_toggles_gridspec[1,0]),\n",
    "}\n",
    "for label, subfig in fig_supp_extra_toggles_subfigures.items():\n",
    "    subfig.add_artist(matplotlib.text.Text(x=0, y=1,\n",
    "        text=f'{label}.', fontsize=font_sizes['subpanel_label'], fontweight='bold', verticalalignment='top',transform=subfig.transSubfigure))\n",
    "# Plot toggles examples\n",
    "fig_supp_extra_toggles_subfigures['A'].subplots_adjust(left=0.1, bottom=0.3, right=0.98, top=0.8)\n",
    "fig_toggle_ex_axes = fig_supp_extra_toggles_subfigures['A'].subplots(ncols=4, sharey=True)\n",
    "for ax, cond in zip(fig_toggle_ex_axes, ['fig.toggles.tandem_reporter_upstream', 'fig.toggles.tandem_reporter_downstream', 'fig.toggles.convergent', 'fig.toggles.divergent']):\n",
    "    idx = toggle_example_mapping[cond]\n",
    "    fig_3b_df = fig_toggles_examples[fig_toggles_examples.idx == idx]\n",
    "    cond = fig_3b_df['condition'].iloc[0]\n",
    "    sns.lineplot(data=fig_toggles_examples[fig_toggles_examples.idx == idx], x='time', y='mRNA_1', linewidth=0.8, color=fig_toggles_palette[cond], ax=ax)\n",
    "    sns.lineplot(data=fig_toggles_examples[fig_toggles_examples.idx == idx], x='time', y='mRNA_2', linewidth=0.8, color=fig_toggles_palette['gray'], ax=ax)\n",
    "    ax.axvspan(10000, 25000, color=fig_toggles_palette['background_gray'])\n",
    "    ax.set_xlim([0,25000])\n",
    "    ax.set_xticks([0,3600 * 3,3600 * 6])\n",
    "    ax.xaxis.set_major_formatter(hours_formatter)\n",
    "    sns.despine(ax=ax)\n",
    "    ax.set_xlabel('')\n",
    "    if cond.endswith('tandem_reporter_upstream'):\n",
    "        plot_genes_unified(ax, [(0.3, 0.4), (0.6, 0.7)], [fig_toggles_palette[cond], fig_toggles_palette['gray']])\n",
    "    if cond.endswith('tandem_reporter_downstream'):\n",
    "        plot_genes_unified(ax, [(0.3, 0.4), (0.6, 0.7)], [fig_toggles_palette['gray'], fig_toggles_palette[cond]])\n",
    "    if cond.endswith('convergent'):\n",
    "        plot_genes_unified(ax, [(0.3, 0.4), (0.7, 0.6)], [fig_toggles_palette['gray'], fig_toggles_palette[cond]])\n",
    "    if cond.endswith('divergent'):\n",
    "        plot_genes_unified(ax, [(0.4, 0.3), (0.6, 0.7)], [fig_toggles_palette['gray'], fig_toggles_palette[cond]])\n",
    "fig_toggle_ex_axes[0].set_ylabel('mRNA count')\n",
    "fig_supp_extra_toggles_subfigures['A'].text(0.53, 0.05, \"Time (hr)\", horizontalalignment='center')\n",
    "\n",
    "# Plot burst dynamics\n",
    "fig_supp_extra_toggles_subfigures['B'].subplots_adjust(left=0.3, bottom=0.3, right=0.95, top=0.90)\n",
    "fig_toggles_interburst_ax = fig_supp_extra_toggles_subfigures['B'].subplots(ncols=1)\n",
    "burst_df = bursting_df[\n",
    "    (bursting_df.alpha == 0.025) &\n",
    "    (bursting_df.burst_threshold == 30) &\n",
    "    (bursting_df.inter_rnap_gene == 1) &\n",
    "    (np.abs(bursting_df.mRNA_deg_factor - 10) < 0.001) &\n",
    "    (np.abs(bursting_df.k_factor - 1.0) < 0.01) &\n",
    "    (bursting_df.hill_n == 2.0) &\n",
    "    (bursting_df.burst_size > 0) &\n",
    "    (bursting_df.inter_burst_time > 0) &\n",
    "    (\n",
    "        (bursting_df.condition == 'fig.toggles.sc_bursts.tandem_reporter_upstream') |\n",
    "        (bursting_df.condition == 'fig.toggles.sc_bursts.tandem_reporter_downstream') |\n",
    "        (bursting_df.condition == 'fig.toggles.sc_bursts.convergent') |\n",
    "        (bursting_df.condition == 'fig.toggles.sc_bursts.divergent')\n",
    "    )\n",
    "].copy()\n",
    "burst_df.condition = burst_df.condition.cat.remove_unused_categories()\n",
    "\n",
    "sns.kdeplot(data=burst_df, x='inter_burst_time', hue='condition', palette=fig_toggles_palette, common_norm=False, log_scale=True,\n",
    "        legend=None, ax=fig_toggles_interburst_ax)\n",
    "sns.despine(ax=fig_toggles_interburst_ax)\n",
    "fig_toggles_interburst_ax.set_xlabel('Inter-burst time (s)')\n",
    "\n",
    "fig_supp_extra_toggles_subfigures['C'].subplots_adjust(left=0.1, bottom=0.3, right=0.98, top=0.8)\n",
    "fig_supp_toggle_sc_axes = fig_supp_extra_toggles_subfigures['C'].subplots(ncols=4, sharey=True, sharex=True, gridspec_kw={'hspace': 0.3})\n",
    "conditions = [f'fig.toggles.sc_bursts.{x}' for x in ['tandem_reporter_upstream', 'tandem_reporter_downstream', 'convergent', 'divergent']]\n",
    "for ax, cond in zip(fig_supp_toggle_sc_axes, conditions):\n",
    "    ax.axhline(y=0, color=main_palette['axis_gray'])\n",
    "    sns.lineplot(data=sc_density_df[\n",
    "        (sc_density_df.condition == cond) &\n",
    "        (sc_density_df.induction == 1.0 / 120.0) &\n",
    "        (np.abs(sc_density_df.mRNA_deg_factor - 10) < 0.001) &\n",
    "        (np.abs(sc_density_df.k_factor - 1.0) < 0.01) &\n",
    "        (sc_density_df.hill_n==1.0) &\n",
    "        (sc_density_df.alpha == 0.025) &\n",
    "        (np.abs(sc_density_df.time - 20000) < 1)\n",
    "    ], x='distance', y='sc_density',\n",
    "       color=fig_supp_toggle_sc_palette[cond],\n",
    "       ax=ax, legend=None)\n",
    "    ax.set_yticks([-0.2,0.2])\n",
    "    ax.set_xticks([x * 0.34 for x in [0,5000,10000]])\n",
    "    ax.set_xticklabels(['0', '5', '10'])\n",
    "    ax.set_xlabel('')\n",
    "    if cond.endswith('tandem_reporter_upstream'):\n",
    "        plot_genes_unified(ax, [(3000 * 0.34, 4000 * 0.34), (6000 * 0.34, 7000 * 0.34)], [fig_supp_toggle_sc_palette[cond], fig_supp_toggle_sc_palette['gray']], plot_gene_body=True, units='xaxis')\n",
    "    if cond.endswith('tandem_reporter_downstream'):\n",
    "        plot_genes_unified(ax, [(3000 * 0.34, 4000 * 0.34), (6000 * 0.34, 7000 * 0.34)], [fig_supp_toggle_sc_palette['gray'], fig_supp_toggle_sc_palette[cond]], plot_gene_body=True, units='xaxis')\n",
    "    if cond.endswith('convergent'):\n",
    "        plot_genes_unified(ax, [(3000 * 0.34, 4000 * 0.34), (7000 * 0.34, 6000 * 0.34)], [fig_supp_toggle_sc_palette['gray'], fig_supp_toggle_sc_palette[cond]], plot_gene_body=True, units='xaxis')\n",
    "    if cond.endswith('divergent'):\n",
    "        plot_genes_unified(ax, [(4000 * 0.34, 3000 * 0.34), (6000 * 0.34, 7000 * 0.34)], [fig_supp_toggle_sc_palette['gray'], fig_supp_toggle_sc_palette[cond]], plot_gene_body=True, units='xaxis')\n",
    "    ax.set_xlim([0,3400])\n",
    "    ax.set_ylim([-0.4, 0.4])\n",
    "    sns.despine(ax=ax)\n",
    "fig_supp_toggle_sc_axes[0].set_ylabel('Supercoiling density')\n",
    "fig_supp_toggle_sc_axes[0].yaxis.labelpad = 0\n",
    "fig_supp_extra_toggles_subfigures['C'].text(0.5, 0.04, 'Genomic distance (kb)', ha='center')\n",
    "fig_supp_extra_toggles.savefig('test.svg')\n",
    "fig_supp_extra_toggles.savefig(rd.outfile(rd.rootdir/'writeups'/'figures'/'modeling_paper'/'fig_supp_extra_toggles.svg', 'fig_toggles'))\n",
    "fig_supp_extra_toggles.savefig(rd.outfile(rd.rootdir/'writeups'/'figures'/'modeling_paper'/'fig_supp_extra_toggles.pdf', 'fig_toggles'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Supplemental figure - Basin dynamics at n = 1.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot basin examples\n",
    "fig_n1_basins_df = fig_toggles_basin_examples[\n",
    "    (fig_toggles_basin_examples.alpha == 0.025) &\n",
    "    (np.abs(fig_toggles_basin_examples.mRNA_deg_factor - 10) < 0.001) &\n",
    "    (np.abs(fig_toggles_basin_examples.k_factor - 1.0) < 0.01) &\n",
    "    (fig_toggles_basin_examples.hill_n==1.0)\n",
    "].copy()\n",
    "fig_n1_basins = plt.figure(figsize=(5, 4))\n",
    "fig_n1_basins.subplots_adjust(left=0.08, bottom=0.10, right=0.68, top=0.85)\n",
    "fig_n1_basins_axes = fig_n1_basins.subplots(nrows=4, ncols=4, sharey=True, sharex=True)\n",
    "basin_times = [10200,12500,17500,25000]\n",
    "conditions = ['fig.toggles.tandem_reporter_upstream', 'fig.toggles.tandem_reporter_downstream', 'fig.toggles.convergent', 'fig.toggles.divergent']\n",
    "for t_idx, cond_idx in itertools.product(range(len(basin_times)), range(len(conditions))):\n",
    "    t = basin_times[t_idx]\n",
    "    cond = conditions[cond_idx]\n",
    "    ax = fig_n1_basins_axes[cond_idx, t_idx]\n",
    "    sub_df = fig_n1_basins_df[fig_n1_basins_df.condition == cond]\n",
    "    # Count the amount of mRNA in each state\n",
    "    mean_state = np.mean(sub_df[f'mRNA_1_at_{t}'] > sub_df[f'mRNA_2_at_{t}'])\n",
    "    sns.kdeplot(data=sub_df,x=f'mRNA_1_at_{t}', y=f'mRNA_2_at_{t}',\n",
    "        cmap=custom_cmap_from_base(fig_toggles_palette[cond], (1.1, 0.8), (0.9, 1.4)),\n",
    "        cut=0, clip=[0,40],fill=True, ax=ax)\n",
    "    ax.axline((0,0), slope=1, linestyle=':', color=main_palette['gray'])\n",
    "    sns.despine(ax=ax)\n",
    "    ax.set_xlim([-2,40])\n",
    "    ax.set_ylim([-2,40])\n",
    "    ax.set_xlabel('')\n",
    "    ax.set_ylabel('')\n",
    "    ax.text(0.25, 0.65, f'{(1.0 - mean_state) * 100:.1f}%', transform=ax.transAxes, fontsize=font_sizes['data_annotation'])\n",
    "    ax.text(0.65, 0.25, f'{mean_state * 100:.1f}%', transform=ax.transAxes, fontsize=font_sizes['data_annotation'])\n",
    "fig_n1_basins.text(0.4, 0.02, \"mRNA 1 count\", horizontalalignment='center')\n",
    "fig_n1_basins.text(0.02, 0.5, \"mRNA 2 count\", horizontalalignment='center', verticalalignment='center', rotation=90)\n",
    "\n",
    "for ax, cond in zip(fig_n1_basins_axes[:,-1], conditions):\n",
    "    ax.text(1.1, 0.5, {\n",
    "                        'fig.toggles.tandem_reporter_upstream': 'Upstream Tandem',\n",
    "                        'fig.toggles.tandem_reporter_downstream': 'Downstream Tandem',\n",
    "                        'fig.toggles.convergent': 'Convergent',\n",
    "                        'fig.toggles.divergent': 'Divergent'\n",
    "                      }[cond], transform=ax.transAxes, fontsize=font_sizes['data_annotation'])\n",
    "\n",
    "fig_n1_basins_time_ax = fig_n1_basins.add_axes((0.08, 0.908, 0.6, 0.1))\n",
    "fig_n1_basins_time_ax.yaxis.set_visible(False)\n",
    "fig_n1_basins_time_ax.set_xlim([0,8])\n",
    "fig_n1_basins_time_ax.set_xticks([1, 3, 5, 7])\n",
    "fig_n1_basins_time_ax.set_xticklabels(['10.2k', '12.5k', '17.5k', '25k'])\n",
    "fig_n1_basins.text(0.4, 0.97, \"n = 1.0\", horizontalalignment='center', fontstyle='italic')\n",
    "fig_n1_basins.text(0.4, 0.93, \"Time (s)\", horizontalalignment='center')\n",
    "fig_n1_basins_time_ax.plot(1, 0, \">k\", transform=fig_n1_basins_time_ax.transAxes, clip_on=False)\n",
    "sns.despine(ax=fig_n1_basins_time_ax, left=True)\n",
    "\n",
    "fig_n1_basins.savefig('test.svg')\n",
    "fig_n1_basins.savefig(rd.outfile(rd.rootdir/'writeups'/'figures'/'modeling_paper'/'fig_supp_n1_basins.svg', 'fig_toggles'))\n",
    "fig_n1_basins.savefig(rd.outfile(rd.rootdir/'writeups'/'figures'/'modeling_paper'/'fig_supp_n1_basins.pdf', 'fig_toggles'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Zinani figure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mean_amplitude(trace):\n",
    "    peaks, _ = scipy.signal.find_peaks(trace ,distance=40)\n",
    "    npeaks, _ = scipy.signal.find_peaks(-trace, distance=40)\n",
    "    total_peaks = np.sort(np.concatenate((peaks, npeaks)))\n",
    "    total_peaks = total_peaks[total_peaks > 200]\n",
    "    return np.mean(np.abs(np.diff(trace[total_peaks])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if (rd.datadir/'projects'/'tangles'/'simulations'/'fig_zinani_examples.gzip').exists():\n",
    "    fig_zinani_examples = pd.read_parquet(rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/'fig_zinani_examples.gzip', 'fig_zinani'))\n",
    "    print(f'Loaded {len(fig_zinani_examples)} total example datapoints from Parquet cached file')\n",
    "else:\n",
    "    with h5py.File(\n",
    "        rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/'modeling_model_zinani_tangles_sims.h5', 'fig_zinani_cache'), 'r') as fig5_raw:\n",
    "        n_examples = 5\n",
    "        example_counts = {}\n",
    "        fig_zinani_examples = []\n",
    "        i = 1\n",
    "        for group in fig5_raw.keys():\n",
    "            for run in fig5_raw[group].values():\n",
    "                run_shape = run['mRNA'].shape\n",
    "                condition = run.attrs['comment'].decode('utf-8')\n",
    "                temperature = run.attrs['temperature']\n",
    "                alpha = run.attrs['coeff.sigma_squared']\n",
    "                c_tup = (condition, alpha, temperature)\n",
    "                if c_tup not in example_counts or example_counts[c_tup] < n_examples:\n",
    "                    fig_zinani_examples.append(pd.DataFrame(data={\n",
    "                        'condition': condition,\n",
    "                        'alpha': alpha,\n",
    "                        'temperature': temperature,\n",
    "                        'time': run['time'],\n",
    "                        'her1_mRNA': run['mRNA'][:,0,0],\n",
    "                        'her7_mRNA': run['mRNA'][:,1,0],\n",
    "                        'her1_protein': run['discrete_components'][:,0,0],\n",
    "                        'her7_protein': run['discrete_components'][:,1,0],\n",
    "                        'hes6_protein': run['discrete_components'][:,2,0],\n",
    "                        'her1_promoter_empty': run['discrete_components'][:,3,0],\n",
    "                        'her7_promoter_empty': run['discrete_components'][:,4,0],\n",
    "                        'her1_promoter_with_11': run['discrete_components'][:,5,0],\n",
    "                        'her1_promoter_with_76': run['discrete_components'][:,6,0],\n",
    "                        'her7_promoter_with_11': run['discrete_components'][:,7,0],\n",
    "                        'her7_promoter_with_76': run['discrete_components'][:,8,0],\n",
    "                        'idx': i\n",
    "                    }))\n",
    "                    i += 1\n",
    "                    if c_tup not in example_counts:\n",
    "                        example_counts[c_tup] = 1\n",
    "                    else:\n",
    "                        example_counts[c_tup] += 1\n",
    "        fig_zinani_examples = pd.concat(fig_zinani_examples)\n",
    "        print(f'Loaded {len(fig_zinani_examples)} example datapoints!')\n",
    "        fig_zinani_examples.to_parquet(rd.outfile(rd.datadir/'projects'/'tangles'/'simulations'/'fig_zinani_examples.gzip', 'fig_zinani_cache'), compression='gzip')\n",
    "tangles_signal_length = len(fig_zinani_examples[fig_zinani_examples.time > 2000].time.unique())\n",
    "if (rd.datadir/'projects'/'tangles'/'simulations'/'fig_zinani_summaries.gzip').exists():\n",
    "    fig_zinani_df = pd.read_parquet(rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/'fig_zinani_summaries.gzip', 'fig_zinani'))\n",
    "    print(f'Loaded {len(fig_zinani_df)} simulation runs from Parquet cached file')\n",
    "else:\n",
    "    with h5py.File(\n",
    "        rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/'modeling_model_zinani_tangles_sims.h5', 'fig_zinani_cache'), 'r') as fig5_raw:\n",
    "        fig_zinani_df = []\n",
    "        i = 0\n",
    "        # Extract the Spearman correlation between the two mRNA values\n",
    "        # alongside storing the mean-centered correlation and auto-correlation\n",
    "        # of the two.\n",
    "        for group in fig5_raw.keys():\n",
    "            group_df = []\n",
    "            for run in fig5_raw[group].values():\n",
    "                run_shape = run['mRNA'].shape\n",
    "                mRNA = run['mRNA'][:,:,:]\n",
    "                for subrun in range(run_shape[2]):\n",
    "                    time = run['time'][:]\n",
    "                    # Perform correlation over the entire range\n",
    "                    her1 = mRNA[:,0,subrun]\n",
    "                    her7 = mRNA[:,1,subrun]\n",
    "                    trimmed_her1 = her1[time > 2000]\n",
    "                    trimmed_her7 = her7[time > 2000]\n",
    "                    mc_her1 = trimmed_her1 - np.mean(trimmed_her1)\n",
    "                    mc_her7 = trimmed_her7 - np.mean(trimmed_her7)\n",
    "                    sc, _ = scipy.stats.spearmanr(her1, her7)\n",
    "                    cross_corr = scipy.signal.correlate(mc_her1, mc_her7)\n",
    "                    autocorr_correction = np.sqrt(\n",
    "                        scipy.signal.correlate(mc_her1, mc_her1, mode='valid')[0] * \n",
    "                        scipy.signal.correlate(mc_her7, mc_her7, mode='valid')[0])\n",
    "                    group_df.append(pd.DataFrame(data={\n",
    "                        'spearman_corr': sc,\n",
    "                        'cross_corr': (cross_corr,),\n",
    "                        'autocorr_correction': autocorr_correction if autocorr_correction != 0.0 else 1.0,\n",
    "                        'mean_amplitude': mean_amplitude(her1),\n",
    "                        'final_her7': her7[-1],\n",
    "                        'condition': run.attrs['comment'].decode('utf-8'),\n",
    "                        'alpha': run.attrs['coeff.sigma_squared'],\n",
    "                        'temperature': run.attrs['temperature'],\n",
    "                        'idx': i}))\n",
    "                    i += 1\n",
    "            fig_zinani_df.append(pd.concat(group_df, ignore_index=True))\n",
    "            print(f'Loaded {len(fig_zinani_df[-1])} simulations...')\n",
    "    # Load matlab simulations\n",
    "    print(f'Loading Matlab simulations...')\n",
    "    fig_zinani_matlab_df = pd.read_csv(rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/'zinani_matlab.csv', 'fig_zinani'))\n",
    "    fig_zinani_matlab_df['time'] = fig_zinani_matlab_df['time_min'] * 60.0\n",
    "    tangles_signal_length = len(fig_zinani_examples[fig_zinani_examples.time > 2000].time.unique())\n",
    "    matlab_signal_length = len(fig_zinani_matlab_df[(fig_zinani_matlab_df.idx == 1) & (fig_zinani_matlab_df.condition == 'Paired')])\n",
    "    tangles_lags = scipy.signal.correlation_lags(tangles_signal_length, tangles_signal_length) * np.diff(fig_zinani_examples.time.unique())[0]\n",
    "    matlab_lags = scipy.signal.correlation_lags(matlab_signal_length,matlab_signal_length) * np.diff(fig_zinani_matlab_df[fig_zinani_matlab_df.idx == 1].time.unique())[0]\n",
    "    # Split each dataframe by signal length\n",
    "    for n in range(int(len(fig_zinani_matlab_df) / matlab_signal_length)):\n",
    "        run_df = fig_zinani_matlab_df.iloc[(n*matlab_signal_length):((n+1)*matlab_signal_length),:]\n",
    "        her1 = run_df['her1']\n",
    "        her7 = run_df['her7']\n",
    "        mc_her1 = her1 - np.mean(her1)\n",
    "        mc_her7 = her7 - np.mean(her7)\n",
    "        sc, _ = scipy.stats.spearmanr(her1, her7)\n",
    "        cross_corr = scipy.signal.correlate(mc_her1, mc_her7)\n",
    "        autocorr_correction = np.sqrt(\n",
    "            scipy.signal.correlate(mc_her1, mc_her1, mode='valid')[0] * \n",
    "            scipy.signal.correlate(mc_her7, mc_her7, mode='valid')[0])\n",
    "        # Resample cross correlation to the same timescale\n",
    "        resample_cross_corr = np.interp(tangles_lags, matlab_lags, cross_corr)\n",
    "        fig_zinani_df.append(pd.DataFrame(data={\n",
    "            'spearman_corr': sc,\n",
    "            'cross_corr': (resample_cross_corr,),\n",
    "            'autocorr_correction': autocorr_correction,\n",
    "            'condition': 'matlab.' + run_df['condition'].iloc[0],\n",
    "            'idx': i\n",
    "        }))\n",
    "        i += 1\n",
    "    fig_zinani_df = pd.concat(fig_zinani_df, ignore_index=True)\n",
    "    print(f'Loaded {len(fig_zinani_df)} total simulations!')\n",
    "    fig_zinani_df.to_parquet(rd.outfile(rd.datadir/'projects'/'tangles'/'simulations'/'fig_zinani_summaries.gzip', 'fig_zinani_cache'), compression='gzip')\n",
    "if (rd.datadir/'projects'/'tangles'/'simulations'/'fig_zinani_matlab_examples.gzip').exists():\n",
    "    fig_zinani_matlab_examples = pd.read_parquet(rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/'fig_zinani_matlab_examples.gzip', 'fig_zinani'))\n",
    "    print(f'Loaded {len(fig_zinani_matlab_examples)} total example datapoints from Parquet cached file')\n",
    "else:\n",
    "    print(f'Loading Matlab examples...')\n",
    "    fig_zinani_matlab_examples = []\n",
    "    fig_zinani_matlab_df = pd.read_csv(rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/'zinani_matlab.csv', 'fig_zinani_cache'))\n",
    "    fig_zinani_matlab_df['time'] = fig_zinani_matlab_df['time_min'] * 60.0\n",
    "    matlab_signal_length = len(fig_zinani_matlab_df[(fig_zinani_matlab_df.idx == 1) & (fig_zinani_matlab_df.condition == 'Paired')])\n",
    "    example_counts = {'Paired': 0, 'Unpaired': 0}\n",
    "    for n in range(int(len(fig_zinani_matlab_df) / matlab_signal_length)):\n",
    "        run_df = fig_zinani_matlab_df.iloc[(n*matlab_signal_length):((n+1)*matlab_signal_length),:]\n",
    "        time = run_df['time']\n",
    "        her1 = run_df['her1']\n",
    "        her7 = run_df['her7']\n",
    "        condition = run_df['condition'].iloc[0]\n",
    "        if example_counts[condition] < 5:\n",
    "            fig_zinani_matlab_examples.append(pd.DataFrame({\n",
    "                'time': time,\n",
    "                'her1': her1,\n",
    "                'her7': her7,\n",
    "                'condition': f'matlab.{condition}',\n",
    "                'idx': example_counts[condition]\n",
    "            }))\n",
    "            example_counts[condition] += 1\n",
    "        if example_counts['Paired'] > 5 and example_counts['Unpaired'] > 5:\n",
    "            break\n",
    "\n",
    "    fig_zinani_matlab_examples = pd.concat(fig_zinani_matlab_examples)\n",
    "    print(f'Loaded {len(fig_zinani_matlab_examples)} example datapoints!')\n",
    "    fig_zinani_matlab_examples.to_parquet(rd.outfile(rd.datadir/'projects'/'tangles'/'simulations'/'fig_zinani_matlab_examples.gzip', 'fig_zinani_cache'), compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Prepare cross-correlation data\n",
    "fig_zinani_crosscorr = fig_zinani_df.groupby(['condition', 'alpha'], dropna=False).apply(lambda df:np.vstack(df.cross_corr / df.autocorr_correction).mean(axis=0)).reset_index()\n",
    "fig_zinani_crosscorr.insert(2, 'time_delay', 11 * [scipy.signal.correlation_lags(tangles_signal_length, tangles_signal_length) * np.diff(fig_zinani_examples.time.unique())[0]])\n",
    "fig_zinani_crosscorr = fig_zinani_crosscorr.rename(columns={0:'cross_correlation'})\n",
    "fig_zinani_crosscorr = fig_zinani_crosscorr.explode(['time_delay', 'cross_correlation'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create the overall figure, gridspec, and add subfigure labels.\n",
    "fig_zinani = plt.figure(figsize=(6.2,6.12))\n",
    "fig_zinani_gridspec = gridspec.GridSpec(3, 6, figure=fig_zinani,\n",
    "    wspace=0.1, hspace=0.1, height_ratios=[1.0,1.0,0.8])\n",
    "fig_zinani_subfigures = {\n",
    "    'A': fig_zinani.add_subfigure(fig_zinani_gridspec[0,:3]),\n",
    "    'B': fig_zinani.add_subfigure(fig_zinani_gridspec[0,3:]),\n",
    "    'C': fig_zinani.add_subfigure(fig_zinani_gridspec[1,:]),\n",
    "    'D': fig_zinani.add_subfigure(fig_zinani_gridspec[2,:2]),\n",
    "    'E': fig_zinani.add_subfigure(fig_zinani_gridspec[2,2:4]),\n",
    "    'F': fig_zinani.add_subfigure(fig_zinani_gridspec[2,4:6])\n",
    "}\n",
    "for label, subfig in fig_zinani_subfigures.items():\n",
    "    subfig.add_artist(matplotlib.text.Text(x=0, y=1, text=f'{label}.', fontsize=font_sizes['subpanel_label'], fontweight='bold', verticalalignment='top',transform=subfig.transSubfigure))\n",
    "fig_zinani_palette = {\n",
    "    'her1': '#5254a3',\n",
    "    'her7': '#9c9ede',\n",
    "    'matlab.Unpaired': '#AAAAAA',\n",
    "    'matlab.Paired': main_palette['gray'],\n",
    "    'fig.zinani.uncoupled': main_palette['light_gray'],\n",
    "    'fig.zinani.fully-coupled': main_palette['gray'],\n",
    "    'fig.zinani.tangles-coupled': main_palette['divergent'],\n",
    "    'gray': main_palette['gray'],\n",
    "    'background_gray': '#DDDDDD'\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_alpha = 0.025\n",
    "# Plot Zinani examples\n",
    "fig_zinani_subfigures['C'].subplots_adjust(left=0.08, bottom=0.25, right=0.98, top=0.9)\n",
    "fig_zinani_mRNA_axes = fig_zinani_subfigures['C'].subplots(ncols=2, sharey=False, gridspec_kw={'wspace': 0.28})\n",
    "for ax, condition in zip(fig_zinani_mRNA_axes, ['fig.zinani.uncoupled', 'fig.zinani.tangles-coupled']):\n",
    "    df = fig_zinani_examples[(fig_zinani_examples.condition == condition) & (fig_zinani_examples.alpha == plot_alpha)]\n",
    "    idx = df.idx.unique()[2]\n",
    "    if condition.endswith('tangles-coupled'):\n",
    "        tangles_idx = idx\n",
    "    sns.lineplot(data=df[df.idx == idx], x='time', y='her1_mRNA', ax=ax, color=fig_zinani_palette['her1'])\n",
    "    sns.lineplot(data=df[df.idx == idx], x='time', y='her7_mRNA', ax=ax, color=fig_zinani_palette['her7'])\n",
    "    sns.despine(ax=ax)\n",
    "    ax.set_xlim([0, 3600*4])\n",
    "    ax.set_xticks([3600 * x for x in [0,2,4]])\n",
    "    ax.xaxis.set_major_formatter(hours_1f_formatter)\n",
    "    ax.set_xlabel('')\n",
    "    ax.set_ylabel('')\n",
    "for ax in fig_zinani_mRNA_axes:\n",
    "    ax.text(0.40, 0.07, 'her1', fontsize=font_sizes['subaxis_title'], color=fig_zinani_palette['her1'], transform=ax.transAxes, fontstyle='italic')\n",
    "    ax.text(0.49, 0.07, 'her7', fontsize=font_sizes['subaxis_title'], color=fig_zinani_palette['her7'], transform=ax.transAxes, fontstyle='italic')\n",
    "    ax.set_ylim(bottom=0)\n",
    "#fig_zinani_mRNA_axes[0].set_ylim([0,25])\n",
    "#fig_zinani_mRNA_axes[1].set_ylim([0,90])\n",
    "fig_zinani_mRNA_axes[0].set_ylabel('mRNA counts')\n",
    "fig_zinani_mRNA_axes[0].yaxis.labelpad = 5\n",
    "fig_zinani_mRNA_axes[0].set_xlabel('Time (hr)')\n",
    "fig_zinani_mRNA_axes[1].set_xlabel('Time (hr)')\n",
    "fig_zinani_mRNA_axes[0].text(x=0.5, y=0.98, s='Gene unpaired', color=fig_zinani_palette['fig.zinani.uncoupled'], horizontalalignment='center', transform=fig_zinani_mRNA_axes[0].transAxes)\n",
    "fig_zinani_mRNA_axes[1].text(x=0.5, y=0.98, s='Gene paired', color=fig_zinani_palette['fig.zinani.tangles-coupled'], horizontalalignment='center', transform=fig_zinani_mRNA_axes[1].transAxes)\n",
    "# Plot Zinani correlation coefficient\n",
    "fig_zinani_subfigures['D'].subplots_adjust(left=0.25, bottom=0.23, right=0.95, top=0.9)\n",
    "fig_zinani_cc_ax = fig_zinani_subfigures['D'].subplots(ncols=1)\n",
    "sns.violinplot(data=fig_zinani_df[\n",
    "    (fig_zinani_df.alpha == plot_alpha) & ((fig_zinani_df.condition == 'fig.zinani.uncoupled') | (fig_zinani_df.condition == 'fig.zinani.tangles-coupled'))\n",
    "],\n",
    "    x='condition', y='spearman_corr', inner='quartile',\n",
    "    order=['fig.zinani.uncoupled', 'fig.zinani.tangles-coupled'],\n",
    "    palette=fig_zinani_palette, saturation=0.9,\n",
    "    bw=0.1, ax=fig_zinani_cc_ax)\n",
    "fig_zinani_cc_ax.set_xlabel('')\n",
    "fig_zinani_cc_ax.set_ylabel('Correlation coefficient')\n",
    "fig_zinani_cc_ax.set_xticklabels(['Gene\\nunpaired','Gene\\npaired'])\n",
    "fig_zinani_cc_ax.yaxis.labelpad = 0.5\n",
    "sns.despine(ax=fig_zinani_cc_ax)\n",
    "# Plot amplitude\n",
    "fig_zinani_subfigures['E'].subplots_adjust(left=0.35, bottom=0.23, right=0.95, top=0.9)\n",
    "fig_zinani_amplitude_ax = fig_zinani_subfigures['E'].subplots(ncols=1)\n",
    "sns.violinplot(data=fig_zinani_df[\n",
    "    (fig_zinani_df.alpha == plot_alpha) & ((fig_zinani_df.condition == 'fig.zinani.uncoupled') | (fig_zinani_df.condition == 'fig.zinani.tangles-coupled'))\n",
    "],\n",
    "    x='condition', y='mean_amplitude', inner='quartile', scale='width', width=1.2,\n",
    "    order=['fig.zinani.uncoupled', 'fig.zinani.tangles-coupled'],\n",
    "    palette=fig_zinani_palette, saturation=0.9,\n",
    "    bw=0.1, ax=fig_zinani_amplitude_ax)\n",
    "fig_zinani_amplitude_ax.set_xlabel('')\n",
    "fig_zinani_amplitude_ax.set_ylabel('Mean oscillation amplitude')\n",
    "fig_zinani_amplitude_ax.set_ylim([0, 40])\n",
    "fig_zinani_amplitude_ax.set_xlim([-0.72, 1.7])\n",
    "fig_zinani_amplitude_ax.set_xticklabels(['Gene\\nunpaired','Gene\\npaired'])\n",
    "sns.despine(ax=fig_zinani_amplitude_ax)\n",
    "# Plot cross-correlation plot\n",
    "fig_zinani_subfigures['F'].subplots_adjust(left=0.3, bottom=0.23, right=0.95, top=0.9)\n",
    "fig_zinani_crosscorr_ax = fig_zinani_subfigures['F'].subplots(ncols=1)\n",
    "fig_zinani_crosscorr_ax.axhline(y=0, color='k')\n",
    "fig_zinani_crosscorr_ax.axvline(x=0, color='#888888', linestyle=':')\n",
    "for condition in ['fig.zinani.uncoupled', 'fig.zinani.tangles-coupled']:\n",
    "    sns.lineplot(data=fig_zinani_crosscorr[(fig_zinani_crosscorr.condition == condition) & (fig_zinani_crosscorr.alpha == plot_alpha)].reset_index(), x='time_delay', y='cross_correlation',\n",
    "        hue='condition', palette=fig_zinani_palette, ax=fig_zinani_crosscorr_ax, ci=None, legend=None)\n",
    "sns.despine(ax=fig_zinani_crosscorr_ax)\n",
    "fig_zinani_crosscorr_ax.set_xlim([-4000,4000])\n",
    "fig_zinani_crosscorr_ax.set_xticks(np.linspace(-3600,3600, 5))\n",
    "fig_zinani_crosscorr_ax.xaxis.set_major_formatter(hours_1f_formatter)\n",
    "#fig_zinani_crosscorr_ax.set_ylim([-.3, 0.75])\n",
    "#fig_zinani_crosscorr_ax.set_yticks(np.arange(-0.25, 1.0, 0.25))\n",
    "fig_zinani_crosscorr_ax.set_xlabel('Time delay (hr)')\n",
    "fig_zinani_crosscorr_ax.set_ylabel('Cross-correlation')\n",
    "fig_zinani_crosscorr_ax.text(900, 0.6, 'Unpaired', color=fig_zinani_palette['fig.zinani.uncoupled'], ha='left')\n",
    "fig_zinani_crosscorr_ax.text(900, 0.5, 'Paired', color=fig_zinani_palette['fig.zinani.tangles-coupled'], ha='left')\n",
    "fig_zinani.savefig('test.svg')\n",
    "fig_zinani.savefig(rd.outfile(rd.rootdir/'writeups'/'figures'/'modeling_paper'/'fig_zinani_mpl.svg', 'fig_zinani'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Alternate zinani main figs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if (rd.datadir/'projects'/'tangles'/'simulations'/'fig_zinani_topo_examples.gzip').exists():\n",
    "    fig_zinani_topo_examples = pd.read_parquet(rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/'fig_zinani_topo_examples.gzip', 'fig_zinani_topo'))\n",
    "    print(f'Loaded {len(fig_zinani_topo_examples)} total example datapoints from Parquet cached file')\n",
    "else:\n",
    "    with h5py.File(\n",
    "        rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/'zinani_topo_sims.h5', 'fig_zinani_topo_cache'), 'r') as zinani_topo_df_raw:\n",
    "        n_examples = 5\n",
    "        example_counts = {}\n",
    "        fig_zinani_topo_examples = []\n",
    "        i = 1\n",
    "        for group in zinani_topo_df_raw.keys():\n",
    "            for run in zinani_topo_df_raw[group].values():\n",
    "                run_shape = run['mRNA'].shape\n",
    "                condition = run.attrs['comment'].decode('utf-8')\n",
    "                temperature = run.attrs['temperature']\n",
    "                topo_type = run.attrs['topo'].decode('utf-8')\n",
    "                alpha = run.attrs['coeff.sigma_squared']\n",
    "                c_tup = (topo_type, condition, alpha, temperature)\n",
    "                if c_tup not in example_counts or example_counts[c_tup] < n_examples:\n",
    "                    fig_zinani_topo_examples.append(pd.DataFrame(data={\n",
    "                        'topo_type': topo_type,\n",
    "                        'condition': condition,\n",
    "                        'alpha': alpha,\n",
    "                        'temperature': temperature,\n",
    "                        'time': run['time'],\n",
    "                        'her1_mRNA': run['mRNA'][:,0,0],\n",
    "                        'her7_mRNA': run['mRNA'][:,1,0],\n",
    "                        'her1_protein': run['discrete_components'][:,0,0],\n",
    "                        'her7_protein': run['discrete_components'][:,1,0],\n",
    "                        'hes6_protein': run['discrete_components'][:,2,0],\n",
    "                        'her1_promoter_empty': run['discrete_components'][:,3,0],\n",
    "                        'her7_promoter_empty': run['discrete_components'][:,4,0],\n",
    "                        'her1_promoter_with_11': run['discrete_components'][:,5,0],\n",
    "                        'her1_promoter_with_76': run['discrete_components'][:,6,0],\n",
    "                        'her7_promoter_with_11': run['discrete_components'][:,7,0],\n",
    "                        'her7_promoter_with_76': run['discrete_components'][:,8,0],\n",
    "                        'idx': i\n",
    "                    }))\n",
    "                    i += 1\n",
    "                    if c_tup not in example_counts:\n",
    "                        example_counts[c_tup] = 1\n",
    "                    else:\n",
    "                        example_counts[c_tup] += 1\n",
    "        fig_zinani_topo_examples = pd.concat(fig_zinani_topo_examples)\n",
    "        print(f'Loaded {len(fig_zinani_topo_examples)} example datapoints!')\n",
    "        fig_zinani_topo_examples.to_parquet(rd.outfile(rd.datadir/'projects'/'tangles'/'simulations'/'fig_zinani_topo_examples.gzip', 'fig_topo_zinani_cache'), compression='gzip')\n",
    "tangles_topo_signal_length = len(fig_zinani_topo_examples[fig_zinani_topo_examples.time > 2000].time.unique())\n",
    "if (rd.datadir/'projects'/'tangles'/'simulations'/'fig_zinani_topo_summaries.gzip').exists():\n",
    "    fig_zinani_topo_df = pd.read_parquet(rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/'fig_zinani_topo_summaries.gzip', 'fig_zinani_topo'))\n",
    "    print(f'Loaded {len(fig_zinani_topo_df)} simulation runs from Parquet cached file')\n",
    "else:\n",
    "    with h5py.File(\n",
    "        rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/'zinani_topo_sims.h5', 'fig_zinani_topo_cache'), 'r') as zinani_topo_df_raw:\n",
    "        fig_zinani_topo_df = []\n",
    "        i = 0\n",
    "        # Extract the Spearman correlation between the two mRNA values\n",
    "        # alongside storing the mean-centered correlation and auto-correlation\n",
    "        # of the two.\n",
    "        for group in zinani_topo_df_raw.keys():\n",
    "            group_df = []\n",
    "            for run in zinani_topo_df_raw[group].values():\n",
    "                run_shape = run['mRNA'].shape\n",
    "                mRNA = run['mRNA'][:,:,:]\n",
    "                for subrun in range(run_shape[2]):\n",
    "                    time = run['time'][:]\n",
    "                    # Perform correlation over the entire range\n",
    "                    her1 = mRNA[:,0,subrun]\n",
    "                    her7 = mRNA[:,1,subrun]\n",
    "                    trimmed_her1 = her1[time > 2000]\n",
    "                    trimmed_her7 = her7[time > 2000]\n",
    "                    mc_her1 = trimmed_her1 - np.mean(trimmed_her1)\n",
    "                    mc_her7 = trimmed_her7 - np.mean(trimmed_her7)\n",
    "                    sc, _ = scipy.stats.spearmanr(her1, her7)\n",
    "                    cross_corr = scipy.signal.correlate(mc_her1, mc_her7)\n",
    "                    autocorr_correction = np.sqrt(\n",
    "                        scipy.signal.correlate(mc_her1, mc_her1, mode='valid')[0] * \n",
    "                        scipy.signal.correlate(mc_her7, mc_her7, mode='valid')[0])\n",
    "                    group_df.append(pd.DataFrame(data={\n",
    "                        'spearman_corr': sc,\n",
    "                        'cross_corr': (cross_corr,),\n",
    "                        'autocorr_correction': autocorr_correction if autocorr_correction != 0.0 else 1.0,\n",
    "                        'mean_amplitude': mean_amplitude(her1),\n",
    "                        'final_her7': her7[-1],\n",
    "                        'topo_type': run.attrs['topo'].decode('utf-8'),\n",
    "                        'condition': run.attrs['comment'].decode('utf-8'),\n",
    "                        'alpha': run.attrs['coeff.sigma_squared'],\n",
    "                        'temperature': run.attrs['temperature'],\n",
    "                        'idx': i}))\n",
    "                    i += 1\n",
    "            fig_zinani_topo_df.append(pd.concat(group_df, ignore_index=True))\n",
    "            print(f'Loaded {len(fig_zinani_topo_df[-1])} simulations...')\n",
    "    fig_zinani_topo_df = pd.concat(fig_zinani_topo_df, ignore_index=True)\n",
    "    print(f'Loaded {len(fig_zinani_topo_df)} total simulations!')\n",
    "    fig_zinani_topo_df.to_parquet(rd.outfile(rd.datadir/'projects'/'tangles'/'simulations'/'fig_zinani_topo_summaries.gzip', 'fig_zinani_topo_cache'), compression='gzip')\n",
    "# Prepare cross-correlation data\n",
    "fig_zinani_topo_crosscorr = fig_zinani_topo_df.groupby(['topo_type', 'condition', 'alpha'], dropna=False).apply(lambda df:np.vstack(df.cross_corr / df.autocorr_correction).mean(axis=0)).reset_index()\n",
    "fig_zinani_topo_crosscorr.insert(2, 'time_delay', 36 * [scipy.signal.correlation_lags(tangles_topo_signal_length, tangles_topo_signal_length) * np.diff(fig_zinani_topo_examples.time.unique())[0]])\n",
    "fig_zinani_topo_crosscorr = fig_zinani_topo_crosscorr.rename(columns={0:'cross_correlation'})\n",
    "fig_zinani_topo_crosscorr = fig_zinani_topo_crosscorr.explode(['time_delay', 'cross_correlation'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for topo_type in fig_zinani_topo_examples.topo_type.unique():\n",
    "    # Create the overall figure, gridspec, and add subfigure labels.\n",
    "    fig_zinani_topo = plt.figure(figsize=(6.2,6.12))\n",
    "    fig_zinani_topo_gridspec = gridspec.GridSpec(3, 6, figure=fig_zinani_topo,\n",
    "        wspace=0.1, hspace=0.1, height_ratios=[1.0,1.0,0.8])\n",
    "    fig_zinani_topo_subfigures = {\n",
    "        'A': fig_zinani_topo.add_subfigure(fig_zinani_topo_gridspec[0,:3]),\n",
    "        'B': fig_zinani_topo.add_subfigure(fig_zinani_topo_gridspec[0,3:]),\n",
    "        'C': fig_zinani_topo.add_subfigure(fig_zinani_topo_gridspec[1,:]),\n",
    "        'D': fig_zinani_topo.add_subfigure(fig_zinani_topo_gridspec[2,:2]),\n",
    "        'E': fig_zinani_topo.add_subfigure(fig_zinani_topo_gridspec[2,2:4]),\n",
    "        'F': fig_zinani_topo.add_subfigure(fig_zinani_topo_gridspec[2,4:6])\n",
    "    }\n",
    "    for label, subfig in fig_zinani_topo_subfigures.items():\n",
    "        subfig.add_artist(matplotlib.text.Text(x=0, y=1, text=f'{label}.', fontsize=font_sizes['subpanel_label'], fontweight='bold', verticalalignment='top',transform=subfig.transSubfigure))\n",
    "    fig_zinani_topo_palette = {\n",
    "        'her1': '#5254a3',\n",
    "        'her7': '#9c9ede',\n",
    "        'matlab.Unpaired': '#AAAAAA',\n",
    "        'matlab.Paired': main_palette['gray'],\n",
    "        'fig.zinani.uncoupled': main_palette['light_gray'],\n",
    "        'fig.zinani.fully-coupled': main_palette['gray'],\n",
    "        'fig.zinani.tangles-coupled': main_palette['divergent'],\n",
    "        'gray': main_palette['gray'],\n",
    "        'background_gray': '#DDDDDD'\n",
    "        }\n",
    "    plot_alpha = 0.025\n",
    "    # Plot Zinani examples\n",
    "    fig_zinani_topo_subfigures['C'].subplots_adjust(left=0.08, bottom=0.25, right=0.98, top=0.9)\n",
    "    fig_zinani_topo_mRNA_axes = fig_zinani_topo_subfigures['C'].subplots(ncols=2, sharey=False, gridspec_kw={'wspace': 0.28})\n",
    "    for ax, condition in zip(fig_zinani_topo_mRNA_axes, ['fig.zinani.uncoupled', 'fig.zinani.tangles-coupled']):\n",
    "        df = fig_zinani_topo_examples[(fig_zinani_topo_examples.condition == condition) & (fig_zinani_topo_examples.alpha == plot_alpha) & (fig_zinani_topo_examples.topo_type == topo_type)]\n",
    "        idx = df.idx.unique()[2]\n",
    "        if condition.endswith('tangles-coupled'):\n",
    "            tangles_idx = idx\n",
    "        sns.lineplot(data=df[df.idx == idx], x='time', y='her1_mRNA', ax=ax, color=fig_zinani_topo_palette['her1'])\n",
    "        sns.lineplot(data=df[df.idx == idx], x='time', y='her7_mRNA', ax=ax, color=fig_zinani_topo_palette['her7'])\n",
    "        sns.despine(ax=ax)\n",
    "        ax.set_xlim([0, 3600*4])\n",
    "        ax.set_xticks([3600 * x for x in [0,2,4]])\n",
    "        ax.xaxis.set_major_formatter(hours_1f_formatter)\n",
    "        ax.set_xlabel('')\n",
    "        ax.set_ylabel('')\n",
    "    for ax in fig_zinani_topo_mRNA_axes:\n",
    "        ax.text(0.40, 0.07, 'her1', fontsize=font_sizes['subaxis_title'], color=fig_zinani_topo_palette['her1'], transform=ax.transAxes, fontstyle='italic')\n",
    "        ax.text(0.49, 0.07, 'her7', fontsize=font_sizes['subaxis_title'], color=fig_zinani_topo_palette['her7'], transform=ax.transAxes, fontstyle='italic')\n",
    "        ax.set_ylim(bottom=0)\n",
    "    #fig_zinani_topo_mRNA_axes[0].set_ylim([0,25])\n",
    "    #fig_zinani_topo_mRNA_axes[1].set_ylim([0,90])\n",
    "    fig_zinani_topo_mRNA_axes[0].set_ylabel('mRNA counts')\n",
    "    fig_zinani_topo_mRNA_axes[0].yaxis.labelpad = 5\n",
    "    fig_zinani_topo_mRNA_axes[0].set_xlabel('Time (hr)')\n",
    "    fig_zinani_topo_mRNA_axes[1].set_xlabel('Time (hr)')\n",
    "    fig_zinani_topo_mRNA_axes[0].text(x=0.5, y=0.98, s='Gene unpaired', color=fig_zinani_topo_palette['fig.zinani.uncoupled'], horizontalalignment='center', transform=fig_zinani_topo_mRNA_axes[0].transAxes)\n",
    "    fig_zinani_topo_mRNA_axes[1].text(x=0.5, y=0.98, s='Gene paired', color=fig_zinani_topo_palette['fig.zinani.tangles-coupled'], horizontalalignment='center', transform=fig_zinani_topo_mRNA_axes[1].transAxes)\n",
    "    # Plot Zinani correlation coefficient\n",
    "    fig_zinani_topo_subfigures['D'].subplots_adjust(left=0.25, bottom=0.23, right=0.95, top=0.9)\n",
    "    fig_zinani_topo_cc_ax = fig_zinani_topo_subfigures['D'].subplots(ncols=1)\n",
    "    sns.violinplot(data=fig_zinani_topo_df[\n",
    "        (fig_zinani_topo_df.topo_type == topo_type) & (fig_zinani_topo_df.alpha == plot_alpha) & ((fig_zinani_topo_df.condition == 'fig.zinani.uncoupled') | (fig_zinani_topo_df.condition == 'fig.zinani.tangles-coupled'))\n",
    "    ],\n",
    "        x='condition', y='spearman_corr', inner='quartile',\n",
    "        order=['fig.zinani.uncoupled', 'fig.zinani.tangles-coupled'],\n",
    "        palette=fig_zinani_topo_palette, saturation=0.9,\n",
    "        bw=0.1, ax=fig_zinani_topo_cc_ax)\n",
    "    fig_zinani_topo_cc_ax.set_xlabel('')\n",
    "    fig_zinani_topo_cc_ax.set_ylabel('Correlation coefficient')\n",
    "    fig_zinani_topo_cc_ax.set_xticklabels(['Gene\\nunpaired','Gene\\npaired'])\n",
    "    fig_zinani_topo_cc_ax.yaxis.labelpad = 0.5\n",
    "    sns.despine(ax=fig_zinani_topo_cc_ax)\n",
    "    # Plot amplitude\n",
    "    fig_zinani_topo_subfigures['E'].subplots_adjust(left=0.35, bottom=0.23, right=0.95, top=0.9)\n",
    "    fig_zinani_topo_amplitude_ax = fig_zinani_topo_subfigures['E'].subplots(ncols=1)\n",
    "    sns.violinplot(data=fig_zinani_topo_df[\n",
    "        (fig_zinani_topo_df.topo_type == topo_type) & (fig_zinani_topo_df.alpha == plot_alpha) & ((fig_zinani_topo_df.condition == 'fig.zinani.uncoupled') | (fig_zinani_topo_df.condition == 'fig.zinani.tangles-coupled'))\n",
    "    ],\n",
    "        x='condition', y='mean_amplitude', inner='quartile', scale='width', width=1.2,\n",
    "        order=['fig.zinani.uncoupled', 'fig.zinani.tangles-coupled'],\n",
    "        palette=fig_zinani_topo_palette, saturation=0.9,\n",
    "        bw=0.1, ax=fig_zinani_topo_amplitude_ax)\n",
    "    fig_zinani_topo_amplitude_ax.set_xlabel('')\n",
    "    fig_zinani_topo_amplitude_ax.set_ylabel('Mean oscillation amplitude')\n",
    "    fig_zinani_topo_amplitude_ax.set_ylim([0, 40])\n",
    "    fig_zinani_topo_amplitude_ax.set_xlim([-0.72, 1.7])\n",
    "    fig_zinani_topo_amplitude_ax.set_xticklabels(['Gene\\nunpaired','Gene\\npaired'])\n",
    "    sns.despine(ax=fig_zinani_topo_amplitude_ax)\n",
    "    # Plot cross-correlation plot\n",
    "    fig_zinani_topo_subfigures['F'].subplots_adjust(left=0.3, bottom=0.23, right=0.95, top=0.9)\n",
    "    fig_zinani_topo_crosscorr_ax = fig_zinani_topo_subfigures['F'].subplots(ncols=1)\n",
    "    fig_zinani_topo_crosscorr_ax.axhline(y=0, color='k')\n",
    "    fig_zinani_topo_crosscorr_ax.axvline(x=0, color='#888888', linestyle=':')\n",
    "    for condition in ['fig.zinani.uncoupled', 'fig.zinani.tangles-coupled']:\n",
    "        sns.lineplot(data=fig_zinani_topo_crosscorr[\n",
    "                (fig_zinani_topo_crosscorr.topo_type == topo_type) &\n",
    "                (fig_zinani_topo_crosscorr.condition == condition) &\n",
    "                (fig_zinani_topo_crosscorr.alpha == plot_alpha)].reset_index(), x='time_delay', y='cross_correlation',\n",
    "            hue='condition', palette=fig_zinani_topo_palette, ax=fig_zinani_topo_crosscorr_ax, ci=None, legend=None)\n",
    "    sns.despine(ax=fig_zinani_topo_crosscorr_ax)\n",
    "    fig_zinani_topo_crosscorr_ax.set_xlim([-4000,4000])\n",
    "    fig_zinani_topo_crosscorr_ax.set_xticks(np.linspace(-3600,3600, 5))\n",
    "    fig_zinani_topo_crosscorr_ax.xaxis.set_major_formatter(hours_1f_formatter)\n",
    "    #fig_zinani_topo_crosscorr_ax.set_ylim([-.3, 0.75])\n",
    "    #fig_zinani_topo_crosscorr_ax.set_yticks(np.arange(-0.25, 1.0, 0.25))\n",
    "    fig_zinani_topo_crosscorr_ax.set_xlabel('Time delay (hr)')\n",
    "    fig_zinani_topo_crosscorr_ax.set_ylabel('Cross-correlation')\n",
    "    fig_zinani_topo_crosscorr_ax.text(900, 0.6, 'Unpaired', color=fig_zinani_topo_palette['fig.zinani.uncoupled'], ha='left')\n",
    "    fig_zinani_topo_crosscorr_ax.text(900, 0.5, 'Paired', color=fig_zinani_topo_palette['fig.zinani.tangles-coupled'], ha='left')\n",
    "    fig_zinani_topo.text(0.5, 0.2, topo_type)\n",
    "    fig_zinani_topo.savefig('test.svg')\n",
    "    fig_zinani_topo.savefig(rd.outfile(rd.rootdir/'writeups'/'figures'/'modeling_paper'/f'fig_zinani_topo_{topo_type}_mpl.svg', 'fig_zinani_topo'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Supplemental figure - Zinani with alpha = 0.02\n",
    "A figure showing the Zinani figure for alpha = 0.02, showing a lack of coupling (e.g. the regimes are important!)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create the overall figure, gridspec, and add subfigure labels.\n",
    "fig_supp_zinani_low_alpha = plt.figure(figsize=(6.2,4.5))\n",
    "fig_supp_zinani_low_alpha_gridspec = gridspec.GridSpec(2, 6, figure=fig_supp_zinani_low_alpha,\n",
    "    wspace=0.1, hspace=0.1)\n",
    "fig_supp_zinani_low_alpha_subfigures = {\n",
    "    'A': fig_supp_zinani_low_alpha.add_subfigure(fig_zinani_gridspec[0,:]),\n",
    "    'B': fig_supp_zinani_low_alpha.add_subfigure(fig_zinani_gridspec[1,:2]),\n",
    "    'C': fig_supp_zinani_low_alpha.add_subfigure(fig_zinani_gridspec[1,2:4]),\n",
    "    'D': fig_supp_zinani_low_alpha.add_subfigure(fig_zinani_gridspec[1,4:6])\n",
    "}\n",
    "for label, subfig in fig_supp_zinani_low_alpha_subfigures.items():\n",
    "    subfig.add_artist(matplotlib.text.Text(x=0, y=1, text=f'{label}.', fontsize=font_sizes['subpanel_label'], fontweight='bold', verticalalignment='top',transform=subfig.transSubfigure))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_alpha = 0.02\n",
    "# Plot Zinani examples\n",
    "fig_supp_zinani_low_alpha_subfigures['A'].subplots_adjust(left=0.08, bottom=0.25, right=0.98, top=0.8)\n",
    "fig_supp_zinani_low_alpha_mRNA_axes = fig_supp_zinani_low_alpha_subfigures['A'].subplots(ncols=2, sharey=False, gridspec_kw={'wspace': 0.28})\n",
    "for ax, condition in zip(fig_supp_zinani_low_alpha_mRNA_axes, ['fig.zinani.uncoupled', 'fig.zinani.tangles-coupled']):\n",
    "    df = fig_zinani_examples[(fig_zinani_examples.condition == condition) & (fig_zinani_examples.alpha == plot_alpha)]\n",
    "    idx = df.idx.unique()[1]\n",
    "    if condition.endswith('tangles-coupled'):\n",
    "        tangles_idx = idx\n",
    "    sns.lineplot(data=df[df.idx == idx], x='time', y='her1_mRNA', ax=ax, color=fig_zinani_palette['her1'])\n",
    "    sns.lineplot(data=df[df.idx == idx], x='time', y='her7_mRNA', ax=ax, color=fig_zinani_palette['her7'])\n",
    "    sns.despine(ax=ax)\n",
    "    ax.set_xlim([0, 15000])\n",
    "    ax.set_xticks([0, 3600 * 2, 3600 * 4])\n",
    "    ax.xaxis.set_major_formatter(hours_1f_formatter)\n",
    "    ax.set_xlabel('')\n",
    "    ax.set_ylabel('')\n",
    "for ax in fig_supp_zinani_low_alpha_mRNA_axes:\n",
    "    ax.text(0.40, 0.07, 'her1', fontsize=font_sizes['subaxis_title'], color=fig_zinani_palette['her1'], transform=ax.transAxes, fontstyle='italic')\n",
    "    ax.text(0.49, 0.07, 'her7', fontsize=font_sizes['subaxis_title'], color=fig_zinani_palette['her7'], transform=ax.transAxes, fontstyle='italic')\n",
    "fig_supp_zinani_low_alpha_mRNA_axes[0].set_ylabel('mRNA counts')\n",
    "fig_supp_zinani_low_alpha_mRNA_axes[0].yaxis.labelpad = 5\n",
    "fig_supp_zinani_low_alpha_mRNA_axes[0].set_xlabel('Time (hr)')\n",
    "fig_supp_zinani_low_alpha_mRNA_axes[1].set_xlabel('Time (hr)')\n",
    "fig_supp_zinani_low_alpha_mRNA_axes[0].text(x=0.5, y=1.1, s='Gene unpaired', color=fig_zinani_palette['fig.zinani.uncoupled'], horizontalalignment='center', transform=fig_supp_zinani_low_alpha_mRNA_axes[0].transAxes)\n",
    "fig_supp_zinani_low_alpha_mRNA_axes[1].text(x=0.5, y=1.1, s='Gene paired', color=fig_zinani_palette['fig.zinani.tangles-coupled'], horizontalalignment='center', transform=fig_supp_zinani_low_alpha_mRNA_axes[1].transAxes)\n",
    "# Plot Zinani correlation coefficient\n",
    "fig_supp_zinani_low_alpha_subfigures['B'].subplots_adjust(left=0.25, bottom=0.23, right=0.95, top=0.9)\n",
    "fig_supp_zinani_low_alpha_cc_ax = fig_supp_zinani_low_alpha_subfigures['B'].subplots(ncols=1)\n",
    "sns.violinplot(data=fig_zinani_df[\n",
    "    (fig_zinani_df.alpha == plot_alpha) & ((fig_zinani_df.condition == 'fig.zinani.uncoupled') | (fig_zinani_df.condition == 'fig.zinani.tangles-coupled'))\n",
    "],\n",
    "    x='condition', y='spearman_corr', inner='quartile',\n",
    "    order=['fig.zinani.uncoupled', 'fig.zinani.tangles-coupled'],\n",
    "    palette=fig_zinani_palette, saturation=0.9,\n",
    "    bw=0.1, ax=fig_supp_zinani_low_alpha_cc_ax)\n",
    "fig_supp_zinani_low_alpha_cc_ax.set_xlabel('')\n",
    "fig_supp_zinani_low_alpha_cc_ax.set_ylabel('Correlation coefficient')\n",
    "fig_supp_zinani_low_alpha_cc_ax.set_xticklabels(['Gene\\nunpaired','Gene\\npaired'])\n",
    "fig_supp_zinani_low_alpha_cc_ax.yaxis.labelpad = 0.5\n",
    "sns.despine(ax=fig_supp_zinani_low_alpha_cc_ax)\n",
    "# Plot amplitude\n",
    "fig_supp_zinani_low_alpha_subfigures['C'].subplots_adjust(left=0.35, bottom=0.23, right=0.95, top=0.9)\n",
    "fig_supp_zinani_low_alpha_amplitude_ax = fig_supp_zinani_low_alpha_subfigures['C'].subplots(ncols=1)\n",
    "sns.violinplot(data=fig_zinani_df[\n",
    "    (fig_zinani_df.alpha == plot_alpha) & ((fig_zinani_df.condition == 'fig.zinani.uncoupled') | (fig_zinani_df.condition == 'fig.zinani.tangles-coupled'))\n",
    "],\n",
    "    x='condition', y='mean_amplitude', inner='quartile', scale='width', width=1.2,\n",
    "    order=['fig.zinani.uncoupled', 'fig.zinani.tangles-coupled'],\n",
    "    palette=fig_zinani_palette, saturation=0.9,\n",
    "    bw=0.1, ax=fig_supp_zinani_low_alpha_amplitude_ax)\n",
    "fig_supp_zinani_low_alpha_amplitude_ax.set_xlabel('')\n",
    "fig_supp_zinani_low_alpha_amplitude_ax.set_ylabel('Mean oscillation amplitude')\n",
    "fig_supp_zinani_low_alpha_amplitude_ax.set_ylim([0, 40])\n",
    "fig_supp_zinani_low_alpha_amplitude_ax.set_xlim([-0.72, 1.7])\n",
    "fig_supp_zinani_low_alpha_amplitude_ax.set_xticklabels(['Gene\\nunpaired','Gene\\npaired'])\n",
    "sns.despine(ax=fig_supp_zinani_low_alpha_amplitude_ax)\n",
    "# Plot cross-correlation plot\n",
    "fig_supp_zinani_low_alpha_subfigures['D'].subplots_adjust(left=0.3, bottom=0.23, right=0.95, top=0.9)\n",
    "fig_supp_zinani_low_alpha_crosscorr_ax = fig_supp_zinani_low_alpha_subfigures['D'].subplots(ncols=1)\n",
    "fig_supp_zinani_low_alpha_crosscorr_ax.axhline(y=0, color='k')\n",
    "fig_supp_zinani_low_alpha_crosscorr_ax.axvline(x=0, color='#888888', linestyle=':')\n",
    "for condition in ['fig.zinani.uncoupled', 'fig.zinani.tangles-coupled']:\n",
    "    sns.lineplot(data=fig_zinani_crosscorr[(fig_zinani_crosscorr.condition == condition) & (fig_zinani_crosscorr.alpha == plot_alpha)].reset_index(), x='time_delay', y='cross_correlation',\n",
    "        hue='condition', palette=fig_zinani_palette, ax=fig_supp_zinani_low_alpha_crosscorr_ax, ci=None, legend=None)\n",
    "sns.despine(ax=fig_supp_zinani_low_alpha_crosscorr_ax)\n",
    "fig_supp_zinani_low_alpha_crosscorr_ax.set_xlim([-4000,4000])\n",
    "fig_supp_zinani_low_alpha_crosscorr_ax.set_xticks(np.linspace(-3600,3600, 5))\n",
    "fig_supp_zinani_low_alpha_crosscorr_ax.xaxis.set_major_formatter(hours_1f_formatter)\n",
    "#fig_supp_zinani_low_alpha_crosscorr_ax.set_ylim([-.3, 0.75])\n",
    "#fig_supp_zinani_low_alpha_crosscorr_ax.set_yticks(np.arange(-0.25, 1.0, 0.25))\n",
    "fig_supp_zinani_low_alpha_crosscorr_ax.set_xlabel('Time delay (hr)')\n",
    "fig_supp_zinani_low_alpha_crosscorr_ax.set_ylabel('Cross-correlation')\n",
    "fig_supp_zinani_low_alpha_crosscorr_ax.text(900, 0.2, 'Unpaired', color=fig_zinani_palette['fig.zinani.uncoupled'], ha='left')\n",
    "fig_supp_zinani_low_alpha_crosscorr_ax.text(900, 0.15, 'Paired', color=fig_zinani_palette['fig.zinani.tangles-coupled'], ha='left')\n",
    "\n",
    "fig_supp_zinani_low_alpha.savefig('test.svg')\n",
    "fig_supp_zinani_low_alpha.savefig(rd.outfile(rd.rootdir/'writeups'/'figures'/'modeling_paper'/'fig_supp_zinani_low_alpha.svg', 'fig_zinani'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Supplemental figure - Matlab results\n",
    "Show the original Zinani simulation results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create the overall figure, gridspec, and add subfigure labels.\n",
    "fig_supp_orig_zinani = plt.figure(figsize=(5.2,3.12))\n",
    "fig_supp_orig_zinani_gridspec = gridspec.GridSpec(2, 2, figure=fig_supp_orig_zinani,\n",
    "    wspace=0.1, hspace=0.1)\n",
    "fig_supp_orig_zinani_subfigures = {\n",
    "    'A': fig_supp_orig_zinani.add_subfigure(fig_supp_orig_zinani_gridspec[0,:]),\n",
    "    'B': fig_supp_orig_zinani.add_subfigure(fig_supp_orig_zinani_gridspec[1,0]),\n",
    "    'C': fig_supp_orig_zinani.add_subfigure(fig_supp_orig_zinani_gridspec[1,1])\n",
    "}\n",
    "for label, subfig in fig_supp_orig_zinani_subfigures.items():\n",
    "    subfig.add_artist(matplotlib.text.Text(x=0, y=1, text=f'{label}.', fontsize=font_sizes['subpanel_label'], fontweight='bold', verticalalignment='top',transform=subfig.transSubfigure))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot Zinani examples\n",
    "fig_supp_orig_zinani_subfigures['A'].subplots_adjust(left=0.12, bottom=0.25, right=0.95, top=0.8)\n",
    "fig_supp_orig_zinani_mRNA_axes = fig_supp_orig_zinani_subfigures['A'].subplots(ncols=2, sharey=False, gridspec_kw={'wspace': 0.28})\n",
    "for ax, condition in zip(fig_supp_orig_zinani_mRNA_axes, ['matlab.Unpaired', 'matlab.Paired']):\n",
    "    df = fig_zinani_matlab_examples[fig_zinani_matlab_examples.condition == condition]\n",
    "    idx = df.idx.unique()[1]\n",
    "    sns.lineplot(data=df[df.idx == idx], x='time', y='her1', ax=ax, color=fig_zinani_palette['her1'])\n",
    "    sns.lineplot(data=df[df.idx == idx], x='time', y='her7', ax=ax, color=fig_zinani_palette['her7'])\n",
    "    sns.despine(ax=ax)\n",
    "    ax.set_xlim([0, 15000])\n",
    "    ax.set_xticks([0, 3600 * 2, 3600 * 4])\n",
    "    ax.xaxis.set_major_formatter(hours_1f_formatter)\n",
    "    ax.set_xlabel('')\n",
    "    ax.set_ylabel('')\n",
    "    ax.set_facecolor(fig_zinani_palette[condition] + '55')\n",
    "fig_supp_orig_zinani_mRNA_axes[0].set_ylabel('mRNA counts')\n",
    "fig_supp_orig_zinani_mRNA_axes[1].set_xlabel('Time (hr)')\n",
    "fig_supp_orig_zinani_mRNA_axes[0].text(x=0.5, y=1.07, s='Uncoupled\\n(Zinani MATLAB)', horizontalalignment='center', transform=fig_supp_orig_zinani_mRNA_axes[0].transAxes)\n",
    "fig_supp_orig_zinani_mRNA_axes[1].text(x=0.5, y=1.07, s='Transcript coupled\\n(Zinani MATLAB)', horizontalalignment='center', transform=fig_supp_orig_zinani_mRNA_axes[1].transAxes)\n",
    "# Plot correlation coefficents\n",
    "fig_supp_orig_zinani_subfigures['B'].subplots_adjust(left=0.25, bottom=0.23, right=0.95, top=0.9)\n",
    "fig_supp_orig_zinani_cc_ax = fig_supp_orig_zinani_subfigures['B'].subplots(ncols=1)\n",
    "sns.violinplot(data=fig_zinani_df[\n",
    "    (fig_zinani_df.condition == 'matlab.Unpaired') |\n",
    "    (fig_zinani_df.condition == 'matlab.Paired')\n",
    "],\n",
    "    x='condition', y='spearman_corr', inner='quartile',\n",
    "    order=['matlab.Unpaired', 'matlab.Paired'],\n",
    "    palette=fig_zinani_palette, saturation=0.7,\n",
    "    bw=0.1, ax=fig_supp_orig_zinani_cc_ax)\n",
    "fig_supp_orig_zinani_cc_ax.set_xlabel('')\n",
    "fig_supp_orig_zinani_cc_ax.set_ylabel('Correlation\\ncoefficient')\n",
    "fig_supp_orig_zinani_cc_ax.set_xticklabels(['Uncoupled','Transcript\\ncoupled'])\n",
    "sns.despine(ax=fig_supp_orig_zinani_cc_ax)\n",
    "# Plot cross-correlation plot\n",
    "fig_supp_orig_zinani_subfigures['C'].subplots_adjust(left=0.2, bottom=0.25, right=0.95, top=0.9)\n",
    "fig_supp_orig_zinani_crosscorr_ax = fig_supp_orig_zinani_subfigures['C'].subplots(ncols=1)\n",
    "for condition in ['matlab.Unpaired', 'matlab.Paired']:\n",
    "    sns.lineplot(data=fig_zinani_crosscorr[fig_zinani_crosscorr.condition == condition], x='time_delay', y='cross_correlation',\n",
    "        hue='condition', palette=fig_zinani_palette, ax=fig_supp_orig_zinani_crosscorr_ax, estimator=None, legend=None)\n",
    "sns.despine(ax=fig_supp_orig_zinani_crosscorr_ax)\n",
    "fig_supp_orig_zinani_crosscorr_ax.set_xlim([-4000,4000])\n",
    "fig_supp_orig_zinani_crosscorr_ax.set_xticks(np.linspace(-3600,3600, 5))\n",
    "fig_supp_orig_zinani_crosscorr_ax.xaxis.set_major_formatter(hours_1f_formatter)\n",
    "#fig_supp_orig_zinani_crosscorr_ax.set_ylim([-.3, 0.75])\n",
    "#fig_supp_orig_zinani_crosscorr_ax.set_yticks(np.arange(-0.25, 1.0, 0.25))\n",
    "fig_supp_orig_zinani_crosscorr_ax.set_xlabel('Time delay (hr)')\n",
    "fig_supp_orig_zinani_crosscorr_ax.set_ylabel('Cross-correlation')\n",
    "fig_supp_orig_zinani_crosscorr_ax.axhline(y=0, color='k')\n",
    "fig_supp_orig_zinani_crosscorr_ax.axvline(x=0, color='#888888', linestyle=':')\n",
    "fig_supp_orig_zinani.savefig('test.svg')\n",
    "fig_supp_orig_zinani.savefig(rd.outfile(rd.rootdir/'writeups'/'figures'/'modeling_paper'/'fig_supp_orig_zinani.svg', 'fig_zinani'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Supplemental figure - Resampling results\n",
    "The figure where we show the various resampling rates and how they affect the correlation coefficient results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if (rd.datadir/'projects'/'tangles'/'simulations'/'fig_zinani_resampling.gzip').exists():\n",
    "    zinani_resampling_df = pd.read_parquet(rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/'fig_zinani_resampling.gzip', 'fig_supp_zinani_resample'))\n",
    "    print(f'Loaded {len(zinani_resampling_df)} total example datapoints from Parquet cached file')\n",
    "else:\n",
    "    zinani_resampling_df = pd.DataFrame(columns=['condition','original_r', 'full_resample_r','100000_resample_r', '10000_resample_r', '5000_resample_r', '1000_resample_r'])\n",
    "    for i in range(1,301):\n",
    "        paired_raw = pd.read_csv(rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/'raw_matlab'/f'GenePaired_Run{i}.csv','fig_supp_zinani_resample_cache'))\n",
    "        unpaired_raw = pd.read_csv(rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/'raw_matlab'/f'GeneUnpaired_Run{i}.csv','fig_supp_zinani_resample_cache'))\n",
    "        result_paired = np.zeros(6)\n",
    "        result_unpaired = np.zeros(6)\n",
    "        result_paired[0],_   = scipy.stats.spearmanr(paired_raw.mh1, paired_raw.mh7)\n",
    "        result_unpaired[0],_ = scipy.stats.spearmanr(unpaired_raw.mh1, unpaired_raw.mh7)\n",
    "        for i, resample_n in enumerate([len(paired_raw.mh1), 100000, 10000, 5000, 1000]):\n",
    "            re_mh1 = scipy.interpolate.interp1d(paired_raw.Time, paired_raw.mh1, kind='previous')(np.linspace(0, 240, resample_n))\n",
    "            re_mh7 = scipy.interpolate.interp1d(paired_raw.Time, paired_raw.mh7, kind='previous')(np.linspace(0, 240, resample_n))\n",
    "            result_paired[i + 1],_ = scipy.stats.spearmanr(re_mh1, re_mh7)\n",
    "        for i, resample_n in enumerate([len(unpaired_raw.mh1), 100000, 10000, 5000, 1000]):\n",
    "            re_mh1 = scipy.interpolate.interp1d(unpaired_raw.Time, unpaired_raw.mh1, kind='previous')(np.linspace(0, 240, resample_n))\n",
    "            re_mh7 = scipy.interpolate.interp1d(unpaired_raw.Time, unpaired_raw.mh7, kind='previous')(np.linspace(0, 240, resample_n))\n",
    "            result_unpaired[i + 1],_ = scipy.stats.spearmanr(re_mh1, re_mh7)\n",
    "        zinani_resampling_df.loc[len(zinani_resampling_df)] = ['paired', *result_paired]\n",
    "        zinani_resampling_df.loc[len(zinani_resampling_df)] = ['unpaired', *result_unpaired]\n",
    "        print('.', end='')\n",
    "    zinani_resampling_df.to_parquet(rd.outfile(rd.datadir/'projects'/'tangles'/'simulations'/'fig_zinani_resampling.gzip', 'fig_supp_zinani_resample_cache'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig_supp_zinani_resampling = plt.figure(figsize=(7.5,2.5))\n",
    "fig_supp_zinani_resampling.subplots_adjust(left=0.1, bottom=0.25, right=0.95, top=0.85)\n",
    "fig_supp_zinani_resampling_axes = fig_supp_zinani_resampling.subplots(ncols=4, sharey=True)\n",
    "for ax, col, title in zip(\n",
    "        fig_supp_zinani_resampling_axes,\n",
    "        ['original_r', 'full_resample_r', '10000_resample_r', '1000_resample_r'],\n",
    "        ['Original', 'Uniform\\nresampling', '10k\\ndownsample', '1k\\ndownsample']):\n",
    "    sns.violinplot(data=zinani_resampling_df, x='condition', y=col, ax=ax, order=['unpaired','paired'], palette={\n",
    "        'paired': fig_zinani_palette['matlab.Paired'],\n",
    "        'unpaired': fig_zinani_palette['matlab.Unpaired'],\n",
    "    })\n",
    "    sns.despine(ax=ax)\n",
    "    ax.set_xticklabels(['Uncoupled', 'Transcript\\ncoupled'])\n",
    "    ax.set_xlabel('')\n",
    "    ax.set_ylabel('')\n",
    "    ax.set_ylim([-0.2,1.0])\n",
    "    ax.set_title(title)\n",
    "fig_supp_zinani_resampling_axes[0].set_ylabel('Correlation\\ncoefficent')\n",
    "fig_supp_zinani_resampling.text(0.52, 0.05, \"Condition\", horizontalalignment='center')\n",
    "fig_supp_zinani_resampling.savefig('test.svg')\n",
    "fig_supp_zinani_resampling.savefig(rd.outfile(rd.rootdir/'writeups'/'figures'/'modeling_paper'/'fig_supp_zinani_resampling.svg', 'fig_supp_zinani_resample'))\n",
    "fig_supp_zinani_resampling.savefig(rd.outfile(rd.rootdir/'writeups'/'figures'/'modeling_paper'/'fig_supp_zinani_resampling.pdf', 'fig_supp_zinani_resample'))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.7.4 ('env': venv)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  },
  "vscode": {
   "interpreter": {
    "hash": "393577709b641b15c2aa1de8cfe0868d00094c5de3ea5b0856cceab616d9a2d4"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
