{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.colors\n",
    "import matplotlib.cm as cm\n",
    "import matplotlib.image\n",
    "import matplotlib.path\n",
    "import matplotlib.patches\n",
    "import matplotlib.ticker\n",
    "from matplotlib import animation\n",
    "import matplotlib.gridspec as gridspec\n",
    "import h5py\n",
    "import pandas as pd\n",
    "from pandas.api.types import CategoricalDtype\n",
    "import numpy as np\n",
    "import svgutils.transform as sg\n",
    "import scipy.signal\n",
    "import scipy.stats\n",
    "\n",
    "from pathlib import Path\n",
    "import re\n",
    "\n",
    "import itertools\n",
    "sns.set_style('ticks')\n",
    "sns.set_context('paper',rc={'font.family': 'sans-serif', 'font.sans-serif':['Helvetica Neue']})\n",
    "\n",
    "import rushd as rd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "main_palette = {\n",
    "    'tandem': '#225A9B',\n",
    "    'tandem_alt': '#19D2BF',\n",
    "    'convergent': '#FFB133',\n",
    "    'divergent': '#FE484E',\n",
    "    'gray': '#666666',\n",
    "    'light_gray': '#888888',\n",
    "    'axis_gray': '#262626'\n",
    "}\n",
    "font_sizes = {\n",
    "    'colorbar_title': 8,\n",
    "    'subaxis_title': 9,\n",
    "    'line_annotation': 8,\n",
    "    'data_annotation': 8,\n",
    "    'base_size': 8,\n",
    "}\n",
    "light_palette = {k:v + '28' for k, v in main_palette.items()}\n",
    "no_yellow_viridis = matplotlib.colors.ListedColormap(cm.get_cmap('viridis', 256)(np.linspace(0,0.8,256)))\n",
    "k_formatter = matplotlib.ticker.FuncFormatter(lambda x, _: f'{x:.0f}' if abs(x) < 1000 else f'{x/1000:.0f}k')\n",
    "fold_formatter = matplotlib.ticker.FuncFormatter(lambda x, _: f'{x}x')\n",
    "hours_formatter = matplotlib.ticker.FuncFormatter(lambda x,_: f'{x/3600:.0f}')\n",
    "hours_1f_formatter = matplotlib.ticker.FuncFormatter(lambda x,_: f'{x/3600:.1f}')\n",
    "kb_formatter = matplotlib.ticker.FuncFormatter(lambda x,_: f'{x/0.34/1000:.1f}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "base_model_df = pd.read_parquet(\n",
    "    rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/'fig_base_model_summary.gzip.0')\n",
    ")\n",
    "for i in range(1,8):\n",
    "    base_model_df = base_model_df.append(\n",
    "        pd.read_parquet(rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/f'fig_base_model_summary.gzip.{i}')),\n",
    "        ignore_index=True\n",
    "    )\n",
    "print(f'Loaded {len(base_model_df)} total simulation runs from Parquet cached file')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot BCs\n",
    "fig_bm_palette = {\n",
    "    'fig.bm.tandem_reporter_upstream': main_palette['tandem'],\n",
    "    'fig.bm.tandem_reporter_downstream': main_palette['tandem_alt'],\n",
    "    'fig.bm.convergent': main_palette['convergent'],\n",
    "    'fig.bm.divergent': main_palette['divergent']\n",
    "}\n",
    "df_bm = base_model_df[\n",
    "    (\n",
    "        (np.abs(base_model_df['coeff.sigma_squared'] - 0.025) < 0.001) &\n",
    "        (base_model_df['rates.sc_dependent'] == 1.0)\n",
    "    ) & (\n",
    "        (base_model_df['condition'] == 'fig.bm.tandem_reporter_upstream') |\n",
    "        (base_model_df['condition'] == 'fig.bm.tandem_reporter_downstream') |\n",
    "        (base_model_df['condition'] == 'fig.bm.convergent') |\n",
    "        (base_model_df['condition'] == 'fig.bm.divergent')\n",
    "    ) & (\n",
    "        base_model_df.is_plasmid == 0.0\n",
    "    )\n",
    "].copy()\n",
    "df_bm['condition'] = df_bm['condition'].cat.remove_unused_categories()\n",
    "df_bm['norm_induction'] = df_bm['induction'] / (1.0 / 120.0)\n",
    "df_bm['constant_mRNA'] = df_bm['constant_mRNA'] / 250.0\n",
    "for stage in range(1,4):\n",
    "    mask = (\n",
    "        (df_bm['condition'] == 'fig.bm.tandem_reporter_upstream') |\n",
    "        (df_bm['condition'] == 'fig.bm.tandem_reporter_downstream')\n",
    "    )\n",
    "    if stage >= 2:\n",
    "        mask = mask | (df_bm['condition'] == 'fig.bm.convergent')\n",
    "    if stage >= 3:\n",
    "        mask = mask | (df_bm['condition'] == 'fig.bm.divergent')\n",
    "    sns.lineplot(data=df_bm[mask],\n",
    "                x='norm_induction', y='constant_mRNA', hue='condition',\n",
    "                palette=fig_bm_palette, legend=False, linewidth=4)\n",
    "    sns.despine()\n",
    "    plt.xlim([0.0,2.5])\n",
    "    plt.ylim([0,1.25])\n",
    "    plt.xticks([0.0,1.0,2.0], ['0x', '1x', '2x'])\n",
    "    plt.xlabel('Adjacent gene fold induction')\n",
    "    plt.ylabel('Reporter output')\n",
    "    plt.text(0.18, 1.12, \"Tandem (←)\", color=main_palette['tandem'])\n",
    "    plt.text(0.2, 0.25, \"Tandem (→)\", color=main_palette['tandem_alt'])\n",
    "    if stage >= 2:\n",
    "        plt.text(1.1, 0.4, \"Convergent\", color=main_palette['convergent'])\n",
    "    if stage >= 3:\n",
    "        plt.text(0.8, 0.9, \"Divergent\", color=main_palette['divergent'])\n",
    "    plt.savefig(rd.outfile(rd.rootdir/'img'/f'seminar_linear_bc_bulk_{stage}.svg'), bbox_inches='tight')\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with h5py.File(rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/'sigma_sweep.h5'), 'r') as sigma_h5:\n",
    "    alpha_sweep_df = pd.concat([\n",
    "        pd.DataFrame({\n",
    "            'sigma': sigma_h5['sweep']['sigma'],\n",
    "            'alpha': sigma_h5['sweep']['s2_coeff'][a_idx],\n",
    "            'energy': sigma_h5['sweep']['values'][a_idx,:],\n",
    "            'torque': sigma_h5['sweep']['values'][a_idx,:] / (1.2 * 2 * np.pi)\n",
    "        }) for a_idx in range(len(sigma_h5['sweep']['s2_coeff']))\n",
    "    ])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(8,2.5))\n",
    "for alpha, color in zip([0.0], [main_palette['gray'], main_palette['gray']]):\n",
    "    closest_alpha_df = alpha_sweep_df[\n",
    "        np.abs(alpha_sweep_df.alpha - alpha) == min(np.abs(alpha_sweep_df.alpha - alpha))\n",
    "    ].copy()\n",
    "    plt.plot(closest_alpha_df['sigma'], closest_alpha_df['torque'], color=color)\n",
    "sns.despine(left=False, bottom=True, top=False, right=True)\n",
    "plt.xlim([-0.2,0.2])\n",
    "plt.ylim([-35,35])\n",
    "plt.xticks([-0.2,0.2])\n",
    "plt.yticks([-30,30])\n",
    "alpha_ax = plt.gca()\n",
    "alpha_ax.tick_params(direction='inout', length=10, pad=1)\n",
    "alpha_ax.yaxis.tick_left()\n",
    "alpha_ax.xaxis.tick_top()\n",
    "alpha_ax.spines['top'].set_position('zero')\n",
    "alpha_ax.spines['left'].set_position('zero')\n",
    "alpha_ax.text(0.22, 0, 'σ', verticalalignment='center')\n",
    "alpha_ax.text(0, 40, 'Torque (pN nm)', horizontalalignment='center')\n",
    "plt.savefig(rd.outfile(rd.rootdir/'img'/'seminar_torque.svg'), bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for alphas, label in zip([[0.0], [0.0, 0.025]], ['energy_initial', 'energy_corrected']):\n",
    "    plt.figure(figsize=(8,2.5))\n",
    "    for alpha, color in zip(alphas, [main_palette['light_gray'], 'k']):\n",
    "        closest_alpha_df = alpha_sweep_df[\n",
    "            np.abs(alpha_sweep_df.alpha - alpha) == min(np.abs(alpha_sweep_df.alpha - alpha))\n",
    "        ].copy()\n",
    "        plt.plot(closest_alpha_df['sigma'], closest_alpha_df['energy'], color=color)\n",
    "    sns.despine(left=False, bottom=True, top=False, right=True)\n",
    "    plt.xlim([-0.2,0.2])\n",
    "    plt.ylim([-250,250])\n",
    "    plt.xticks([-0.2,0.2])\n",
    "    plt.yticks([-208.4,208.4], ['-30','30'])\n",
    "    alpha_ax = plt.gca()\n",
    "    alpha_ax.tick_params(direction='inout', length=10, pad=1)\n",
    "    alpha_ax.yaxis.tick_left()\n",
    "    alpha_ax.xaxis.tick_top()\n",
    "    alpha_ax.spines['top'].set_position('zero')\n",
    "    alpha_ax.spines['left'].set_position('zero')\n",
    "    alpha_ax.text(0.22, 0, 'σ', verticalalignment='center')\n",
    "    alpha_ax.text(0, 280, 'Energy (kcal / mol)', horizontalalignment='center')\n",
    "    plt.savefig(rd.outfile(rd.rootdir/'img'/f'seminar_{label}.svg'), bbox_inches='tight')\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bursting_df = pd.read_parquet(\n",
    "    rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/'fig_bursting.gzip')\n",
    ")\n",
    "print(f'Loaded {len(bursting_df)} total simulation runs from Parquet cached file')\n",
    "available_induction = bursting_df[bursting_df.condition == 'fig.bm.sc_bursts.convergent'].induction\n",
    "selected_induction = [np.min(available_induction), 1/120.0, np.max(available_induction)]\n",
    "bursting_df_selected = bursting_df[\n",
    "    ((bursting_df.induction) == selected_induction[0]) |\n",
    "    ((bursting_df.induction) == selected_induction[1]) |\n",
    "    ((bursting_df.induction) == selected_induction[2])\n",
    "]\n",
    "conditions = [f'fig.bm.sc_bursts.{x}' for x in ['convergent', 'divergent', 'tandem_reporter_upstream', 'tandem_reporter_downstream']]\n",
    "bm_bursting_df = bursting_df_selected[\n",
    "    (bursting_df_selected.burst_threshold == 30) &\n",
    "    (bursting_df_selected.inter_rnap_gene == 2) &\n",
    "    (bursting_df_selected.burst_size > 0) &\n",
    "    (bursting_df_selected.burst_duration > 0) &\n",
    "    (bursting_df_selected.inter_burst_time > 0) &\n",
    "    (bursting_df_selected.alpha == 0.025) &\n",
    "    (\n",
    "        ((bursting_df_selected.condition) == conditions[0]) |\n",
    "        ((bursting_df_selected.condition) == conditions[1]) |\n",
    "        ((bursting_df_selected.condition) == conditions[2]) |\n",
    "        ((bursting_df_selected.condition) == conditions[3])\n",
    "    )\n",
    "].copy()\n",
    "bm_bursting_df.condition = bm_bursting_df.condition.cat.remove_unused_categories()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for stage in range(1,5):\n",
    "    sns.kdeplot(data=bm_bursting_df[(bm_bursting_df.condition == 'fig.bm.sc_bursts.convergent') & (bm_bursting_df.induction == 0.0)], x='burst_size',\n",
    "        common_norm=False, color=main_palette['gray'], linewidth=3,\n",
    "        legend=None, log_scale=True)\n",
    "    plt.text(10, 1.1, \"Reporter only\", color=main_palette['gray'])\n",
    "    if stage >= 2:\n",
    "        sns.kdeplot(data=bm_bursting_df[(bm_bursting_df.condition == 'fig.bm.sc_bursts.convergent') & (bm_bursting_df.induction == 1.0 / 120.0)], x='burst_size',\n",
    "            common_norm=False, color=main_palette['convergent'], linewidth=3,\n",
    "            legend=None, log_scale=True)\n",
    "        plt.text(10, 1.0, \"Convergent\", color=main_palette['convergent'])\n",
    "    if stage >= 3:\n",
    "        sns.kdeplot(data=bm_bursting_df[(bm_bursting_df.condition == 'fig.bm.sc_bursts.divergent') & (bm_bursting_df.induction == 1.0 / 120.0)], x='burst_size',\n",
    "            common_norm=False, color=main_palette['divergent'], linewidth=3,\n",
    "            legend=None, log_scale=True)\n",
    "        plt.text(200, 1.2, \"Divergent\", color=main_palette['divergent'])\n",
    "    if stage >= 4:\n",
    "        sns.kdeplot(data=bm_bursting_df[(bm_bursting_df.condition == 'fig.bm.sc_bursts.tandem_reporter_upstream') & (bm_bursting_df.induction == 1.0 / 120.0)], x='burst_size',\n",
    "            common_norm=False, color=main_palette['tandem'], linewidth=3,\n",
    "            legend=None, log_scale=True)\n",
    "        plt.text(150, 1.45, \"Upstream tandem\", color=main_palette['tandem'])\n",
    "        sns.kdeplot(data=bm_bursting_df[(bm_bursting_df.condition == 'fig.bm.sc_bursts.tandem_reporter_downstream') & (bm_bursting_df.induction == 1.0 / 120.0)], x='burst_size',\n",
    "            common_norm=False, color=main_palette['tandem_alt'], linewidth=3,\n",
    "            legend=None, log_scale=True)\n",
    "        plt.text(7, 1.5, \"Downstream tandem\", color=main_palette['tandem_alt'])\n",
    "    plt.xlim([1, 3e3])\n",
    "    plt.ylim([0,1.8])\n",
    "    plt.xlabel('Burst size')\n",
    "    sns.despine()\n",
    "    plt.savefig(rd.outfile(rd.rootdir/'img'/f'seminar_burst_size_{stage}.svg'), bbox_inches='tight')\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for stage in range(1,5):\n",
    "    sns.kdeplot(data=bm_bursting_df[(bm_bursting_df.condition == 'fig.bm.sc_bursts.convergent') & (bm_bursting_df.induction == 0.0)], x='inter_burst_time',\n",
    "        common_norm=False, color=main_palette['gray'], linewidth=3,\n",
    "        legend=None, log_scale=True)\n",
    "    plt.text(160, 1.0, \"Reporter only\", color=main_palette['gray'])\n",
    "    if stage >= 2:\n",
    "        sns.kdeplot(data=bm_bursting_df[(bm_bursting_df.condition == 'fig.bm.sc_bursts.convergent') & (bm_bursting_df.induction == 1.0 / 120.0)], x='inter_burst_time',\n",
    "            common_norm=False, color=main_palette['convergent'], linewidth=3,\n",
    "            legend=None, log_scale=True)\n",
    "        plt.text(400, 0.4, \"Convergent\", color=main_palette['convergent'])\n",
    "    if stage >= 3:\n",
    "        sns.kdeplot(data=bm_bursting_df[(bm_bursting_df.condition == 'fig.bm.sc_bursts.divergent') & (bm_bursting_df.induction == 1.0 / 120.0)], x='inter_burst_time',\n",
    "            common_norm=False, color=main_palette['divergent'], linewidth=3,\n",
    "            legend=None, log_scale=True)\n",
    "        plt.text(30, 1.45, \"Divergent\", color=main_palette['divergent'])\n",
    "    if stage >= 4:\n",
    "        sns.kdeplot(data=bm_bursting_df[(bm_bursting_df.condition == 'fig.bm.sc_bursts.tandem_reporter_upstream') & (bm_bursting_df.induction == 1.0 / 120.0)], x='inter_burst_time',\n",
    "            common_norm=False, color=main_palette['tandem'], linewidth=3,\n",
    "            legend=None, log_scale=True)\n",
    "        sns.kdeplot(data=bm_bursting_df[(bm_bursting_df.condition == 'fig.bm.sc_bursts.tandem_reporter_downstream') & (bm_bursting_df.induction == 1.0 / 120.0)], x='inter_burst_time',\n",
    "            common_norm=False, color=main_palette['tandem_alt'], linewidth=3,\n",
    "            legend=None, log_scale=True)\n",
    "        plt.text(30, 1.38, \"Upstream tandem\", color=main_palette['tandem'])\n",
    "        plt.text(400, 0.8, \"Downstream tandem\", color=main_palette['tandem_alt'])\n",
    "    plt.xlim([15, 5e3])\n",
    "    plt.ylim([0,1.55])\n",
    "    plt.xlabel('Inter-burst time (s)')\n",
    "    sns.despine()\n",
    "    plt.savefig(rd.outfile(rd.rootdir/'img'/f'seminar_inter_burst_time_{stage}.svg'), bbox_inches='tight')\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig_zinani_examples = pd.read_parquet(rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/'fig_zinani_examples.gzip', 'fig_zinani'))\n",
    "print(f'Loaded {len(fig_zinani_examples)} total example datapoints from Parquet cached file')\n",
    "fig_zinani_matlab_examples = pd.read_parquet(rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/'fig_zinani_matlab_examples.gzip', 'fig_zinani'))\n",
    "print(f'Loaded {len(fig_zinani_matlab_examples)} total example datapoints from Parquet cached file')\n",
    "fig_zinani_palette = {\n",
    "    'her1': '#5254a3',\n",
    "    'her7': '#9c9ede',\n",
    "    'matlab.Unpaired': '#b5cf6b',\n",
    "    'matlab.Paired': '#637939',\n",
    "    'fig.zinani.uncoupled': '#b5cf6b',\n",
    "    'fig.zinani.fully-coupled': '#637939',\n",
    "    'fig.zinani.tangles-coupled': main_palette['divergent'],\n",
    "    'gray': main_palette['gray'],\n",
    "    'background_gray': '#DDDDDD'\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_alpha = 0.025\n",
    "# Plot Zinani examples\n",
    "df = fig_zinani_examples[(fig_zinani_examples.condition == 'fig.zinani.uncoupled') & (fig_zinani_examples.alpha == plot_alpha)]\n",
    "idx = df.idx.unique()[1]\n",
    "sns.lineplot(data=df[df.idx == idx], x='time', y='her1_mRNA', color=fig_zinani_palette['her1'])\n",
    "sns.lineplot(data=df[df.idx == idx], x='time', y='her7_mRNA', color=fig_zinani_palette['her7'])\n",
    "sns.despine()\n",
    "plt_ax = plt.gca()\n",
    "plt_ax.set_xlim([0, 15000])\n",
    "plt_ax.set_xticks([0, 3600 * 2, 3600 * 4])\n",
    "plt_ax.xaxis.set_major_formatter(hours_1f_formatter)\n",
    "plt_ax.set_xlabel('')\n",
    "plt_ax.set_ylabel('')\n",
    "plt_ax.set_facecolor(fig_zinani_palette['fig.zinani.uncoupled'] + '55')\n",
    "plt.ylabel('mRNA counts')\n",
    "plt.xlabel('Time (hr)')\n",
    "plt.title('Dimer coupling only')\n",
    "plt.savefig(rd.outfile(rd.rootdir/'img'/'zinani_uncoupled.svg'), bbox_inches='tight')\n",
    "plt.show()\n",
    "\n",
    "matlab_df = fig_zinani_matlab_examples[fig_zinani_matlab_examples.condition == 'matlab.Paired']\n",
    "idx = matlab_df.idx.unique()[1]\n",
    "sns.lineplot(data=matlab_df[matlab_df.idx == idx], x='time', y='her1', color=fig_zinani_palette['her1'])\n",
    "sns.lineplot(data=matlab_df[matlab_df.idx == idx], x='time', y='her7', color=fig_zinani_palette['her7'])\n",
    "sns.despine()\n",
    "plt_ax = plt.gca()\n",
    "plt_ax.set_xlim([0, 15000])\n",
    "plt_ax.set_xticks([0, 3600 * 2, 3600 * 4])\n",
    "plt_ax.xaxis.set_major_formatter(hours_1f_formatter)\n",
    "plt_ax.set_xlabel('')\n",
    "plt_ax.set_ylabel('')\n",
    "plt_ax.set_facecolor(fig_zinani_palette['matlab.Paired'] + '55')\n",
    "plt.ylabel('mRNA counts')\n",
    "plt.xlabel('Time (hr)')\n",
    "plt.title('Transcript coupled (Zinani)')\n",
    "plt.savefig(rd.outfile(rd.rootdir/'img'/'zinani_transcript_coupled.svg'), bbox_inches='tight')\n",
    "plt.show()\n",
    "\n",
    "df = fig_zinani_examples[(fig_zinani_examples.condition == 'fig.zinani.tangles-coupled') & (fig_zinani_examples.alpha == plot_alpha)]\n",
    "idx = df.idx.unique()[1]\n",
    "sns.lineplot(data=df[df.idx == idx], x='time', y='her1_mRNA', color=fig_zinani_palette['her1'])\n",
    "sns.lineplot(data=df[df.idx == idx], x='time', y='her7_mRNA', color=fig_zinani_palette['her7'])\n",
    "sns.despine()\n",
    "plt_ax = plt.gca()\n",
    "plt_ax.set_ylim([0,90])\n",
    "plt_ax.set_xlim([0, 15000])\n",
    "plt_ax.set_xticks([0, 3600 * 2, 3600 * 4])\n",
    "plt_ax.xaxis.set_major_formatter(hours_1f_formatter)\n",
    "plt_ax.set_xlabel('')\n",
    "plt_ax.set_ylabel('')\n",
    "plt_ax.set_facecolor(fig_zinani_palette['fig.zinani.tangles-coupled'] + '55')\n",
    "plt.ylabel('mRNA counts')\n",
    "plt.xlabel('Time (hr)')\n",
    "plt.title('Biophysical and dimer coupling')\n",
    "plt.savefig(rd.outfile(rd.rootdir/'img'/'zinani_biophysically_coupled.svg'), bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def dox_string_to_float(s):\n",
    "    # Handle '0'\n",
    "    if s == '0':\n",
    "        return 0.0\n",
    "    # Handle E-1.5 and friends\n",
    "    if s.startswith('E'):\n",
    "        return 10**float(s[1:])\n",
    "    # Handle BASEeEXP (1e-1.5)\n",
    "    e_split = s.lower().split('e')\n",
    "    if len(e_split) == 2:\n",
    "        return float(e_split[0]) * (10**float(e_split[1]))\n",
    "    raise RuntimeError(f\"Unable to convert dox string: '{s}' to float.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "simulation_df = []\n",
    "for filename in Path(rd.rootdir/'output'/'julia_summaries').glob('*'):\n",
    "    with h5py.File(rd.infile(filename), 'r') as file:\n",
    "        for run in file.values():\n",
    "            simulation_df.append(pd.DataFrame(data={'eGFP': run['final_mRNA'][0,:], 'mRuby2': run['final_mRNA'][1,:]}))\n",
    "            simulation_df[-1]['topo_rate'] = run.attrs['rates.topo']\n",
    "            simulation_df[-1]['induction'] = run.attrs['gene.base_rate'][0]\n",
    "            simulation_df[-1]['condition'] = run.attrs['comment'].decode('utf-8')\n",
    "            simulation_df[-1]['is_plasmid'] = run.attrs['bcs.is_circular']\n",
    "            simulation_df[-1]['sc_dependent'] = run.attrs['rates.sc_dependent']\n",
    "simulation_df = pd.concat(simulation_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import our three datasets\n",
    "n_gene_regex = re.compile(r'^293T\\.(?P<condition>\\d-gene(?:\\.\\w+)?)\\.(?P<dox>.*)-dox\\.(?P<rep>\\d)-rep_Single Cells')\n",
    "control_regex = re.compile(r'^293T\\.(?P<condition>[^.]+)\\.(?P<rep>\\d)-rep_Single Cells')\n",
    "cell_transfection_regex = re.compile(r'^export_(?P<celltype>[^.]+)\\.(?P<condition>\\dg\\.\\w+)\\.(?P<dox>.*)dox\\.r(?P<rep>\\d)_Single Cells')\n",
    "pb_regex = re.compile(r'^export_(?P<celltype>[^.]+)\\.(?P<condition>(\\d[gc]\\.\\w+|\\w+))(.\\d)?\\.(?P<dox>.*)dox\\.r(?P<rep>\\d)_Single Cells')\n",
    "first_transfection_dataset = []\n",
    "control_dataset = []\n",
    "cell_transfection_dataset = []\n",
    "pb_dataset = []\n",
    "for file in Path(rd.rootdir/'output'/'flow'/'2021.05.07-T2.0').glob('*.csv'):\n",
    "    exp_match = n_gene_regex.match(file.stem)\n",
    "    if exp_match is not None:\n",
    "        first_transfection_dataset.append(pd.read_csv(rd.infile(file)))\n",
    "        first_transfection_dataset[-1]['celltype'] = '293T'\n",
    "        first_transfection_dataset[-1]['condition'] = exp_match.group('condition')\n",
    "        dox_str = exp_match.group('dox')\n",
    "        first_transfection_dataset[-1]['dox'] = dox_string_to_float(dox_str)\n",
    "        first_transfection_dataset[-1]['replicate'] = int(exp_match.group('rep'))\n",
    "for file in Path(rd.rootdir/'output'/'flow'/'2021.05.07-T2.0').glob('*.csv'):\n",
    "    control_match = control_regex.match(file.stem)\n",
    "    if control_match is not None:\n",
    "        control_dataset.append(pd.read_csv(rd.infile(file)))\n",
    "        control_dataset[-1]['celltype'] = '293T'\n",
    "        control_dataset[-1]['condition'] = control_match.group('condition')\n",
    "\n",
    "for file in Path(rd.rootdir/'output'/'flow'/'2021.10.18-transfection'/'export').glob('*.csv'):\n",
    "    match = cell_transfection_regex.match(file.stem)\n",
    "    if match is not None:\n",
    "        dataset = pd.read_csv(rd.infile(file))\n",
    "        if dataset['mRuby2-A'].dtype != np.int64:\n",
    "            continue\n",
    "        cell_transfection_dataset.append(dataset)\n",
    "        cell_transfection_dataset[-1]['celltype'] = match.group('celltype')\n",
    "        cell_transfection_dataset[-1]['condition'] = match.group('condition').replace('g.','-gene.')\n",
    "        cell_transfection_dataset[-1]['dox'] = dox_string_to_float(match.group('dox'))\n",
    "        cell_transfection_dataset[-1]['replicate'] = int(match.group('rep'))\n",
    "for file in Path(rd.rootdir/'output'/'flow'/'2021.11.07-flow_tangles_integration').glob('*.csv'):\n",
    "    match = pb_regex.match(file.stem)\n",
    "    if match is not None:\n",
    "        pb_dataset.append(pd.read_csv(rd.infile(file)))\n",
    "        pb_dataset[-1]['celltype'] = match.group('celltype')\n",
    "        # Fixup condition\n",
    "        condition = match.group('condition')\n",
    "        if condition[1] == 'c':\n",
    "            condition = condition[0] + 'g' + condition[2:]\n",
    "        pb_dataset[-1]['condition'] = condition.replace('g.','-gene.')\n",
    "        pb_dataset[-1]['dox'] = dox_string_to_float(match.group('dox'))\n",
    "        pb_dataset[-1]['replicate'] = int(match.group('rep'))\n",
    "first_transfection_dataset = pd.concat(first_transfection_dataset, ignore_index=True)\n",
    "control_dataset = pd.concat(control_dataset, ignore_index=True)\n",
    "cell_transfection_dataset = pd.concat(cell_transfection_dataset, ignore_index=True)\n",
    "pb_dataset = pd.concat(pb_dataset, ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Postprocess\n",
    "mRuby2_control_intensity = sorted(control_dataset[(control_dataset.condition == 'iRFP670') & (control_dataset['iRFP670-A'] > 2e2)]['mRuby2-A'])\n",
    "mRuby2_gate = mRuby2_control_intensity[int(len(mRuby2_control_intensity) * 0.95)]\n",
    "gated_transfection = cell_transfection_dataset[\n",
    "    (cell_transfection_dataset['iRFP670-A'] > 2e2) &\n",
    "    (cell_transfection_dataset['mRuby2-A'] > mRuby2_gate)].copy()\n",
    "gated_initial_dataset = first_transfection_dataset[first_transfection_dataset['iRFP670-A'] > 2e2].copy()\n",
    "gated_pb = pb_dataset[pb_dataset['iRFP670-A'] > 2e2].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_df = gated_transfection[(gated_transfection['mRuby2-A'] > 1) & (gated_transfection['eGFP-A'] > 1)].copy()\n",
    "filter_set = plot_df['dox'].unique()[1::2]\n",
    "plot_df = plot_df[plot_df['dox'].apply(lambda x: x==0 or x in filter_set)]\n",
    "# Filter out the EGFP/mRuby2 conditions\n",
    "plot_df = plot_df[plot_df['condition'].apply(lambda x: x[1] == '-')]\n",
    "plot_df['log mRuby2-A'] = np.log10(plot_df['mRuby2-A'])\n",
    "plot_df['log eGFP-A'] = np.log10(plot_df['eGFP-A'])\n",
    "plot_df['recentered_mRuby2'] = plot_df['log mRuby2-A'] - 2\n",
    "plot_df['recentered_EGFP'] = plot_df['log eGFP-A'] - 1.2\n",
    "g = sns.jointplot(data=plot_df[(plot_df.condition == '3-gene.Divergent') & ((plot_df.dox == 0.0) | (plot_df.dox == np.max(plot_df.dox)))],\n",
    "              x='recentered_EGFP', y='recentered_mRuby2', hue='dox', fill=True, kind='kde', alpha=0.5, legend=None,\n",
    "              palette = {0.0:main_palette['gray'], np.max(plot_df.dox):main_palette['divergent']}, height=1.5)\n",
    "g.ax_marg_x.remove()\n",
    "g.ax_joint.text(0.2, 3.0, 'Reporter only', color=main_palette['gray'])\n",
    "g.ax_joint.text(2.5, 0.6, 'Adjacent\\nexpression', color=main_palette['divergent'])\n",
    "g.ax_joint.set_ylim([0,3.5])\n",
    "g.ax_joint.set_xlim([-0.5,4])\n",
    "g.ax_joint.set_xlabel('Adjacent expression')\n",
    "g.ax_joint.set_ylabel('Reporter expression')\n",
    "sns.despine()\n",
    "plt.savefig(rd.outfile(rd.rootdir/'img'/'exp_divergent_paperscale.svg'), bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_df = simulation_df.copy()\n",
    "plot_df = plot_df[(plot_df['condition'] == '2_gene.divergent') & (plot_df['is_plasmid'] == 1.0) & (plot_df['induction'] < .008) & (plot_df['sc_dependent'] == 1.0)]\n",
    "plot_df['log mRuby2'] = np.log10(plot_df['mRuby2'] + 1)\n",
    "plot_df['log eGFP'] = np.log10(plot_df['eGFP'] + 1)\n",
    "plot_df['recentered_EGFP'] = plot_df['log eGFP'] + 0.5\n",
    "g = sns.jointplot(data=plot_df[(np.abs(plot_df.induction - 1.481e-3) < 1e-4) | (np.abs(plot_df.induction - 6.249e-3) < 1e-4)],\n",
    "              x='recentered_EGFP', y='log mRuby2', hue='induction', fill=True, kind='kde', alpha=0.5, legend=None,\n",
    "              palette = {plot_df.induction.unique()[2]:main_palette['gray'], plot_df.induction.unique()[3]:main_palette['divergent']})\n",
    "g.ax_marg_x.remove()\n",
    "g.ax_joint.text(-0.3, 1.6, 'Reporter-only', color=main_palette['gray'])\n",
    "g.ax_joint.text(2.5, 2.6, 'Adjacent\\nexpression', color=main_palette['divergent'])\n",
    "g.ax_joint.set_ylim([0,3.5])\n",
    "g.ax_joint.set_xlim([-0.5,4])\n",
    "g.ax_joint.set_xlabel('Simulated log adjacent gene')\n",
    "g.ax_joint.set_ylabel('Simulated log reporter gene')\n",
    "sns.despine()\n",
    "plt.savefig(rd.outfile(rd.rootdir/'img'/'sim_divergent.svg'), bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if (rd.datadir/'projects'/'tangles'/'simulations'/'fig_base_model_summary.gzip.0').exists():\n",
    "    base_model_df = pd.read_parquet(\n",
    "        rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/'fig_base_model_summary.gzip.0', 'fig_base_model'), memory_map=True\n",
    "    )\n",
    "    for i in range(1,8):\n",
    "        base_model_df = base_model_df.append(\n",
    "            pd.read_parquet(rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/f'fig_base_model_summary.gzip.{i}', 'fig_base_model'), memory_map=True),\n",
    "            ignore_index=True\n",
    "        )\n",
    "    print(f'Loaded {len(base_model_df)} total simulation runs from Parquet cached file')\n",
    "else:\n",
    "    base_model_df = []\n",
    "    with h5py.File(\n",
    "        rd.infile(rd.datadir/'projects'/'tangles'/'simulations'/'modeling_model_bc_topo_selection_summaries.h5', 'fig_base_model_cache'), 'r') as summaries:\n",
    "        for group in summaries.keys():\n",
    "            group_df = []\n",
    "            for run in summaries[group].values():\n",
    "                if not run.attrs['comment'].decode('utf-8').startswith('fig.bm'):\n",
    "                    continue\n",
    "                group_df.append(pd.DataFrame(data={\n",
    "                    'inducible_mRNA': run['final_mRNA'][0,:],\n",
    "                    'constant_mRNA': run['final_mRNA'][1,:]}))\n",
    "                group_df[-1]['induction'] = run.attrs['gene.base_rate'][0]\n",
    "                group_df[-1]['condition'] = run.attrs['comment'].decode('utf-8')\n",
    "                group_df[-1]['is_plasmid'] = run.attrs['bcs.is_circular']\n",
    "                for term in [\n",
    "                    'rates.sc_dependent', 'coeff.sigma_squared',\n",
    "                    'coeff.mRNA_drag_exponent', 'coeff.mRNA_drag_coeff',\n",
    "                    'rnap.max_velocity', 'rnap.stall_torque', 'rnap.stall_width']:\n",
    "                    group_df[-1][term] = run.attrs[term]\n",
    "                gene_endpoints = sorted(np.concatenate((\n",
    "                    run.attrs['gene.start'],\n",
    "                    run.attrs['gene.end'])))\n",
    "                group_df[-1]['spacing'] = int((gene_endpoints[2] - gene_endpoints[1]) / 0.34)\n",
    "            base_model_df.append(pd.concat(group_df, ignore_index=True))\n",
    "            print(f'Loaded {len(base_model_df[-1])} simulations...')\n",
    "    base_model_df = pd.concat(base_model_df, ignore_index=True)\n",
    "    print(f'Loaded {len(base_model_df)} total simulation runs!')\n",
    "    base_model_df['condition'] = pd.Categorical(base_model_df['condition'])\n",
    "    for i, split in enumerate(np.array_split(base_model_df, 8)):\n",
    "        split.to_parquet(rd.outfile(rd.datadir/'projects'/'tangles'/'simulations'/f'fig_base_model_summary.gzip.{i}', 'fig_base_model_cache'), compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Randomly downsample (replace later)\n",
    "base_model_df = base_model_df.sample(frac=0.3, random_state=2022)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot distribution data\n",
    "fig_joint_dist = plt.figure(figsize=(7.30,3.5))\n",
    "fig_joint_dist_gridspec = gridspec.GridSpec(1, 2, figure=fig_joint_dist,\n",
    "    wspace=0.1, hspace=0.15)\n",
    "fig_joint_dist_subfigures = {\n",
    "    'A': fig_joint_dist.add_subfigure(fig_joint_dist_gridspec[0,0]),\n",
    "    'B': fig_joint_dist.add_subfigure(fig_joint_dist_gridspec[0,1]),\n",
    "}\n",
    "fig_joint_palette = {\n",
    "    'fig.bm.tandem_reporter_upstream': main_palette['tandem'],\n",
    "    'fig.bm.tandem_reporter_downstream': main_palette['tandem_alt'],\n",
    "    'fig.bm.convergent': main_palette['convergent'],\n",
    "    'fig.bm.divergent': main_palette['divergent']\n",
    "}\n",
    "fig_joint_dist_noise_axes = []\n",
    "for is_plasmid, subfig, noise_offset in zip([0,1], [fig_joint_dist_subfigures[x] for x in 'AB'], [[-0.06,0], [0.3,0]]):\n",
    "    df_bm_dist = base_model_df[\n",
    "        (\n",
    "            base_model_df['is_plasmid'] == is_plasmid\n",
    "        ) & (\n",
    "            (base_model_df['condition'] == 'fig.bm.tandem_reporter_upstream') |\n",
    "            (base_model_df['condition'] == 'fig.bm.tandem_reporter_downstream') |\n",
    "            (base_model_df['condition'] == 'fig.bm.convergent') |\n",
    "            (base_model_df['condition'] == 'fig.bm.divergent')\n",
    "        ) & (\n",
    "            np.abs(base_model_df['induction'] - 8.0e-3) < 1e-4\n",
    "        ) & (\n",
    "            (base_model_df['rates.sc_dependent'] == 1.0) &\n",
    "            (np.abs(base_model_df['coeff.sigma_squared'] - 0.025) < 0.001)\n",
    "        )\n",
    "    ].copy()\n",
    "    df_bm_dist['condition'] = df_bm_dist['condition'].cat.remove_unused_categories()\n",
    "    df_bm_dist['norm_reporter']  = df_bm_dist['constant_mRNA'] / 340.0\n",
    "    df_bm_dist['norm_inducible'] = df_bm_dist['inducible_mRNA'] / 340.0\n",
    "    subfig.subplots_adjust(left=0.2, bottom=0.25, right=0.9, top=0.8)\n",
    "    fig_joint_dist_dist_axes = subfig.subplots(ncols=2, sharey=False, gridspec_kw={'wspace': 0.5, 'width_ratios':[5, 1]})\n",
    "    for condition, noise_yloc in zip(['fig.bm.convergent', 'fig.bm.divergent', 'fig.bm.tandem_reporter_downstream', 'fig.bm.tandem_reporter_upstream'], [1,1,0.9,1.1]):\n",
    "        df_plot_dist = df_bm_dist[df_bm_dist.condition == condition]\n",
    "        # Compute distribution noise\n",
    "        ext_noise = np.var((df_plot_dist.norm_inducible + df_plot_dist.norm_reporter) / 2.0)\n",
    "        int_noise = np.var((df_plot_dist.norm_inducible - df_plot_dist.norm_reporter) / 2.0)\n",
    "        fig_joint_dist_dist_axes[1].plot(noise_yloc, int_noise / ext_noise, 'o', color=fig_joint_palette[condition])\n",
    "        sns.kdeplot(data=df_plot_dist,\n",
    "                    x='norm_inducible', y='norm_reporter', color=fig_joint_palette[condition], alpha=0.9, linewidths=0.8,\n",
    "                    ax=fig_joint_dist_dist_axes[0], levels=5 if is_plasmid else 8)\n",
    "    sns.despine(ax=fig_joint_dist_dist_axes[0])\n",
    "    fig_joint_dist_dist_axes[0].set_aspect('equal')\n",
    "    sns.despine(ax=fig_joint_dist_dist_axes[1], bottom=True)\n",
    "    # Configure the noise ratio plot\n",
    "    fig_joint_dist_dist_axes[1].xaxis.set_visible(False)\n",
    "    fig_joint_dist_dist_axes[1].text(0.5, 0.95, 'Noise\\nratio', ha='center', ma='center', transform=fig_joint_dist_dist_axes[1].transAxes)\n",
    "    fig_joint_dist_dist_axes[1].set_xlim([0.8,1.2])\n",
    "    fig_joint_dist_dist_axes[1].set_yscale('log')\n",
    "    fig_joint_dist_dist_axes[0].set_xlim([0.0, 1.0])\n",
    "    fig_joint_dist_dist_axes[0].set_ylim([0.0, 1.0])\n",
    "    fig_joint_dist_dist_axes[0].set_xticks(np.linspace(0.0,1.0,5))\n",
    "    fig_joint_dist_dist_axes[0].set_yticks(np.linspace(0.0,1.0,5))\n",
    "\n",
    "    fig_joint_dist_dist_axes[0].set_xlabel('Inducible gene output')\n",
    "    fig_joint_dist_dist_axes[0].set_ylabel('Reporter output')\n",
    "\n",
    "    o_x = noise_offset[0]\n",
    "    o_y = noise_offset[1]\n",
    "    fig_joint_dist_dist_axes[0].annotate(text='', xytext=(0.25 + o_x, 0.05 + o_y), xy=(0.1 + o_x, 0.2 + o_y),\n",
    "        arrowprops={'facecolor': 'black', 'edgecolor': '#000000ff', 'width': 0.8, 'headwidth': 5.0, 'headlength': 5.0, 'shrink': 0.0, 'capstyle': 'round'})\n",
    "    fig_joint_dist_dist_axes[0].annotate(text='', xytext=(0.25 + o_x, 0.05 + o_y), xy=(0.4 + o_x, 0.2 + o_y),\n",
    "        arrowprops={'facecolor': 'black', 'edgecolor': '#000000ff', 'width': 0.8, 'headwidth': 5.0, 'headlength': 5.0, 'shrink': 0.0, 'capstyle': 'round'})\n",
    "    fig_joint_dist_dist_axes[0].text(0.18 + o_x, 0.05 + o_y, 'int.', horizontalalignment='right', transform=fig_joint_dist_dist_axes[0].transData, fontsize=font_sizes['data_annotation'], fontstyle='italic')\n",
    "    fig_joint_dist_dist_axes[0].text(0.32 + o_x, 0.05 + o_y, 'ext.', horizontalalignment='left', transform=fig_joint_dist_dist_axes[0].transData, fontsize=font_sizes['data_annotation'], fontstyle='italic')\n",
    "    subfig.text(0.25 + o_x, 0.20 + o_y, 'noise', horizontalalignment='center', transform=fig_joint_dist_dist_axes[0].transData, fontsize=font_sizes['data_annotation'], fontstyle='italic')\n",
    "    fig_joint_dist_noise_axes.append(fig_joint_dist_dist_axes[1])\n",
    "fig_joint_dist_subfigures['A'].text(0.5, 0.85, 'Linear boundary conditions', ha='center')\n",
    "fig_joint_dist_subfigures['B'].text(0.5, 0.85, 'Circular boundary conditions', ha='center')\n",
    "fig_joint_dist_noise_axes[0].set_ylim((0.1, 50))\n",
    "fig_joint_dist_noise_axes[0].set_yticks([0.1,1.0,10.0])\n",
    "fig_joint_dist_noise_axes[0].set_yticklabels(['0.1','1.0', '10.0'])\n",
    "fig_joint_dist_noise_axes[1].set_ylim((0.01, 0.5))\n",
    "fig_joint_dist_noise_axes[1].set_yticks([0.01,0.1])\n",
    "fig_joint_dist_noise_axes[1].set_yticklabels(['0.01','0.1'])\n",
    "fig_joint_dist.savefig('test.svg')\n",
    "fig_joint_dist.savefig(rd.outfile(rd.rootdir/'img'/'presentation_joint_distributions.svg'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams['animation.ffmpeg_path'] = r'C:\\Users\\ChemeGrad2019\\Downloads\\ffmpeg-2020-12-09-git-7777e5119a-full_build\\bin\\ffmpeg.exe'\n",
    "#sns.set_context('talk',rc={'font.family': 'sans-serif', 'font.sans-serif':['Helvetica Neue']})\n",
    "ffmpeg_writer = animation.FFMpegWriter(fps=30, codec='libx264', extra_args=['-crf', '18', '-pix_fmt', 'yuv420p'])\n",
    "movie_examples = h5py.File(rd.infile(rd.rootdir/'output'/'seminar_examples.h5'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "movie_examples.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "raw_points = '-1,0 -1,0.5 -0.1,0.5 0,0.5 0.1,0.5 1,0.5 1,0 1,-0.5 0.1,-0.5 0,-0.5 -0.1,-0.5 -1,-0.5 -1,0 0,0'\n",
    "points = np.array([(float(point.split(',')[0]), float(point.split(',')[1])) for point in raw_points.split(' ')])\n",
    "scaled_points = np.copy(points) * 0.6\n",
    "scaled_points[:,1] -= 0.5\n",
    "all_points = np.concatenate((scaled_points, points), axis=0)\n",
    "all_points[:,1] += 0.15\n",
    "ellipse_path = matplotlib.path.Path(all_points,\n",
    "    codes=([matplotlib.path.Path.MOVETO] + [matplotlib.path.Path.CURVE4]*12 + [matplotlib.path.Path.CLOSEPOLY]) * 2\n",
    ")\n",
    "def gen_animation(data_group, gene_colors, title, subtitle, filename, n_frames, dpi=120, mRNA_remap=None, tspan=None):\n",
    "    genes = list(zip(data_group.attrs['gene.start'], data_group.attrs['gene.end']))\n",
    "    gene_directions = [1 - 2 * (gene[0] > gene[1]) for gene in genes]\n",
    "\n",
    "    times = np.array(data_group['time'])\n",
    "    mRNA_data = np.array(data_group['mRNA'])\n",
    "    if tspan is None:\n",
    "        tspan = (0, max(data_group['time']))\n",
    "    max_length = data_group.attrs['bcs.length']\n",
    "    max_gene_length = max([abs(gene[1] - gene[0]) for gene in genes])\n",
    "\n",
    "    fig = plt.figure(figsize=(1900 / dpi,950 / dpi))\n",
    "    fig.text(0.28, 0.9, title, ha='center')\n",
    "    fig.text(0.28, 0.85, subtitle, ha='center', style='italic')\n",
    "    plot_ax = plt.subplot2grid((2,2), (0,0))\n",
    "    sigma_ax = plt.subplot2grid((2,2), (1,0))\n",
    "    mRNA_ax = plt.subplot2grid((2,2), (0,1), rowspan=2)\n",
    "\n",
    "    plot_ax.set_xlim([0,max_length])\n",
    "    plot_ax.set_ylim([-20, 100 + max_gene_length])\n",
    "    plot_ax.spines['top'].set_visible(False)\n",
    "    plot_ax.spines['right'].set_visible(False)\n",
    "    plot_ax.spines['left'].set_visible(False)\n",
    "    plot_ax.tick_params(left=False, labelleft=False)\n",
    "    plot_ax.set_xticks(np.arange(0.0, 12.5, 2.5) * 1000 * 0.34)\n",
    "    plot_ax.xaxis.set_major_formatter(matplotlib.ticker.NullFormatter())\n",
    "\n",
    "    sigma_ax.set_xlim([0,max_length])\n",
    "    sigma_ax.set_ylim([-0.6, 0.6])\n",
    "    sigma_ax.spines['top'].set_visible(False)\n",
    "    sigma_ax.spines['right'].set_visible(False)\n",
    "    sigma_ax.set_xticks(np.arange(0.0, 12.5, 2.5) * 1000 * 0.34)\n",
    "    sigma_ax.xaxis.set_major_formatter(kb_formatter)\n",
    "    sigma_ax.set_xlabel('Distance (kb)')\n",
    "    sigma_ax.set_ylabel('σ')\n",
    "    sigma_ax.text(50, 0.4, 'Overwound')\n",
    "    sigma_ax.text(50, -0.5, 'Underwound')\n",
    "\n",
    "    mRNA_ax.set_xlim(tspan)\n",
    "    mRNA_ax.set_ylim([0,np.max(data_group['mRNA'])])\n",
    "    mRNA_ax.spines['top'].set_visible(False)\n",
    "    mRNA_ax.spines['right'].set_visible(False)\n",
    "    mRNA_ax.set_xlabel('Time (hr)')\n",
    "    mRNA_ax.set_ylabel('# mRNAs')\n",
    "    mRNA_ax.set_xticks([0, 10000, 3600 * 4, 3600 * 8])\n",
    "    mRNA_ax.xaxis.set_major_formatter(hours_1f_formatter)\n",
    "\n",
    "    fig.tight_layout()\n",
    "\n",
    "    sigma_ax.axhline(0, color=main_palette['axis_gray'])\n",
    "    sigma_line = sigma_ax.plot([], [], linewidth=3, color='#3d8c7d')[0]\n",
    "\n",
    "    mRNA_ax.axvspan(10000, max(times), color=main_palette['light_gray'] + '55')\n",
    "    if mRNA_remap is None:\n",
    "        gene_lines = [mRNA_ax.plot([], [], color=color)[0] for gene, color in zip(genes, gene_colors)]\n",
    "    else:\n",
    "        gene_lines = [mRNA_ax.plot([], [], color=color)[0] for gene, color in zip(genes, [gene_colors[mRNA_remap[i]] for i in range(len(gene_colors))])]\n",
    "\n",
    "    xax_transform = plot_ax.get_xaxis_transform()\n",
    "    for gene, gene_dir, color in zip(genes, gene_directions, gene_colors):\n",
    "        base = gene[0] - (max_length * 0.34 * 0.01)\n",
    "        plot_ax.plot([base, base], [20/(120 + max_gene_length), 0.23], linewidth=4, solid_capstyle='projecting', color=color, transform=xax_transform, clip_on=False)\n",
    "        plot_ax.plot([base, base + gene_dir * (max_length * 0.34 * 0.13)], [0.23, 0.23], linewidth=4, solid_capstyle='projecting', color=color, transform=xax_transform, clip_on=False)\n",
    "        plot_ax.add_artist(matplotlib.patches.FancyArrow(base + gene_dir * (max_length * 0.34 * 0.13), 0.23, gene_dir, 0,\n",
    "            width=0.08, head_width=0.08, head_length=max_length * 0.34 * 0.08, color=color, transform=xax_transform, clip_on=False))\n",
    "\n",
    "\n",
    "    plot_ax.plot([0, max_length], [0, 0], color=main_palette['light_gray']+'55', linewidth=14, zorder=0, solid_capstyle='butt')\n",
    "    \n",
    "    gene_plots = [plot_ax.plot(gene, [0, 0], color=color,linewidth=14, zorder=0, solid_capstyle='butt')[0] for gene, color in zip(genes, gene_colors)]\n",
    "\n",
    "    mRNA_lines = matplotlib.collections.LineCollection([], colors='#d18cca')\n",
    "\n",
    "    polymerases = plot_ax.plot([], [], marker=ellipse_path, ls='',\n",
    "                        mec='#3d8c7d', mfc='#82d4bd', mew=2, ms=15, zorder=2.5)[0]\n",
    "\n",
    "    def init():\n",
    "        mRNA_lines.set_linewidth(3)\n",
    "        plot_ax.add_collection(mRNA_lines)\n",
    "        return (*gene_plots, mRNA_lines)\n",
    "    \n",
    "    def animate(i):\n",
    "        t = tspan[0] + (tspan[1] - tspan[0]) * (i / n_frames)\n",
    "        possible_idxes = np.where(times < t)[0]\n",
    "        t_idx = possible_idxes[-1] if len(possible_idxes) > 0 else 0\n",
    "        frame_vals = {k:data_group[k][:,t_idx] for k in ['rnap_location', 'phi', 'mRNA_length', 'mRNA']}\n",
    "        n_polymerases = np.sum(frame_vals['rnap_location'] >= 0)\n",
    "\n",
    "        # Plot mRNA data values\n",
    "\n",
    "        # Plot mRNA values (on the right-hand plot)\n",
    "        gene_time_mask = (times < t)\n",
    "        for i, gene_line in enumerate(gene_lines):\n",
    "            gene_line.set_data(times[gene_time_mask], mRNA_data[i,gene_time_mask])\n",
    "\n",
    "        # Set visibility\n",
    "        polymerases.set_visible(n_polymerases > 0)\n",
    "        mRNA_lines.set_visible(n_polymerases > 0)\n",
    "\n",
    "        # Plot polymerases and genes coming off polymerases.\n",
    "        if n_polymerases > 0:\n",
    "            rnap_locs = np.concatenate((\n",
    "                np.zeros(1,),\n",
    "                frame_vals['rnap_location'][:n_polymerases],\n",
    "                np.ones(1,) * max_length\n",
    "            ))\n",
    "            phi = np.concatenate((\n",
    "                np.zeros(1,),\n",
    "                frame_vals['phi'][:n_polymerases],\n",
    "                np.zeros(1,)\n",
    "            ))\n",
    "            sigma = np.diff(phi) / np.diff(rnap_locs) / -1.85\n",
    "\n",
    "            # Set supercoiling line\n",
    "            sigma_line.set_data(\n",
    "                np.concatenate(([rnap_locs[0]], [loc for loc in rnap_locs[1:-1] for _ in range(2)], [rnap_locs[-1]])),\n",
    "                [s for s in sigma for _ in range(2)]\n",
    "            )\n",
    "\n",
    "            points = np.array([rnap_locs, np.zeros_like(rnap_locs)]).T.reshape(-1,1,2)\n",
    "            segments = np.concatenate([points[:-1], points[1:]], axis=1)\n",
    "            polymerases.set_data(rnap_locs[1:-1], np.zeros_like(rnap_locs[1:-1]))\n",
    "            mRNA_base = np.array([rnap_locs[1:-1], np.zeros(rnap_locs[1:-1].shape)]).T.reshape(-1,1,2)\n",
    "            mRNA_tail = np.array([rnap_locs[1:-1], frame_vals['mRNA_length'][:n_polymerases] / 5]).T.reshape(-1,1,2)\n",
    "            mRNA_segments = np.concatenate([mRNA_base, mRNA_tail], axis=1)\n",
    "            mRNA_lines.set_segments(mRNA_segments)\n",
    "        else:\n",
    "            sigma_line.set_data([0, max_length], [0,0])\n",
    "        return [sigma_line, polymerases, mRNA_lines] + list(gene_lines)\n",
    "    anim = animation.FuncAnimation(fig, animate, init_func=init, frames=n_frames, interval=33, blit=True)\n",
    "    anim.save(str(filename), dpi=dpi, writer=ffmpeg_writer, progress_callback= lambda i,n: print(f'Saving frame {i} of {n}') if i % 250 == 0 else None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gen_animation(movie_examples['tangles_full_run.000018'], [main_palette['tandem'], main_palette['gray']], 'Tandem upstream syntax', 'upstream dominance',\n",
    "    rd.outfile(rd.rootdir/'output'/'presentation_short_tandem_upstream.mp4'), 25*20, tspan=[0,3600*8], mRNA_remap={0:1, 1:0}, dpi=270)\n",
    "gen_animation(movie_examples['tangles_full_run.000022'], [main_palette['gray'], main_palette['tandem_alt']], 'Tandem downstream syntax', 'upstream dominance',\n",
    "    rd.outfile(rd.rootdir/'output'/'presentation_short_tandem_downstream.mp4'), 25*20, tspan=[0,3600*8], dpi=270)\n",
    "gen_animation(movie_examples['tangles_full_run.000025'], [main_palette['gray'], main_palette['convergent']], 'Convergent syntax', 'either-or bursting',\n",
    "    rd.outfile(rd.rootdir/'output'/'presentation_short_convergent.mp4'), 25*20, tspan=[0,3600*8], dpi=270)\n",
    "gen_animation(movie_examples['tangles_full_run.000027'], [main_palette['gray'], main_palette['divergent']], 'Divergent syntax', 'coordinated busrting',\n",
    "    rd.outfile(rd.rootdir/'output'/'presentation_short_divergent.mp4'), 25*20, tspan=[0,3600*8], dpi=270)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load Kasey's lentivirus Cellprofiler dataset\n",
    "lentivirus_primary_objects = pd.read_csv(rd.infile(rd.datadir/'projects'/'tangles'/'images'/'viral_analysisIdentifyPrimaryObjects.csv', 'lentivirus_images'))\n",
    "lentivirus_conditions = pd.read_csv(rd.infile(rd.datadir/'projects'/'tangles'/'images'/'viral_analysisImage.csv')).set_index('ImageNumber')\n",
    "lentivirus_data = lentivirus_primary_objects.join(lentivirus_conditions, 'ImageNumber')\n",
    "lentivirus_data['Intensity_MaxIntensity_Red'] *= 4.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(3,2))\n",
    "sns.kdeplot(data=lentivirus_data, x='Intensity_MaxIntensity_Red', hue='Metadata_Condition', palette={'tandem': main_palette['tandem'], 'divergent': main_palette['divergent']}, legend=None)\n",
    "sns.despine()\n",
    "plt.xlabel('Max cell intensity')\n",
    "plt.text(0.02, 1.7, 'Tandem', color=main_palette['tandem'])\n",
    "plt.text(0.02, 1.5, 'Divergent', color=main_palette['divergent'])\n",
    "plt.savefig(rd.outfile(rd.rootdir/'img'/'presentation_viral_context.svg', 'lentivirus_images'), bbox_inches='tight')\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.7.4 ('env': venv)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
