\documentclass[11pt]{article}
\usepackage[utf8]{inputenc}
\usepackage[labelfont=bf,small]{caption}
\usepackage[compact]{titlesec}
\usepackage{appendix}
\usepackage[noblocks]{authblk}
\usepackage{graphicx,amsmath,booktabs,subcaption,placeins,mathtools,multirow} % All of the classics
\usepackage{tikz}
\usepackage[activate={true,nocompatibility},
    final,
    tracking=true,
    kerning=true,
    spacing=true,
    factor=1100,
    stretch=10,
    shrink=10]{microtype}
    \microtypecontext{spacing=nonfrench}
\usepackage[colorlinks,
    linkcolor=teal,
    citecolor=teal,
    filecolor=teal,
    urlcolor=teal]{hyperref}
\usepackage[margin=1in]{geometry} % 1in margins
\usepackage[version=4]{mhchem}

\DeclareCaptionLabelFormat{withincaption}{#2)}
\captionsetup{subrefformat=withincaption}

% -- Use the Charter fonts as base fonts, fixup math-mode display
\usepackage[charter]{mathdesign}
\usepackage[scaled=.96,osf]{XCharter}
\linespread{1.04}
\usepackage[backend=biber,style=nature,backref=true]{biblatex} %author-year for in-text content
%\hyphenpenalty=750
\usepackage{cleveref}
\crefname{appendix}{supplemental section}{supplemental sections}
%-------------------end standard preamble-------------------------
\newcommand{\units}[2]{\frac{\text{#1}}{\text{#2}}\,}
\newcommand{\unit}[1]{\; \text{#1}\,}

\titlespacing{\section}{0pt}{2ex}{1ex}
\titleformat{\subsection}[display]{\bfseries}{}{0em}{}
\titlespacing{\subsection}{0pt}{0.1em}{0.1em}

\begin{document}
\begin{figure}[h]
    \centering
    \includegraphics{figures/modeling_paper/graphical_abstract}
    \caption{Supercoiling-mediated feedback drives emergent regulatory behaviors that depend on the surrounding genetic context.} \label{fig:graphical_abstract}
\end{figure}
\begin{figure}[h]
    \centering
    \includegraphics[width=\textwidth]{figures/modeling_paper/variable_supercoiling_explanation}
    \caption{In the limit of fast supercoiling relaxation relative to polymerase motion, the supercoiling density is constant in the region between polymerases. Four key variables define the location of each polymerase: it's linear distance \(z\) along the genome, the length of the nascent mRNA transcript \(x\), the rotation of the polymerase \(\theta\), and the local DNA excess twist \(\phi\). The tradeoff between RNAP rotation and DNA rotation generates supercoiling upstream and downstream, with the drag generated by the nascent mRNA primarily balancing the torque caused by generated supercoils. Using an energy model responsive to local supercoiling, we can derive supercoiling-dependent initiation terms to model differential polymerase loading rates.}
    \label{fig:key_variables_diagram}
\end{figure}
\begin{figure}[h]
    \centering
    {\includegraphics[width=\textwidth]{figures/modeling_paper/fig_base_model.pdf}
    \phantomsubcaption\label{fig:base_orientations}
    \phantomsubcaption\label{fig:linear_bc_distributions}
    \phantomsubcaption\label{fig:circular_bc_distributions}
    \phantomsubcaption\label{fig:linear_fold_induction}
    \phantomsubcaption\label{fig:circular_fold_induction}
    \phantomsubcaption\label{fig:intergene_spacing_cartoon}
    \phantomsubcaption\label{fig:reporter_output_by_spacing_fold_induction}}
    \phantomsubcaption\label{fig:base_model_sc_density}
\end{figure}
\begin{figure}
    \ContinuedFloat
    \caption{Supercoiling-dependent polymerase motion and polymerase initiation predict context-dependent circuit behavior.
        \subref{fig:base_orientations} Two genes were placed into four different orientations as a testbed for understanding context-dependence driven by supercoiling. All four orientations include a reporter gene (gray) and an inducible gene (colored).
        \subref{fig:linear_bc_distributions} For equal basal expression rates between the reporter and inducible gene for the systems with linear boundary conditions, gene orientation dramatically affects the expression distributions. The two tandem orientations behave nearly symmetrically, whereas the convergent and divergent orientations show enhanced intrinsic and extrinsic noise, respectively.
        \subref{fig:circular_bc_distributions} With circular boundary conditions and equal basal expression rates, extrinsic noise dominates the expression distributions for all orientations. The convergent and divergent orientations show especially strong extrinsic noise.
        \subref{fig:linear_fold_induction} Reporter output for the linear boundary condition orientations show non-monotonic behavior as a function of adjacent induction. The upstream tandem and divergent orientations
        \subref{fig:circular_fold_induction} For circular boundary conditions, reporter output tends to increase as a function of fold induction.
        \subref{fig:intergene_spacing_cartoon} Modifying the inter-gene spacing tunes circuit behavior by changing the amount of accumulated supercoils needed to affect polymerase stalling and supercoiling-dependent initiation.
        \subref{fig:reporter_output_by_spacing_fold_induction} Deviations in reporter output relative to the maximum intergene-spacing is shown as a function of intergene-spacing. The convergent orientation is most strongly affected by inter-gene spacing.
        \subref{fig:base_model_sc_density} The supercoiling density across the linear constructs are shown as a function of induction of the inducible gene. At 0-fold induction, positive and negative supercoiling accumulates upstream and downstream of the reporter gene, respectively. Context-dependent behavior is demonstrated when the adjacent gene is activated.
    } \label{fig:top:orientation_bc_behavior}
\end{figure}
\begin{figure}[htbp]
    \centering
    {\includegraphics{figures/modeling_paper/fig_sc_behavior}
    \phantomsubcaption\label{fig:sc_examples_convergent_divergent}
    \phantomsubcaption\label{fig:sc_density_convergent_divergent}
    \phantomsubcaption\label{fig:burst_dynamics_convergent_divergent}
    \phantomsubcaption\label{fig:sc_examples_tandem}
    \phantomsubcaption\label{fig:sc_density_tandem}
    \phantomsubcaption\label{fig:burst_dynamics_tandem}
    \phantomsubcaption\label{fig:cross_correlation_cartoon}
    \phantomsubcaption\label{fig:orientation_cross_correlation}
    \phantomsubcaption\label{fig:output_distribution_by_orientation_dynamics}
    \phantomsubcaption\label{fig:noise_by_orientation}
    }
\end{figure}
\begin{figure}[htbp]
    \ContinuedFloat
    \caption{Our supercoiling model predicts emergent coupling and noise behavior at a single-cell level.
        \subref{fig:sc_examples_convergent_divergent} mRNA counts from example simulation runs for the convergent and divergent syntaxes are shown. The adjacent gene (gray) is enabled after ten thousand seconds (2.8 hours).%TODO: add labels above each plot "Convergent", "Divergent"
        \subref{fig:sc_density_convergent_divergent} The mean supercoiling density across the convergent and divergent ensembles is shown before and after adjacent gene activation.
        \subref{fig:burst_dynamics_convergent_divergent} The ensemble distribution of burst size (number of polymerases within a burst) and inter-burst time (gap between successive bursts) is compared for the convergent and divergent syntaxes.
        \subref{fig:sc_examples_tandem} mRNA counts from example runs are shown for the upstream and downstream tandem syntaxes.
        \subref{fig:sc_density_tandem} The mean supercoiling density across the tandem ensembles shows strong positive and negative supercoiling accumulation post-induction. %TODO: move the t<2.8 hr and add more sigma labels.
        \subref{fig:burst_dynamics_tandem} The ensemble distribution of burst size and inter-burst time is compared for the two tandem syntaxes.
        \subref{fig:cross_correlation_cartoon} The cross-correlation of two signals \(f(t), g(t)\) at a time offset \(\tau\) can be calculated by `sliding' one mean-centered signal relative to the other mean-centered and integrating the product of the resulting signals.
        \subref{fig:orientation_cross_correlation} The cross-correlation between the two genes is shown for the equal-induction case across the four syntaxes. The convergent and divergent syntaxes showed the strongest cross-correlation, with the convergent case showing periodic behavior and the divergent showing strong correlated expression.
        \subref{fig:output_distribution_by_orientation_dynamics} Distributions of the reporter output before \textit{(dotted)} and after \textit{(solid)} induction of the adjacent gene shows changes in both the mean and standard deviation due to adjacent expression.
        \subref{fig:noise_by_orientation} Ensemble noise behavior for the four simulated syntaxes is shown by plotting the standard deviation of the reporter gene as a function of time.
    }
    \label{fig:top:single_cell_noise_correlation}
\end{figure}
\begin{figure}[htbp]
    \centering
    {\includegraphics{figures/modeling_paper/fig_toggles}
    \phantomsubcaption\label{fig:toggle_cartoon}
    \phantomsubcaption\label{fig:toggle_basin_stability_over_time}
    \phantomsubcaption\label{fig:toggle_stable_frac_n_2.0}
    \phantomsubcaption\label{fig:toggle_burst_size}
    \phantomsubcaption\label{fig:toggle_basin_stability}
    \phantomsubcaption\label{fig:toggle_vs_topo_rate}
    \phantomsubcaption\label{fig:toggle_half_life_vs_mRNA_deg}
    }
\end{figure}
\begin{figure}[htbp]
    \ContinuedFloat
    \caption{Toggle switches implemented as a mutually-inhibitory pair of genes show context-dependent stability. All plots represent simulations where the Hill coefficient has been set to \(n = 2.0\).
    \subref{fig:toggle_cartoon} Simulated toggle switches are regulated both by a mutually-inhibitory interaction at the mRNA level and via supercoiling-dependent phenomena.
    \subref{fig:toggle_basin_stability_over_time} The ensemble mRNA count distributions are shown as a function of syntax at four selected time points. Initially after the second gene is enabled, most of the simulations remain in the starting basin. As time progresses, the ensemble reaches and fluctuates around an equilibrium determined by circuit syntax.
    \subref{fig:toggle_stable_frac_n_2.0} The stability, measured as the fraction of simulations in the ensemble that have never escaped the initial starting basin, of the four starting states of the system are plotted as a function of time. The convergent and tandem-upstream systems demonstrate the highest stability.
    \subref{fig:toggle_burst_size} Expression burst size distributions is shown as a function of circuit syntax. Syntaxes with high stability also demonstrate high burst sizes.
    \subref{fig:toggle_basin_stability} The fraction of simulation runs that remain in the initial stable basin is plotted for several relevant values of the Hill coefficent \(n\). The tandem toggle switch shows asymmetric state stability, where the state with the upstream gene active is more stable than the opposite state.
    \subref{fig:toggle_vs_topo_rate} The stability of the four starting states of the toggle systems are plotted as a function of topoisomerase relaxation rate. High topoisomerase activity tends to decrease the stability of the system, due to reduced supercoiling accumulation.
    \subref{fig:toggle_half_life_vs_mRNA_deg} The half life at different values of the mRNA degradation rate are shown. As the mRNA degradation rate principally sets the average number of mRNA molecules, high degradation rates lead to systems with low overall mRNA concentration and concordant stochastic instability.
} \label{fig:top:toggle_switch}
\end{figure}
\begin{figure}[htbp]
    \centering
    {\includegraphics{figures/modeling_paper/fig_zinani.pdf}
    \phantomsubcaption\label{fig:her1_her7_cartoon}
    \phantomsubcaption\label{fig:zinani_summary_cartoon}
    \phantomsubcaption\label{fig:zinani_mRNA_behavior}
    \phantomsubcaption\label{fig:zinani_correlation_coeff}
    \phantomsubcaption\label{fig:zinani_oscillation_amplitude}
    \phantomsubcaption\label{fig:zinani_cross_correlation}
    }
\end{figure}
\begin{figure}[htbp]
    \ContinuedFloat
    \caption{The experimental observations of the \textit{her1}-\textit{her7} clock circuit in zebrafish\parencite{zinaniPairingSegmentationClock2021} show good agreement with the predictions of supercoiling-mediated regulation.
        \subref{fig:her1_her7_cartoon} Schematic drawing of the mutually-inhibitory \textit{her1}-\textit{her7} system. Both a \textit{her1-her1} dimer or a \textit{hes6-her7} dimer can bind to either promoter, preventing transcription.
        \subref{fig:zinani_summary_cartoon} Coupling between \textit{her1}, \textit{her7} genes on the same allele appears necessary for proper zebrafish somite formation \parencite{zinaniPairingSegmentationClock2021}. Disruption of this intra-allele coupling leads to mutant phenotypes.
        \subref{fig:zinani_mRNA_behavior} The number of \textit{her1} and \textit{her7} mRNAs is shown in example simulation runs for two coupling conditions. The gene unpaired system retains the dimer-driven regulation but is not simulated with biophysical feedback. The gene paired system adds in supercoiling-driven biophysical coupling. The initial spike of mRNA in the gene-paired case is an artifact of starting from an all-zero initial condition; this initial condition was chosen to match literature modeling work \parencite{zinaniPairingSegmentationClock2021}.
        \subref{fig:zinani_correlation_coeff} The correlation between the \textit{her1} and \textit{her7} mRNA counts is shown for the entire ensemble of the coupling conditions. The gene-paired case shows a larger correlation between the two clock genes than the gene-unpaired case.
        \subref{fig:zinani_oscillation_amplitude} The ensemble oscillation amplitude for the two simulated cases demonstrates that biophysically coupling arising from gene-pairing enhances the oscillation amplitude.
        \subref{fig:zinani_cross_correlation} The ensemble cross-correlation between the \textit{her1} and \textit{her7} mRNA counts as a function of coupling type shows that the biophysically-coupled model has the largest minima and maxima. A large maxima at \(\tau = 0\) combined with large roughly-symmetric minima can support the strong cyclic behavior observed experimentally in zebrafish. % All axes in hours
    } \label{fig:top:her1_her7}
\end{figure}
\end{document}