%\documentclass[3p]{elsarticle} % <- Post Elsevier conversion
\documentclass[11pt]{article} % <- Pre Elsevier conversion
\usepackage[utf8]{inputenc}
\usepackage[labelfont=bf,small]{caption}
\usepackage[compact]{titlesec}
\usepackage{appendix}
\usepackage[noblocks]{authblk} % <- Pre Elsevier conversion
\usepackage{graphicx,amsmath,booktabs,subcaption,placeins,mathtools,multirow} % All of the classics
\usepackage{tikz}
\usepackage[activate={true,nocompatibility},
    final,
    tracking=true,
    kerning=true,
    spacing=true,
    factor=1100,
    stretch=10,
    shrink=10]{microtype}
    \microtypecontext{spacing=nonfrench}
\usepackage[colorlinks,
    linkcolor=teal,
    citecolor=teal,
    filecolor=teal,
    urlcolor=teal]{hyperref}
\usepackage[margin=1in]{geometry} % 1in margins <- Overriden by Elsesvier
\usepackage[version=4]{mhchem}

\DeclareCaptionLabelFormat{withincaption}{#2)}
\captionsetup{subrefformat=withincaption}

% -- Use the Charter fonts as base fonts, fixup math-mode display
% Use osf for small caps
\usepackage[charter]{mathdesign}
\usepackage[scaled=.96,lf]{XCharter}
%\usepackage[scaled=.96,lf]{XCharter}
%\linespread{1.04}
\usepackage[backend=biber,style=nature,backref=true,url=false]{biblatex} %author-year for in-text content
%\hyphenpenalty=750
\usepackage{cleveref}
\crefname{appendix}{supplemental section}{supplemental sections}
%-------------------end standard preamble-------------------------
\newcommand{\units}[2]{\frac{\text{#1}}{\text{#2}}\,}
\newcommand{\unit}[1]{\; \text{#1}\,}

\titlespacing{\section}{0pt}{2ex}{1ex}
\titleformat{\subsection}[display]{\bfseries}{}{0em}{}
\titlespacing{\subsection}{0pt}{0.1em}{0.1em}

% LAYOUT TWEAKS - reduce separation between text and floats
\setlength{\textfloatsep}{8.0pt plus 3.0pt minus 3.0pt}
\setlength{\intextsep}{8.0pt plus 2.5pt minus 2.5pt}

\title{Supercoiling-mediated feedback rapidly couples and tunes transcription}
\author{Christopher P. Johnstone}
\author{Kate E. Galloway}
\affil{Department of Chemical Engineering, MIT, 25 Ames St., Cambridge, MA 02139, USA}
% ^ pre elseiver
%\author[1]{Christopher P. Johnstone}
%\ead{cjohnsto@mit.edu}
%\author[1]{Kate E. Galloway\corref{cor1}}
%\ead{katiegal@mit.edu}
%\affiliation[1]{organization={Department of Chemical Engineering, MIT},
%                addressline={25 Ames St.},
%                city={Cambridge},
%                state={MA},
%                postcode={02139},
%                country={USA}}
%\cortext[cor1]{Corresponding author}
% ^ post elsevier
\date{}

\addbibresource{main_library.bib}

\begin{document}
\maketitle

\begin{abstract} \noindent
    % Basic background
    Transcription induces a wave of DNA supercoiling, altering the binding affinity of RNA polymerases and reshaping the biochemical landscape of gene regulation. 
     % Detailed background
     As supercoiling rapidly diffuses, transcription dynamically reshapes the regulation of proximal genes, forming a complex feedback loop. The resulting intergene coupling may provide a mechanism to control transcriptional variance in engineered gene networks and explain the behavior of co-localized native circuits.
    % General problem
    However, a theoretical framework is needed for integrating  both biophysical and biochemical transcriptional regulation to investigate the role of supercoiling-mediated feedback within multi-gene systems.
    % Main result
    Here, we model transcriptional regulation under the influence of supercoiling-mediated polymerase dynamics, allowing us to identify patterns of expression that result from physical intergene coupling and explore integration of this biophysical model with a set of canonical biochemical gene regulatory systems.  
    % How did the results further the field?
    We find that gene syntax---the relative ordering and orientation of genes---defines the expression profiles, variance, burst dynamics, and intergene correlation of two-gene systems. By applying our model to both a synthetic toggle switch and the endogenous zebrafish segmentation network, we find that supercoiling can enhance or weaken conventional biochemical regulatory strategies such as mRNA- and protein-mediated feedback loops. 
    % Put results in context
    Together, our results suggest that supercoiling couples behavior between neighboring genes, representing a novel regulatory mechanism. 
    % Broader perspective
    Integrating biophysical regulation into the analysis and design of gene regulation provides a framework for enhanced understanding of native networks and engineering of synthetic gene circuits.
    %idk, is this needed here? We could mention silencing, tight gene clusters, multi-gene viruses, etc as possible targets. Also some interesting reflections on transcription not as a products but as a process. Gene A doesn't care that gene B is coding or not, all that matter is that it induces or doesn't some wake of supercoiling, which may explain pervasive noncoding transcription (this might be a discussion point). 
\end{abstract}

\section{Introduction}
Cells coordinate complex behaviors through precise spatiotemporal control of gene expression. To rapidly advance gene and cell-based therapies, synthetic biology aims to harness the power of native biology by constructing synthetic gene regulatory networks capable of dynamically prescribing cellular processes, states, and identities \parencite{chenSyntheticBiologyAdvancing2012,beitzSyntheticGeneCircuits2022,purnickSecondWaveSynthetic2009,elowitzBuildLifeUnderstand2010}.
Synthetic networks process diverse inputs into complex logical and temporal responses \parencite{weinbergLargescaleDesignRobust2017,xieMultiInputRNAiBasedLogic2011,taborSyntheticGeneticEdge2009}.
From oscillators to pulse generators, synthetic circuits can  precisely coordinate dynamic patterns of gene expression across populations of cells to control cell fate \parencite{gardnerConstructionGeneticToggle2000,elowitzSyntheticOscillatoryNetwork2000,strickerFastRobustTunable2008,daninoSynchronizedQuorumGenetic2010,maSyntheticMammalianSignaling2022,parkEngineeringEpigeneticRegulation2019,bashorUsingEngineeredScaffold2008,gallowayDynamicallyReshapingSignaling2013}.
However, rational \textit{de novo} design of synthetic circuits remains challenging. Despite extensive biomolecular modeling, integration of single genetic elements into systems often leads to emergent behaviors, requiring iterative design-build-test cycles to achieve the desired performance \parencite{jonesEndoribonucleasebasedFeedforwardController2020,freiCharacterizationMitigationGene2020,qianResourceCompetitionShapes2017}. Compounding the challenge, transcription exhibits significant extrinsic and intrinsic noise \parencite{toNoiseCanInduce2010,zopfCellCycleDependenceTranscription2013,desaiDNArepairPathwayCan2021}.
In particular, the stochastic nature of transcription makes coordinating expression across multiple genetic elements challenging \parencite{rodriguezIntrinsicDynamicsHuman2019,rodriguezTranscriptionLivingCells2020,quartonUncouplingGeneExpression2020}.
Spatial variation in the nucleus and biochemical dynamics in condensates may contribute to bursting but provide limited parameters for tuning transcriptional noise \parencite{henningerRNAMediatedFeedbackControl2020,guoPolIIPhosphorylation2019}.
Alternatively, mechanical sources of gene regulation offer one potential mechanism by which to understand and harness transcriptional noise to improve the predictable design of gene circuits \parencite{johnstoneEngineeringCellularSymphonies2021,anconaTranscriptionalBurstsNonequilibrium2019a,kimLongDistanceCooperativeAntagonistic2019,elhoudaiguiBacterialGenomeArchitecture2019a,meyerTorsionMediatedInteractionAdjacent2014}.

The mechanical forces of DNA supercoiling powerfully shape transcriptional variance \parencite{desaiDNArepairPathwayCan2021,chongMechanismTranscriptionalBursting2014}.
In the process of transcription, RNA polymerases induce a leading wave of positive DNA supercoiling \parencite{wuTranscriptionGeneratesPositively1988,liuSupercoilingDNATemplate1987}, reshaping the local structure of chromatin \parencite{acharNegativeSupercoilGene2020,tevesTranscriptiongeneratedTorsionalStress2014,naughtonTranscriptionFormsRemodels2013,guoHighresolutionGenomewideMapping2021}.
At the kilobase scale, chromatin structure correlates with gene regulation \parencite{hsiehResolving3DLandscape2020,rowleyEvolutionarilyConservedPrinciples2017, krietensteinUltrastructuralDetailsMammalian2020}.
In yeast and human cells, transcriptionally-induced supercoiling demarks bounds of gene activity \parencite{acharNegativeSupercoilGene2020,naughtonTranscriptionFormsRemodels2013,kouzineTranscriptiondependentDynamicSupercoiling2013}.
In particular, transcriptional activity dictates the strength of contact domains, indicating a role for transcription in forming and maintaining interactions at the kilobase scale \parencite{rowleyOrganizationalPrinciples3D2018,rowleyEvolutionarilyConservedPrinciples2017}.
Together these data suggest that the process of transcription drives formation of supercoiling-linked, kilobase-scale structures that feed back into transcriptional regulation of gene expression.
As supercoiling rapidly diffuses across long distances \parencite{loenhoutDynamicsDNASupercoils2012}, transcriptional activity at one site may impact the overall activity and dynamics of transcription of proximal genes \parencite{sevierPropertiesGeneExpression2018, sevierCollectivePolymeraseDynamics2022,tripathiDNASupercoilingmediatedCollective2021}.
Understanding how supercoiling induces coupling between neighboring genes provides the opportunity to improve the predictable design of transgenic systems from simple reporters to sophisticated dynamic circuits.

% Removed - to be uploaded as a separate file.
%\begin{figure}[h]
%    \centering
%    \includegraphics{figures/modeling_paper/graphical_abstract}
%    \caption{Graphical abstract: Supercoiling-mediated feedback drives emergent regulatory behaviors. Supercoiling-mediated feedback induces syntax-specific expression profiles and transcriptional variance. Via supercoiling-mediated feedback, circuit syntax can modulate toggle switch stability, and dynamically couples transcription through coordinated bursting. } \label{fig:graphical_abstract}
%\end{figure}

Here we develop a model of transcriptional regulation that integrates DNA supercoiling to examine how the orientation and placement of neighboring genes affects expression. Extending from a model of supercoiling-dependent polymerase motion \parencite{sevierPropertiesGeneExpression2018}, we developed our model to include the effects of supercoiling on polymerase binding and initiation. Specifically, we model RNA polymerase binding and initiation as a function of DNA supercoiling, such that underwound DNA favors RNA polymerase binding and overwound DNA limits binding. To extract experimentally testable predictions, we  apply our model to simple two-gene systems that include a constitutive reporter and an inducible gene. Using these two-gene systems, we find that DNA supercoiling strongly influences the profile of gene expression and that influence is defined by syntax---the relative orientation and position of genetic elements---and the enclosing boundary conditions. In addition to regulating the output of simulated genes, supercoiling-dependent feedback tunes the size and frequency of transcriptional bursts. To investigate how these tunable parameters may impact synthetic gene circuits, we applied our model to a canonical gene circuit, a synthetic toggle switch constructed with different syntaxes.
We find that circuit syntax affects the stability of states and sets biochemical parameters required for bistability including repressor cooperativity and RNA stability. Finally, we explored how DNA supercoiling might support transcriptional coordination within the native genome of zebrafish (\textit{Danio rerio}) to enable somite segmentation. We find that DNA supercoiling acts as a mechanism for coordinating expression between divergently expressed genes in the segmentation network. Supercoiling-dependent feedback supports tight regulation of these proximal segmentation genes, providing a molecular mechanism for the precise coordination of gene expression observed during somite formation \parencite{zinaniPairingSegmentationClock2021}. Thus, supercoiling-mediated feedback represents a novel, testable regulatory mechanism that can both explain native behaviors and guide synthetic designs.



% LAYOUT TWEAKS - allow graphical abstract to float into model and methods
\FloatBarrier

\FloatBarrier
\section{Results}
\FloatBarrier

Simulating the behavior of native and synthetic circuits under the influence of transcription-induced feedback requires a model that integrates explicitly-modeled RNA polymerase (RNAP) motion and RNA- and protein-mediated feedback mechanisms.
Our method combines three modeling levels: an ordinary differential equation system that simulates the continuous progression of polymerases loaded onto DNA, a core stochastic system that models supercoiling-dependent polymerase initiation, and a user-specified stochastic layer that allows for simulation of other modes of transcriptional regulation such as the promoter-repressive and -activating interactions that are often included in synthetic circuits (see STAR Methods for details on model development).

As RNA polymerases move along chromatin, positive supercoiling is generated ahead of the polymerase and negative supercoiling is generated behind. Supercoiling generation occurs because the drag of the nascent mRNA rotating with the polymerase is balanced by torque arising from the over- and under-winding of the DNA (\cref{fig:key_variables_diagram}). On the length scale of tens of kilobases, supercoiling diffusion is negligible, causing the supercoiling density, \(\sigma\), to take on a constant value between polymerases. Here, we model supercoiling as affecting gene expression in two ways---excessive supercoiling can both stall polymerases due to excess torque and enhance or suppress polymerase binding at promoter regions due to supercoiling-dependent initiation (\cref{fig:key_variables_diagram}, STAR Methods).

\begin{figure}[htb]
    \centering
    \includegraphics{figures/modeling_paper/variable_supercoiling_explanation}
    \caption{In the limit of fast supercoiling relaxation relative to polymerase motion, the supercoiling density is constant in the region between polymerases and can be calculated from the slope of the linearly-interpolated \(\phi(z)\) graph. The relaxed DNA twist frequency \(\omega_0\) has value \(\omega_0 = 1.85\) radians/nm. Four key variables define the location of each polymerase: its linear distance \(z\) along the genome, the length of the nascent mRNA transcript \(x\), the rotation of the polymerase \(\theta\), and the local DNA excess twist \(\phi\). The tradeoff between RNAP rotation and DNA rotation generates supercoiling upstream and downstream, with the drag generated by the nascent mRNA primarily balancing the torque caused by generated supercoils. Using an energy model responsive to local supercoiling, we can derive supercoiling-dependent initiation terms to model differential polymerase loading rates.}
    \label{fig:key_variables_diagram}
\end{figure}

\subsection{Gene syntax and boundary conditions define DNA supercoiling dynamics, expression profiles, and noise}
\begin{figure}[ht]
    \centering
    {\includegraphics[width=\textwidth]{figures/modeling_paper/fig_base_model.pdf}
    \phantomsubcaption\label{fig:base_orientations}
    \phantomsubcaption\label{fig:bc_induction_sweep}
    \phantomsubcaption\label{fig:circular_bc_distributions}
    \phantomsubcaption\label{fig:linear_bc_distributions}
    \phantomsubcaption\label{fig:base_model_sc_density}
    \phantomsubcaption\label{fig:reporter_output_by_spacing_fold_induction}}
\end{figure}
\begin{figure}
    \ContinuedFloat
    \caption{Supercoiling-dependent feedback induces syntax-specific expression profiles.
        \subref{fig:base_orientations} Two-gene circuits serve as a testbed for investigating supercoiling-mediated feedback. All four syntaxes include a reporter gene (colored) and an inducible gene (gray).
        \subref{fig:bc_induction_sweep} Two classes of boundary conditions are simulated. Linear boundary conditions are simulated with adjacent ``walls'' that prevent supercoiling propagation, whereas circular boundary conditions allow supercoiling generated at one gene to freely affect other genes in either direction around the circle. At right, mean reporter expression is plotted as a function of the level of induction of the adjacent gene for circular and linear boundary conditions. Reporter output is normalized to the uninduced expression case by dividing mRNA counts by a constant value per boundary condition case (10 mRNAs for the circular case and 250 mRNAs for the linear case).
        \subref{fig:circular_bc_distributions} Gene expression distributions of circular-boundary-condition simulations are shown, where the adjacent induced gene is equally induced relative to the reporter gene. For each of the four syntaxes, the expression variance can be decomposed into intrinsic and extrinsic noise components; the ratio of intrinsic to extrinsic noise is shown on the right. Reporter output is normalized by dividing mRNA counts by a constant value (10 mRNAs).
        \subref{fig:linear_bc_distributions} Gene expression distributions and the intrinsic to extrinsic noise ratios are shown for linear boundary conditions. Reporter output is normalized by dividing counts by a constant value (250 mRNAs).
        \subref{fig:base_model_sc_density} The mean supercoiling density of linear constructs are shown as a function of induction of the inducible gene. Induction (colored line) displays syntax-specific behavior compared to the uninduced case (dashed line).
        \subref{fig:reporter_output_by_spacing_fold_induction} Reporter output is shown as a function of intergene spacing (\(\Delta x\)) at five different induction levels. Reporter output is normalized by dividing mRNA counts by a constant value (250 mRNAs).
    } \label{fig:top:orientation_bc_behavior}
\end{figure}

In order to characterize the behavior of supercoiling-mediated feedback, we simulated a series of two-gene systems. These experimentally accessible two-gene systems allow us to test and understand the core design considerations---syntax, relevant boundary conditions, and other experimentally-tunable parameters---within a well-defined and controlled system. Our two-gene systems consist of a reporter gene that is constitutively active and an adjacent, inducible gene placed in either a tandem orientation with the reporter upstream, tandem orientation with the reporter downstream, convergent orientation, or divergent orientation (\cref{fig:base_orientations}).

Varying syntax, we examined how boundary conditions affect the expression profiles of the reporter and inducible genes. The type of boundary condition determines how transcriptionally-generated supercoiling propagates to adjacent genes. Experimentally, plasmid constructs and genomically-integrated cassettes allow for interrogation of circular and linear boundary conditions, respectively.  Plotting reporter expression as a function of adjacent gene induction, we observed that circular boundary conditions show a monotonic increase in reporter output that scales by syntax (\cref{fig:bc_induction_sweep}). With circular boundary conditions, both the convergent and divergent syntaxes show a large enhancement of expression relative to the uninduced case, whereas the tandem cases showed only a slight enhancement. This phenomena occurs because, on a circle, the convergent and divergent syntaxes differ only in the relative lengths of the intergenic and inter-promoter regions; as such, negative supercoiling density will accumulate in the inter-promoter region, enhancing expression in the convergent and divergent syntaxes. In contrast, linear boundary conditions show diverse behaviors. The upstream-tandem and divergent syntaxes maintain high expression levels while the convergent and downstream-tandem syntaxes show decreasing reporter expression with increasing induction of the neighboring gene.

Transcriptional noise substantially contributes to variance of gene expression \parencite{quartonUncouplingGeneExpression2020}, often confounding circuit designs. Thus, designing circuits to respond to or suppress noise may improve circuit performance. To investigate how syntax impacts expression profiles and noise, we simulated the ensemble behavior for circular and linear two-gene systems at equal transcriptional induction (gray dotted line in \cref{fig:bc_induction_sweep}). To examine different forms of noise, we decompose the population variance into an \emph{extrinsic} noise component that describes how ``all'' genes co-vary within a cell and an \emph{intrinsic} noise component that describes the inter-gene variance within a cell. Then, we define the noise ratio as the intrinsic noise divided by the extrinsic noise. For the circular boundary condition simulations, we found that all four ensembles are dominated by extrinsic noise (\cref{fig:circular_bc_distributions}). In contrast, while the tandem syntaxes with linear boundary conditions show approximately equal intrinsic and extrinsic noise, the linear convergent and divergent populations showed diverging shifts in the variance distribution (\cref{fig:linear_bc_distributions}). Despite similar levels of extrinsic noise, the divergent syntax minimizes intrinsic variation between the two genes while the convergent syntax maximizes intergene variation. 
Moving forward, we used the linear set of boundary conditions to analyze system behaviors as we observe the richest set of behaviors under these boundary conditions.

To understand the mechanisms that support syntax-specific expression profiles (\cref{fig:linear_bc_distributions}), we examined the ensemble supercoiling density of our two-gene systems. Putatively, differences in supercoiling across the two-gene systems give rise to differences in RNAP initiation and thus impact gene expression. To observe supercoiling across the two-gene systems, we averaged the supercoiling density across the simulated ensemble. To examine how induction of the adjacent gene changes supercoiling density, we compared the profiles for when the adjacent gene is uninduced (zero-fold induction) and induced (one-fold induction). As expected, the uninduced cases uniformly show that positive supercoiling accumulates downstream of the constitutively-active reporter gene while negative supercoiling accumulates upstream (\cref{fig:base_model_sc_density}, for other induction levels, see \cref{fig:supp:sc_density_induction}). Differences in basal expression of the reporter correspond to differences in supercoiling profile across the different syntaxes. While each uninduced case generates an equal magnitude of positive and negative supercoiling upstream and downstream, the distance from the reporter gene to the boundaries differs depending on syntax, leading to the shorter segment having the higher magnitude of supercoiling \emph{density}. Upon induction of the adjacent gene, supercoiling accumulates within the intergenic regions in a syntax-specific manner.

Supercoiling density determines both supercoiling-dependent initiation and polymerase stalling. Changing the intergene spacing directly tunes the transcriptional activity required to reach a specified supercoiling density and thus reach different expression profiles. To understand the spacing-driven deviations in reporter behavior, we simulated our linear two-gene circuits with different intergene spacings from 500 bp to 10 kb and plotted the reporter output of each circuit (\cref{fig:reporter_output_by_spacing_fold_induction}). We found that the convergent and tandem-upstream syntaxes showed minimal spacing-dependent changes in reporter output; the reporter output in these two syntaxes was mostly dependent on the level of adjacent induction. In contrast, the reporter output in the divergent and tandem-downstream syntaxes depended on inter-gene spacing. With weak adjacent induction in the divergent case (blue curves), decreasing the inter-gene spacing \emph{decreases} reporter output. We attribute this decrease to the effect of polymerase stalling; a small inter-gene spacing increases the local supercoiling density which can lead to polymerase stalling. At high adjacent induction (red curves), strong supercoiling-driven polymerase initiation overcomes the stalling effect, leading to a relatively distance-insensitive divergent reporter output. Finally, in the tandem-downstream syntax, in addition to the strong suppression of reporter output at high adjacent induction, a short inter-gene spacing further reduces reporter output by increasing the supercoiling density. Thus, depending on syntax, supercoiling-dependent feedback can be tuned by changing inter-gene spacing.


% LAYOUT TWEAKS - allow figures to float around within results
%\FloatBarrier
\subsection{DNA supercoiling dynamics confer rapid, tunable coupling between adjacent genes}

\begin{figure}[htbp]
    \centering
    {\includegraphics{figures/modeling_paper/fig_sc_behavior}
    \phantomsubcaption\label{fig:sc_examples}
    \phantomsubcaption\label{fig:sc_density}
    \phantomsubcaption\label{fig:cross_correlation_cartoon}
    \phantomsubcaption\label{fig:orientation_cross_correlation}
    \phantomsubcaption\label{fig:output_distribution_by_orientation_dynamics}
    \phantomsubcaption\label{fig:noise_by_orientation}
    }
\end{figure}
\begin{figure}[htbp]
    \ContinuedFloat
    \caption{Supercoiling-dependent feedback induces dynamic coupling and altered variance between genes.
        \subref{fig:sc_examples} mRNA counts over time from example individual simulations for the four syntaxes are shown. Simulations are initialized with only the reporter gene (colored) active, with the adjacent gene (gray) enabled with equal basal expression after ten thousand seconds (2.8 hours).
        \subref{fig:sc_density} The average ensemble supercoiling density is shown both before and after adjacent gene induction.
        \subref{fig:cross_correlation_cartoon} The cross-correlation of two signals \(f(t), g(t)\) at a time offset \(\tau\) can be calculated by `sliding' one mean-centered signal relative to the other mean-centered signal and integrating the product of the resulting signals.
        \subref{fig:orientation_cross_correlation} The cross-correlation between the two genes is shown for the equal-induction case across the four syntaxes. The convergent and divergent syntaxes showed the strongest cross-correlation, with the convergent case showing periodic behavior and the divergent case showing strong correlated expression.
        \subref{fig:output_distribution_by_orientation_dynamics} Distributions of the reporter output before \textit{(dotted)} and after \textit{(solid)} induction of the adjacent gene show changes in both the mean and standard deviation due to adjacent expression.
        \subref{fig:noise_by_orientation} Ensemble noise behavior for the four simulated syntaxes is shown by plotting the standard deviation of the reporter gene across the ensemble of simulations as a function of time.
    }
    \label{fig:top:single_cell_behavior}
\end{figure}



To explore the impact of DNA supercoiling beyond the ensemble steady state behavior, we investigated the dynamic behavior of our system. To understand the initiation dynamics of the system, we initialized two-gene constructs with only the constitutive reporter gene active. After a settling period (2.8 hours), we induced transcription of the adjacent gene (\cref{fig:sc_examples}).
We observe that expression dynamics vary extensively by syntax. In the convergent syntax, we observe anticorrelated dynamics, with ``either-or'' promoter activity. In contrast, the divergent syntax supports high levels of expression from both genes. When comparing the two tandem orientations, we observe strong biasing of expression towards the upstream gene with stochastic bursts of expression from the downstream gene. While this upstream dominance does not completely disable the downstream gene, activation of the upstream gene reduces the average expression of the downstream reporter. 

To ensure that this observed behavior was not an artifact of enabling adjacent induction later in the simulation, we need to verify that the two-gene ensembles are \emph{ergodic}---not kinetically trapped or otherwise dependent on initial condition. To start, we examined the ensemble expression distribution at the simulation endpoint (\cref{fig:supp:fig_examples_ensemble_behavior}), and found that we recapitulated the population behavior observed in \cref{fig:linear_bc_distributions}. As an additional check, we plotted the ensemble-averaged supercoiling density both before (at 2.3hr) and after (at 11 hours) adjacent gene induction (\cref{fig:sc_density}). If the system is ergodic, then the ensemble behavior observed before and after the adjacent gene is turned on (snapshots of one population separated in time) should match ensemble behavior of simulations where the adjacent gene is time-independently active or inactive (snapshots of separate populations). \Cref{fig:sc_density} does replicate the behavior observed in \cref{fig:base_model_sc_density}, indicating that the two-gene circuit populations are ergodic.
 

To expand our understanding of the emergent supercoiling-dependent dynamics, we examined the temporal correlation of transcription between both genes. To quantify the correlation and extract temporal patterns, we computed the cross-correlation between the gene outputs following induction. The cross-correlation of two signals is itself a function of a time offset; the cross-correlation at some offset time \(\tau\) can be thought of as the Pearson correlation coefficient between the two signals where one has been shifted by \(\tau\) (\cref{fig:cross_correlation_cartoon}, see \cref{sec:supp:cross_correlation} for more detail). In particular, periodic but out-of-phase signals appear as strong negative and positive peaks on a cross-correlation plot.  The time offset of peaks corresponds to the phase offset between the signals.

The four syntaxes show starkly different cross-correlation behaviors (\cref{fig:orientation_cross_correlation}).  The convergent system shows a large anti-correlation at zero time offset with positive correlation peaks at offsets around \(\pm 2\) hours. This combination suggests a periodic but out-of-phase behavior between the two genes with a period of around two hours, confirming that our ensemble behaves similarly to the example simulation in \cref{fig:sc_examples}. In contrast, the divergent syntax shows a strong positive peak at zero time offset, showing strong aperiodic, but correlated, behavior. The tandem syntax showed weak correlation between genes. 

We then examined how adjacent induction affects the reporter output distributions. Visualizing the entire ensemble reporter output distribution per condition in \cref{fig:output_distribution_by_orientation_dynamics} at equal induction, we observe that adjacent induction changes both the mean and variance of the reporter output. In particular, the width of the distributions change before and after induction, suggesting a change in the noise profile. At high adjacent induction, we observe that the upstream-tandem and divergent cases show enhanced transcription, with the downstream-tandem and convergent cases effectively turning off (\cref{fig:supp:sc_distributions_high_alpha_induction}).

As noise impacts the properties of native and synthetic gene networks, we quantified the width of the these distributions by plotting the standard deviation of the ensemble reporter output as a function of time (\cref{fig:noise_by_orientation}). Prior to induction of the second gene, all four systems display similar standard deviations. Syntax differences emerge upon induction of the second gene. We found that syntax strongly modulates the noise behavior of the reporter. In particular, the downstream-tandem and convergent syntaxes show a strong \emph{increase} in noise levels while the upstream-tandem and divergent syntaxes show a small \emph{decrease} in noise levels. These changes in noise may results from differences in transcriptional burst sizes and frequencies.

\begin{figure}[htbp]
    \centering
    {\includegraphics{figures/modeling_paper/fig_bursts.pdf}
    \phantomsubcaption\label{fig:perturbation_overview}
    \phantomsubcaption\label{fig:burst_time_cartoon}
    \phantomsubcaption\label{fig:burst_size_dynamics}
    \phantomsubcaption\label{fig:interburst_time_dynamics}
    }
    \caption{Burst dynamics vary in different models of DNA supercoiling.
        \subref{fig:perturbation_overview} Two model perturbations are compared to the base model, one where polymerase initiation is impossible once the supercoiling density is extreme, and one where nucleosomes provide buffering against positive supercoiling. Each model tunes either the function relating torque and supercoiling density (middle) or the polymerase initiation energy function (bottom); changes relative to the base model are marked with a gray background.
        \subref{fig:burst_time_cartoon} Bursts are defined as a group of consecutive polymerase loading events. The size of a burst is defined as the number of loaded polymerases, whereas inter-burst time is defined as the gap between successive bursts.
        \subref{fig:burst_size_dynamics} The ensemble distribution of burst size is shown for the different orientations for each of the polymerase conditions.
        \subref{fig:interburst_time_dynamics} The ensemble distribution of inter-burst time is shown for the different orientations for each of the polymerase conditions.
    }
    \label{fig:top:bursts}
\end{figure}

%\FloatBarrier
\subsection{Burst dynamics vary in different models of DNA supercoiling }

Transcription occurs in bursts of activity, and native and synthetic mechanisms can modify burst dynamics \parencite{desaiDNArepairPathwayCan2021,chongMechanismTranscriptionalBursting2014,poppAlteringTranscriptionFactor2021}.
In our base model, transcriptional bursting arises from the stochastic addition of polymerases. We hypothesized that burst dynamics may explain the distribution dynamics observed in \cref{fig:output_distribution_by_orientation_dynamics,fig:noise_by_orientation}. While our base model provides a species-agnostic approach to investigate supercoiling-mediated feedback, the mechanical regulation of eukaroytic chromatin may introduce additional complexity and tunability to bursting dynamics. For example, nucleosomes can serve as a reservoir of negative supercoiling by stabilizing the wrapped double-loop of DNA \parencite{leSynergisticCoordinationChromatin2019}.
Furthermore, at hypernegative or hyperpositive supercoiling densities, local chromatin structure may be disrupted causing structures such as R-loops and G-quadruplexes to preferentially form. These structures can block polymerase binding and motion \parencite{stolzInterplayDNASequence2019}.
Even when including our second-order correction term as in \cref{eq:second_order_sc_initation}, our base model predicts favorable supercoiling-dependent polymerase initiation at the hypernegative value of \(\sigma = -0.1\) (\cref{fig:perturbation_overview}). To address these complexities, we implemented two alternate models which extend our torque-response and polymerase initiation energy equations. 

To simulate the effects of structures forming at hypernegative or hyperpositive supercoiling densities, we penalize polymerase initiation by adding a near-infinite positive energy penalty to polymerase initiation at hypernegative and hyperpositive supercoiling densities (\cref{fig:perturbation_overview}) roughly matching the density at which structures such as R-loops form (\(\sigma < -0.06, \sigma > 0.125\), \parencite{stolzInterplayDNASequence2019}). Putatively, these barriers enable us to simulate the formation of stable structures within the DNA that strongly penalize RNAP binding while remaining agnostic to any specific molecular structure. We find that penalization of extreme supercoiling globally reduces supercoiling density (\cref{fig:supp:energy_well_sc_density}) but maintains the qualitative syntax-dependent behavior of the steady-state, linear systems. However, syntax-specific differences are mostly eliminated for circular templates (\cref{fig:supp:energy_well_induction_sweep,fig:supp:energy_well_joint_distribution}).

To model nucleosomes, we updated our torque-response function to phenomenologically match \textit{in vitro} experimental measurements \parencite{leSynergisticCoordinationChromatin2019}. Nucleosomes putatively ``buffer'' the effects of positive supercoiling by unbinding and releasing stored negative supercoiling. We accounted for this buffering by extending a zero-torque plateau within the region \(0 \leq \sigma \leq 0.031\) (\cref{fig:perturbation_overview}). We found that the nucleosome buffering does not significantly impact the steady-state expression profile or supercoiling density observed in \cref{fig:top:orientation_bc_behavior} (\cref{fig:supp:nucleosome_sc_density,fig:supp:nucleosome_induction_sweep,fig:supp:nucleosome_joint_distribution}).


With these three models, we simulated the four linear syntaxes and recorded polymerase initiation events per gene. We then examined the distribution of both \emph{burst size}, which we define as the number of polymerase initiation events during a burst, and \emph{inter-burst time}, the amount of time separating two consecutive bursts for the reporter gene (\cref{fig:burst_time_cartoon}, see \cref{sec:supp:burst_threshold} for more detail). Upon induction of the adjacent gene, we find burst dynamics differ by syntax (\cref{fig:burst_size_dynamics}). In both the base model and the nucleosome-buffering model, the downstream-tandem and convergent syntaxes show a reduced burst size when compared to the divergent and upstream-tandem syntaxes. We attribute this reduction in burst size to accumulated positive supercoiling at the site of the reporter gene's promoter. In contrast, induction of the adjacent gene in the divergent syntax increases the burst size of the reporter gene, putatively due to enhanced loading of polymerases facilitated by accumulated negative supercoiling. With penalization of hypernegative supercoiling density, the divergent syntax shows a decrease in burst size, potentially due a shift in the tradeoff between polymerase stalling and supercoiling-dependent initiation. 

Examining the inter-burst time distributions, induction of the neighboring gene shifts the upstream-tandem and divergent syntaxes to shorter inter-burst times (\cref{fig:interburst_time_dynamics}).  The downstream-tandem and convergent syntaxes shift to longer inter-burst times upon induction of the adjacent gene. Interestingly, this qualitative observation holds \emph{independently} of the choice of torque-energy model, suggesting that changes in inter-burst time---and its inverse, burst frequency---is a syntax-dependent effect that is largely insensitive to additional chromatin perturbations. All in all, these results suggest that syntax provides a method of orthogonally tuning burst frequency.

From these observations, we can now understand the population-level behaviors observed in \cref{fig:output_distribution_by_orientation_dynamics,fig:noise_by_orientation}. 
The increase of noise and reduction of mean expression value in the downstream-tandem and convergent cases occurs concomitant with an \emph{increase} in inter-burst time (\cref{fig:interburst_time_dynamics}). This is expected; if bursts occur rarely, stochastic fluctuations will have an outsized effect on each individual simulation, enhancing population variance. In contrast, the decrease of noise in the upstream-tandem and divergent syntaxes is matched by a decrease in the inter-burst time (\cref{fig:interburst_time_dynamics}). Because bursts happen more frequently, the ensemble reporter output is more stable as we approach the limit of constant transcription.
Taken as a whole, syntax provides a powerful design parameter for inducing and tuning time-dependent behaviors between genes and shaping output gene distributions.

\FloatBarrier
\subsection{Optimizing toggle switch performance and stability through circuit syntax}
\begin{figure}[htbp]
    \centering
    {\includegraphics{figures/modeling_paper/fig_toggles}
    \phantomsubcaption\label{fig:toggle_cartoon}
    \phantomsubcaption\label{fig:toggle_basin_stability_over_time}
    \phantomsubcaption\label{fig:toggle_stable_frac_n_2.0}
    \phantomsubcaption\label{fig:toggle_burst_size}
    \phantomsubcaption\label{fig:toggle_half_life_vs_mRNA_deg}
    }
\end{figure}
\begin{figure}[htbp]
    \ContinuedFloat
    \caption{Toggle switches implemented as a mutually-inhibitory pair of genes show syntax-specific stability.
    \subref{fig:toggle_cartoon} Schematic of a synthetic toggle switch composed of mutual transcriptional repressors, A and B, which are expressed from a promoter negatively regulated by the opposite gene. Repression follows a Hill function (center), which shows cooperativity based on the value of \(n\). Reactions where n is greater than 1 show cooperativity. Simulated toggle switches are regulated both by a mutually-inhibitory interaction at the mRNA level and via supercoiling-dependent phenomena.
    \subref{fig:toggle_basin_stability_over_time} The ensemble mRNA count distributions are shown as a function of syntax at four selected time points.   All plots represent simulations where the Hill coefficient has been set to \(n = 2.0\).
    \subref{fig:toggle_stable_frac_n_2.0} The stability, measured as the percentage of simulations in the ensemble that have never escaped the initial starting basin, of the four starting states of the system plotted as a function of time. 
    \subref{fig:toggle_burst_size} Expression burst size distributions of the initially-active gene A are plotted as a function of circuit syntax. 
    \subref{fig:toggle_half_life_vs_mRNA_deg} The half life at different values of the mRNA degradation rate are shown. As the mRNA degradation rate principally sets the average number of mRNA molecules, high degradation rates lead to systems with low overall mRNA concentration and concordant stochastic instability.
} \label{fig:top:toggle_switch}
\end{figure}

From oscillators to pulse generators, synthetic circuits aim to precisely coordinate dynamic patterns of gene expression. Yet, emergent dynamics, mediated through DNA supercoiling, may support or impede the performance of dynamic circuits. To examine how supercoiling-mediated feedback influences a dynamic circuit, we applied our model to the classic repressor-mediated toggle switch \parencite{gardnerConstructionGeneticToggle2000}. Toggle switches are well-characterized both theoretically \parencite{gardnerConstructionGeneticToggle2000} and experimentally \parencite{gardnerConstructionGeneticToggle2000,yeungBiophysicalConstraintsArising2017,zhuSyntheticMultistabilityMammalian2021}. The behavior of a simple toggle under the additional influence of supercoiling-mediated feedback provides an ideal testbed to understand how syntax influences circuit performance.

A toggle switch can be constructed with two genes that mutually repress each other (\cref{fig:toggle_cartoon}). Ideally, such a toggle switch exhibits bistability, generating two stable basins. If modeled with continuous, noise-free equations, a toggle switch will remain within one of the basins based on the initial conditions \parencite{gardnerConstructionGeneticToggle2000}. However, if we treat the mRNA concentration discretely with a stochastic simulation, the system escapes the stable basin with a certain probability, depending on the size of fluctuations relative to the steady-state values. How does supercoiling-mediated feedback impact this probability of escape? How might supercoiling-mediated feedback interact with mutual transcriptional repression and alter toggle switch behavior?
To answer these questions, we simulate the behavior of a two-gene toggle switch with our model for various circuit syntaxes. To establish the conventional dual repressor system used for toggle switches, we abstracted the regulatory interaction of the repressors using a Hill function to define the base promoter initiation rates:
\begin{equation}
    r_A = r_0 \frac{K_A}{K_A + [B]^n} \qquad r_B = r_0 \frac{K_B}{K_B + [A]^n}
\end{equation}
for \(r_0 = 1/160 \,\text{s}^{-1}\) and \(n = 2.0\). Here, we do not explicitly model protein production. Rather, transcriptional repression directly depends on the mRNA counts of represessors. This parsimonious model allows us to understand the behavior of the system without introducing additional rate constants. We chose the half-max value \(K\)---the mRNA count at which the promoter activity is half that of \(r_0\)---to approximately match the mean steady-state value of either stable state to ensure that the toggle switch operates in the regime of maximum sensitivity (see \cref{sec:supp:choice_of_k} for details). In order to compare the behavior of the toggle switches, we initialized toggle switches of different syntaxes within one of the stable basins (gene A) and induced the second gene (gene B) after 2.8 hours (see \cref{fig:supp_toggles_run_examples} for example runs). The system then evolves under simultaneous mutual inhibition from expression of the repressors as well as from supercoiling-dependent feedback.

Circuit syntax specifies unique toggle switch dynamics that can be understood by visualizing the distribution of mRNA counts over time (\cref{fig:toggle_basin_stability_over_time}). Initially, nearly all of the simulations in the ensemble lie along the axis corresponding to the initially-active gene, gene A. As time progresses, each ensemble approaches and fluctuates around an equilibrium.
The convergent syntax approaches an equilibrium where approximately half of the population distributes into each state. In this syntax, activation of either gene causes positive supercoiling to accumulate in the intergenic region, enhancing negative feedback between genes and thus between states. With divergent syntax, the toggle distributions converge toward monostablity with low differential expression of either gene. Accumulation of negative supercoiling between genes enhances polymerase loading, weakening negative feedback. Finally, the vast majority of the simulations in the tandem syntax remain in or transition to the upstream-active state, demonstrating upstream dominance. We find that these results qualitatively hold for varying values of \(n\), the repressor cooperativity coefficient (\cref{fig:supp:toggles_hill_n_sweep}). We find that  even in the absence of cooperativity, \(n = 1.0\), the convergent syntax shows some level of bistability (\cref{fig:supp:n1_toggle_distributions}), indicating that supercoiling-mediated feedback introduces a degree of nonlinearity that can reinforce toggle switch function. 

To quantify these distribution results, we computed the \emph{stable fraction} of the ensemble, defined as the fraction of simulations that have never left the initial starting basin at a certain simulation time.
The stable fraction monotonically decreases toward zero with time, as simulations that cross into the other stable basin are no longer counted as stable even if they return to the original basin.
We observe substantial syntax differences in the dynamics of the stable fraction of the ensemble (\cref{fig:toggle_stable_frac_n_2.0}). While the tandem orientations represent the extremes of stability, the convergent and divergent syntaxes exhibit intermediate stabilities.
As expected, different burst dynamics characterize toggle switch behaviors, varying by syntax. In particular, we observe that the divergent syntax displays \emph{reduced} burst size when compared to the other syntaxes (\cref{fig:toggle_burst_size}). In contrast, the inter-burst time distributions do not significantly vary with syntax (\cref{fig:supp_toggles_interburst_time}), suggesting that burst size, not burst frequency, is a key parameter tuned by supercoiling-mediated feedback. We hypothesize that the divergent toggle switch may be governed by a conflicting interaction at the promoter level between supercoiling-mediated feedback and mutual inhibition.
Overall, these trends suggest that the observed toggle switch behavior emerges through correlated (or anti-correlated) transcription and changes in burst size, suggesting that toggle switch behavior could be tuned orthogonally by supercoiling-mediated feedback.

For any stochastic system, the steady-state number of molecules influences the stability of the system. As the reservoir of molecules grows larger, the size of fluctuations relative to the total concentration decreases. For toggle switches, we expect that as the number of steady-state mRNA molecules grows, we should approach the theoretical, continuous solution that predicts that no state-switching occurs. To examine this expectation, we modified the simulated mRNA degradation rate, scaling \(K\) as described in \cref{sec:supp:choice_of_k}, and plotted the resulting half-lives (\cref{fig:toggle_half_life_vs_mRNA_deg}).
As the mRNA degradation rate goes to zero, we increase the reservoir size and observe that the half life for all syntaxes approaches the simulation upper-limit on the half life (\cref{fig:toggle_half_life_vs_mRNA_deg}).  Interestingly, increasing mRNA degradation rates reduces the asymmetry in the half-lives between the tandem upstream and tandem downstream syntaxes. These results suggest that state switching increases as mRNA degradation rate increases as expected.

%Finally, topoisomerases regulate DNA topology including DNA supercoiling in the genome, reducing genomic stress.
%By relaxing accumulated supercoiling across the genome, activity of topoisomerases decreases the average amount of supercoiling. Thus, simulated topoisomerase activity may temporarily relax constraints generated through transcriptionally-induced DNA supercoiling.
%We examined how the rate of topoisomerase relaxation events affected the stability of the simulated toggle switches. Divergent toggle switches appear insensitive to a large range of topoisomerase rates (\cref{fig:toggle_vs_topo_rate}). We observe that the most stable systems---the upstream tandem and convergent syntaxes---exhibit topoisomerase-sensitive behavior, showing a minor increase in stability at high topoisomerase rates. Thus, while topoisomerases play an important role across the genome, supercoiling-dependent regulatory behavior may be relatively insensitive to topoisomerase activity.

% LAYOUT TWEAKS - allow figures to float around within results
%\FloatBarrier
\subsection{DNA supercoiling tightly coordinates expression of proximal segmentation genes}

DNA supercoiling provides a mechanism for the precise coordination of colocalized genes. Through colocalization, native circuits may incorporate transcription-linked feedback mechanisms to reduce noise and tune cell-state specific output in tightly-regulated, dynamic processes such as somite formation. In zebrafish, proper somite segmentation requires precise coordination of two clock genes, \textit{her1} and \textit{her7}.   \textit{her1} and \textit{her7} form an inhibitory feedback loop encoded in a divergent syntax (\cref{fig:her1_her7_cartoon}). Proper somite formation requires one intact allele of \textit{her1} and \textit{her7}, provided these genes are expressed from the same locus \parencite{zinaniPairingSegmentationClock2021}.  Mutant zebrafish embryos where \emph{her1} and \emph{her7} are only expressed from separate loci eliminate any supercoiling-mediated coupling while retaining the dimer-mediated inhibitory feedback loop. \Textcite{zinaniPairingSegmentationClock2021} found that in the gene-unpaired embryos, transcriptional coordination between genes is lost and proper somite segmentation is disrupted (\cref{fig:zinani_summary_cartoon}). Consequently, physical colocalization represents an important feature supporting transcriptional coordination between genes and proper somitogenesis, which may be mediated by supercoiling-mediated feedback.

\begin{figure}[htbp]
    \centering
    {\includegraphics{figures/modeling_paper/fig_zinani.pdf}
    \phantomsubcaption\label{fig:her1_her7_cartoon}
    \phantomsubcaption\label{fig:zinani_summary_cartoon}
    \phantomsubcaption\label{fig:zinani_mRNA_behavior}
    \phantomsubcaption\label{fig:zinani_cross_correlation}
    \phantomsubcaption\label{fig:zinani_correlation_coeff}
    \phantomsubcaption\label{fig:zinani_oscillation_amplitude}
    }
\end{figure}
\begin{figure}[htbp]
    \ContinuedFloat
    \caption{Supercoiling-mediated feedback supports robust transcriptional coordination within the zebrafish segmentation clock circuit.
        \subref{fig:her1_her7_cartoon} Schematic of the mutually-inhibitory \textit{her1}-\textit{her7} system. Either a \textit{her1-her1} dimer or a \textit{hes6-her7} dimer can bind to either promoter, preventing transcription of the downstream gene.
        \subref{fig:zinani_summary_cartoon} Coupling between \textit{her1}-\textit{her7} genes on the same allele supports proper zebrafish somite formation \parencite{zinaniPairingSegmentationClock2021}. Disruption of this intra-allele coupling through unpaired mutations in \textit{her1}-\textit{her7} leads to loss of proper segmentation.
        \subref{fig:zinani_mRNA_behavior} Example simulations of \textit{her1} and \textit{her7} mRNA levels are shown as a function of gene syntax and topoisomerase activity. The initial spike of mRNA in the gene-paired, topoisomerase-active case is a result of starting from an initial condition without repressors; this initial condition was chosen to match literature modeling work \parencite{zinaniPairingSegmentationClock2021}.
        \subref{fig:zinani_cross_correlation} The ensemble cross-correlation between the \textit{her1} and \textit{her7} mRNA counts is shown across pairing and topoisomerase conditions.  The large maxima at \(\tau = 0\) combined with large roughly-symmetric minima observed in the gene-paired cases signal the strong cyclic behavior observed experimentally in zebrafish.
        \subref{fig:zinani_correlation_coeff} The distribution of correlation coefficients between the \textit{her1} and \textit{her7} mRNA counts is shown for the various pairing and topoisomerase conditions. 
        \subref{fig:zinani_oscillation_amplitude} The oscillation amplitude over the ensemble is shown for the various conditions. 
    } \label{fig:top:her1_her7}
\end{figure}
Based on our above results from two-gene systems, we hypothesized the feedback from DNA supercoiling supports coordination between discrete transcripts expressed from divergent promoters. Using our full computational model, we replicated the previously developed stochastic reaction network. Importantly, \textit{her1} and \textit{her7} are regulated in a binary fashion; the promoters are either completely off when bound by a dimer or expressed at their basal rate when unbound. We simulated two cases: an \emph{unpaired} system where the simulated genes were separated by a large distance (1Mb) to prevent supercoiling interactions, and a \emph{gene paired} system where \textit{her1} and \textit{her7} were spaced at their genomically-active locations. In the paired system, linear boundary conditions were used with boundaries chosen at the nearest adjacent genes on each side in the zebrafish genome.

Strikingly, the gene-paired case shows strong periodic levels of mRNA expression (\cref{fig:zinani_mRNA_behavior}). In fact, such levels of periodicity are not observed even in the original computational model presented by \textcite{zinaniPairingSegmentationClock2021} (\cref{fig:supp:zinani_matlab_mRNA,fig:supp:zinani_matlab_cross_correlation}); we confirmed that this is not simply an artifact of the uniform time resampling performed in order to compare our model behavior to the literature model (\cref{fig:supp:zinani_matlab_correlation}). The level of periodicity appears sensitive to the second-order polymerase initiation model. Performing simulations below the behavior transition at \(\alpha = 0.02\) (see \cref{fig:top:orientation_bc_behavior,fig:top:alpha_sweep}), the system displays weak periodicity (\cref{fig:supp:zinani_alpha_mRNA,fig:supp:zinani_alpha_cross_correlation,fig:supp:zinani_alpha_correlation,fig:supp:zinani_alpha_oscillation_amplitude}). 

Examining the time-dependent nature of the \textit{her1}-\textit{her7} system, we plotted the ensemble cross-correlation for paired and unpaired genes (\cref{fig:zinani_cross_correlation}). Here, we found that in addition to the enhanced positive correlation peak at a time delay of \(\tau = 0\) seconds, the paired case showed exceptionally strong, nearly-symmetric cross-correlation at positive and negative time offsets. Such cross-correlation is the hallmark of a periodic signal. Thus, both individual examples (\cref{fig:zinani_mRNA_behavior}) and ensemble behavior (\cref{fig:zinani_cross_correlation}) show that supercoiling-mediated feedback provides a strong mechanistic driver of intergene coordination in the \textit{her1}-\textit{her7} clock circuit that is inaccessible to solely dimer-mediated regulation.

In order to confirm that these results apply across the ensemble, we examined the ensemble correlation between the counts of \textit{her1} and \textit{her7} mRNA (\cref{fig:zinani_correlation_coeff}). We found that while the unpaired case shows minimal correlation, the biophysically-coupled case shows strong correlation between the two clock genes. We attribute this strong, periodic correlation to the additional biophysical coupling conferred by the divergent syntax. Notably, we observed that our model predicts an increase in the amplitude of oscillations (\cref{fig:zinani_oscillation_amplitude}). \textit{In vivo}, loss of gene pairing reduces oscillation amplitude, leading to improper segmentation \parencite{zinaniPairingSegmentationClock2021}. 

Because this periodic behavior depends on biophysical coupling, we investigated if the periodicity was robust to topoisomerase activity, nucleosome buffering, or hypernegative chromatin structure formation. We found that intergenic topoisomerase relaxation did not abrogate the periodic behavior, with strong correlated oscillations still visible (\cref{fig:zinani_cross_correlation,fig:zinani_correlation_coeff,fig:zinani_oscillation_amplitude}). However, intergenic topoisomerase activity did reduce the cross-correlation after a few periods, indicating that topoisomerase activity contributes to variance in oscillation frequency while still supporting strongly correlated \emph{her1} and \emph{her7} expression. Intragenic topoisomerase activity showed similar behaviors (\cref{fig:supp:intragenic_cross_correlation}). Finally, we found that while penalization of hypernegative supercoiling densities eliminated periodic behavior (\cref{fig:supp:energy_well_zinani_examples,fig:supp:energy_well_zinani_crosscorr,fig:supp:energy_well_zinani_correlation_coefficient,fig:supp:energy_well_zinani_oscillation_amplitude}), periodicity was maintained in the presence of nucleosome buffering (\cref{fig:supp:nucleosome_zinani_examples,fig:supp:nucleosome_zinani_crosscorr,fig:supp:nucleosome_zinani_correlation_coefficient,fig:supp:nucleosome_zinani_oscillation_amplitude}).
Thus, we propose that supercoiling-mediated feedback offers a mechanism to support robust oscillations in the \textit{her1} and \textit{her7} network for proper somite formation. 


\FloatBarrier
\section{Discussion}
%big picture resummary broad to specific
Transcription induces significant variance in gene expression. At a single-cell level, individual genes are expressed stochastically, with most genes experiencing relatively long periods of quiescence punctuated by bursts of polymerase activity. Phenomenological models of this process based on stochastic probability distributions can provide some insights, but defining the mechanically-regulated physical factors that influence RNA polymerase dynamics  will improve existing models of gene regulation and support enhanced design of transgenic systems. Thus, we aimed to integrate a biophysical model of DNA supercoiling into RNA polymerase binding and transcription to examine how DNA supercoiling-mediated feedback influences patterns of gene expression in two-gene systems. Importantly, we sought to use this model to define a set of experimentally-testable predictions as well as lay the groundwork for future modeling across multigenic loci and circuits. 
 
%what we did
In this work, we developed a model of supercoiling-mediated feedback that captures emergent coupling between neighboring genes to influence expression levels as well as dynamics. This model allowed us to tractably compute polymerase activity at the scale of synthetic circuits (\cref{fig:top:orientation_bc_behavior}). Within supercoiling-mediated feedback, we included both supercoiling-dependent polymerase motion terms and supercoiling-dependent polymerase initiation terms. Importantly, this computational framework lays the groundwork for understanding how DNA supercoiling functions as a regulatory mechanism that can be integrated with canonical biochemical models of gene regulation. Using this model, we extracted insights into how mechanical and biochemical regulation combine to generate diverse profiles of expression and support or impede the performance of gene networks.

%what we found..
%[First sentence summarizes fig 3 and 4 then goes into supporting detail; need to discuss fit with existing literature modeling and/or relevant experimental systems]
We find that induction of neighboring genes significantly influences the transcriptional activity of both genes (\cref{fig:top:orientation_bc_behavior}). Syntax-specific differences in DNA supercoiling dynamics, expression profiles, and noise emerge due to physical coupling. We find that such a system regulated by this biophysical coupling is responsive to expression level changes in adjacent genes, with both mean expression and population variance changing as a function of gene orientation (\cref{fig:top:single_cell_behavior}). The observed supercoiling-mediated feedback is itself dependent on inter-gene spacing, mRNA degradation rate, and other variables tunable in an experimental setting. Generally, accumulated negative supercoiling leads to correlated bursting, which occurs concomitant to a decrease in intrinsic noise. In contrast, accumulated positive supercoiling can lead to anti-correlated bursting, which instead enhances intrinsic noise. While the tandem syntax does not lead to large supercoiling accumulation in the intergenic region, we observe upstream dominance, where the upstream gene is more highly expressed than the downstream gene.

Our prediction of burst dynamics (\cref{fig:top:bursts}) complements theoretical and experimental investigations of cooperative interactions of RNA polymerases arising from the beneficial cancellation of positive and negative supercoiling generated by adjacent polymerases \parencite{sevierCollectivePolymeraseDynamics2022,kimLongDistanceCooperativeAntagonistic2019}. We predict that syntax causes a two- to three-fold change in burst size (\cref{fig:burst_size_dynamics}), but can shift inter-burst time by an order of magnitude (\cref{fig:interburst_time_dynamics}). In addition, the syntax-specific trends in inter-burst time are not strongly affected by other potential mechanical regulators of eukaroytic chromatin---nucleosomes and strained structures such as R-loops---suggesting that syntax may robustly control gene expression through differences in burst dynamics.
While we do not directly examine transcription elongation rates, we similarly predict syntax-specific differences in expression dynamics, but observe distinct syntax-specific behaviors in our model \parencite{sevierCollectivePolymeraseDynamics2022,tripathiDNASupercoilingmediatedCollective2021}.  We also find that intergenic distance only weakly affects supercoiling feedback \parencite{tripathiDNASupercoilingmediatedCollective2021}.
 In alignment with experimental work, we find that  positive supercoiling accumulates in the intergenic region of convergently-oriented native genes \parencite{guoHighresolutionGenomewideMapping2021}. When combined with sequencing methods that precisely measure nascent mRNA transcription \parencite{mellorInterleavedGenome2016}, these methods may provide a window into experimental systems in order to test theoretical predictions of our work and others.


%reflections on circuit tuning
 The fast timescale of supercoiling-mediated feedback offers access to a uniquely tunable and orthogonal form of gene regulation. In contrast to regulatory mechanisms dependent on relatively long timescales, such as mRNA- and protein-mediated systems, supercoiling-mediated feedback occurs at the timescale of seconds. Polymerases can stall and unstall each other within seconds, while local polymerase loading rates can vary over the course of minutes. By combining the fast dynamic feedback with slower classic feedback mechanisms, circuit regulation can be selectively stabilized or destabilized. We found that specification of syntax within a simple two-gene toggle switch generated diverse behaviors including: a reasonably stable switch, a hypersensitive toggle with hysteresis, and an asymmetric system that preferentially decays towards a single target state (\cref{fig:top:toggle_switch}).
As a rapid mechanism for coordinating transcriptional dynamics, supercoiling-dependent feedback may support intergenic coordination in native systems. Examining the zebrafish segmentation clock, we find that addition of supercoiling-mediated feedback recapitulates the synchronized, periodic expression of the clock genes, \textit{her1-her7} (\cref{fig:top:her1_her7}). 

Our model integrates supercoiling-mediated biophysical feedback with classic gene regulation motifs that are well-studied in native and synthetic contexts. This unified framework brings us closer to an understanding of how supercoiling contributes to transcriptional regulation. We offer testable predictions about the performance of genetic circuits. The predicted changes in reporter output, supercoiling density, and burst dynamics observed in \cref{fig:top:orientation_bc_behavior,fig:top:single_cell_behavior,fig:top:bursts} are experimentally accessible with modern sequencing and single-cell imaging technology \parencite{guoHighresolutionGenomewideMapping2021,mellorInterleavedGenome2016,patelDNASupercoilingRestricts2022}. Experimental verification of our theoretical results will aid in constructing a mechanistic understanding of how transcription-induced supercoiling couples expression. Harnessing these insights will enable gene regulation at the level of transcription, providing a robust method to control expression dynamics, levels, and noise.

%limitations; this paragraph should explicitly state the limitation of the model as we see it that way the reviewer knows that we know and doesn't say "what about this? you're neglecting this important thing"
% model RNAP binding, initiation, elongation as one step (and it's more complicated, but we chose a simple model to begin). 
%no explicit modeling of nucleosomes(which you do address below already). 
%limits-sensitivity to alpha; alpha might not be constant across the genome and may vary by sequence, nucleosome density, and occupancy of SMC complexes. 
\subsection{Limitations of the study}

In deriving our model, we made several simplifying assumptions. Our derivation of the energy function for supercoiling-dependent polymerase initiation adds relevant molecular detail to our model. The second-order correction term reflects the asymptotic relationship observed in \textit{in vitro}  assays between torque and supercoiling density for underwound DNA \parencite{leSynergisticCoordinationChromatin2019}. Inclusion of this correction supports the periodic behavior of the native \emph{her1-her7} clock circuit, suggesting this term may accurately capture regulation \textit{in vivo}. For simplicity, we model the dynamic processes of RNAP binding, initiation, and pause release as a single reaction. In real biological systems, each of these processes may vary across the genome by sequence and by the presence of nucleosomes, transcription factors, and other DNA-binding proteins. Formation of supercoiled structures shows sequence bias \textit{in vitro} which may impact \textit{in vivo} structures and gene regulation \parencite{kimDNASequenceEncodes2018}.  While we do approximate the behavior of the system in the presence of hypernegative/hyperpositive chromatin structures and nucleosomes, we do not explicitly model the presence of these and instead perturb our torque and polymerase binding energy functions. This implicit simulation method may not accurately capture the discrete nature of these phenomena, especially in the case of positive supercoiling waves displacing nucleosomes. Furthermore, while real topoisomerases dynamically relax chromatin and are often recruited to sites of active transcription \parencite{baranelloRNAPolymeraseII2016}, we use a simplified model that instantly relaxes DNA in a non-specific manner. We find that our systems are mostly insensitive to this simple model of topoisomerase activity; future work using a different topoisomerase activity model may reveal additional roles for supercoiling relaxation. Additionally, our model excludes polymerase collision, premature termination, and the impact of 3D structures and loop domains formed by protein complexes such as CCCTC-binding factor (CTCF) and other structural maintenance of chromosomes (SMC) proteins. More broadly, we assume that regions of simulated chromatin are uniformly accessible and have equal torque responses during polymerase elongation; these assumptions may fail at the boundaries of chromatin domains. Finally, we also neglect the speed of supercoiling diffusion. While this is expected to be a negligible effect at the scale considered in this work, supercoiling diffusion remains slow at the scale of hundreds of kilobases to megabases.


%future directions

%big take-home message (if they can only remember one what should they know and remind them why important; 3 sentence)


\section{STAR Methods}

\subsection{Resource availability}

\subsubsection{Lead contact}
Further information and requests for resources, datasets, and code should be directed to and will be fulfilled by the lead contact, Kate E. Galloway (katiegal@mit.edu).
\subsubsection{Materials availability}
This study did not generate new unique reagents.
\subsubsection{Data and Code Availability}
\begin{itemize}
    \item All unprocessed and preprocessed simulation data reported in this paper has been deposited at Zenodo and is publicly available as of the date of publication. DOIs are listed in the key resources table.
    \item All original code has been deposited at Zenodo and is publicly available as of the date of publication. DOIs are listed in the key resources table.
    \item Any additional information required to reanalyze the data reported in this paper is available from the lead contact upon request.
\end{itemize}

\subsection{Method details}


\subsubsection{Supercoiling-dependent transcription model}

Here, supercoiling is defined as the amount of excess twist \(\phi\) relative to relaxed DNA. Relaxed DNA rotates  one full revolution per \(\approx10\) basepairs (1bp \(\approx 0.34\) nm); thus its relaxed twist is \(\omega_0 = 1.85\) radians/nm. Supercoiling density can generally be defined as a varying function of genomic location \(\sigma(z)\). However, in a region of constant supercoiling density, we can use the excess DNA twist \(\phi_i, \phi_{i+1}\) at the endpoints \(z_i, z_{i+1}\) to define the supercoiling density as:
\begin{equation}
    \sigma = \frac{\phi_i - \phi_{i+1}}{\omega_0 (z_{i+1} - z_{i})}
\label{eq:supercoiling_density}
\end{equation}
For a polymerase with \(\phi_1 > 0\) between endpoints with \(\phi_0 = \phi_2 = 0\), \cref{eq:supercoiling_density} implies that the supercoiling density is \emph{positive} in front of the polymerase and \emph{negative} behind the polymerase (\cref{fig:key_variables_diagram}). Following on the work of \textcite{sevierPropertiesGeneExpression2018}, we assume that on the length scales of synthetic and native circuits of interest (\(O(\approx10 \unit{kb})\)), the supercoiling density is constant in all regions between polymerases and other barriers. Because supercoiling diffusion and plectoneme hopping \parencite{loenhoutDynamicsDNASupercoils2012} occur at rates faster than transcription (supercoiling diffusion: \(\mathcal{D} \approx O(0.5 \units{kb$^2$}{s})\) versus transcription rate: \(v_0 \approx 0.05 \units{kb}{s}\))\parencite{munizRNAPolymeraseII2021}, the supercoiling generated by a polymerase will diffuse outward far more rapidly than polymerases can move. As in previous reported work, we make a pseudo-steady assumption for inter-RNAP supercoiling---assuming that \cref{eq:supercoiling_density} holds between polymerases---over the relatively small (\(\sim 10\) kb) genomic distances considered in this work in order to simplify the resulting model.

How does transcription both drive the process of supercoiling generation and react to changes in local supercoiling? Under the assumption of supercoiling relaxation, each polymerase is defined by four variables---the one-dimensional genomic location of the polymerase \(z_i\), the length of the nascent RNA \(x_i\), the excess twist at the location of the polymerase \(\phi_i\), and the rotation angle of the polymerase \(\theta_i\) (\cref{fig:key_variables_diagram}). Then, two governing equations define the motion of all polymerases \parencite{sevierPropertiesGeneExpression2018}. First, we equate linear polymerase motion with the rotational motion required to track the DNA groove:
\begin{equation}
    \underbrace{\omega_0 \frac{dz_i}{dt}}_{\mathclap{\text{RNAP velocity}}} = \overbrace{\frac{d\theta_i}{dt}}^{\mathclap{\text{RNAP rotation}}} + \underbrace{\frac{d\phi_i}{dt}}_{\mathclap{\text{supercoiling generation}}}
\label{eq:linear_rotational_balance}
\end{equation}
where the change in \(\theta_i\) represents polymerase rotation and the change in \(\phi_i\) represents local rotation of the DNA.
The second equation provides a torque balance between DNA-mediated torques on the left hand side and torque caused by drag acting on the nascent RNA:
\begin{equation}
    \underbrace{\tau(\sigma(z_i, \phi_{i-1}, \phi_{i+1}))}_{\mathclap{\text{torque acting on RNAP}}} + \overbrace{\chi \frac{d\phi_i}{dt}}^{\mathclap{\text{supercoiling restoring force}}}= \underbrace{\eta x_i^n \frac{d\theta_i}{dt}}_{\mathclap{\text{nascent RNA drag}}}
\label{eq:torque_balance}
\end{equation}
To develop a final system of ordinary differential equations, we still must define the torque response function \(\tau(\sigma)\) and the polymerase velocity function \(\frac{dz}{dt} = v(\tau)\). With these two functions, \cref{eq:linear_rotational_balance,eq:torque_balance} can be solved as in \textcite{sevierPropertiesGeneExpression2018}. Here, we use Marko's torque-response model of supercoiling which accounts for the thermodynamic behavior of both non-buckled, twisted DNA and buckled, plectonemic DNA (see \cref{eq:marko_torque,eq:marko_energy} in \cref{sec:appendix:model}) \parencite{markoTorqueDynamicsLinking2007}. The resulting \(\tau(\sigma)\) function exhibits a phase transition, where the torque response is nearly constant at intermediate values of \(\sigma\) where the DNA is transitioning from a locally-twisted phase to a plectonemic-phase (\cref{fig:supp:torque_diagram}).

For the velocity response of a polymerase experiencing a torque \(\tau_f\) in front and \(\tau_b\) behind, we model polymerase stalling as:
\begin{equation}
    v(\tau_f, \tau_b) = \frac{v_0}{(1 + e^{(|\tau_f| - \tau_s)/\tau_w})(1 + e^{(|\tau_b| - \tau_s)/\tau_w})}
\label{eq:velocity_response}
\end{equation}
where the stall torque \(\tau_s = 12 \unit{pN nm}\) and stall-width \(\tau_w = 3 \unit{pN nm}\) define a sigmoidal stall-response curve.
As shown in \cref{fig:hyperparam_stall_torque,fig:hyperparam_stall_width}, our results are only weakly dependent on the specific choice of \(\tau_s\) and \(\tau_w\).
Importantly, our selected phenomenological term will stall polymerase motion if \emph{either} the torque upstream or downstream exceeds the stall torque \(\tau_s\). Some models choose a stalling equation that only stalls if the difference between the upstream and downstream torque exceeds a stall torque \parencite{tripathiDNASupercoilingmediatedCollective2021}; we chose this form, reasoning that the DNA unwinding and rewinding process opposed, respectively, by upstream and downstream torque could independently stall. When simulated, the difference between these stalling models is small in practice; polymerases at the start or end of the burst encounter both high upstream and downstream torques \emph{and} a high torque difference, whereas polymerases in the middle of a burst experience both lower adjacent torques and a lower torque difference.

Taking the above equations together, we can simulate the coupled motion of an arbitrary number of polymerases as a single coupled ODE system. We further examine different experimental systems by implementing different boundary conditions that allow us to simulate both plasmid systems and genomically-integrated systems (see \cref{sec:appendix:bcs}).

% LAYOUT TWEAKS - let model summary figure float within model and methods
% \FloatBarrier
\subsubsection{Supercoiling-dependent initiation model}
While the described differential equation system can simulate polymerase motion, we need a way to model the addition of polymerases to simulated genes. A simple strategy is to assume a supercoiling-independent initiation rate and use a stochastic simulation method to randomly add polymerases to transcriptional start sites at a certain fixed rate. However, this simple model assumes that polymerases can bind equally well to initiation sites independent of local supercoiling, missing supercoiling-dependent dynamics. In order to include supercoiling in a polymerase initiation model, we relate the basal expression rate to a corresponding base energy term. We can then additively introduce extra energy costs for polymerase binding under different local supercoiling conditions. Under the approximation that the direct energetic cost of locally melting the DNA to fit in the RNAP groove dwarfs the relative change in unwinding energy caused by supercoiling, the majority of the energetic cost comes from inserting supercoiling ahead and behind the inserted polymerase. Under this assumption, the first-order supercoiling energetic correction can be written as:
\begin{equation}
    E_\text{sc} = 1.2  \cdot 2\pi \cdot \tau(\sigma)
\label{eq:first_order_sc_initation}
\end{equation}
 Is this a good approximation? We can estimate the energetic cost of local melting, and find that neglecting local melting leads to a minor change in the resulting energy as seen in \cref{fig:supp:energy_with_melting}. A full derivation of \cref{eq:first_order_sc_initation} is given in \cref{sec:sc_initation}.

While this first-order energetic term introduces much-needed behavior to the modeled system---where locally high positive supercoiling decreases the RNAP initiation rate and locally negative supercoiling increases the RNAP initiation rate---at extreme values of \(\sigma\), this energetic term gives aphysical predictions. In particular, under highly negative supercoiling densities, the energetics of polymerase loading become increasingly favorable, with loading sometimes occurring more than two orders of magnitude faster when compared to relaxed DNA. To correct for this behavior, we add a second-order (quadratic) term that constrains polymerase loading at highly positive or negative local supercoiling:
\begin{equation}
    E_\text{sc} = 1.2  \cdot 2\pi \cdot \tau(\sigma) + \alpha \cdot \tau_0 \cdot \sigma^2
\label{eq:second_order_sc_initation}
\end{equation}
for \(\tau_0\), the relevant scale factor in the \(\tau(\sigma)\) equation (\cref{eq:marko_torque,eq:critical_tau_values} from \parencite{markoTorqueDynamicsLinking2007}) and \(\alpha\), a positive tunable parameter. As the \(\tau(\sigma)\) equation is linear in \(\sigma\) outside of the phase-transition region, this added \(\sigma^2\) term can be contextualized as an additional term in the Taylor expansion of the physically-realistic \(E_\text{sc}(\sigma)\) equation. This form of the binding energy enables us to qualitatively match the experimentally observed asymptotic behavior between torque and supercoiling for underwound DNA \parencite{leSynergisticCoordinationChromatin2019}.

For these three models of supercoiling-dependent initiation,we found that the supercoiling-independent initiation model predicted only small changes in reporter output (\cref{fig:supp:initation_order_comparison}). Comparing the first- and second-order models, we found that a critical value of \(\alpha\) existed, \(\alpha \approx 0.2\), above which the second-order model demonstrated emergent non-monotonic behavior (\cref{fig:top:alpha_sweep}). At low values of \(\alpha\), the second-order model behaves similarly to the first-order model, so we used \(\alpha = 0.025\) for this work. Increasing \(\alpha\) beyond this chosen value appears to scale down reporter output without qualitatively modifying behavior (\cref{fig:top:alpha_sweep}).

When simulating the ODE model, the rate of stochastic polymerase initiation, \(r_\text{initiation}\), varies continuously based on the local supercoiling density \(\sigma\) at the transcription start site as:
\begin{equation}
    r_\text{initiation} = r_\text{base rate} \cdot e^{- E_\text{sc} / (k_B T)}
\label{eq:intermediate_init_rate}
\end{equation}


\subsubsection{Additional discrete reaction model}
Many of the native and synthetic systems of interest include mechanisms of gene regulation that rely on other regulatory species. In order to analyze these types of systems using our supercoiling model, we extended our model to simultaneously simulate arbitrary discrete stochastic equations---such as those commonly used in the literature to model protein production, degradation, dimerization, and more. This addition allowed us to model discrete events otherwise not accounted for in the continuous model.

Importantly, this framework allows us to simulate the activity of topoisomerases. While the total amount of supercoiling (the integral of supercoiling density) is conserved, topoisomerases can be modeled as stochastic events that redistribute supercoiling in certain regions. Here, we simulated the removal of supercoiling in either intergenic or intragenic regions. Removal of supercoiling in intergenic regions is performed by updating the rotation of polymerases on adjacent genes to make the intergenic supercoiling density zero. After relaxation of supercoiling in intragenic regions, polymerases on the gene experience a constant supercoiling density which conserves overall supercoiling. With the exception of the segmentation gene network simulations in \cref{fig:top:her1_her7} and related supplemental figures, we found that inclusion of topoisomerase relaxation did not appreciably change the observed results. Thus, unless otherwise stated, simulations were performed without topoisomerase relaxation.

In addition, we allowed the base initiation rate of genes to vary as an arbitrary function of all species concentrations in the model (\(S_i\)), such that \cref{eq:intermediate_init_rate} becomes:
\begin{equation}
    r_\text{initiation} = r(S_i) \cdot e^{- E_\text{sc} / (k_B T)}
\label{eq:final_init_rate}
\end{equation}
By combining discrete reactions with the ability to dynamically change polymerase initiation rates, we are able to simulate a wide range of phenomena.
For example, a cooperative repressive interaction between some repressor protein \(R\) and a promoter could be modeled using a repressive Hill function:
\begin{equation*}
    r_\text{initiation}(R) = \frac{1}{1 + \left(\frac{R}{K}\right)^n} \cdot e^{-E_\text{sc} / (k_B T)}
\end{equation*}
More generally, we can use stochastic formulations of other regulatory mechanisms and test how these mechanisms behave in concert with supercoiling-mediated feedback.

\subsubsection{Boundary conditions} \label{sec:appendix:bcs}

Key to our simulations is calculating the supercoiling density across the domain using \cref{eq:supercoiling_density}.
For simulations using linear boundary conditions, we use the left and right edges as boundaries, assigning excess twist \(\phi = 0\) at both boundary locations. Then, the supercoiling density can be defined between every polymerase. The location of the boundary conditions for the simulations is described in \cref{tab:supp:simulated_loc_distances}.

For simulations using circular boundary conditions, we must define how generated supercoiling ``wraps around'' the edges of the simulation. To do this, we choose an arbitrary origin, and order polymerases based on their (clockwise) position from the origin. In \cref{tab:supp:simulated_loc_distances}, the first boundary location is used as this origin location and the second boundary location is the length of the circle, relative to the origin. As in the linear case, when there are zero polymerases loaded, the supercoiling density is uniformly 0. In addition, for circular boundary conditions, the supercoiling density is \emph{also} uniformly 0; for assumed supercoiling relaxation, a single polymerase can never accumulate negative or positive supercoiling.

For two or more polymerases, we create a list of excess twists, wrapping as \(\phi_{n}, \phi_1, \cdots, \phi_n, \phi_1\). We additionally project the locations of the wrapped polymerases past the origin (e.g., the position of the wrap-around \(\phi_n\) is placed at \(z_\text{total length} - z_n\)), then compute the supercoiling density as in the linear case.

\subsubsection{ODE and stochastic simulation}
The core ordinary differential equations were simulated using a Tsitouras's explicit Runge-Kutte 4-5 order method \parencite{tsitourasRungeKuttaPairs2011}.
Normally, one implements stochastic simulations using a time-jumping method such as Gillespie's method. However, because the propensity of our stochastic events changes \emph{continuously} with the continuous simulation, we need a stochastic solver that can be applied within the continuous integrator loop. Here, we used \texttt{DifferentialEquations.jl}, a performant Julia package that allows layered differential and stochastic equations
\parencite{rackauckasDifferentialEquationsJlPerformant2017}.

Due to the stochastic nature of these simulations, we simulated ensembles of simulations, choosing an ensemble size that was large enough to show the desired population behavior. These simulations then were post-processed, by either taking the average over each ensemble or by directly showing population distributions, summarized by a kernel density estimate (smoothed histograms provided by the Python package \texttt{seaborn}). The specific meaning and ensemble size for each plot is summarized below.

\begin{table}[h]
\centering
\begin{tabular}{@{}p{0.4\textwidth}rp{0.4\textwidth}@{}}
\toprule
Plot description & \(n\) & Figures \\
\midrule
\multirow{4}{0.4\textwidth}{Each point comprising the plotted curves is an average over \(n\) simulations.}&  750 & \cref{fig:supp:energy_well_induction_sweep,fig:supp:nucleosome_induction_sweep} \\
& 2000 & \cref{fig:bc_induction_sweep,fig:hyperparam_drag_coeff,fig:hyperparam_drag_exponent,fig:hyperparam_stall_torque,fig:hyperparam_stall_width,fig:supp:initation_order_comparison,fig:supp_alpha_sweep_induction,fig:supp_alpha_sweep_linear,fig:supp_alpha_sweep_circular} \\
& 5000 & \cref{fig:reporter_output_by_spacing_fold_induction} \\
& 10000 & \cref{fig:toggle_stable_frac_n_2.0,fig:toggle_half_life_vs_mRNA_deg,fig:supp:toggles_hill_n_sweep} \\
\midrule
\multirow{3}{0.4\textwidth}{Each line shows the average supercoiling density profile over an ensemble of \(n\) simulations.} & 100 & \cref{fig:base_model_sc_density,fig:supp:sc_density_induction,fig:supp_toggles_mean_sc_density} \\
& 250 & \cref{fig:sc_density,fig:supp:energy_well_sc_density,fig:supp:nucleosome_sc_density} \\
\\
\midrule
The standard deviation of gene output for an ensemble of \(n\) simulations is shown as a function of time. & 250 & \cref{fig:noise_by_orientation} \\
\midrule
\multirow{3}{0.4\textwidth}{The average cross-correlation over an ensemble of \(n\) simulations is plotted.} & 250 & \cref{fig:orientation_cross_correlation} \\
& 2000 & \cref{fig:zinani_cross_correlation,fig:supp:zinani_alpha_cross_correlation,fig:supp:intragenic_cross_correlation,fig:supp:energy_well_zinani_crosscorr,fig:supp:nucleosome_zinani_crosscorr} \\
& 10000 & \cref{fig:supp:zinani_matlab_cross_correlation} \\
\midrule
\multirow{5}{0.4\textwidth}{Distribution profiles are shown over an ensemble of \(n\) simulations. One-dimensional distributions are shown as a kernel density estimate (smoothed histogram) or violin plot. Two-dimensional distributions are shown as filled or unfilled isocurves.} &  100 & \cref{fig:burst_size_dynamics} (base model),\cref{fig:interburst_time_dynamics} (base model), \cref{fig:supp:fig_examples_ensemble_behavior,fig:burst_threshold_burst_size,fig:burst_threshold_interburst_time,fig:supp_toggles_interburst_time} \\
& 250 & \cref{fig:output_distribution_by_orientation_dynamics,fig:supp:sc_distributions_high_alpha_induction},\cref{fig:burst_size_dynamics} (model perturbations), \cref{fig:interburst_time_dynamics} (model perturbations) \\
& 750 & \cref{fig:supp:energy_well_joint_distribution,fig:supp:nucleosome_joint_distribution} \\
& 2000 & \cref{fig:circular_bc_distributions,fig:linear_bc_distributions,fig:zinani_correlation_coeff,fig:zinani_oscillation_amplitude,fig:supp_alpha_sweep_distribution,fig:supp:zinani_alpha_correlation,fig:supp:zinani_alpha_oscillation_amplitude,fig:supp:zinani_topo_correlation_coefficent,fig:supp:zinani_topo_oscillation_amplitude,fig:supp:energy_well_zinani_correlation_coefficient,fig:supp:energy_well_zinani_oscillation_amplitude,fig:supp:nucleosome_zinani_correlation_coefficient,fig:supp:nucleosome_zinani_oscillation_amplitude} \\
& 10000 & \cref{fig:toggle_basin_stability_over_time,fig:toggle_burst_size,fig:supp:n1_toggle_distributions,fig:supp:zinani_matlab_correlation,fig:supp:zinani_matlab_correlation} \\
\midrule
\multirow{3}{0.4\textwidth}{An example simulation was randomly selected from an ensemble of \(n\) simulations and plotted.} & 250 & \cref{fig:sc_examples} \\
& 2000 & \cref{fig:zinani_mRNA_behavior,fig:supp:zinani_alpha_mRNA,fig:supp:intragenic_zinani_examples,fig:supp:energy_well_zinani_examples,fig:supp:nucleosome_zinani_examples} \\
& 10000 & \cref{fig:supp_toggles_run_examples,fig:supp:zinani_matlab_mRNA} \\
\bottomrule
\end{tabular}
\caption{Key ensemble size and interpretive description of each plot.}
\end{table}

\FloatBarrier
\subsubsection{Computing cross-correlation} \label{sec:supp:cross_correlation}
We computed the cross-correlation between the gene outputs following induction (\cref{fig:cross_correlation_cartoon}), normalized by the geometric mean of the auto-correlation of these outputs.  Strong negative or positive peaks evenly spaced around \(\tau = 0\) is a hallmark of periodic behavior, with the time offset of the peak encoding the phase offset between the signals. For two signals \(f(t)\) and \(g(t)\), the normalized cross-correlation at a certain time offset \(\tau\) is bounded between \(\pm1\) and can be thought of as the correlation coefficient between the unshifted version of one of the signals (\(f(t)\)) and the other signal shifted in the time axis by \(\tau\) (\(g(t + \tau)\)). Mathematically, we use:
\begin{equation}
\text{cross-correlation}(\tau) = \frac{(f \star g)(\tau)}{\sqrt{(f \star f)(0) \cdot (g \star g)(0)}}
\end{equation}
for
\begin{equation}
    (f \star g)(\tau) = \sum_t (f(t) - \langle f(t)\rangle)(g(t+\tau) - \langle g(t)\rangle)
\end{equation}
In fact, for \(\tau = 0\), the cross-correlation of the two signals is exactly the Pearson correlation coefficient. The shape of the cross-correlation curve can also reveal periodic and other time-dependent correlative behaviors.

\subsubsection{Burst threshold choice} \label{sec:supp:burst_threshold}
We calculated burst size and inter-burst time by using a burst threshold. From polymerase initiation times, we calculate the time between successive polymerase additions. Intra-addition times greater than the burst threshold form the boundaries between different expression bursts. We define the burst size to be the number of polymerases included in a burst, and the inter-burst time to be those intra-polymerase-addition times greater than the burst threshold. For the main text, we used a burst threshold of 30 seconds; specifically, this means that bursts ended if 30 seconds passed without a new polymerase being added.

In \cref{fig:burst_threshold_burst_size,fig:burst_threshold_interburst_time}, we reanalyze the data presented in \cref{fig:burst_size_dynamics,fig:interburst_time_dynamics} for different burst thresholds. We find that using a burst threshold of twenty or sixty seconds does not significantly affect the qualitative results observed. However, using a ten second burst threshold does dramatically shift the resulting burst size and inter-burst time distributions, with the inter-burst time distribution becoming concentrated around ten seconds. This indicates that a ten-second burst time is too short and incorrectly separates bursts.

\subsubsection{Toggle-switch parameter selection} \label{sec:supp:choice_of_k}
The half-max value \(K\), the mRNA count at which the promoter activity is half that of \(r_0\), is chosen here to approximately match the mean steady-state expression of the steady states. The mean steady-state value is identified using simulations where only one of the toggle switch genes is enabled; this allows us to directly account for the influence of supercoiling-mediated feedback on the steady state mRNA concentration. With this choice of \(K\), we ensure that the toggle switch operates in the regime of maximum sensitivity (e.g. the stable basin steady-state value is in the middle of the sigmoidal repression curve).

 In \cref{fig:toggle_half_life_vs_mRNA_deg}, we tune the mRNA degradation rate, which directly impacts the mean steady-state value. If the mRNA degradation rate is doubled, we expect that the mean steady-state value should decrease to half its original value. To compare between these otherwise disparate conditions, we scaled the \(K\) value alongside the mRNA degradation rate, dividing by the fold increase in the mRNA degradation rate.

\subsubsection{Relevant simulation parameters}

\begin{table}[h]
\centering
\begin{tabular}{@{}lllll@{}}
\toprule
Drag coefficient & Drag exponent & Stall torque & Stall width \\
\multicolumn{1}{c}{\(\eta\) (pN nm$^{\beta - 1}$)} &  \multicolumn{1}{c}{\(\beta\)} &  \multicolumn{1}{c}{\(\tau_s\) (pN nm)} & \multicolumn{1}{c}{\(\tau_w\) (pN nm)} & Figures\\
\midrule
1/20 & 1 & 12 & 3 & All except \cref{fig:hyperparam_drag_coeff,fig:hyperparam_drag_exponent,fig:hyperparam_stall_torque,fig:hyperparam_stall_width} \\
1/50\(\leftrightarrow\) 1/5 & 1 & 12 & 3 & \Cref{fig:hyperparam_drag_coeff} \\
1/20 & 1\(\leftrightarrow\)3 & 12 & 3 & \Cref{fig:hyperparam_drag_exponent} \\
1/20 & 1 & 8\(\leftrightarrow\)16 & 3 & \Cref{fig:hyperparam_stall_torque} \\
1/20 & 1 & 12 & 1\(\leftrightarrow\)5 & \Cref{fig:hyperparam_stall_width} \\
\bottomrule
\end{tabular}
\caption{Key dynamic constants.}
\end{table}

\begin{table}[h]
\centering
\begin{tabular}{@{}lllll@{}}
\toprule
Type & Boundary locations & Reporter gene & Inducible gene & Figures \\
\midrule
Upstream tandem & 0, 10000 &  3000\(\rightarrow\)4000 & 6000\(\rightarrow\)7000 & \multirow{4}{3cm}{\Cref{fig:bc_induction_sweep,fig:circular_bc_distributions,fig:linear_bc_distributions,fig:base_model_sc_density,fig:top:single_cell_behavior,fig:top:bursts,fig:supp:initation_order_comparison,fig:hyperparam_drag_coeff,fig:hyperparam_drag_exponent,fig:hyperparam_stall_torque,fig:hyperparam_stall_width,fig:top:alpha_sweep,fig:supp:sc_density_induction,fig:supp:sc_distributions_high_alpha_induction,fig:supp:fig_examples_ensemble_behavior,fig:burst_threshold_burst_size,fig:burst_threshold_interburst_time,fig:supp:energy_well_sc_density,fig:supp:energy_well_induction_sweep,fig:supp:energy_well_joint_distribution,fig:supp:nucleosome_sc_density,fig:supp:nucleosome_induction_sweep,fig:supp:nucleosome_joint_distribution}}\\
Downstream tandem & 0, 10000 &  6000\(\rightarrow\)7000 & 3000\(\rightarrow\)4000 \\
Convergent & 0, 10000 & 7000\(\rightarrow\)6000 & 3000\(\rightarrow\)4000 \\
Divergent & 0, 10000 & 6000\(\rightarrow\)7000 & 4000\(\rightarrow\)3000 \\
\midrule
Type & Boundary locations & Reporter gene & Inducible gene & Figures \\
\midrule
Upstream tandem & 0, \(8000 + \Delta x\) &  3000\(\rightarrow\)4000 & \(4000 + \Delta x\) \\ &&& \(\rightarrow 5000 + \Delta x\) & \multirow{6}{3cm}{\Cref{fig:reporter_output_by_spacing_fold_induction}}\\
Downstream tandem & 0, \(8000 + \Delta x\) &  \(4000 + \Delta x\)& 3000\(\rightarrow\)4000 \\
 && \(\rightarrow 5000 + \Delta x\)  \\
Convergent & 0, \(8000 + \Delta x\) & \(5000 + \Delta x\)& 3000\(\rightarrow\)4000 \\
 && \(\rightarrow 4000 + \Delta x\)  \\
Divergent & 0, \(8000 + \Delta x\) & \(4000 + \Delta x\)& 4000\(\rightarrow\)3000 \\
 && \(\rightarrow 5000 + \Delta x\)  \\
\midrule
Type & Boundary locations & Gene A & Gene B & Figures \\
\midrule
Upstream tandem & 0, 10000 &  3500\(\rightarrow\)4500 & 5500\(\rightarrow\)6500 & \multirow{4}{3cm}{\Cref{fig:top:toggle_switch,fig:supp_toggles_interburst_time,fig:supp_toggles_interburst_time,fig:supp:toggles_hill_n_sweep,fig:supp:n1_toggle_distributions}}\\
Downstream tandem & 0, 10000 &  5500\(\rightarrow\)6500 & 3500\(\rightarrow\)4500 \\
Convergent & 0, 10000 & 3500\(\rightarrow\)4500 & 6500\(\rightarrow\)5500 \\
Divergent & 0, 10000 & 4500\(\rightarrow\)3500 & 5500\(\rightarrow\)6500 \\
\midrule
Type & Boundary locations & \(her1\) & \(her7\) & Figures \\
\midrule
Gene-unpaired & 0, 10\(^9\)  (free ends) & 6405\(\rightarrow\)6017 & \((10^9 - 9864)\)  & \multirow{3}{3cm}{\Cref{fig:top:her1_her7,fig:supp:zinani_alpha_mRNA,fig:supp:zinani_alpha_cross_correlation,fig:supp:zinani_alpha_correlation,fig:supp:zinani_alpha_oscillation_amplitude,fig:supp:intragenic_zinani_examples,fig:supp:intragenic_cross_correlation,fig:supp:zinani_topo_correlation_coefficent,fig:supp:zinani_topo_oscillation_amplitude,fig:supp:energy_well_zinani_examples,fig:supp:energy_well_zinani_crosscorr,fig:supp:energy_well_zinani_correlation_coefficient,fig:supp:energy_well_zinani_oscillation_amplitude,fig:supp:nucleosome_zinani_examples,fig:supp:nucleosome_zinani_crosscorr,fig:supp:nucleosome_zinani_correlation_coefficient,fig:supp:nucleosome_zinani_oscillation_amplitude}}\\
                                                                 &&& \(\rightarrow (10^9 - 8549)\) \\
Gene-paired & 0, 34393 & 12422\(\rightarrow\)6017 & 24529\(\rightarrow\)25844 \\
\bottomrule
\end{tabular}
\caption{Simulated circuit locations and distances. Locations are all given in basepairs (with 0.34 nanometers per base pair). For the given genes, the first number given is the simulated promoter location and start of the gene body, and the second number is a simulated termination site.}
\label{tab:supp:simulated_loc_distances}
\end{table}

\FloatBarrier
\subsection{Acknowledgements}
Research reported in this manuscript was supported by National Institute of General Medical Sciences of the National Institutes of Health under award number R35-GM143033. The authors acknowledge the MIT SuperCloud and Lincoln Laboratory Supercomputing Center \parencite{reutherInteractiveSupercomputing402018} for providing HPC resources that have contributed to the research results reported within this paper. We also thank Adam Beitz, Patrick Han,  Ross Jones, Sneha Kabaria, Brittany Lende, Kasey Love, Christian Otero, Emma Peterman, Jack Suggs, and Nathan Wang for helpful suggestions throughout and for their effort in editing this manuscript.

\subsubsection{Author contributions}
C.P.J. and K.E.G. conceptualized the work. C.J. developed the mathematical model, wrote the simulation and data analysis software, and visualized and synthesized results. K.G. supervised the project and acquired financial support for this project. C.P.J. and K.E.G. discussed results and reviewed and edited the manuscript.

\subsubsection{Declaration of interests}
The authors declare no competing interests.

\printbibliography

\clearpage
\appendix
\titleformat{\subsection}[hang]{\bfseries}{\thesubsection}{0em}{\;\;}
\renewcommand{\appendixpagename}{Supplemental information}
% Reset supplemental figures to using S syntax.
\renewcommand{\thefigure}{S\arabic{figure}}
\renewcommand{\theequation}{S\arabic{equation}}
\setcounter{equation}{0}
\setcounter{figure}{0}
\appendixpage

\section{Supplemental figures}
\FloatBarrier
\begin{figure}[htbp]
    \centering
    {\includegraphics{figures/modeling_paper/fig_supp_hyper_torque.pdf}
    \phantomsubcaption\label{fig:supp:torque_diagram}
    \phantomsubcaption\label{fig:supp:energy_with_melting}
    \phantomsubcaption\label{fig:hyperparam_drag_coeff}
    \phantomsubcaption\label{fig:hyperparam_drag_exponent}
    \phantomsubcaption\label{fig:hyperparam_stall_torque}
    \phantomsubcaption\label{fig:hyperparam_stall_width}
    }
\end{figure}
\begin{figure}[htbp]
    \ContinuedFloat
    \caption{Related to \Cref{fig:key_variables_diagram,fig:top:orientation_bc_behavior}. Reporter output of two-gene circuits with linear boundary conditions (as in \cref{fig:top:orientation_bc_behavior}) demonstrates similar qualitative behavior when underlying model parameters are changed.
        \subref{fig:supp:torque_diagram} The torque response predicted by Marko's model (\cref{eq:marko_torque}) exhibits phase behavior, with a transition occurring between a over- and under-twisted phase to a largely plectonemic phase. During the transition, the torque response is constant.
        \subref{fig:supp:energy_with_melting} For the supercoiling-dependent initiation model, the energy that it takes to introduce supercoiling upstream and downstream by unwinding the DNA dominates the energy expression. Including a term representing the energy of local melting of the double helix at the location of the polymerase only gives a minor correction.
        \subref{fig:hyperparam_drag_coeff} Varying the nascent mRNA drag coefficient from the chosen value \(\eta = 0.05\) rescales but does not dramatically change reporter output behavior. Larger drag coefficients increase generated supercoiling by introducing additional torque into the system.
        \subref{fig:hyperparam_drag_exponent} Varying the exponent on the nascent mRNA length in the drag coefficient from the choice of linear drag \(\beta = 1.0\) also rescales but does not dramatically change reporter output behavior. Choices of \(\beta > 1.0\) tends to increase generated supercoiling due to longer nascent mRNAs having an increased drag profile.
        \subref{fig:hyperparam_stall_torque} Varying the RNAP stall torque from the chosen \(\tau_s = 12\) pN nm has a minor effect on reporter output.
        \subref{fig:hyperparam_stall_width} Varying the width of the stalling function from the chosen \(\tau_w = 3\) pN nm has no noticeable effect on reporter output.
    }
\end{figure}

\begin{figure}[htbp]
    \centering
    {\includegraphics[width=\textwidth]{figures/modeling_paper/fig_supp_alpha_sweep.pdf}
    \phantomsubcaption\label{fig:supp:initation_order_comparison}
    \phantomsubcaption\label{fig:supp:alpha_reporter_counts}
    \phantomsubcaption\label{fig:supp_alpha_sweep_induction}
    \phantomsubcaption\label{fig:supp_alpha_sweep_energy}
    \phantomsubcaption\label{fig:supp_alpha_sweep_linear}
    \phantomsubcaption\label{fig:supp_alpha_sweep_circular}
    \phantomsubcaption\label{fig:supp_alpha_sweep_distribution}
    }
\end{figure}
\begin{figure}[htbp]
    \ContinuedFloat
    \caption{Related to \Cref{fig:top:orientation_bc_behavior}.
    As the value of the quadratic energy correction coefficient \(\alpha\) varies, two regimes of supercoiling-dependent behaviors emerge.
        \subref{fig:supp:initation_order_comparison} Reporter output is shown, normalized to the zero-fold induction case for the four tested syntaxes for the three different polymerase initiation models.
        \subref{fig:supp:alpha_reporter_counts} For the second-order model, we chose the value of \(\alpha\) that defined the transition between regimes.
        \subref{fig:supp_alpha_sweep_induction} For \(\alpha \leq 0.02\), supercoiling-dependent initiation appears to dominate, leading to both the maintenance of reporter output in the linear divergent case and near-equality of the convergent and divergent cases with circular boundary conditions. For \(\alpha \geq 0.025\), emergent non-monotonic behavior appears, leading to complex behaviors with linear boundary conditions and a separation between the response of the convergent and divergent syntaxes with circular boundary conditions.
        \subref{fig:supp_alpha_sweep_energy} The additional polymerase binding energy as a function of supercoiling density is plotted for various values of \(\alpha\).
        \subref{fig:supp_alpha_sweep_linear} \subref{fig:supp_alpha_sweep_circular} Reporter behavior of the four syntaxes under linear and circular boundary conditions are shown, in comparison to \cref{fig:bc_induction_sweep}.
        \subref{fig:supp_alpha_sweep_distribution} The overall distribution of the four syntaxes as a function of alpha is shown, in comparison to \cref{fig:linear_bc_distributions}.
    }
    \label{fig:top:alpha_sweep}
\end{figure}

\begin{figure}[htbp]
    \centering
    {\includegraphics{figures/modeling_paper/fig_supp_dynamics_details.pdf}
    \phantomsubcaption\label{fig:supp:sc_density_induction}
    \phantomsubcaption\label{fig:supp:fig_examples_ensemble_behavior}
    \phantomsubcaption\label{fig:supp:sc_distributions_high_alpha_induction}
    \phantomsubcaption\label{fig:burst_threshold_burst_size}
    \phantomsubcaption\label{fig:burst_threshold_interburst_time}
    }
\end{figure}
\begin{figure}[htbp]
    \ContinuedFloat
    \caption{Related to \cref{fig:top:single_cell_behavior,fig:top:bursts}.
    \subref{fig:supp:sc_density_induction} Supercoiling density profiles are plotted for a wide range of adjacent gene induction levels across the four syntaxes considered. At low adjacent induction of the adjacent gene, positive and negative supercoiling mainly accumulates upstream and downstream, respectively, of the reporter gene. At 1-fold (equal base rates) induction, the supercoiling distributions are nearly symmetric in the convergent and divergent syntaxes. At high adjacent induction, the adjacent gene expression dominates the generated supercoiling profiles.
    \subref{fig:supp:fig_examples_ensemble_behavior} The distribution of inducible and reporter mRNA counts are shown for the simulation ensembles presented in \cref{fig:top:single_cell_behavior}. The different initial starting condition does not affect the final distribution state when compared to \cref{fig:linear_bc_distributions}, indicating that the final state reached is independent of initial state.
    \subref{fig:supp:sc_distributions_high_alpha_induction}The ensemble distribution of the four linear circuit syntaxes is shown for \(\alpha = 0.03\) and high induction of the adjacent gene (1.8 fold).
    \subref{fig:burst_threshold_burst_size},\subref{fig:burst_threshold_interburst_time} Burst size and inter-burst time distributions are shown at equal-induction as a function of threshold time and circuit syntax. At the lowest threshold time, the burst size and inter-burst time distributions qualitatively differ, with only a small percentage of bursts having an inter-burst time larger than 20 seconds. This lack of larger inter-burst times suggests that a burst threshold of at least 20 seconds gives useful predictions.
    }
\end{figure}

\begin{figure}[htbp]
    \centering
    {\includegraphics{figures/modeling_paper/fig_supp_perturb_energy_well}
    \phantomsubcaption\label{fig:supp:energy_well_sc_density}
    \phantomsubcaption\label{fig:supp:energy_well_induction_sweep}
    \phantomsubcaption\label{fig:supp:energy_well_joint_distribution}
    \phantomsubcaption\label{fig:supp:energy_well_zinani_examples}
    \phantomsubcaption\label{fig:supp:energy_well_zinani_crosscorr}
    \phantomsubcaption\label{fig:supp:energy_well_zinani_correlation_coefficient}
    \phantomsubcaption\label{fig:supp:energy_well_zinani_oscillation_amplitude}
    }
\end{figure}
\begin{figure}
    \ContinuedFloat
    \caption{Related to \Cref{fig:top:orientation_bc_behavior,fig:top:bursts,fig:top:her1_her7}.
        Simulation of RNAP binding penalty at highly positive and negative supercoiling densities reduces syntax-specific differences and impedes transcriptional coupling between \textit{her1}and \textit{her7}.
        \subref{fig:supp:energy_well_sc_density} The average supercoiling density for both the induced and induced case is shown. The average supercoiling density is smaller than observed in \cref{fig:sc_density,fig:supp:nucleosome_sc_density} due to the inhibition of polymerase initiation at extreme values of supercoiling density.
        \subref{fig:supp:energy_well_induction_sweep} Reporter output as a function of adjacent gene induction and circuit syntax is shown for both circular and linear boundary conditions.
        \subref{fig:supp:energy_well_joint_distribution} The population distribution is shown as a function of boundary condition and circuit syntax for the equally-induced case. The noise ratio of each is defined as the intrinsic noise (off-diagonal component) divided by the extrinsic noise (on-diagonal component).
        \subref{fig:supp:energy_well_zinani_examples} Four example simulation traces of the zebrafish segmentation gene network are shown, for two topoisomerase conditions and the gene-unpaired and gene-paired orientations. No clear periodicity is observed.
        \subref{fig:supp:energy_well_zinani_crosscorr} The ensemble cross-correlation for the zebrafish segmentation gene network is shown. In this restricted polymerase initiation case, strong periodicity is not observed.
        \subref{fig:supp:energy_well_zinani_correlation_coefficient} The ensemble distribution of the correlation coefficient between \textit{her1} and \textit{her7} is shown. Restricting polymerase initiation eliminates \emph{her1}/\emph{her7} correlation, even in the gene-paired case.
        \subref{fig:supp:energy_well_zinani_oscillation_amplitude} The ensemble oscillation amplitude is shown for each topoisomerase and gene pairing condition.
    }
    \label{fig:top:supp_perturb_energy_well}
\end{figure}

\begin{figure}[htbp]
    \centering
    {\includegraphics{figures/modeling_paper/fig_supp_perturb_nucleosomes}
    \phantomsubcaption\label{fig:supp:nucleosome_sc_density}
    \phantomsubcaption\label{fig:supp:nucleosome_induction_sweep}
    \phantomsubcaption\label{fig:supp:nucleosome_joint_distribution}
    \phantomsubcaption\label{fig:supp:nucleosome_zinani_examples}
    \phantomsubcaption\label{fig:supp:nucleosome_zinani_crosscorr}
    \phantomsubcaption\label{fig:supp:nucleosome_zinani_correlation_coefficient}
    \phantomsubcaption\label{fig:supp:nucleosome_zinani_oscillation_amplitude}
    }
\end{figure}
\begin{figure}
    \ContinuedFloat
    \caption{Related to \Cref{fig:top:orientation_bc_behavior,fig:top:bursts,fig:top:her1_her7}.
        Nucleosome buffering of positive supercoiling displays weak effect on expression profile of two-gene systems at steady state and on the \textit{her1}-\textit{her7} somite network.
        \subref{fig:supp:nucleosome_sc_density} The average supercoiling density for both the induced and induced case is shown.
        \subref{fig:supp:nucleosome_induction_sweep} Reporter output as a function of adjacent gene induction and circuit syntax is shown for both circular and linear boundary conditions.
        \subref{fig:supp:nucleosome_joint_distribution} The population distribution is shown as a function of boundary condition and circuit syntax for the equally-induced case. The noise ratio of each is defined as the intrinsic noise (off-diagonal component) divided by the extrinsic noise (on-diagonal component).
        \subref{fig:supp:nucleosome_zinani_examples} Four example simulation traces of the zebrafish segmentation gene network are shown, for two topoisomerase conditions and the gene-unpaired and gene-paired orientations.
        \subref{fig:supp:nucleosome_zinani_crosscorr} The ensemble cross-correlation for the zebrafish segmentation gene network with nucleosome buffering is shown. Robust periodicity in the gene-paired case remains in the presence of nucleosomes.
        \subref{fig:supp:nucleosome_zinani_correlation_coefficient} The ensemble distribution of the correlation coefficient between \textit{her1} and \textit{her7} is shown. In the presence of nucleosomes, the gene-paired case has a significantly increased correlation coefficient.
        \subref{fig:supp:nucleosome_zinani_oscillation_amplitude} The ensemble oscillation amplitude is shown for each topoisomerase and gene pairing condition.
    }
    \label{fig:top:supp_perturb_nucleosomes}
\end{figure}


\begin{figure}[htbp]
    \centering
    {\includegraphics{figures/modeling_paper/fig_supp_extra_toggles.pdf}
    \phantomsubcaption\label{fig:supp_toggles_run_examples}
    \phantomsubcaption\label{fig:supp_toggles_mean_sc_density}
    \phantomsubcaption\label{fig:supp_toggles_interburst_time}
    \phantomsubcaption\label{fig:supp:n1_toggle_distributions}
    \phantomsubcaption\label{fig:supp:toggles_hill_n_sweep}
    }
    \caption{Related to \Cref{fig:top:toggle_switch}.
        \subref{fig:supp_toggles_run_examples} Four example simulation runs for the toggle switches are placed in the four orientations. 
        \subref{fig:supp_toggles_mean_sc_density} The mean ensemble supercoiling density is plotted for the four different circuit syntaxes.
        \subref{fig:supp_toggles_interburst_time} The distribution of inter-burst times for the four different toggle switch architectures is shown.
        \subref{fig:supp:n1_toggle_distributions} The ensemble mRNA count distributions are shown as a function of syntax at four selected time points for \(n = 1.0\). Even as early as 200 seconds after induction of the second gene, we see that the tandem-downstream and divergent syntaxes are quickly approaching their equilibrium ensemble distributions.
        \subref{fig:supp:toggles_hill_n_sweep} The basin stability for the four toggle-switch architectures is shown explicitly as a function of hill coefficient, \(n\).
    }
\end{figure}

% ---- Zinani supplements ----
\begin{figure}[htbp]
    \centering
    {\includegraphics{figures/modeling_paper/fig_supp_zinani.pdf}
    \phantomsubcaption\label{fig:supp:zinani_matlab_mRNA}
    \phantomsubcaption\label{fig:supp:zinani_matlab_cross_correlation}
    \phantomsubcaption\label{fig:supp:zinani_matlab_correlation}
    \phantomsubcaption\label{fig:supp:zinani_alpha_mRNA}
    \phantomsubcaption\label{fig:supp:zinani_alpha_cross_correlation}
    \phantomsubcaption\label{fig:supp:zinani_alpha_correlation}
    \phantomsubcaption\label{fig:supp:zinani_alpha_oscillation_amplitude}
    \phantomsubcaption\label{fig:supp:intragenic_zinani_examples}
    \phantomsubcaption\label{fig:supp:intragenic_cross_correlation}
    \phantomsubcaption\label{fig:supp:zinani_topo_correlation_coefficent}
    \phantomsubcaption\label{fig:supp:zinani_topo_oscillation_amplitude}
    }
\end{figure}
\begin{figure}[htbp]
    \ContinuedFloat
    \caption{Related to \Cref{fig:top:her1_her7}.
       \subref{fig:supp:zinani_matlab_mRNA} Using the authors' MATLAB implementation of the model in \textcite{zinaniPairingSegmentationClock2021}, we simulated ensembles of 10,000 runs in order to compare to our simulation work. Unlike our implementation, which does not contain extra delays not due to polymerase motion and stalling, the original model uses a fixed mRNA production delay time of 9 minutes and a fixed protein production delay time of 1 minute. Example runs of the original model are shown. The transcript-coupled case shows correlated but relatively aperiodic behavior.
        \subref{fig:supp:zinani_matlab_cross_correlation} The ensemble cross-correlation shows that the uncoupled and transcript coupled cases in the original model show very similar correlation structures, showing that both have similar amounts of periodicity in the original model.
        \subref{fig:supp:zinani_matlab_correlation} Using Zinani's original MATLAB code, we computed the correlation coefficient distributions using both their original sampling method (far-left) and with the dataset resampled uniformly in time. The qualitative results between these two sampling methods remain the same, with the original sampling method slightly under-reporting \textit{her1}-\textit{her7} correlation due to its reliance on Gillespie-sampled timepoints. We further found that down-sampling the uniformly-sampled dataset did not noticeably change the correlation distributions.
        Re-simulation of the \textit{her1-her7} system presented in \cref{fig:top:her1_her7} is shown for \(\alpha = 0.02\), a choice that occurs in the first regime of supercoiling-dependent initiation behavior. Strikingly, the strong correlated behavior seen in \cref{fig:top:her1_her7} is absent at this choice of \(\alpha\), suggesting that supercoiling-dependent initiation is an important phenomena in the synchronization of \textit{her1} and \textit{her7} activity.
        \subref{fig:supp:zinani_alpha_mRNA} The number of \textit{her1} and \textit{her7} mRNAs are shown in the two different simulation contexts.
        \subref{fig:supp:zinani_alpha_cross_correlation} For \(\alpha = 0.02\), the biophysically-coupled case no longer shows strong periodic behavior in the cross-correlation plots, instead showing similar asymmetric cross-correlation behavior.
        \subref{fig:supp:zinani_alpha_correlation} In contrast to \cref{fig:zinani_correlation_coeff}, for \(\alpha = 0.02\), the gene paired system now shows a weaker correlated expression.
        \subref{fig:supp:zinani_alpha_oscillation_amplitude} The mean oscillation amplitude is shown for the gene unpaired and gene paired cases. The gene paired case still shows an increase in average oscillation amplitude.
        \subref{fig:supp:intragenic_zinani_examples} Example mRNA traces are shown for the zebrafish segmentation network with intragenic topoisomerase relaxation.
        \subref{fig:supp:intragenic_cross_correlation} The cross correlation of clock gene expression is shown for intragenic topoisomerase relaxation.
        \subref{fig:supp:zinani_topo_correlation_coefficent} The correlation coefficient between \textit{her1} \and \textit{her7} is shown, compared across the three different topoisomerase activity conditions.
        \subref{fig:supp:zinani_topo_oscillation_amplitude} The size of oscillations is compared across the three different topoisomerase activity conditions.
    }
\end{figure}


\FloatBarrier

{\small \noindent \textbf{Supplemental Movie 1:} An annotated example convergent simulation is shown. The top left panel shows currently loaded polymerases with growing nascent mRNAs. The bottom left panel shows the current supercoiling density plot. The right panel shows the number of mRNAs as a trace through time. The reporter gene (colored) is enabled for the length of the simulation, whereas the adjacent gene is only induced after 2.8 hours.}

{\small \noindent \textbf{Supplemental Movie 2:} An example simulation run is shown for the tandem upstream syntax. Upstream dominance, where the upstream gene has higher mean expression, is demonstrated.}

{\small \noindent \textbf{Supplemental Movie 3:} An example simulation run is shown for the tandem downstream syntax. Upstream dominance is again demonstrated.}

{\small \noindent \textbf{Supplemental Movie 4:} An example simulation run is shown for the divergent syntax. The divergent syntax demonstrates coordinated bursting.}

\FloatBarrier
\section{Model design and derivation}
\label{sec:appendix:model}
Our model builds upon several prior papers and relevant mathematical frameworks. To ensure access and provide definitions needed for implementation of our model, we have aggregated all of these relevant equations and simulation constants together. From this foundation, we discuss how we have extended our model to examine how DNA supercoiling generates coupling between proximal genes.

\subsection{Literature supercoiling model}
Why do we care about supercoiling at all? From statistical mechanics, the thermal energy unit \(k_B T\), defined by the product of the Boltzmann constant and the temperature of the system, defines an energy accessible by fluctuations; useful chemical reactions and interactions often occur above this fluctuation energy. Converting into units relevant to supercoiling at \(T = 300 \unit{K}\), \(k_b T = 4.1 \unit{pN nm}\),  Many of the later quantities we will solve for will have magnitude around 10-100 pN nm---well above the thermal energy----so can determining system behavior.

When we derived the supercoiling density, \(\sigma\), in \cref{eq:supercoiling_density}, we defined \(\phi\) to be the local excess twist. While this term is most easily identified with the local DNA excess twist, it is truly a linking number constraint that takes into account both twist, local stretching of bonds that causes more or fewer complete rotations to occur over a set distance, and writhe, where mesoscopic plectonemes store supercoiling. These two forms of supercoiling are physically interconvertable, but are not energetically identical. This energetic difference is accounted for in the underlying statistical mechanical model (\cref{fig:supp:torque_diagram}).

The statistical mechanics model depends on the following constants:
\begin{table}[h]
    \centering
    \begin{tabular}{@{}lll@{}}
        \toprule
        Variable & Value & Reference \\
        \midrule
        \(k_b\), Boltzmann constant & \(0.01381 \units{pN nm}{K}\) & - \\
        \(\omega_0\), relaxed DNA twist frequency & 1.85 \(\units{1}{nm}\) & \parencite{sevierPropertiesGeneExpression2018}, others \\
        \midrule
        A, bend persistence length & \(50\)nm & \parencite{markoTorqueDynamicsLinking2007} \\
        C, twist persistence length (twisted) & \(95 \pm 10\)nm & \parencite{markoTorqueDynamicsLinking2007} \\
        P, twist persistence length (plectonemic) & \(24 \pm 3\)nm & \parencite{markoTorqueDynamicsLinking2007} \\
        \(f\), applied force on DNA & \(1 \unit{pN}\) & \parencite{sevierPropertiesGeneExpression2018} \\
        \bottomrule
    \end{tabular}
    \caption{Key values for the statistical mechanics model}
    \label{tab:constants}
\end{table}
\FloatBarrier
Especially key measurable constants are the bend persistence length \(A\), the twist persistence length of stretched DNA \(C\), and the twist persistence length of the plectonemic
state \(P\). Because plectonemes convert twist into writhe, we expect \(P < C\).

Marko defines the following rescaled versions of \(P\) and \(C\), \(p,c\):
\[c = k_B T C \omega_0^2 \qquad p = k_B T P \omega_0^2\]
which in turn can be used to define helper constants \(g, c_s\):
\[g = f - \sqrt{\frac{k_B T f}{A}} \qquad c_s = c \left[1 - \frac{C}{4A} \sqrt{\frac{k_B T}{A f}}\right]\]

With these defined, Marko's statistical-mechanics model calculates the torque as a function of supercoiling density (\cref{fig:supp:torque_diagram}) \parencite{markoTorqueDynamicsLinking2007}:
\begin{equation}
    \tau(\sigma) = \begin{dcases} \tau_S \sigma & |\sigma| < |\sigma_s| \\ \tau_0 \text{sgn}(\sigma) & |\sigma_s| \leq |\sigma | \leq |\sigma_p| \\ \tau_P \sigma & |\sigma_p| \leq |\sigma|\end{dcases}
    \label{eq:marko_torque}
\end{equation}
for critical supercoiling values \(\sigma_s, \sigma_p\) separating the phase regimes:
\begin{equation}
    |\sigma_s| = \frac{1}{c_s} \sqrt{\frac{2pg}{1 - \frac{p}{c_s}}} \qquad |\sigma_p| = \frac{1}{p} \sqrt{\frac{2pg}{1 - \frac{p}{c_s}}}
    \label{eq:critical_sigma_values}
\end{equation}
with corresponding critical torque values \(\tau_S, \tau_P\):
\begin{equation}
    \tau_S = \frac{c_s}{\omega_0} \qquad \tau_0 = \sqrt{\frac{2 pg}{\omega_0^2 \left(1 - \frac{p}{c_s}\right)}} \qquad \tau_P = \frac{p}{\omega_0}
    \label{eq:critical_tau_values}
\end{equation}

Using the torque equation \cref{eq:marko_torque}, Marko expands the free energy of the DNA per unit length, \(S\), as a power series in \(\sigma\):
\begin{equation}
    S = \begin{dcases}
        -g + \frac{1}{2} c_s \sigma^2 & |\sigma| < |\sigma_s| \\
        \frac{-g}{1 - \frac{p}{c_s}} + \sqrt{\frac{2 p g}{1 - \frac{p}{c_s}}} |\sigma| & |\sigma_s| < |\sigma| < |\sigma_p| \\
        \frac{1}{2} p \sigma^2 & |\sigma| > |\sigma_p|
    \end{dcases}
    \label{eq:marko_energy}
\end{equation}
While the cubic term can also be calculated and accounts for asymmetry between under- and over-winding,  \textcite{markoTorqueDynamicsLinking2007} argues that this effect is small for realistic physiological conditions.

\subsection{Literature dynamic model}
Following \textcite{sevierPropertiesGeneExpression2018}, we adopt a dynamics model for modeling the motion of polymerases over the DNA. This model depends on the following constants:

\begin{table}[h]
    \centering
    \begin{tabular}{@{}lll@{}}
        \toprule
        Variable & Value & Reference \\
        \midrule
        \(\chi\), DNA twist mobility  & \(0.01381 \unit{s pN nm}\) & \parencite{sevierPropertiesGeneExpression2018} \\
        \(\eta\), mRNA drag coefficent  & \(1/20 \unit{pN}\) & \parencite{sevierPropertiesGeneExpression2018} \\
        \(n\), Drag scaling exponent  & \(1\) & \parencite{sevierPropertiesGeneExpression2018} \\
        \(v_0\), Maximum polymerase velocity  & \(20 \units{nm}{s}\) & \parencite{sevierPropertiesGeneExpression2018} \\
        \(\tau_s\), Polymerase stall torque  & \(12 \unit{pN nm}\) & \parencite{sevierPropertiesGeneExpression2018} \\
        \(\Delta \tau_s\), Polymerase stall width  & \(3 \unit{pN nm}\) & This work \\
        \(\delta\), Polymerase radius  & \(15 \unit{nm}\) & \parencite{sevierPropertiesGeneExpression2018} \\
        \bottomrule
    \end{tabular}
    \caption{Key dynamic model values}
    \label{tab:dynamic_model_constants}
\end{table}
\FloatBarrier
Briefly, for each polymerase, this model tracks the position \(z_i\), transcript length \(x_i\), and excess DNA twist \(\phi_i\) of each RNAP. We create a \emph{linking number constraint} by specifying \(\phi\) at any genomic location, which means that each RNAP sets an linking number constraint. We use dynamic equations to model the motion of RNAP against either fixed or free boundary conditions.

As mentioned in the main text, the dynamic equations are:

\begin{equation*}
    \omega_0 \frac{d z_i}{dt} = \frac{d \theta_i}{dt} + \frac{d \phi_i}{dt} \qquad \qquad \tau(z_i, \phi_{i-1}, \phi_{i+1}) = \eta x_i^n \frac{d\theta_i}{dt} - \chi \frac{d\phi_i}{dt}
\end{equation*}

By solving the left equation for \(\frac{d \theta_i}{dt}\), the polymerase angular velocity, and substituting it into the right equation, we get:
\[\tau(z_i, \phi_{i-1}, \phi_{i+1}) = \omega_0 \frac{dz_i}{dt} \eta x_i^n - \eta x_i^n \frac{d\phi_i}{dt} - \chi \frac{d\phi_i}{dt}\]
Solving for \(\frac{d\phi_i}{dt}\), we get a dynamic equation for the differential linking number constraint:

\begin{equation}
    \frac{d\phi_i}{dt} = \omega_0 \frac{dz_i}{dt} \frac{\eta x_i^n}{\chi + \eta x_i^n} - \frac{\tau(z_i, \phi_{i-1}, \phi_{i+1})}{\chi + \eta x_i^n}
\end{equation} \label{eq:phi_deriv}

To model polymerase stalling, we use a parameterized function discussed in \cref{eq:velocity_response} that explicitly depends on a stall torque and a torque width \(\tau_w\) over which stalling begins to occur.

Equations \ref{eq:phi_deriv} and \ref{eq:velocity_response} along with the definition of linear velocity \(\frac{dx_i}{dt} = \frac{dz_i}{dt}\) together represent a complete coupled set of ODEs that can be simulated, as long as the following constants are defined:


\subsection{Model extension: supercoiling-dependent initiation} \label{sec:sc_initation}
Reflecting on the desired properties of a model based on literature observations, we extended the models from Sevier and Marko to account for supercoiling-dependent initiation. The rest of this appendix details this process, in addition to the initial model formulation and useful simplifications.

\subsubsection{Literature observations and modeling goals}
A naked DNA bead experiment (\textcite{revyakinPromoterUnwindingPromoter2004}) found that positive and negative supercoiling affected RNAP polymerase binding rates, with negative supercoiling encouraging binding and positive supercoiling discouraging binding. When a polymerase bound, approximately 1.2 turns of DNA unwound (13 base pairs, or 4.42nm). For a typical bacterial promoter, RNAP binding occurred reversibly. The mean time between RNAP binding events was measured and related to a kinetic model. The mean time between binding events defines a promoter-on rate. Once the DNA torque reaches the constant-torque regime, the promoter-on rate becomes constant.

Ideally, our model would replicate these results, with negative supercoiling promoting RNAP binding and negative supercoiling inhibiting RNAP binding, but would only depend on the local supercoiling density. Successful phenomenological models such as \textcite{elhoudaiguiBacterialGenomeArchitecture2019a} often use sigmodial curves that are most sensitive in the range from \(\sigma = -0.1\) to \(\sigma = 0.1\); ideally, our model would be sensitive in a similar regime.

\FloatBarrier
\subsubsection{Dual linking number constraint model}
To model the unwinding induced by RNAP binding, we add two new linking number constraints, spaced 13 base pairs apart, and analyze the energy required to perform this unwinding. To start, all of these have specific linear locations \(z\) and rotations \(\phi\); we will relax this restriction later.

\begin{figure}[h]
    \centering
    \begin{tikzpicture}
        \draw (-5.1, 0) -- (-2.9, 0);
        \draw (-5, -.2) -- (-5, .2);
        \draw (-3, -.2) -- (-3, .2);
        \node [below] at (-4.88, -.2) {$\phi_{i-1}$};
        \node [below] at (-2.99, -.2) {$\phi_i$};
        \node [above] at (-4, 0) {$\sigma_0$};

        \draw [ultra thick,->,>=stealth] (-2.5, 0) -- (-1.5, 0);

        \draw (-1.1, 0) -- (1.1, 0);
        \draw (-1, -.2) -- (-1, .2);
        \draw (1, -.2) -- (1, .2);
        \draw (.3, -.2) -- (.3, .2);
        \draw (-.1, -.2) -- (-.1, .2);
        \node [below] at (-.88, -.2) {$\phi_{i-1}$};
        \node [below] at (1.04, -.2) {$\phi_i$};
        \node [below] at (.43, -.2) {$\phi''$};
        \node [below] at (-.03, -.2) {$\phi'$};
        \node [above] at (.65, -.04) {$\sigma_f$};
        \node [above] at (-.5, 0) {$\sigma_b$};
    \end{tikzpicture}
    \caption{Diagram of the addition of two linking number constraints}
    \label{fig:lnc_diagram}
\end{figure}

We then put an angular restriction on the two new linking number constraints. By subtracting off the excess angle \(\phi\) that the undisturbed DNA has prior to RNAP binding, we can define:
\begin{equation}
    \Delta \phi' = \phi' - \left(\phi_{i - 1} + \frac{z' - z_{i-1}}{z_i - z_{i-1}} (\phi_i - \phi_{i-1})\right)
\end{equation}

\begin{equation}
    \Delta \phi'' = \phi'' - \left(\phi_{i - 1} + \frac{z'' - z_{i-1}}{z_i - z_{i-1}} (\phi_i - \phi_{i-1})\right)
\end{equation}

For unwinding of the region bound to the RNAP to occur, we impose two constraints that ensure that the 13 base pairs that interact with the RNAP are fully unwound:
\begin{equation}
    |\Delta \phi'| + |\Delta \phi''| = 1.2 \cdot 2 \pi
\end{equation}
\begin{equation}
    \Delta \phi' - \Delta \phi'' = 1.2 \cdot 2\pi \label{eq:dphi_relation}
\end{equation}
For the intermediate region, this implies that there are two different ways to unwind the DNA; either to rotate the leading linking number constraint backwards or rotate the trailing linking number constraint forwards. There are also many intermediate solutions where \emph{both} of the bounding linking number constraints move. The first restriction is there to ensure that we use the solution that minimizes total amount of rotation needed.

Given a complete energy model that related \(\tau\) over the complete range of unwinding \(\Delta \phi', \Delta \phi''\) values, we could directly write the binding energy as
\begin{equation}
    \Delta E = \overline \tau' \Delta \phi' + \overline \tau'' \Delta \phi''
    \label{eq:direct_torque_calc}
\end{equation}
and minimize the energy cost with respect to \(\Delta \phi'\) or \(\Delta \phi''\), giving a unique solution for any specific geometry. As Marko's statistical mechanical model is likely not valid in the limit of complete unwinding, we instead move forward with this model by estimating the energy cost in both the ``exterior'' region and within the ``interior'', 13-bp region.

\subsubsection{Energy estimation in the exterior region}
If we assume that the energy surface is locally linear (e.g. in the exterior region, there is a small change in supercoiling density), then we can relate the instantaneous torque to the change in energy:
\[ \tau = \frac{1}{\omega_0} \frac{\partial S(\sigma)}{\partial \sigma}\]
for \(S\), the energy per unit length. This means that:
\begin{equation}
    \Delta E = (z_2 - z_1) \tau \omega_0 \Delta \sigma = (z_2 - z_1) \frac{\partial S}{\partial \sigma} \Delta \sigma
\end{equation}
By expanding the definition of \(\Delta \sigma\) for an arbitrary region bounded by two linking number constraints, we can actually recover the normal energy-torque definition:
\begin{equation}
    \Delta E_\text{external} = (z_2 - z_1) \tau \omega_0 \frac{\Delta\phi_2 - \Delta\phi_1}{\omega_0 (z_2 - z_1)} = \tau (\Delta \phi_2 - \Delta\phi_1) \label{eq:external_de}
\end{equation}
Note that this implies that the energy cost of linking number constraint rotation is independent of the region width; under a locally linear assumption, the energy cost depends only on the local instantaneous torque and the rotation angle. When is this valid? The change in supercoiling density in the exterior region is relatively low, \(O(0.05)\) as long as the region of interest is at least \(250bp\) away from the nearest linking number constraint on one of its sides. For promoter regions analyzed here, space between gene bodies ensures that this assumption holds.

\subsubsection{Energy estimation in the interior region (local melting)}
We can generally write the energy it takes to unwind the DNA in the interior region as an integral over the linear free energy density, \(S\):
\[\Delta E_\text{internal} =  \int_{\sigma_0}^{\sigma_0 + \Delta \sigma_\text{RNAP}} \Delta z_\text{RNAP} \frac{\partial S}{\partial \sigma} d\sigma\]

Given the well-defined structure of the RNAP-DNA complex and the fact that under all realistic physiological conditions \(\Delta \sigma_\text{RNAP} \gg \sigma_0\), we assume that the end-state energy is a constant, but unknown \(S_\text{unwound}\):
\begin{equation}
    \Delta E_\text{internal} = \Delta z_\text{RNAP} \left(S_\text{unwound} - S(\sigma_0)\right)
\end{equation}
While \(S_\text{unwound}\) is unknown, it is a constant with respect to all promoters, so this energetic term is implicitly already included in any promoter base rate. This means that the external energy depends only on the free energy density, evaluated at the local, initial supercoiling density \(\sigma_0\):
\begin{equation}
    \Delta E_\text{internal} = -\Delta z_\text{RNAP} S(\sigma_0) \label{eq:internal_de}
\end{equation}

\subsubsection{Order of magnitude energy analysis}
Combining the results of \cref{eq:external_de,eq:internal_de} and plugging in \cref{eq:dphi_relation}, we have an estimated binding energy:
\begin{equation}
    \Delta E = \tau(\sigma_0) 1.2 \cdot 2\pi - \Delta z_\text{RNAP} S(\sigma_0)
\end{equation}

When we plot this energetic term in \cref{fig:supp:energy_with_melting}, we see that the internal energetic term is minimal when compared to the external energetic term (local melting) at the supercoiling densities considered here, so we use the simplified form

\begin{equation}
    \Delta E = \tau(\sigma_0) \cdot 1.2 \cdot 2\pi
\end{equation}
as presented in \cref{eq:first_order_sc_initation}.

\end{document}
